---
title: "Single Cell Analysis with Seurat using singleCellTK Package"
author: "`r params$authors`"
date: "`r format(Sys.time(), '%B %d, %Y')`"
params:
  subtitle: !r NULL
  authors: !r NULL
  sce: !r NULL
  biological.group: !r NULL
  phenotype.groups: !r NULL
  selected.markers: !r NULL
  clustering.resolution: 0.8
  variable.features: 2000
  pc.count: 50
  outputPath: !r NULL
  showSession: TRUE
  pdf: TRUE
  jackStraw: FALSE
output:
  html_document:
    toc: yes
    toc_float: yes
    number_sections: yes
    theme: cosmo
    code_folding: hide
    self_contained: false
  pdf_document:
    toc: yes
subtitle: "`r params$subtitle`"
editor_options:
  chunk_output_type: console
---

```{r, include = TRUE, message=FALSE, warning=FALSE, echo=FALSE}
#Loading required libraries
library(Seurat)
library(dplyr)
library(cowplot)
library(RColorBrewer)
library(ggplot2)
library(knitr)
library(kableExtra)
library(SingleCellExperiment)
library(scater)
library(gridExtra)
library(grid)
library(ggpubr)
library(patchwork)
library(singleCellTK)

#Setting up parameters and loading data
group <- params$biological.group
selected_phenotypes <- params$phenotype.groups
Pre_selected_markers <- params$selected.markers
seurat.selected.resolution <- as.numeric(params$clustering.resolution)
seurat.selected.features.numbers <- as.numeric(params$variable.features)
initial.PC.numbers <- as.numeric(params$pc.count)
data <- params$sce
sampleNames <- unique(colData(data)[[group]])
outputPath <- params$outputPath
showSession <- params$showSession
pdf <- params$pdf

numCells <- ncol(data)
numFeatures <- nrow(data)
numSamples <- length(sampleNames)

# Knitr Global Options
dev <- ifelse(isTRUE(pdf), c("png"), c("png", "pdf"))
opts_chunk$set(
  echo = TRUE,
  cache = FALSE,
  cache.lazy = FALSE,
  cache.comments = FALSE,
  fig.align = "center",
  fig.keep = "all",
  dev = dev,
  warning = FALSE
)
```

<!-- Compute -->

<!-- Normalization -->
```{r, echo=FALSE, results='asis', warning=FALSE, message=FALSE}
resNormalization <- knitr::knit_child(paste0(getwd(), "/inst/rmarkdown/seurat/", 'reportSeuratNormalizeData.Rmd'), quiet = TRUE)
```

<!-- Feature Selection -->
```{r, echo=FALSE, results='asis', warning=FALSE, message=FALSE}
runHVG <- TRUE
plotHVG <- TRUE
resFeatureSelection <- knitr::knit_child(paste0(getwd(), "/inst/rmarkdown/seurat/", 'reportSeuratFeatureSelection.Rmd'), quiet = TRUE)
```

<!-- Scale Data -->
```{r, echo=FALSE, results='asis', warning=FALSE, message=FALSE}
resScaleData <- knitr::knit_child(paste0(getwd(), "/inst/rmarkdown/seurat/", 'reportSeuratScaleData.Rmd'), quiet = TRUE)
```

<!-- Dimensionality Reduction -->
```{r, echo=FALSE, results='asis', warning=FALSE, message=FALSE}
runDimRed <- TRUE
plotJackStraw <- TRUE
plotElbowPlot <- TRUE
plotHeatmaps <- FALSE
significant_PC <- 10

resDimRed <- knitr::knit_child(paste0(getwd(), "/inst/rmarkdown/seurat/", 'reportSeuratDimRed.Rmd'), quiet = TRUE)
```

<!-- Clustering -->
```{r, echo=FALSE, results='asis', warning=FALSE, message=FALSE}
runClustering <- TRUE
minResolution <- 0.3
maxResolution <- 1.5
numClusters <- 10
showClusterDesc <- TRUE

resClustering <- knitr::knit_child(paste0(getwd(), "/inst/rmarkdown/seurat/", 'reportSeuratClustering.Rmd'), quiet = TRUE)
```

# Summary {-}
<!-- Summary -->
<!-- some variables still need to be included, e.g. from deg -->
> The input data included *`r numFeatures`* features and *`r numCells`* cells from *`r numSamples`* samples divided through the group *`r group`* namely *`r sampleNames`*. The *`r assayNames(data)[1]`* feature count matrix from the data was normalized using *LogNormalize* method which normalized as well as log-transformed the data. From *`r numFeatures`* total features, a subset of *`r seurat.selected.features.numbers`* features that exhibited high cell-to-cell variation were selected for the downstream analysis to better capture the biological variation using *vst* (mean-to-variance) feature selection method. This subset of features was then scaled (z-score) using *linear model* method and trimmed to a maximum and minimum value of *10* and *-10* respectively. PCA was run on this scaled subset of features and *`r initial.PC.numbers`* components were computed from which top *`r significant_PC`* significant components were selected for subsequent clustering analysis. Using the *louvain* algorithm that uses community-based detection to identify clusters with resolution set to *`r seurat.selected.resolution`*, clustering was performed on the scaled subset of features that identified *`r numClusters`* clusters in the data. These clusters were then used to run differential expression using *wilcox* test and identified *`r 2665`* genes differentially expressed between clusters at significance criteria of *q-value < 0.05*. The gene markers that defined the main biological group *`r group`* were identified using *wilcox* test that highlighted *2651* marker genes at *q-value < 0.05* significance criteria. Additionatlly, markers *`r Pre_selected_markers`* from the input pre-selected markers *`r Pre_selected_markers`* were detected and visualized over UMAP plot.

# Seurat Run

## Introduction
[Seurat](https://satijalab.org/seurat/) is a popular R package designed for the analysis of single-cell genomics data and offers a multitude of methods for processing and visualization. The singleCellTK package integrates these methods in an intuitive user-interface and all of which are also accessible through console-based wrapper functions. This comprehensive report summarizes the Seurat workflow for the input data by running all of the steps from the proposed pipeline including Normalization, Feature Selection, Scaling, Dimensionality Reduction, Clustering and Differential Expression, along with all of the supporting visualizations. A summary of the input data and results computed from this data is described below:

<!-- Populate RMD with Results -->
```{r, echo=FALSE, results='asis', warning=FALSE, message=FALSE}
cat(resNormalization, sep = '\n')
cat(resFeatureSelection, sep = '\n')
cat(resScaleData, sep = '\n')
cat(resDimRed, sep = '\n')
cat(resClustering, sep = '\n')
```

<!-- Save Data File -->
```{r}
saveRDS(data, "data.rds")
```