---
title: "GSVA"
author: "Irzam Sarfraz"
output:
  html_document:
    toc: true
    toc_depth: 2
bibliography: references.bib
csl: ieee.csl
---
```{r develSetting, include=FALSE}
doEval = TRUE
```
## Introduction
The ***GSVA*** (Gene Set Variation Analysis) ([HÃ¤nzelmann, S., Castelo, R. & Guinney, J. GSVA: gene set variation analysis for microarray and RNA-Seq data. BMC Bioinformatics 14, 7 (2013).](https://doi.org/10.1186/1471-2105-14-7)) is a popular method for ***GSE*** (Gene Set Enrichment) and allows the identification of changes in pathway activity
in RNA-Seq data. Overall, ***gene sets*** can be uploaded or selected from the available databases and the ***GSVA*** method can be run on the selected ***gene sets*** which can eventually be visualized using ***Violin*** or ***Heatmap*** plots.

To view detailed instructions on how to use the method from the toolkit, please select 'Interactive Analysis' for using GSVA in shiny application or 'Console Analysis' for using this method on R console from the tabs below:


````{=html}
<head>
<meta name="viewport" content="width=device-width, initial-scale=1">
</head>
<body>
````


## Workflow Guide

````{=html}
<div class="tab">
  <button class="tablinks" onclick="openTab(event, 'interactive')" id="ia-button">Interactive Analysis</button>
  <button class="tablinks" onclick="openTab(event, 'console')" id="console-button">Console Analysis</button>
</div>

<div id="interactive" class="tabcontent">
````

Before running the ***GSVA*** algorithm, the gene sets must be selected or uploaded through the ***Import Gene Sets*** sub-tab from the ***Data*** tab in the toolkit menu. An overall workflow guide on how to import/upload ***gene sets*** and consequently using them with ***GSVA*** sub-tab is described below. ***This workflow guide assumes that the *SingleCell Data* has already been uploaded through the *Import Single Cell Data*** sub-tab.**<br>

![](ui_screenshots/pathway_analysis/capa_1.png)

1. Select the ***Data*** tab from the ***singleCellTK*** top menu.<br>

2. Select the ***Import Gene Sets*** from the drop-down menu.<br>

![](ui_screenshots/pathway_analysis/capa_2.png)

3. Users can choose  the ***Upload a GMT file*** option to upload gene sets directly stored in a GMT file or choose ***Select from a database*** option to select gene sets from multiple databases integrated within the toolkit or lastly choose the ***Paste in your gene set*** option to paste gene identifiers directly into the user-interface to create a gene set.<br>

4. (a) If ***Upload a GMT file*** option is selected, users can upload a .GMT file containing the gene sets. <br>

![](ui_screenshots/pathway_analysis/capa_3.png)

4. (b) If ***Select from a database*** option is selected, the available gene sets are populated. Users can select the gene sets and click on the ***Upload*** button to get them to use in the downstream analysis. <br>

![](ui_screenshots/pathway_analysis/capa_4.png)

5. Once ***gene sets*** are uploaded, the selected gene sets are listed in this table. <br>

![](ui_screenshots/pathway_analysis/capa_5.png)

6. To run the ***GSVA*** algorithm, go to ***Cell Annotation & Pathway Analysis*** tab. <br>

![](ui_screenshots/pathway_analysis/capa_6.png)

7. Select ***GSVA*** option from the drop-down menu. <br>

![](ui_screenshots/pathway_analysis/capa_7.png)

8. Select ***assay*** to use with ***GSVA***. <br>
9. Select ***GSVA*** as the method.
10. Select ***gene sets*** to use with ***GSVA***. Only uploaded ***gene sets*** from the ***Import Gene Sets*** page are available here. <br>
11. Select a ***phenotype variable*** to use with the ***plots***. <br>
12. Select if a ***Heatmap*** should be plotted or a ***Violin** plot. <br>
13. Press ***Run*** to start computation. <br>
14. Once computation is complete, users use the ***Save Pathways*** button to store the pathways in the internal object for use in downstream analysis or use the ***Download Pathway Results*** option to download the results to the local drive. <br>

![](ui_screenshots/pathway_analysis/capa_8.png)

15. Results can be visualized by selecting the ***Plot*** tab **(16)** or the ***Results Table*** tab **(17)**. <br>



````{=html}
</div>
<div id="console" class="tabcontent">
````

**1. Upload or import gene sets:**
```{r, eval=FALSE}
# Import a GMT file
gmtFilePath <- "C:/Users/HP/Desktop/gmtfile.GMT" # Specify user-defined location for the GMT file
sce <- importGeneSetsFromGMT(inSCE = sce, file = gmtFilePath, by = "rownames", collectionName = "Collection1")

# Alternatively
# Select from an existing database
sce <- importGeneSetsFromMSigDB(inSCE = sce, categoryIDs = "C1", by = "rownames")
```

**2. Run GSVA using `gsvaSCE` function:**
```{r, eval=FALSE}
gsvaRes <- gsvaSCE(inSCE = sce, useAssay = "logcounts", pathwayNames = c("C1"))
```

The parameters to the `gsvaSCE` function above include: <br>
`inSCE`: input `SingleCellExperiment` object that contains the stored gene sets <br>
`useAssay`: name of the assay to use for GSVA computation <br>
`pathwayNames`: names of the pathways to use with GSVA (previously stored in the input object) <br>

**3. Visualize the results:**
```{r, eval=FALSE}
# Fit Limma & Populate Results
conditionFactor <- factor(colData(sce)[, "Biological_Condition"])
      if(nlevels(conditionFactor)<=1){
        stop("Condition variable must have atleast two levels!")
      }
fit <- limma::lmFit(gsvaRes, stats::model.matrix(~conditionFactor))
fit <- limma::eBayes(fit)
toptableres <- limma::topTable(fit, number = nrow(gsvaRes))
temptable <- cbind(rownames(toptableres), toptableres)
rownames(temptable) <- NULL
colnames(temptable)[1] <- "Pathway"
     

tempgsvares <- gsvaRes
topPaths <- 10

# Visualize violin plot
tempgsvares <- tempgsvares[1:min(49, topPaths, nrow(tempgsvares)), , drop = FALSE]
gsvaPlot(inSCE = sce, gsvaData = tempgsvares, plotType = "Violin", condition = "Biological_Condition")

# Visualize heatmap plot
gsvaPlot(inSCE = sce, gsvaData = tempgsvares, plotType = "Heatmap", condition = "Biological_Condition")
```

````{=html}
  <details>
  <summary><b>Example</b></summary>
````

```{r "normalization_console_example_1", message = FALSE, warning = FALSE}
  # Load singleCellTK & pbmc3k example data
  library(singleCellTK)
  sce <- importExampleData(dataset = "pbmc3k")
 #hallmark c7,c8
  #runnormalization
  sce <- importGeneSetsFromMSigDB(inSCE = sce, categoryIDs = "C1", by = "rownames")
  sce <- scaterlogNormCounts(inSCE = sce, useAssay = "counts", assayName = "logcounts")
  sce <- seuratFindHVG(inSCE = sce, useAssay = "logcounts")
  sce <- seuratPCA(inSCE = sce, useAssay = "logcounts", reducedDimName = "pca")
  sce <- seuratFindClusters(inSCE = sce, useAssay = "logcounts", useReduction = "pca", resolution = 1.5)
  colData(sce)$Biological_Condition <- colData(sce)$Seurat_louvain_Resolution1.5
  sce <- sce[, 1:200]
  gsvaRes <- gsvaSCE(inSCE = sce, useAssay = "logcounts", pathwayNames = c("C1"))
  
  conditionFactor <- factor(colData(sce)[, "Biological_Condition"])
      if(nlevels(conditionFactor)<=1){
        stop("Condition variable must have atleast two levels!")
      }
fit <- limma::lmFit(gsvaRes, stats::model.matrix(~conditionFactor))
fit <- limma::eBayes(fit)
toptableres <- limma::topTable(fit, number = nrow(gsvaRes))
temptable <- cbind(rownames(toptableres), toptableres)
rownames(temptable) <- NULL
colnames(temptable)[1] <- "Pathway"
     

tempgsvares <- gsvaRes
topPaths <- 10

# Visualize violin plot
tempgsvares <- tempgsvares[1:min(49, topPaths, nrow(tempgsvares)), , drop = FALSE]
gsvaPlot(inSCE = sce, gsvaData = tempgsvares, plotType = "Violin", condition = "Biological_Condition")

# Visualize heatmap plot
gsvaPlot(inSCE = sce, gsvaData = tempgsvares, plotType = "Heatmap", condition = "Biological_Condition")
```

````{=html}
    </details> 
      </div>
<script>
document.getElementById("ia-button").click();
</script>
</body>
````

## References