---
title: "Quality Control in R console"
---
<style>
body {
text-align: justify}
</style>

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Introduction

Performing comprehensive quality control (QC) is necessary to remove poor quality cells for downstream analysis of single-cell RNA sequencing (scRNA-seq) data. Within droplet-based scRNA-seq data, droplets containing cells must be differentiated from empty droplets. Therefore, assessment of the data is required, for which various QC algorithms have been developed. In singleCellTK, we have written convenience functions for several of these tools. In this guide, we will demonstrate how to use these functions to perform quality control on unfiltered, droplet-level data.

The package can be loaded using the `library` command.

```{r load_package, eval=TRUE, message=FALSE, warning=FALSE}
library(singleCellTK)
library(dplyr)
```


## Running quality control of droplet raw count data with singleCellTK
### Load PBMC4k data from 10X

singleCellTK takes in a `SingleCellExperiment` object from the [SingleCellExperiment](https://bioconductor.org/packages/release/bioc/html/SingleCellExperiment.html) package. We will utilize the 10X PBMC 4K dataset as an example. For the quality control of droplet-based counts data, we will install the dataset from the 10X Genomics website using the `BiocFileCache`(https://www.bioconductor.org/packages/release/bioc/html/BiocFileCache.html) package.

```{r, message=FALSE, echo=TRUE, eval=TRUE, warning=FALSE}
# Install BiocFileCache if is it not already
if (!requireNamespace("BiocFileCache", quietly = TRUE)) {
  if (!requireNamespace("BiocManager", quietly = TRUE)) {
    install.packages("BiocManager")
  }
  BiocManager::install("BiocFileCache")
}

library("BiocFileCache")
bfc <- BiocFileCache::BiocFileCache("raw_data", ask = FALSE)
raw.path <- bfcrpath(bfc, file.path(
  "http://cf.10xgenomics.com/samples",
  "cell-exp/2.1.0/pbmc4k/pbmc4k_raw_gene_bc_matrices.tar.gz"
))
untar(raw.path, exdir = file.path(tempdir(), "pbmc4k"))

fname <- file.path(tempdir(), "pbmc4k/raw_gene_bc_matrices/GRCh38")
pbmc4k.droplet <- DropletUtils::read10xCounts(fname, col.names = TRUE)

### change the sample column
names(colData(pbmc4k.droplet)) <- c("sample", "Barcode")
colData(pbmc4k.droplet)$sample <- rep("pbmc4k", ncol(colData(pbmc4k.droplet)))
```

singleCellTK also supports the importing of single-cell data from the following platforms: [10X CellRanger](https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/what-is-cell-ranger), [STARSolo](https://github.com/alexdobin/STAR/blob/master/docs/STARsolo.md), [BUSTools](https://www.kallistobus.tools/), [SEQC](https://github.com/ambrosejcarr/seqc), [DropEST](https://github.com/hms-dbmi/dropEst), and [AnnData](https://github.com/theislab/anndata). To load your own input data, please refer [Import data into SCTK](./import_scRNAseq_data_as_SCE_Console.html) for deailed instruction. 

### runDropletQC

All droplet-based QC functions are able to be run under the wrapper function runDropletQC. By default all possible QC algorithms will be run.

```{r rundropletqc, eval=TRUE, message = TRUE}
pbmc4k.droplet <- runDropletQC(pbmc4k.droplet)
```

If users choose to only run a specific set of algorithms, they can specify which to run with the algorithms parameter.

When quality control functions are run in singleCellTK, the output of the function is stored in the colData slot of the SingleCellExperiment object.

```{r, coldata_drop, echo=TRUE, eval=FALSE}
head(colData(pbmc4k.droplet), 5)
```

```{r, coldata_drop2, echo=TRUE, eval=TRUE}
df.matrix <- head(colData(pbmc4k.droplet), 5)
df.matrix %>%
  knitr::kable(format = "html") %>% kableExtra::kable_styling() %>%
  kableExtra::scroll_box(width = "80%")
```

A summary of all outputs is shown below:
```{r table2, fig.wide = TRUE, echo=FALSE, message=FALSE, warnings=FALSE, results='asis', fig.align="center"}
tabl <- "
| QC output                              | Description                                                       | Methods             | Package/Tool  |
|:---------------------------------------|:------------------------------------------------------------------|:--------------------|:--------------|
| dropletUtils_emptyDrops_total          | Total counts                                                      | runEmptyDrops       | DropletUtils  |
| dropletUtils_emptyDrops_logprob        | The log-probability of droplet being empty                        | runEmptyDrops       | DropletUtils  |
| dropletUtils_emptyDrops_pvalue         | Monte Carlo p-value of droplet being empty                        | runEmptyDrops       | DropletUtils  |
| dropletUtils_emptyDrops_limited        | Whether a lower p-value could be obtained by increasing niters    | runEmptyDrops       | DropletUtils  |
| dropletUtils_emptyDrops_fdr            | p-value of droplet being empty, corrected for false detection rate| runEmptyDrops       | DropletUtils  |
| dropletUtils_BarcodeRank_Knee          | Whether total UMI count value is higher than knee point           | runBarcodeRankDrops | DropletUtils  |
| dropletUtils_BarcodeRank_Inflection    | Whether total UMI count value is higher than inflection point     | runBarcodeRankDrops | DropletUtils  |
"
cat(tabl)
```

### Plotting QC metrics
Upon running runDropletQC, the QC outputs will need to be plotted. For each QC method, singleCellTK contains specialized plotting functions that may be used for the generation of QC plots.

#### EmptyDrops

It is crucial to distinguish the data occurring from real cells and empty droplets containing ambient RNA. SCTK employs the [EmptyDrops](https://rdrr.io/github/MarioniLab/DropletUtils/man/emptyDrops.html) algorithm from the [DropletUtils](https://bioconductor.org/packages/release/bioc/html/DropletUtils.html) package to test for empty droplets. 

The wrapper function `plotEmptyDropsResults` can be used to plot the results from the EmptyDrops algorithm. This will visualize the empty droplets, by plotting the total UMI counts against the log probability for each barcode.

```{r run_emptydrops, message=FALSE}
emptyDropsResults <- plotEmptyDropsResults(
  inSCE = pbmc4k.droplet,
  axisLabelSize = 20,
  sample = NULL,
  fdrCutoff = 0.01,
  dotSize = 0.5,
  defaultTheme = TRUE
)
```

```{r scatter_emptydrops, fig.height = 4, fig.width = 8, fig.align="center"}
emptyDropsResults$scatterEmptyDrops
```
<br />
Data points are colored by FDR values, where we see a small portion of the dataset contains barcodes that do not meet the threshold.

#### BarcodeRanks

[BarcodeRanks](https://rdrr.io/bioc/DropletUtils/man/barcodeRanks.html) from the [DropletUtils](https://bioconductor.org/packages/release/bioc/html/DropletUtils.html) package computes barcode rank statistics and identifies the knee and inflection points on the total count curve. The knee and inflection points on the curve represent the difference between empty droplets and cell-containing droplets with much more RNA. The lower parameter is again the lower bound of the total UMI count, in which all barcodes below the lower bound are assumed to be empty droplets.

```{r run_kneeplot_barcoderank, message=FALSE, warning=FALSE, fig.align="center", fig.height = 4, fig.width = 8}
plotBarcodeRankScatter(
  inSCE = pbmc4k.droplet,
  title = "BarcodeRanks Rank Plot",
  legendSize = 14
)
```

The total UMI count of each barcode is plotted against its rank, where we see a steep dropoff of UMI counts around the inflection point, where we see a separation between cell containing and empty droplets.

### Filtering the dataset

SingleCellExperiment objects can be subset by its colData using `subsetSCECols`. The `colData` parameter takes in an expression in character vector form which will be used to identify a subset of columns using variables found in the colData of the SingleCellExperiment object. For example, if x is a numeric vector in colData, then “x < 5” will return all columns with x less than 5. The `index` parameter takes in a vector of indices which should be kept, while `bool` takes in a vector of TRUE or FALSE which should be the same length as the number of columns in the SingleCellExperiment object.

```{r, echo = FALSE}
pbmc4k.droplet.prefilt <- pbmc4k.droplet
```

```{r}
#Before filtering:
dim(pbmc4k.droplet)
```

```{r, "filtering droplet QC", message=FALSE, warning=FALSE}
pbmc4k.droplet <- subsetSCECols(pbmc4k.droplet, colData = 'dropletUtils_BarcodeRank_Inflection == 1')
pbmc4k.droplet <- subsetSCECols(pbmc4k.droplet, colData = '!is.na(pbmc4k.droplet$dropletUtils_emptyDrops_fdr)')
pbmc4k.droplet <- subsetSCECols(pbmc4k.droplet, colData = 'pbmc4k.droplet$dropletUtils_emptyDrops_fdr < 0.01')
```

```{r}
#After filtering:
dim(pbmc4k.droplet)
```

We can compare the average total UMI counts per cell before and after cell filtration:

```{r, "Count Violin after filtering", eval = TRUE, echo = TRUE, warning=FALSE, message=FALSE, fig.height = 4, fig.width = 6, fig.align="center"}
p1 <- plotSCEViolinColData(pbmc4k.droplet.prefilt, coldata = "sum", summary = "mean", title = "Pre-filter", ylab = "Total counts")
p2 <- plotSCEViolinColData(pbmc4k.droplet, coldata = "sum", summary = "mean", title = "Post-filter", ylab = "Total counts")
plot(cowplot::plot_grid(p1, p2, ncol = 2))
```

For additional cell-filtered quality control with singleCellTK, please refer to “Running quality control of cell-filtered single cell data with singleCellTK”.


### Running individual QC methods
Instead of running all quality control methods on the dataset at once, users may elect to execute QC methods individually. The parameters as well as the outputs to individual QC functions are described in detail as follows:

#### runEmptyDrops
The wrapper function `runEmptyDrops` can be used to separately run the EmptyDrops algorithm on its own.

In runEmptyDrops, the `lower` parameter is the lower bound of the total UMI count, in which all barcodes below the lower bound are assumed to be empty droplets. The `niters` parameter is the number of iterations the function will run for the calculation. `testAmbient` indicates whether results should be returned for barcodes that have a total UMI count below what is specified in `lower`.

```{r runemptydrops, eval = FALSE}
pbmc4k.droplet <- runEmptyDrops(
  inSCE = pbmc4k.droplet,
  useAssay = "counts",
  lower = 100,
  niters = 10000
)
```


#### runBarcodeRankDrops
The wrapper function `runBarcodeRankDrops` can be used to separately run the BarcodeRanks algorithm on its own.

The `lower` parameter is again the lower bound of the total UMI count, in which all barcodes below the lower bound are assumed to be empty droplets.
```{r run_barcoderanks, message=FALSE, eval=FALSE}
pbmc4k.droplet <- runBarcodeRankDrops(
  inSCE = pbmc4k.droplet,
  useAssay = "counts",
  fitBounds = NULL, df = 20
)
```
<br />
<br />




## Running quality control of cell-filtered single cell data with singleCellTK
### Load PBMC data from 10X

We will use a filtered form of the PBMC 3K and 6K dataset from the package [TENxPBMCData](http://bioconductor.org/packages/release/data/experiment/html/TENxPBMCData.html), which is available from the `importExampleData` function. We will combine these datasets together into a single SingleCellExperiment object.


```{r load_data2, eval=TRUE, message=FALSE}
pbmc3k <- importExampleData(dataset = "pbmc3k")
pbmc6k <- importExampleData(dataset = "pbmc6k")

pbmc.combined <- cbind(pbmc3k, pbmc6k)

sample.vector = colData(pbmc.combined)$sample
```

singleCellTK also supports the importing of single-cell data from the following platforms: [10X CellRanger](https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/what-is-cell-ranger), [STARSolo](https://github.com/alexdobin/STAR/blob/master/docs/STARsolo.md), [BUSTools](https://www.kallistobus.tools/), [SEQC](https://github.com/ambrosejcarr/seqc), [DropEST](https://github.com/hms-dbmi/dropEst), and [AnnData](https://github.com/theislab/anndata). To load your own input data, please refer [Import data into SCTK](./import_scRNAseq_data_as_SCE_Console.html) for deailed instruction. 

### Run Dimensionality Reduction

SCTK utilizes dimensionality reduction techniques such as TSNE and UMAP for visualizing single-cell data. The user can modify the dimensions by adjusting the parameters within the function. The `logNorm` parameter should be set to TRUE for normalization prior to running dimensionality reduction.

The `sample` parameter may be specified if multiple samples exist in the SingleCellExperiment object. Here, we will use the sample vector stored in the colData of the SingleCellExperiment object.


```{r dimensionalityreduction, eval=TRUE, warning=FALSE, message=FALSE}
# UMAP:
pbmc.combined <- getUMAP(inSCE = pbmc.combined, useAssay = "counts", logNorm = TRUE, sample = sample.vector)

# TSNE (not run):
# pbmc <- getTSNE(inSCE=pbmc, useAssay="counts", logNorm = TRUE, sample = colData(pbmc)$sample)
```

### runCellQC 

All of the droplet-based QC algorithms are able to be run under the wrapper function `runCellQC`. By default all possible QC algorithms will be run.

Users may set a `sample` parameter if you would like to compare between multiple samples. Here, we will use the sample vector stored in the SingleCellExperiment object.

If the user wishes, a list of gene sets can be applied to the function to determine the expression of a set of specific genes. A gene list imported into the SingleCellExperiment object using `importGeneSets` functions can be set as `collectionName`. Additionally, a pre-made list of genes can be used to determine the level of gene expression per cell. A list containing gene symbols may be set as `geneSetList`, or the user may instead use the `geneSetCollection` parameter to supply a `GeneSetCollection` object from the [GSEABase](https://bioconductor.org/packages/release/bioc/html/GSEABase.html) package.


```{r load_geneset, eval=TRUE,  message = FALSE}
pbmc.combined <- importGeneSetsFromGMT(inSCE = pbmc.combined, collectionName = "mito", file = system.file("extdata/mito_subset.gmt", package = "singleCellTK"))

set.seed(12345)
pbmc.combined <- runCellQC(pbmc.combined, sample = sample.vector, collectionName = "mito")
```

If users choose to only run a specific set of algorithms, they can specify which to run with the `algorithms` parameter.

When quality control functions are run in singleCellTK, the output of the function is stored in the `colData` slot of the `SingleCellExperiment` object.

```{r, coldata_cell, echo=TRUE, eval=FALSE}
head(colData(pbmc.combined), 5)
```

```{r, eval=TRUE, echo = FALSE}
df.matrix <- head(colData(pbmc.combined), 2)
df.matrix %>%
  knitr::kable(format = "html") %>% kableExtra::kable_styling() %>%
  kableExtra::scroll_box(width = "80%")
```

A summary of all outputs is shown below:
```{r table3, fig.wide = TRUE, echo=FALSE, message=FALSE, warnings=FALSE, results='asis', fig.align="center"}
tabl <- "
| QC output                              | Description                                                   | Methods           | Package/Tool  |
|----------------------------------------|:--------------------------------------------------------------|:------------------|:--------------|
| sum                                    | Total counts                                                  | runPerCellQC      | scater        |
| detected                               | Total features                                                | runPerCellQC      | scater        |
| percent_top                            | % Expression coming from top features                         | runPerCellQC      | scater        |
| subsets_                               | sum, detected, percent_top calculated on specified gene list  | runPerCellQC      | scater        |
| scrublet_score                         | Doublet score                                                 | runScrublet       | scrublet      |
| scrublet_call                          | Doublet classification based on threshold                     | runScrublet       | scrublet      |
| scDblFinder_doublet_score              | Doublet score                                                 | runScDblFinder   | scran         |
| doubletFinder_doublet_score            | Doublet score                                                 | runDoubletFinder  | DoubletFinder |
| doubletFinder_doublet_label_resolution | Doublet classification based on threshold                     | runDoubletFinder  | DoubletFinder |
| scds_cxds_score                        | Doublet score                                                 | runCxds           | SCDS          |
| scds_cxds_call                         | Doublet classification based on threshold                     | runCxds           | SCDS          |
| scds_bcds_score                        | Doublet score                                                 | runBcds           | SCDS          |
| scds_bcds_call                         | Doublet classification based on threshold                     | runBcds           | SCDS          |
| scds_hybrid_score                      | Doublet score                                                 | runCxdsBcdsHybrid | SCDS          |
| scds_hybrid_call                       | Doublet classification based on threshold                     | runCxdsBcdsHybrid | SCDS          |
| decontX_contamination                  | Ambient RNA contamination                                     | runDecontX        | celda         |
| decontX_clusters                       | Clusters determined in dataset based on underlying algorithm  | runDecontX        | celda         |
"
cat(tabl)
```

The names of the reducedDims of the `SingleCellExperiment` object are stored in the `reducedDims` slot.

```{r, reduceddimnames_cell}
reducedDims(pbmc.combined)
```
<br />


### Plotting QC metrics
Upon running runCellQC, the QC outputs will need to be plotted. For each QC method, singleCellTK contains specialized plotting functions that may be used for the generation of QC plots.

#### General QC metrics
##### runPerCellQC

SingleCellTK utilizes the [scater](https://bioconductor.org/packages/release/bioc/html/scater.html) package to compute cell-level QC metrics.

The wrapper function plotRunPerCellQCResults can be used to plot the general QC outputs.

```{r plot_percellqc, message=FALSE, warning=FALSE}
runpercellqc.results <- plotRunPerCellQCResults(inSCE = pbmc.combined, sample = sample.vector, combinePlot = "all", axisSize = 8, axisLabelSize = 9, titleSize = 20)
```

```{r fig.wide = TRUE, fig.height = 10, fig.width = 10}
runpercellqc.results
```

#### Doublet detection

Doublets hinder cell-type identification by appearing as a distinct transcriptomic state, and need to be removed for downstream analysis. singleCellTK contains various doublet detection tools that the user may choose from.

##### Scrublet
[Scrublet](https://github.com/swolock/scrublet/blob/master/examples/scrublet_basics.ipynb) aims to detect doublets by creating simulated doublets from combining transcriptomic profiles of existing cells in the dataset.

The wrapper function `plotScrubletResults` can be used to plot the results from the Scrublet algorithm.

The names of the reducedDims of the `SingleCellExperiment` object are stored in the `reducedDims` slot. Here, we will use the UMAP coordinates generated from `getUMAP` above.

```{r, reduceddimnames_cell2}
reducedDims(pbmc.combined)
```


```{r umap_scrubletplots, message=FALSE, warning=FALSE}
scrublet.results <- plotScrubletResults(
  inSCE = pbmc.combined,
  reducedDimName = "UMAP",
  sample = colData(pbmc.combined)$sample, 
  combinePlot = "all",
  titleSize = 13,
  axisLabelSize = 13,
  axisSize = 13,
  legendSize = 13,
  legendTitleSize = 13
)
```

```{r plot_umap_scrublet, fig.wide = TRUE, fig.height = 12, fig.width = 10}
scrublet.results 
```

##### ScDblFinder
[ScDblFinder](https://rdrr.io/bioc/scDblFinder/) is a doublet detection algorithm. ScDblFinder aims to detect doublets by creating a simulated doublet from existing cells and projecting it to the same PCA space as the cells.

The wrapper function plotScDblFinderResults can be used to plot the QC outputs from the ScDblFinder algorithm.

```{r umap_scDblFinder_plots, message=FALSE, warning=FALSE}
scDblFinder.results <- plotScDblFinderResults(
  inSCE = pbmc.combined, sample = colData(pbmc.combined)$sample, 
  reducedDimName = "UMAP", combinePlot = "all",
  titleSize = 13,
  axisLabelSize = 13,
  axisSize = 13,
  legendSize = 13,
  legendTitleSize = 13
)
```

```{r plot_umap_scDblFinder, fig.wide = TRUE, fig.height = 12, fig.width = 10}
scDblFinder.results
```

##### DoubletFinder
[DoubletFinder](https://github.com/chris-mcginnis-ucsf/DoubletFinder) is a doublet detection algorithm which depends on the single cell analysis package Seurat.

The wrapper function plotDoubletFinderResults can be used to plot the QC outputs from the DoubletFinder algorithm.

```{r, umap_doubletfinder_plots, message=FALSE, warning=FALSE}
doubletFinderResults <- plotDoubletFinderResults(
  inSCE = pbmc.combined,
  sample = colData(pbmc.combined)$sample, 
  reducedDimName = "UMAP",
  combinePlot = "all",
  titleSize = 13,
  axisLabelSize = 13,
  axisSize = 13,
  legendSize = 13,
  legendTitleSize = 13
)
```

```{r plot_umap_doubletfinder, fig.wide = TRUE, fig.height = 12, fig.width = 11}
doubletFinderResults 
```

##### SCDS, CXDS
CXDS, or co-expression based doublet scoring, is an algorithm in the [SCDS](https://www.bioconductor.org/packages/release/bioc/html/scds.html) package which employs a binomial model for the co-expression of pairs of genes to determine doublets.

The wrapper function `plotCxdsResults` can be used to plot the QC outputs from the CXDS algorithm.

```{r umap_cxds_plots, umap_cxds_plots, warning=FALSE, message=FALSE}
cxdsResults <- plotCxdsResults(
  inSCE = pbmc.combined,
  sample = colData(pbmc.combined)$sample, 
  reducedDimName = "UMAP", combinePlot = "all",
  titleSize = 13,
  axisLabelSize = 13,
  axisSize = 13,
  legendSize = 13,
  legendTitleSize = 13
)
```

```{r plot_umap_cxds, fig.wide = TRUE, fig.height = 12, fig.width = 11}
cxdsResults
```


##### SCDS, BCDS
BCDS, or binary classification based doublet scoring, is an algorithm in the [SCDS](https://www.bioconductor.org/packages/release/bioc/html/scds.html) package which uses a binary classification approach to determine doublets.

The wrapper function plotBCDSResults can be used to plot the QC outputs from the BCDS algorithm

```{r umap_bcds_plots, eval=TRUE, warning=FALSE, message=FALSE}
bcdsResults <- plotBcdsResults(
  inSCE = pbmc.combined,
  sample = colData(pbmc.combined)$sample, 
  reducedDimName = "UMAP", combinePlot = "all",
  titleSize = 13,
  axisLabelSize = 13,
  axisSize = 13,
  legendSize = 13,
  legendTitleSize = 13
)
```

```{r plot_umap_bcds, fig.wide = TRUE, fig.height = 12, fig.width = 11}
bcdsResults
```


##### SCDS, CXDS-BCDS hybrid
The CXDS-BCDS hybrid algorithm, uses both CXDS and BCDS algorithms from the [SCDS](https://www.bioconductor.org/packages/release/bioc/html/scds.html) package.

The wrapper function plotScdsHybridResults can be used to plot the QC outputs from the CXDS-BCDS hybrid algorithm.

```{r umap_hybrid_plots, eval=TRUE, warning=FALSE, message=FALSE}
bcdsCxdsHybridResults <- plotScdsHybridResults(
  inSCE = pbmc.combined, sample = colData(pbmc.combined)$sample, 
  reducedDimName = "UMAP", combinePlot = "all",
  titleSize = 13,
  axisLabelSize = 13,
  axisSize = 13,
  legendSize = 13,
  legendTitleSize = 13
)
```

```{r plot_umap_hybrid, fig.wide = TRUE, fig.height = 10, fig.width = 11}
bcdsCxdsHybridResults
```

#### Ambient RNA detection
##### DecontX
In droplet-based single cell technologies, ambient RNA that may have been released from apoptotic or damaged cells may get incorporated into another droplet, and can lead to contamination. [decontX](https://rdrr.io/bioc/celda/man/decontX.html), available from the [celda](https://bioconductor.org/packages/release/bioc/html/celda.html), is a Bayesian method for the identification of the contamination level at a cellular level.

The wrapper function plotDecontXResults can be used to plot the QC outputs from the DecontX algorithm.

```{r umap_decontx_plots, warning=FALSE, message=FALSE, fig.wide = TRUE, fig.height = 10, fig.width = 14}
decontxResults <- plotDecontXResults(
  inSCE = pbmc.combined, sample = colData(pbmc.combined)$sample, 
  reducedDimName = "UMAP", combinePlot = "all",
  titleSize = 13,
  axisLabelSize = 13,
  axisSize = 13,
  legendSize = 13,
  legendTitleSize = 13
)
```

```{r plots_umap_decontx, fig.wide = TRUE, fig.height = 8, fig.width = 9}
decontxResults
```

### Filtering the dataset

SingleCellExperiment objects can be subset by its colData using `subsetSCECols`. The usage of each argument is described in the **Quality Control of Droplet Counts Data** section above. 

```{r}
#Before filtering:
dim(pbmc.combined)
```

Remove barcodes with high mitochondrial gene expression:
```{r}
pbmc.combined <- subsetSCECols(pbmc.combined, colData = 'subsets_mito_percent < 20')
```

Remove detected doublets from Scrublet:
```{r}
pbmc.combined <- subsetSCECols(pbmc.combined, colData = 'scrublet_call == FALSE')
```

Remove cells with high levels of ambient RNA contamination:
```{r}
pbmc.combined <- subsetSCECols(pbmc.combined, colData = 'decontX_contamination < 0.5')
```

```{r}
#After filtering:
dim(pbmc.combined)
```

### Running individual QC methods
Instead of running all quality control methods on the dataset at once, users may elect to execute QC methods individually. The parameters as well as the outputs to individual QC functions are described in detail as follows:

#### runPerCellQC
The wrapper function `runPerCellQC` can be used to separately compute QC metrics on its own.

In this function, the `inSCE` parameter is the input SingleCellExperiment object, while the `useAssay` parameter is the assay object that in the SingleCellExperiment object the user wishes to use.

If the user wishes, a list of gene sets can be applied to the function to determine the expression of a set of specific genes. A gene list imported into the SingleCellExperiment object using `importGeneSets` functions can be set as `collectionName`. Additionally, a pre-made list of genes can be used to determine the level of gene expression per cell. A list containing gene symbols may be set as `geneSetList`, or the user may instead use the `geneSetCollection` parameter to supply a `GeneSetCollection` object from the [GSEABase](https://bioconductor.org/packages/release/bioc/html/GSEABase.html) package.

The QC outputs are `sum`, `detected`, and `percent_top_X`.

`sum` contains the total number of counts for each cell.

`detected` contains the total number of features for each cell.

`percent_top_X` contains the percentage of the total counts that is made up by the expression of the top X genes for each cell.

The `subsets_` columns contain information for the specific gene list that was used. For instance, if a gene list containing mitochondrial genes named `mito` was used, `subsets_mito_sum` would contains the total number of mitochondrial counts for each cell.

```{r, eval=FALSE, message=FALSE}
pbmc.combined <- runPerCellQC(
  inSCE = pbmc.combined,
  useAssay = "counts",
  collectionName = "mito")
```

#### runScrublet
The wrapper function `runScrublet` can be used to separately run the Scrublet algorithm on its own.

The `sample` parameter indicates what sample each cell originated from. It can be set to `NULL` if all cells in the dataset came from the same sample.

Scrublet also has a large set of parameters that the user can adjust; please refer to the Scrublet website for more details.

The Scrublet outputs are `scrublet_score`, which is a numeric variable of the likelihood that a cell is a doublet, and the `scrublet_label`, which is the assignment of whether the cell is a doublet.

```{r run_scrublet, eval = FALSE, message=FALSE}
pbmc.combined <- runScrublet(
  inSCE = pbmc.combined,
  sample = colData(pbmc.combined)$sample, 
  useAssay = "counts"
)
```

#### runScDblFinder
The wrapper function `runBarcodeRankDrops` can be used to separately run the ScDblFinder algorithm on its own. The `nNeighbors` parameter is the number of nearest neighbor used to calculate the density for doublet detection. `simDoublets` is used to determine the number of simulated doublets used for doublet detection.

The output of ScDblFinder is a `scDblFinder_doublet_score`. The doublet score of a droplet will be higher if the it is deemed likely to be a doublet.

```{r run_scDblFinder, eval = FALSE, message=FALSE}
pbmc.combined <- runScDblFinder(inSCE = pbmc.combined, sample = colData(pbmc.combined)$sample, useAssay = "counts")
```

#### runDoubletFinder
The wrapper function `runDoubletFinder` can be used to separately run the DoubletFinder algorithm on its own. `runDoubletFinder` relies on a parameter (in Seurat) called resolution to determine cells that may be doublets. Users will be able to manipulate the resolution parameter through `seuratRes`. If multiple numeric vectors are stored in `seuratRes`, there will be multiple label/scores. The `seuratNfeatures ` parameter determines the number of features that is used in the `FindVariableFeatures ` function in Seurat. `seuratPcs ` parameter determines the number of dimensions used in the `FindNeighbors ` function in Seurat. The `formationRate ` parameter is the estimated doublet detection rate in the dataset. aims to detect doublets by creating simulated doublets from combining transcriptomic profiles of existing cells in the dataset.

The DoubletFinder outputs are `doubletFinder_doublet_score `, which is a numeric variable of the likelihood that a cell is a doublet, and the `doubletFinder_doublet_label `, which is the assignment of whether the cell is a doublet.

```{r run_doubletfinder, eval = FALSE, message=FALSE}
pbmc.combined <- runDoubletFinder(
  inSCE = pbmc.combined, useAssay = "counts",
  sample = colData(pbmc.combined)$sample, 
  seuratRes = c(1.0), seuratPcs = 1:15,
  seuratNfeatures = 2000,
  formationRate = 0.075, seed = 12345
)
```

#### runCXDS
The wrapper function `runCxds ` can be used to separately run the CXDS algorithm on its own. In `runCxds `, the `ntop ` parameter is the number of top variance genes to consider. The `binThresh ` parameter is the minimum counts a gene needs to have to be included in the analysis. `verb ` determines whether progress messages will be displayed or not. `retRes ` will determine whether the gene pair results should be returned or not. The user may set the estimated number of doublets with estNdbl.

The output of `runCxds ` is the doublet score, `scds_cxds_score`.

```{r run_cxds, eval = FALSE, message=FALSE}
pbmc.combined <- runCxds(
  inSCE = pbmc.combined, sample = colData(pbmc.combined)$sample, 
  ntop = 500, binThresh = 0,
  verb = FALSE, retRes = FALSE, estNdbl = FALSE
)
```

#### runBCDS
The wrapper function `runBcds ` can be used to separately run the BCDS algorithm on its own. In `runBcds `, the `ntop` parameter is the number of top variance genes to consider. The `srat` parameter is the ratio between original number of cells and simulated doublets. The `nmax` parameter is the maximum number of cycles that the algorithm should run through. If set to `tune`, this will be automatic. The `varImp` parameter determines if the variable importance should be returned or not.

The output of `runBcds ` is scds_bcds_score, which is the likelihood that a cell is a doublet.

```{r run_bcds, eval = FALSE, message=FALSE}
pbmc.combined <- runBcds(
  inSCE = pbmc.combined, seed = 12345, sample = colData(pbmc.combined)$sample, 
  ntop = 500, srat = 1, nmax = "tune", varImp = FALSE
)
```

#### runCxdsBcdsHybrid
The wrapper function `runCxdsBcdsHybrid` can be used to separately run the CXDS-BCDS hybrid algorithm on its own.

All parameters from the `runBCDS` and `runBCDS` functions may be applied to this function in the cxdsArgs and bcdsArgs parameters, respectively. The output of `runCxdsBcdsHybrid` is the doublet score, `scds_hybrid_score`.

```{r run_hybrid, eval = FALSE, message=FALSE}
pbmc.combined <- runCxdsBcdsHybrid(
  inSCE = pbmc.combined, sample = colData(pbmc.combined)$sample, 
  seed = 12345, nTop = 500
)
```

#### runDecontX
The outputs of `runDecontX` are `decontX_contamination` and `decontX_clusters`.

The wrapper function `runDecontX` can be used to separately run the DecontX algorithm on its own. `decontX_contamination` is a numeric vector which characterizes the level of contamination in each cell. Clustering is performed as part of the `runDecontX` algorithm. `decontX_clusters` is the resulting cluster assignment, which can also be labeled on the plot.

```{r run_decontx, eval = FALSE, message=FALSE}
pbmc.combined <- runDecontX(pbmc.combined, sample = colData(pbmc.combined)$sample, useAssay = "counts")
```

# Session Information

```{r}
sessionInfo()
```