---
title: "GSVA"
author: "Irzam Sarfraz"
---
```{r develSetting, include=FALSE}
doEval = TRUE
```
## Introduction
The ***GSVA*** (Gene Set Variation Analysis) ([HÃ¤nzelmann, S., Castelo, R. & Guinney, J. GSVA: gene set variation analysis for microarray and RNA-Seq data. BMC Bioinformatics 14, 7 (2013).](https://doi.org/10.1186/1471-2105-14-7)) is a popular method for ***GSE*** (Gene Set Enrichment) and allows the identification of changes in pathway activity
in RNA-Seq data. The ***Cell Annotation & Pathway Analysis*** tab in the ***singleCellTK***
offers the computation and visualization of pre-selected gene sets through ***GSVA*** sub-tab. The gene sets 
can be selected from several databases integrated within the toolkit or can be uploaded/created by
the users.

## General Worflow
The figure below describes the workflow of using ***GSVA*** through ***singleCellTK*** user interface:
![](ui_screenshots/pathway_analysis/capa_workflow.png)

## Workflow Guide
Before running the ***GSVA*** algorithm, the gene sets must be selected or uploaded through the ***Import Gene Sets*** sub-tab from the ***Data*** tab in the toolkit menu. An overall workflow guide on how to import/upload ***gene sets*** and consequently using them with ***GSVA*** sub-tab is described below. ***This workflow guide assumes that the *SingleCell Data* has already been uploaded through the *Import Single Cell Data*** sub-tab.**<br>

![](ui_screenshots/pathway_analysis/capa_1.png)

1. Select the ***Data*** tab from the ***singleCellTK*** top menu.<br>

2. Select the ***Import Gene Sets*** from the drop-down menu.<br>

![](ui_screenshots/pathway_analysis/capa_2.png)

3. Users can choose  the ***Upload a GMT file*** option to upload gene sets directly stored in a GMT file or choose ***Select from a database*** option to select gene sets from multiple databases integrated within the toolkit or lastly choose the ***Paste in your gene set*** option to paste gene identifiers directly into the user-interface to create a gene set.<br>

4. (a) If ***Upload a GMT file*** option is selected, users can upload a .GMT file containing the gene sets. <br>

![](ui_screenshots/pathway_analysis/capa_3.png)

4. (b) If ***Select from a database*** option is selected, the available gene sets are populated. Users can select the gene sets and click on the ***Upload*** button to get them to use in the downstream analysis. <br>

![](ui_screenshots/pathway_analysis/capa_4.png)

5. Once ***gene sets*** are uploaded, the selected gene sets are listed in this table. <br>

![](ui_screenshots/pathway_analysis/capa_5.png)

6. To run the ***GSVA*** algorithm, go to ***Cell Annotation & Pathway Analysis*** tab. <br>

![](ui_screenshots/pathway_analysis/capa_6.png)

7. Select ***GSVA*** option from the drop-down menu. <br>

![](ui_screenshots/pathway_analysis/capa_7.png)

8. Select ***assay*** to use with ***GSVA***. <br>
9. Select ***GSVA*** as the method.
10. Select ***gene sets*** to use with ***GSVA***. Only uploaded ***gene sets*** from the ***Import Gene Sets*** page are available here. <br>
11. Select a ***phenotype variable*** to use with the ***plots***. <br>
12. Select if a ***Heatmap*** should be plotted or a ***Violin** plot. <br>
13. Press ***Run*** to start computation. <br>
14. Once computation is complete, users use the ***Save Pathways*** button to store the pathways in the internal object for use in downstream analysis or use the ***Download Pathway Results*** option to download the results to the local drive. <br>

![](ui_screenshots/pathway_analysis/capa_8.png)

15. Results can be visualized by selecting the ***Plot*** tab **(16)** or the ***Results Table*** tab **(17)**. <br>

