---
title: "Import Datasets - UI"
---

## Introduction

This page is the landing page of the Single Cell ToolKit (SCTK) application. Here SCTK allows users to import multiple datasets at once, and the datasets can be in various forms, including the output from preprocessing tools (listed later), the text files of count matrices and annotations, the RDS file saving a SingleCellExperiment object or the example dataset ready online.  
The UI contains mainly three sections for importing and another section for data summary. Through the first and the second sections, users can choose the type of a single data source and specify the location of the data from this single source, respectively. A table in the third section summarizes files and samples that are specified from all sources. Finally, all data can be imported by clicking on the import button right below this table. 

![workflow](ui_screenshots/import/import_ui_workflow.png)\

## Importing Single-Cell Datasets

<img src="ui_screenshots/import/import_ui_source.png" style="border: 1px solid rgba(0,0,0,0.2)"></img>

For a single source of data, users need to choose one option from the radio buttons listed in the screenshot. The next step will depend on the option chosen.  

### Import from a preprocessing tool

![source1](ui_screenshots/import/import_ui_source1.png)\

SCTK supports importing data from the following pre-processing tools in our UI workflow:  
- [CellRanger](https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/what-is-cell-ranger) (versions 2 and 3)  
- [STARSolo](https://github.com/alexdobin/STAR/blob/master/docs/STARsolo.md)  
- [BUStools](https://github.com/BUStools/bustools)  
- [Seqc](https://github.com/ambrosejcarr/seqc)  
- [Optimus](https://data.humancellatlas.org/pipelines/optimus-workflow)  

Choose the specific preprocessing tool that matches your dataset, and then press "**Add a Sample**". A corresponding dialog modal will pop out with description that instructs you to select the directory that stores your data.  

#### Cell Ranger

The import procedure for Cell Ranger versions 2 and 3 are the same. To import a Cell Ranger sample, the user must select how they would like to import the sample: select a base directory that holds multiple sample directories, select a single sample directory, or select a directory holding the data files for a sample. Depending on this selection, the user must then provide the apprpriate directory paths and sample names and then click "OK."  

![cell ranger](ui_screenshots/import/import_ui_cellRanger.png)\

#### STARsolo

To import a STARsolo sample, the user must select the parent directory of the **`"Gene/"`** directory for the sample they want to import, name the sample, and click "OK."  

![STARsolo](ui_screenshots/import/import_ui_STARsolo.png)\

#### BUStools

To import a BUStools sample, the user must point to the **`"genecount/"`** directory for the sample they want to import, name the sample, and click "OK."  

![BUStools](ui_screenshots/import/import_ui_BUStools.png)\

#### SEQC

To import a SEQC sample, the user must point to the directory that contains the **barcodes, gene, and matrix files** for the sample they want to import and name the sample. This tool additionally needs the user to specify a **"Sample ID"** which is what the name of the sample is as it appears in the names of its data files (e.g. for a file called `"pbmc_1k_sparse_counts_barcodes.csv"` the sample ID is `"pbmc_1k"`). Then the user must click "OK."  

![SEQC](ui_screenshots/import/import_ui_SEQC.png)\

#### Optimus

To import an Optimus sample, the user must select the directory that contains the following four directories for the sample they want to import: **`"call-MergeCountFiles"`**, **`"call-MergeCellMetrics"`**, **`"call-MergeGeneMetrics"`**, **`"call-RunEmptyDrops"`**. Then the user must prove a sample name and click "OK."  

![Optimus](ui_screenshots/import/import_ui_Optimus.png)\

### Import from flat files (CSV, TSV)

To import tab-separated data, the user must provide **a counts file** (.csv or .tsv) and specify what type of input assay it is. The user can optionally also provide **cell annotation** and **feature** files (also .csv or .tsv). The user must then click **"Add to Sample List"** to add it to the table of samples.  
After choosing a single file for uploading, user must wait until the progress bar runs to the right end and says "Upload complete". Sometimes the progress bar runs to the end but it takes some time for the system to finalize reading a large file.  

![Flat files](ui_screenshots/import/import_ui_source2.png)\

### Upload `SingleCellExperiment` or `Seurat` object stored in an RDS file

RDS is a file format that stores a single object from R environment. To import a single-cell dataset stored in a **`SingleCellExperiment`** (SCE) object from an RDS file, the user can simply select which **RDS** file to import and click **"Add to Sample List"**.  
After choosing a single file for uploading, user must wait until the progress bar runs to the right end and says "Upload complete". Sometimes the progress bar runs to the end but it takes some time for the system to finalize reading a large file.  

<img src="ui_screenshots/import/import_ui_source3.png" style="border: 1px solid rgba(0,0,0,0.2)"></img>

### Import example datasets

Users can also opt to import and work with published example datasets. To do so, the user must select which dataset they would like to use from the dropdown menu and click **"Add to Sample List"**. For more information about the chosen dataset, users can click on the hyperlink of the title and open the reference website in a browser.  

<img src="ui_screenshots/import/import_ui_source4.png" style="border: 1px solid rgba(0,0,0,0.2)"></img>

### Complete Upload

After successfully choosing each source of dataset, it will be inserted to a new line of the table at the buttom of this UI page, in "**3. Import**" section. This is a wait-list for all the samples of interests. To finally import all datasets of interests, users must click on **"Import"** button.  
Before importing, users can click on the "`[X]`" button on the right to remove the selection of the corresponding single dataset, or click on **"Clear Samples"** to reset all selections. This only clears this wait-list for importing, but will not work on any dataset already imported.  
Note that it is normal that in the "Location" field, some paths might happen to point to a temporary cache.  

<img src="ui_screenshots/import/import_ui_sampleList.png" style="border: 1px solid rgba(0,0,0,0.2)"></img>
