---
title: "UI Tutorial"
author: "Yichen Wang"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, eval = FALSE)
```

## Introduction

Single Cell Toolkit (singleCellTK, SCTK) is a package that works on single-cell RNA-seq (scRNAseq) dataset. SCTK allows users to import multiple datasets, perform quality control and a series of preprocessing, get clustering on cells and markers of clusters, and run various downstream analysis. Meanwhile, SCTK also wraps curated workflows for [celda](https://www.camplab.net/celda/) and [Seurat](https://satijalab.org/seurat/). 

This tutorial will take the real-world scRNAseq dataset as an example, which consists of 2,700 Peripheral Blood Mononuclear Cells (PBMCs) collected from a healthy donor. This dataset (PBMC3K) is available from 10X Genomics and can be found on the 10X website.

In the UI, users' data are stored in the same way as in the [console analysis](console_analysis_tutorial.html). Basically, there will be different versions of expression matrices, low-dimensional representations of the original data, and the metadata for cell and feature annotation. 

## Start the UI

### Online Web Server

Users in the Boston University network can access the online deloyment of SCTK server via: [http://sctk.bu.edu/](http://sctk.bu.edu/)

### Local Shiny Server

After successfully [installing](installation.html) the R package `singleCellTK`, users can easily start a local UI server by running the following commands in the R console:

```{R}
library(singleCellTK)
singleCellTK()
```

If users are running the commands in [RStudio](https://www.rstudio.com/), a web browser should automatically pop up, with the UI displaying. Users should also be able to see a local server URL (e.g. `http://127.0.0.1:XXXX`) in the console. If the browser does not pop up, users can copy the URL into a preferred browser and press Enter button.

Another developmental way of starting the UI locally is to open an editor tab for the server file `inst/shiny/server.R` in the RStudio. And on the top of the editing panel, click on "Run App" button. This approach allows developers to start an APP of the latest edition without installation. 

### Docker server

TODO

## Import Data

After starting the UI, the landing page will be for importing the scRNAseq data. While SCTK has the support for importing from many different sources, to import the PBMC3K dataset previously mentioned, users should follow these steps:

![import](ui_screenshots/ui_tutorial/import.png)\

1. In step 1 **"Choose data source"**, check the radio button **"Import example datasets"**.
2. In step 2 **"Choose Example Dataset"**, select the option **"PBMC 3K (10X)"**, and click on button **"Add To Sample List"**.
3. In step 3 **"Import"**, users will see a new row added for the dataset, in the table **"Samples to Import"**. Then users should click on button **"Import"** right below the table, in order to successfully load the data in to the toolkit.

For importing dataset from other type of sources, such as preprocessing tools like cellranger or flat text files, or importing multiple batches of dataset, please refer to the [detailed documentation of Importing](import_data.html).

## Quality Control (QC)

Quality control and filtering of cells is often needed before down-stream analyses such as dimensionality reduction and clustering. Typical filtering procedures include exclusion of poor quality cells with low numbers of counts/UMIs, estimation and removal of ambient RNA, and identification of potential doublet/multiplets. Many tools and packages are available to perform these operations and users are free to apply their tool(s) of choice in SCTK UI. 

Below is a quick example of how to perform standard QC before heading to the downstream analyses. If your data is already QC'ed or you decide to skip this step, you can move to [Feature Selection section](#variable-feature-selection)

### Running QC methods

![import](ui_screenshots/ui_tutorial/qc_run.png)\

1. Users need to select **"QC & Filtering"** at the top navigation panel to enter this section. 
2. In this tutorial, we perform General **"QC Metrics"**, decontamination algorithm [**"decontX"**](https://bioconductor.org/packages/release/bioc/html/scDblFinder.html) and doublet detection algorithm [**"scDblFinder"**](https://bioconductor.org/packages/release/bioc/html/scDblFinder.html), by checking the corresponding method in the left panel of this page. 
3. After making the selection of methods, users can then click on **"Run"** button at the bottom of the left panel. 

![import](ui_screenshots/ui_tutorial/qc_plot.png)\

Right after running, the visualization of QC results will automatically show up in the right main panel, with tabs for different methods. 

Please find [the detailed documentation of the QC UI](ui_qc.html) for more explanation of each parameter. 

### Filtering

After examining the distributions of various QC metrics, poor quality cells will need to be removed. Typically, thresholds for QC metrics should exclude cells that are outliers of the distribution (i.e. long tails in the violin or density plots). Here is how we limit the data to cells with 1. at least 600 counts, 2. at least 300 genes detected, 3. doublet score no more than 0.8, and 4. at most 5% of detected counts are mitochondrial:

![import](ui_screenshots/ui_tutorial/filter_enter.png)\

1. Enter the filtering section by clicking on sub-tab **"Filtering"**.

![import](ui_screenshots/ui_tutorial/filter_add.png)\

2. Click on button **"Add a Filter"** in the first grey panel **"Select Cell Filtering Criteria"**.
3. In the pop up, select one of the cell metadata variables that stands for the QC metrics. It should be `total`, `detected`, `scDblFinder_doublet_score` and `mito_percent` for each criteria. 
4. Set the cut-off for each criteria. If you want to keep cells with value `total` greater than 600, you would check **"Greater than"**, and enter the number `600`. Similarly, if you want to keep cells with value `scDblFinder_doublet_score` less than 0.8, you would check **"Less than"** and enter the number `0.8`. 
5. For every criteria, after entering the cut-off, click on button **OK** to add the filter "to the cart". 
6. Repeat step 2-5 to add the four filters. After each criteria is added, you can see a new row inserted to the table, as shown in the dashed box in the screenshot below. 

![import](ui_screenshots/ui_tutorial/filter_run.png)\

7. Finally, click on **"Filter"** button at the bottom of the current view to apply all filters at one time. 

![import](ui_screenshots/ui_tutorial/filter_summary.png)\

After successfully filtering the data, you can see a comparison table shows up at the bottom. You can apply new filters by repeating the steps above, but it will only be applied on the filtered data instead of the original data. For more detail about filtering, please refer to the [Filtering documentation](filtering.html)

## Normalization

After removing cells of low quality, we next need to normalize the feature expression matrix. In this tutorial, we apply a global-scaling normalization method from [scater](https://bioconductor.org/packages/release/bioc/html/scater.html), `"logNormCounts"`. It normalizes the feature expression for each cell by the total expression, multiplies this by a scale factor, and log-transforms the result. Afterwards, we recommend running a z-score scaling on the log-normalized matrix prior to dimensionality reduction.  

![norm](ui_screenshots/ui_tutorial/norm.png)\

1. Users should enter the section for normalization from the **"Normalization & Batch Correction"** tab at the top navigation panel.
2. In the **"Normalization"** sub-tab, choose the method `"Scater - LogNormCounts"` from the left panel.
3. Users can directly press the **"Run"** button at the bottom right without taking any more action. 

![norm](ui_screenshots/ui_tutorial/scale.png)\

1. For scaling the matrix, TODO

For detail about all normalization methods, please refer to [the Normalization Documentation](cnsl_normalization.html)

>SCTK also supports a number of batch correction methods. Here we don't present the workflow because PBMC3k dataset only has one batch. For examples, please refer to the [document of batch correction](batch_correction.html).

## Variable Feature Selection

Selecting highly variable genes (HVG) for downstream analyses is recommended, since a subset of variable features can speed up the computation and reduce the noise being introduced. SCTK wraps the methods used in [Seurat](https://satijalab.org/seurat/) and [Scran](https://bioconductor.org/packages/release/bioc/html/scran.html). We recommend keeping at least 2,000-5,000 HVGs. In the example code, we use the "VST" method from Seurat, and select the top 2,000 genes. 





