<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Generation of comprehensive quality control metrics with SCTK • singleCellTK</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/yeti/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../../bootstrap-toc.css">
<script src="../../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../../pkgdown.css" rel="stylesheet">
<script src="../../pkgdown.js"></script><link href="../../extra.css" rel="stylesheet">
<script src="../../extra.js"></script><meta property="og:title" content="Generation of comprehensive quality control metrics with SCTK">
<meta property="og:description" content="singleCellTK">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../../index.html">singleCellTK</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">2.2.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../../articles/articles/introduction_to_SCTK.html">General Workflow</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Shell Command Line Analysis
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../../articles/articles/cmd_import_scRNAseq_data_as_SCE.html">Import data into SCTK</a>
    </li>
    <li>
      <a href="../../articles/articles/cmd_qc.html">Quality Control</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    R Console Analysis
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../../articles/articles/cnsl_import_scRNAseq_data_as_SCE.html">Import data into SCTK</a>
    </li>
    <li>
      <a href="../../articles/articles/cnsl_dropletqc.html">Quality Control of droplet data</a>
    </li>
    <li>
      <a href="../../articles/articles/cnsl_cellqc.html">Quality Control of cell data</a>
    </li>
    <li>
      <a href="../../articles/articles/cnsl_normalization.html">Normalization</a>
    </li>
    <li>
      <a href="../../articles/articles/cnsl_batch_correction.html">Batch Correction</a>
    </li>
    <li>
      <a href="../../articles/articles/cnsl_feature_selection.html">Feature Selection</a>
    </li>
    <li>
      <a href="../../articles/articles/cnsl_dimensionality_reduction.html">Dimensionality Reduction</a>
    </li>
    <li>
      <a href="../../articles/articles/cnsl_clustering.html">Clustering</a>
    </li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Differential Expression</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../../articles/articles/cnsl_differential_expression.html">Differential Expression</a>
        </li>
        <li>
          <a href="../../articles/articles/cnsl_find_marker.html">Find Marker via DE</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Enrichment Analysis</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../../articles/articles/cnsl_gsva.html">GSVA</a>
        </li>
        <li>
          <a href="../../articles/articles/cnsl_enrichR.html">EnrichR</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Visualization</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../../articles/articles/cnsl_visualization.html">General Visualization</a>
        </li>
        <li>
          <a href="../../articles/articles/cnsl_heatmap.html">Heatmap Visualization</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Curated Workflows</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../../articles/articles/cnsl_seurat_curated_workflow.html">Seurat</a>
        </li>
        <li>
          <a href="../../articles/articles/cnsl_celda_curated_workflow.html">Celda</a>
        </li>
      </ul>
</li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Shiny UI Analysis
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Data</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../../articles/articles/ui_import_data.html">Import Data</a>
        </li>
        <li>
          <a href="../../articles/articles/ui_import_genesets.html">Import Gene Sets</a>
        </li>
        <li>
          <a href="../../articles/articles/ui_import_annotation.html">Annotation</a>
        </li>
        <li>
          <a href="../../articles/articles/ui_export_data.html">Export Data</a>
        </li>
        <li>
          <a href="../../articles/articles/ui_delete_data.html">Delete Data</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">QC &amp; Filtering</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../../articles/articles/ui_qc.html">Quality Control</a>
        </li>
        <li>
          <a href="../../articles/articles/ui_filtering.html">Filtering</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Normalization &amp; Batch Correction</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../../articles/articles/ui_normalization.html">Normalization</a>
        </li>
        <li>
          <a href="../../articles/articles/ui_batch_correction.html">Batch Correction</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Feature Selection &amp; Dimensionality Reduction</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../../articles/articles/ui_feature_selection.html">Feature Selection</a>
        </li>
        <li>
          <a href="../../articles/articles/ui_dimensionality_reduction.html">Dimensionality Reduction</a>
        </li>
      </ul>
</li>
    <li>
      <a href="../../articles/articles/ui_clustering.html">Clustering</a>
    </li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Differential Expression &amp; Cell Type Labeling</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../../articles/articles/ui_differential_expression.html">Differential Expression</a>
        </li>
        <li>
          <a href="../../articles/articles/ui_find_marker.html">Find Marker</a>
        </li>
        <li>
          <a href="../../articles/articles/ui_cell_type_labeling.html">Cell Type Labeling</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Pathway &amp; Enrichment Analysis</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../../articles/articles/ui_gsva.html">GSVA</a>
        </li>
        <li>
          <a href="../../articles/articles/ui_enrichR.html">EnrichR</a>
        </li>
      </ul>
</li>
    <li>
      <a href="../../articles/articles/ui_sample_size_calc.html">Sample Size Calculator</a>
    </li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Curated Workflows</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../../articles/articles/ui_seurat_curated_workflow.html">Seurat</a>
        </li>
        <li>
          <a href="../../articles/articles/ui_celda_curated_workflow.html">Celda</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Visualization</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../../articles/articles/ui_cellViewer.html">cellViewer</a>
        </li>
        <li>
          <a href="../../articles/articles/ui_heatmap.html">Generic Heatmap</a>
        </li>
      </ul>
</li>
  </ul>
</li>
<li>
  <a href="../../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/compbiomed/singleCellTK">
    <span class="fas fa-github"></span>
     
  </a>
</li>
<li>
  <a href="https://twitter.com/camplab1">
    <span class="fas fa-twitter"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="cmd_qc_files/header-attrs-2.8/header-attrs.js"></script><script src="cmd_qc_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Generation of comprehensive quality control metrics with SCTK</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/compbiomed/singleCellTK/blob/master/vignettes/articles/cmd_qc.Rmd"><code>vignettes/articles/cmd_qc.Rmd</code></a></small>
      <div class="hidden name"><code>cmd_qc.Rmd</code></div>

    </div>

    
    
<div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<p>This pipeline will import data from single-cell preprocessing algorithms (e.g. CellRanger, HCA Optimus, Alevin), generate various quality control metrics (e.g. general metrics, doublet scores, contamination estimates) using multiple tools, and output results in standard data containers (e.g. SingleCellExperiment, Seurat object, AnnData). For data generated with microfluidic devices, the first major step after UMI counting is to detect cell barcodes that represent droplets containing a true cell and exclude empty droplets that only contain ambient RNA. We use the terms “Droplet” matrix to denote a count matrix that still contains empty droplets, “Cell” matrix to denote a count matrix of cells where empty droplets have been excluded but no other filtering has been performed, and “FilteredCell” matrix to indicate a count matrix where poor quality cells have also been excluded. The Droplet and Cell matrices have also been called “raw” and “filtered” matrices, respectively, by tools such as CellRanger. However, using the term “filtered” can be ambiguous as other forms of cell filtering can be applied beyond empty droplets (e.g. excluding poor-quality cells based on low number of UMIs). Both the original droplet matrix and the filtered cell matrix can be QC’ed in this pipeline. However, QC of the droplex matrix is specific for single cell data generated from microfluidic devices (e.g. 10X). To run the pipeline, users can manually <a href="http://sctk.camplab.net/v2.2.0/index.html">install</a> the singleCellTK package along with Python dependencies that may be potentially used (mainly Scrublet and AnnData). Alternatively, users can run the Docker version of the pipeline which is described in detail further down in this page.</p>
</div>
<div id="running-the-pipeline" class="section level1">
<h1 class="hasAnchor">
<a href="#running-the-pipeline" class="anchor"></a>Running the pipeline</h1>
<div id="running-sctk-qc-after-manual-installation-of-singlecelltk" class="section level2">
<h2 class="hasAnchor">
<a href="#running-sctk-qc-after-manual-installation-of-singlecelltk" class="anchor"></a>Running SCTK-QC after manual installation of singleCellTK</h2>
<p>To run the pipeline script, users will need to download the ‘SCTK_runQC.R’ <a href="https://github.com/compbiomed/singleCellTK/blob/master/exec/SCTK_runQC.R">here</a>.</p>
<p>An simple example to run this pipeline on the Cellranger V3 dataset is shown below:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1"></a>Rscript SCTK_runQC.R \</span>
<span id="cb1-2"><a href="#cb1-2"></a><span class="op">-</span>b <span class="op">/</span>base<span class="op">/</span>path \</span>
<span id="cb1-3"><a href="#cb1-3"></a><span class="op">-</span>P CellRangerV3 \</span>
<span id="cb1-4"><a href="#cb1-4"></a><span class="op">-</span>s SampleName \</span>
<span id="cb1-5"><a href="#cb1-5"></a><span class="op">-</span>o Output_Directory \</span>
<span id="cb1-6"><a href="#cb1-6"></a><span class="op">-</span>S <span class="ot">TRUE</span> \</span>
<span id="cb1-7"><a href="#cb1-7"></a><span class="op">-</span>F R,Python,FlatFile,HTAN</span></code></pre></div>
<p>This pipeline enables different ways to import CellrangerV2/CellrangerV3 data for flexibility. Also, the pipeline is compatible with datasets generated by other algorithms. Please refer to the section <strong>Inputs</strong> for more details.</p>
<p>Besides, the pipeline contains various parameters to control the process of quality control. Please refer to the section <strong>Parameters</strong> for more details.</p>
</div>
<div id="running-sctk-qc-with-docker" class="section level2">
<h2 class="hasAnchor">
<a href="#running-sctk-qc-with-docker" class="anchor"></a>Running SCTK-QC with Docker</h2>
<div id="downloading-docker" class="section level3">
<h3 class="hasAnchor">
<a href="#downloading-docker" class="anchor"></a>Downloading docker</h3>
<p>If you have not used docker before, you can follow the instruction to install and set up docker in <a href="https://docs.docker.com/docker-for-windows/">Windows</a>, <a href="https://docs.docker.com/docker-for-mac/">Mac</a> or <a href="https://runnable.com/docker/install-docker-on-linux">Linux</a>.</p>
</div>
<div id="running-sctk-qc-pipeline-using-docker-image" class="section level3">
<h3 class="hasAnchor">
<a href="#running-sctk-qc-pipeline-using-docker-image" class="anchor"></a>Running SCTK-QC pipeline using docker image</h3>
<p>The Docker image can be obtained by running:</p>
<pre><code>docker pull campbio/sctk_qc:2.0.0</code></pre>
<p>The usage of each argument is the same as running command line analysis. Here is an example code to perform quanlity control of CellRangerV3 data singleCellTK docker:</p>
<pre><code>docker run --rm -v /path/to/data:/SCTK_docker \
-it campbio/sctk_qc:2.0.0 \
-b /SCTK_docker/cellranger \
-P CellRangerV3 \
-s pbmc_100x100 \
-o /SCTK_docker/result/tenx_v3_pbmc \
-g /SCTK_docker/mitochondrial_human_symbol.gmt \
-S TRUE \
-F R,Python,Flatfile,HTAN</code></pre>
<p>Docker image will not access files in your host file system by default. To get access to the files on your machine, you can properly set up a mount volume. Noted that the transcriptome data and GMT file needed to be accessible to the container via mounted volume. In the below example, mount volumn is enabled for accessing input and output directory using argument -v. The transcriptome and GMT files inside <em>/path/to/data</em> is now available in <em>/SCTK_docker</em> folder inside the docker. To learn more about mounted volumes, please check out <a href="https://docs.docker.com/storage/volumes/">this post</a>.</p>
<p>Please refer to the section <strong>Parameters</strong> for more details about parameters.</p>
</div>
<div id="running-sctk-qc-pipeline-docker-image-with-singularity" class="section level3">
<h3 class="hasAnchor">
<a href="#running-sctk-qc-pipeline-docker-image-with-singularity" class="anchor"></a>Running SCTK-QC pipeline docker image with singularity</h3>
<p>The Singularity image can easily be built using Docker Hub as a source:</p>
<pre><code>singularity pull docker://campbio/sctk_qc:2.0.0</code></pre>
<p>The usage of singleCellTK Singularity image is very similar to that of Docker. In Singularity 3.0+, the mount volume is <a href="https://singularity.lbl.gov/docs-mount">automatically overlaid</a>.</p>
<p>It’s recommended to re-set the home directory when you run singularity. Singularity will mount $HOME path on your machine by default, which might contain your personal R/Python library folder. If we don’t re-set the home to mount, singularity will try to use R/Python libraries which are not built within the singularity image and cause some conflicts. You can point to some “sanitized home”, which is different from $HOME path on your machine, using argument <a href="https://singularity.lbl.gov/faq#solution-1-specify-the-home-to-mount">-H/–home</a>. Besides, you can use argument –bind/-B to specify your own mount volume, which is the path that contains the dataset and will be used to store the output of QC pipeline. The example is shown as below:</p>
<pre><code>singularity run --home=/PathToSanitizedHome \
--bind /PathToData:/data sctk_qc_2.0.0.sif \
-P CellRangerV3 \
-s gencodev34_pbmc_1k_v3 \
-b /data/gencodev34_pbmc_1k_v3
-o /data/result/gencodev34_pbmc_1k_v3 \
-S TRUE \
-F R,Python,FlatFile,HTAN \
-n 15 \
-T MulticoreParam</code></pre>
</div>
<div id="important-note-about-docker-image" class="section level3">
<h3 class="hasAnchor">
<a href="#important-note-about-docker-image" class="anchor"></a>Important note about docker image</h3>
<p>One important note about this docker image: please run the docker image on a machine / node which has a cpu with the following architecture: broadwell, haswell, skylake, cascadelake or the latest architecture. This can avoid having the “illegal operation” issue of <strong>scrublet</strong> package. Because this python package are compiled by SIMD instructions that are compatible with these cpu architectures. Please specify cpu architecture as one of the following: broadwell, haswell, skylake, cascadelake or latest architecture. One of the example is shown below:</p>
<pre><code>#!/bin/bash
#$ -cwd
#$ -j y
#$ -pe omp 16
#$ -l cpu_arch=broadwell|haswell|skylake|cascadelake

singularity run --home=/PathToSanitizedHome \ ### this also works for 'docker run'
--bind /PathToData:/data sctk_qc_2.0.0.sif \
-P CellRangerV3 \
-s gencodev34_pbmc_1k_v3 \
-b /data/gencodev34_pbmc_1k_v3
-o /data/result/gencodev34_pbmc_1k_v3 \
-S TRUE \
-F R,Python,FlatFile,HTAN \
-n 15 \
-T MulticoreParam</code></pre>
</div>
</div>
</div>
<div id="parameters" class="section level1">
<h1 class="hasAnchor">
<a href="#parameters" class="anchor"></a>Parameters</h1>
<div id="table-of-parameters" class="section level2">
<h2 class="hasAnchor">
<a href="#table-of-parameters" class="anchor"></a>Table of Parameters</h2>
<p>The pipeline contains various parameters to control the process of quality control. The function of each parameter is shown below:</p>
<div id="required-arguments" class="section level3">
<h3 class="hasAnchor">
<a href="#required-arguments" class="anchor"></a>Required arguments</h3>
<p>The required arguments are as follows:</p>
<table class="table">
<colgroup>
<col width="50%">
<col width="50%">
</colgroup>
<thead><tr class="header">
<th>Parameter</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>-b, –basePath (<strong>required</strong>)</td>
<td>Base path for the output from the preprocessing algorithm.</td>
</tr>
<tr class="even">
<td>-P, –preproc (<strong>required</strong>)</td>
<td>Algorithm used for preprocessing. One of One of ‘CellRangerV2’, ‘CellRangerV3’, ‘BUStools’, ‘STARSolo’, ‘SEQC’, ‘Optimus’, ‘DropEst’, ‘SceRDS’, ‘CountMatrix’ and ‘AnnData’.</td>
</tr>
<tr class="odd">
<td>-s, –sample (<strong>required</strong>)</td>
<td>Name of the sample. This will be prepended to the cell barcodes.</td>
</tr>
<tr class="even">
<td>-o, –directory (<strong>required</strong>)</td>
<td>Output directory. A new subdirectory will be created with the name “sample”. R, Python, and FlatFile directories will be created under the “sample” directory containing the data containers with QC metrics. Default “.”. More information about output directory structure is explained in “Understanding outputs” section below.</td>
</tr>
<tr class="odd">
<td>-F, –outputFormat (<strong>required</strong>)</td>
<td>The output format of this QC pipeline. Currently, it supports SCE, Seurat, FlatFile, AnnData and HTAN (manifest files that meets HTAN requirement).</td>
</tr>
<tr class="even">
<td>-S, –splitSample (<strong>required</strong>)</td>
<td>Save a SingleCellExperiment object for each sample. Default is TRUE. If FALSE, the data of all samples will be combined into one SingleCellExperiment object and this object will be outputed</td>
</tr>
</tbody>
</table>
</div>
<div id="optional-arguments" class="section level3">
<h3 class="hasAnchor">
<a href="#optional-arguments" class="anchor"></a>Optional arguments</h3>
<p>The optional arguments are as follows. Their usage depend on type of data and user-defined behaviour.</p>
<table class="table">
<colgroup>
<col width="50%">
<col width="50%">
</colgroup>
<thead><tr class="header">
<th>Parameter</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>-g, –gmt</td>
<td>GMT file containing gene sets for quality control.</td>
</tr>
<tr class="even">
<td>-t, –delim (required when -g is specified)</td>
<td>Delimiter used in GMT file. Default ".</td>
</tr>
<tr class="odd">
<td>-G, –genome</td>
<td>The name of genome reference. This is only required for CellRangerV2 data.</td>
</tr>
<tr class="even">
<td>-y, –yaml</td>
<td>YAML file used to specify parameters of QC functions called by singleCellTK QC pipeline. Please check “Specify parameters for QC algorithms” section for details.</td>
</tr>
<tr class="odd">
<td>-c, –cellData</td>
<td>The full path of the RDS file or Matrix file of the cell matrix. This would be use only when –preproc is SceRDS or CountMatrix</td>
</tr>
<tr class="even">
<td>-r, –rawData</td>
<td>The full path of the RDS file or Matrix file of the droplet matrix. This would be provided only when –preproc is SceRDS or CountMatrix</td>
</tr>
<tr class="odd">
<td>-C, –cellPath</td>
<td>The directory contains matrix.mtx.gz, features.tsv.gz and barcodes.tsv.gz files originally generated by 10x CellrangerV2 or CellrangerV3 (files in the filtered_feature_bc_matrix directory). This argument only works when –preproc is CellRangerV2 or CellRangerV3. Default is NULL. If ‘base_path’ is NULL, ‘cellPath’ or ‘rawPath’ should be specified</td>
</tr>
<tr class="even">
<td>-R, –rawPath</td>
<td>The directory contains matrix.mtx.gz, features.tsv.gz and barcodes.tsv.gz files originally generated by 10x CellrangerV2 or CellrangerV3 (files in the raw_feature_bc_matrix directory). This argument only works when –preproc is CellRangerV2 or CellRangerV3. Default is NULL. If ‘base_path’ is NULL, ‘cellPath’ or ‘rawPath’ should be specified</td>
</tr>
<tr class="odd">
<td>-d, –dataType</td>
<td>Type of data as input. Default is Both, which means taking both droplet and cell matrix as input. If set as ‘Droplet’, it will only processes droplet data. If set as ‘Cell’, it will only processes cell data</td>
</tr>
<tr class="even">
<td>-D, –detectCells</td>
<td>Detect cells from droplet matrix. Default is FALSE. This argument is only eavluated when -d is ‘Droplet’. If set as TRUE, cells will be detected and cell matrixed will be subset from the droplet matrix. Also, quality control will be performed on the detected cell matrix</td>
</tr>
<tr class="odd">
<td>-m, –cellDetectMethod</td>
<td>Methods to detect cells from droplet matrix. Default is ‘EmptyDrops’. This argument is only eavluated when -D is ‘TRUE’. Other options could be ‘Knee’ or ‘Inflection’. More information is provided in the section below</td>
</tr>
<tr class="even">
<td>-n, –numCores</td>
<td>Number of cores used to run the pipeline. By default is 1. Parallel computing is enabled if -n is greater than 1</td>
</tr>
<tr class="odd">
<td>-T, –parallelType</td>
<td>Type of parallel computing used for parallel computing. Parallel computing used in this pipeline depends on “BiocParallel” package. Default is ‘MulticoreParam’. It can be ‘MulticoreParam’ or ‘SnowParam’. This argument will be evaluated only when numCores &gt; 1</td>
</tr>
<tr class="even">
<td>-i, –studyDesign</td>
<td>The txt file containing the desrciption of the study design. Default is NULL. This would be shown at the begining the html report of cell and droplet QC</td>
</tr>
<tr class="odd">
<td>-L, –subTitle</td>
<td>The subtitle used in the cell and droplet QC HTML report. Default is None. The subtitle can contain information of the sample, like sample name, etc. If -S is set as TRUE, the length of subsitle should be the same as the length of samples. If -S is set as FALSE, the length of subtitle should be one or NULL</td>
</tr>
<tr class="even">
<td>-M, –detectMitoLevel</td>
<td>Detect mitochondrial gene expression level. If TRUE, the pipeline will examinate mito gene expression level automatically without the need of importing user defined gmt file. Default is TRUE</td>
</tr>
<tr class="odd">
<td>-E, –mitoType</td>
<td>Type of mitochondrial gene set to be loaded when –detectMitoLevel is set to TRUE. Possible choices are: ‘human-ensemble’, ‘human-symbol’, ‘human-entrez’, ‘human-ensemble_transcriptID’, ‘mouse-ensemble’, ‘mouse-symbol’, ‘mouse-entrez’ and ‘mouse-ensemble_transcriptID’. They consists of two parts: species and gene id type, which are connected by ‘-’</td>
</tr>
</tbody>
</table>
</div>
</div>
<div id="specify-parameters-using-yaml-file" class="section level2">
<h2 class="hasAnchor">
<a href="#specify-parameters-using-yaml-file" class="anchor"></a>Specify parameters using yaml file</h2>
<p>User can specify parameters for QC algorithms in this pipeline with a yaml file (supplied with -y/–yamlFile argument). The current supported QC algorithms including doublet dection (bcds, cxds, cxds_bcds_hybrid, doubletFinder, doubletCells and scrublet), decontamination (decontX), emptyDrop detection (emptyDrops) and barcodeRankDrops (barcodeRanks). A summary of each function is shown below:</p>
<center>
<img src="qc_yamlParameters.png" title="fig:" alt="QC yaml parameters">
</center>
<p>An example of QC parameters yaml file is shown below:</p>
<pre><code>---
Params:   ### should not be omitted
  bcds:
    ntop: 600

  cxds:
    ntop: 600

  cxds_bcds_hybrid:
    nTop: 600

  decontX:
    maxIter: 600

  emptyDrops:
    lower: 50
    niters: 5000
    testAmbient: True

  barcodeRanks:
    lower: 50</code></pre>
<p>The format of yaml file can be found <a href="https://docs.ansible.com/ansible/latest/reference_appendices/YAMLSyntax.html">here</a>. The parameters should be consistent with the parameters of each QC function in singleCellTK package. Parameters that are not defined in this yaml file will use the default value. Please refer <a href="../../reference/index.html">reference</a> for detailed information about arguments in each QC function.</p>
</div>
<div id="parallel-computing" class="section level2">
<h2 class="hasAnchor">
<a href="#parallel-computing" class="anchor"></a>Parallel computing</h2>
<p>SCTK QC pipeline enables parallel computing to speed up the analysis. Parallel computing is enabled by setting -n/–numCores &gt; 1. The -n/–numCores is used to set the number of cores used for the pipeline.</p>
<p>The backend of parallel computing is supported by “BiocParallel” package. Therefore, users can select different types of parallel evaluation by setting -T/–parallelType argument. Default is “MulticoreParam”. Currently, “MulticoreParam” and “SnowParam” is supported for -T argument. However, “MulticoreParam” is not supported by Windows system. Winows user can choose “SnowParam” as the backend of parallel computing.</p>
</div>
</div>
<div id="description-of-input-data" class="section level1">
<h1 class="hasAnchor">
<a href="#description-of-input-data" class="anchor"></a>Description of input data</h1>
<div id="performing-qc-on-different-combinations-of-the-droplet-and-cell-matrix" class="section level2">
<h2 class="hasAnchor">
<a href="#performing-qc-on-different-combinations-of-the-droplet-and-cell-matrix" class="anchor"></a>Performing QC on different combinations of the droplet and cell matrix</h2>
<p>User can choose to run QC pipeline on only one of the droplet count or cell count matrix, instead of running on both. In this case, the pipeline will only take single input and perform quality control on it. One of the example is shown below:</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1"></a>Rscript SCTK_runQC.R \</span>
<span id="cb8-2"><a href="#cb8-2"></a><span class="op">-</span>b <span class="op">/</span>base<span class="op">/</span>path \</span>
<span id="cb8-3"><a href="#cb8-3"></a><span class="op">-</span>P Preprocessing_Algorithm \</span>
<span id="cb8-4"><a href="#cb8-4"></a><span class="op">-</span>s SampleName \</span>
<span id="cb8-5"><a href="#cb8-5"></a><span class="op">-</span>o Output_Directory \</span>
<span id="cb8-6"><a href="#cb8-6"></a><span class="op">-</span>S <span class="ot">TRUE</span> \</span>
<span id="cb8-7"><a href="#cb8-7"></a><span class="op">-</span>F R,Python,FlatFile,HTAN \</span>
<span id="cb8-8"><a href="#cb8-8"></a><span class="op">-</span>d Droplet \</span>
<span id="cb8-9"><a href="#cb8-9"></a><span class="op">-</span>D <span class="ot">TRUE</span> \</span>
<span id="cb8-10"><a href="#cb8-10"></a><span class="op">-</span>m EmptyDrops</span></code></pre></div>
<p>-d argument is used to specify which count matrix to used in the QC pipeline. Default is “Both”, which means take both droplet and cell count data as input.</p>
<p>If -d argument is set as “Droplet”, the QC pipeline will only take droplet count matrix as input and perform quality control. You can choose whether to detect cells from the droplet matrix by setting -D as TRUE. If yes, cell count matrix will be detected and the pipeline will also perform quality control on this matrix and output the result. You could further define the method used to detect cells from droplet matrix by setting -m argument. -m could be one of “EmptyDrops”, “Knee” or “Inflection”. “EmptyDrops” will keep cells that pass the “runEmptyDrops” function test. “Knee” and “Inflection” will keep cells that pass the knee or inflection point returned from “runBarcodeRankDrops” function.</p>
<p>If -d argument is set as “Cell”, the QC pipeline will only take cell count matrix as input and perform quality control. A figure showing the analysis steps and outputs of different inputs is shown below:</p>
<center>
<img src="qc_singleInput.png" alt="QC single input"> 
</center>
</div>
<div id="importing-data-from-different-preprocessing-tools" class="section level2">
<h2 class="hasAnchor">
<a href="#importing-data-from-different-preprocessing-tools" class="anchor"></a>Importing data from different preprocessing tools</h2>
<div id="import-data-from-cellranger" class="section level3">
<h3 class="hasAnchor">
<a href="#import-data-from-cellranger" class="anchor"></a>Import data from Cellranger</h3>
<p>This pipeline enables different ways to import CellrangerV2/CellrangerV3 data for flexibility.</p>
<ol style="list-style-type: decimal">
<li>If the cellranger data set is saved in the default <a href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/output/overview">cellranger output directory</a>, you can load the data by running following code:</li>
</ol>
<p>For CellRangerV3:</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1"></a>Rscript SCTK_runQC.R \</span>
<span id="cb9-2"><a href="#cb9-2"></a><span class="op">-</span>b <span class="op">/</span>base<span class="op">/</span>path \</span>
<span id="cb9-3"><a href="#cb9-3"></a><span class="op">-</span>P CellRangerV3 \</span>
<span id="cb9-4"><a href="#cb9-4"></a><span class="op">-</span>s SampleName \</span>
<span id="cb9-5"><a href="#cb9-5"></a><span class="op">-</span>o Output_Directory \</span>
<span id="cb9-6"><a href="#cb9-6"></a><span class="op">-</span>S <span class="ot">TRUE</span> \</span>
<span id="cb9-7"><a href="#cb9-7"></a><span class="op">-</span>F R,Python,FlatFile,HTAN</span></code></pre></div>
<p>For CellRangerV2, the reference used by cellranger needs to be specified by -G/–genome:</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1"></a>Rscript SCTK_runQC.R \</span>
<span id="cb10-2"><a href="#cb10-2"></a><span class="op">-</span>b <span class="op">/</span>base<span class="op">/</span>path \</span>
<span id="cb10-3"><a href="#cb10-3"></a><span class="op">-</span>P CellRangerV2 \</span>
<span id="cb10-4"><a href="#cb10-4"></a><span class="op">-</span>s SampleName \</span>
<span id="cb10-5"><a href="#cb10-5"></a><span class="op">-</span>o Output_Directory \</span>
<span id="cb10-6"><a href="#cb10-6"></a><span class="op">-</span>S <span class="ot">TRUE</span> \</span>
<span id="cb10-7"><a href="#cb10-7"></a><span class="op">-</span>G hg19 \</span>
<span id="cb10-8"><a href="#cb10-8"></a><span class="op">-</span>F R,Python,FlatFile,HTAN</span></code></pre></div>
<p>As shown in the <strong>Import data into SCTK</strong> section, -b specify the base path and usually it’s the folder where you ran 10x cellranger count. -s specify the sample name, which has to be same as the name of the sample folder under the base folder. The folder layout would look like the following:</p>
<pre><code>├── BasePath
└── SampleName 
    ├── outs
    |   ├── filtered_feature_bc_matrix
    |   |   ├── barcodes.tsv.gz
    |   |   ├── features.tsv.gz
    |   |   └── matrix.mtx.gz
    |   ├── raw_feature_bc_matrix
    |   |   ├── barcodes.tsv.gz
    |   |   ├── features.tsv.gz
    |   |   └── matrix.mtx.gz
        ...</code></pre>
<ol start="2" style="list-style-type: decimal">
<li>If the cellranger count output have been moved out of the default <a href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/output/overview">cellranger output directory</a>, you can specified the path to droplet and cell count matrix using arguments -R and -C:</li>
</ol>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1"></a>Rscript SCTK_runQC.R \</span>
<span id="cb12-2"><a href="#cb12-2"></a><span class="op">-</span>P CellRangerV2 \</span>
<span id="cb12-3"><a href="#cb12-3"></a><span class="op">-</span>C <span class="op">/</span>path<span class="op">/</span>to<span class="op">/</span>cell<span class="op">/</span>matrix \</span>
<span id="cb12-4"><a href="#cb12-4"></a><span class="op">-</span>R <span class="op">/</span>path<span class="op">/</span>to<span class="op">/</span>droplet<span class="op">/</span>matrix \</span>
<span id="cb12-5"><a href="#cb12-5"></a><span class="op">-</span>s SampleName \</span>
<span id="cb12-6"><a href="#cb12-6"></a><span class="op">-</span>o Output_Directory \</span>
<span id="cb12-7"><a href="#cb12-7"></a><span class="op">-</span>S <span class="ot">TRUE</span> \</span>
<span id="cb12-8"><a href="#cb12-8"></a><span class="op">-</span>F R,Python,FlatFile,HTAN</span></code></pre></div>
<p>In this case, you <strong>must</strong> skip -b arguments and you can also skip -G argument for CellRangerV2 data.</p>
</div>
<div id="import-data-from-rds-or-matrix-stored-in-a-text-file" class="section level3">
<h3 class="hasAnchor">
<a href="#import-data-from-rds-or-matrix-stored-in-a-text-file" class="anchor"></a>Import data from RDS or matrix stored in a text file</h3>
<ol style="list-style-type: decimal">
<li>If your data in stored as a SingleCellExperiment in RDS file, singleCellTK also supports that type of input. To run quality control with RDS file as input, run the following code:</li>
</ol>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1"></a>Rscript SCTK_runQC.R \</span>
<span id="cb13-2"><a href="#cb13-2"></a><span class="op">-</span>P SceRDS \</span>
<span id="cb13-3"><a href="#cb13-3"></a><span class="op">-</span>s Samplename \</span>
<span id="cb13-4"><a href="#cb13-4"></a><span class="op">-</span>o Output_Directory \</span>
<span id="cb13-5"><a href="#cb13-5"></a><span class="op">-</span>S <span class="ot">TRUE</span> \</span>
<span id="cb13-6"><a href="#cb13-6"></a><span class="op">-</span>F R,Python,FlatFile,HTAN \</span>
<span id="cb13-7"><a href="#cb13-7"></a><span class="op">-</span>r <span class="op">/</span>path<span class="op">/</span>to<span class="op">/</span>rds<span class="op">/</span>file<span class="op">/</span>droplet.RDS \</span>
<span id="cb13-8"><a href="#cb13-8"></a><span class="op">-</span>c <span class="op">/</span>path<span class="op">/</span>to<span class="op">/</span>rds<span class="op">/</span>file<span class="op">/</span>cell.RDS</span></code></pre></div>
<ol start="2" style="list-style-type: decimal">
<li>If your input is stored in txt file as a matrix, which has barcodes as colnames and genes as rownames, run the following code to start the quality control pipeline:</li>
</ol>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1"></a>Rscript SCTK_runQC.R \</span>
<span id="cb14-2"><a href="#cb14-2"></a><span class="op">-</span>P CountMatrix \</span>
<span id="cb14-3"><a href="#cb14-3"></a><span class="op">-</span>s Samplename \</span>
<span id="cb14-4"><a href="#cb14-4"></a><span class="op">-</span>o Output_Directory \</span>
<span id="cb14-5"><a href="#cb14-5"></a><span class="op">-</span>S <span class="ot">TRUE</span> \</span>
<span id="cb14-6"><a href="#cb14-6"></a><span class="op">-</span>r <span class="op">/</span>path<span class="op">/</span>to<span class="op">/</span>matrix<span class="op">/</span>file<span class="op">/</span>droplet.txt \</span>
<span id="cb14-7"><a href="#cb14-7"></a><span class="op">-</span>c path<span class="op">/</span>to<span class="op">/</span>matrix<span class="op">/</span>file<span class="op">/</span>cell.txt</span></code></pre></div>
</div>
<div id="methods-to-run-pipeline-on-data-set-generated-by-other-algorithms" class="section level3">
<h3 class="hasAnchor">
<a href="#methods-to-run-pipeline-on-data-set-generated-by-other-algorithms" class="anchor"></a>Methods to run pipeline on data set generated by other algorithms</h3>
<p>If your data is preprocessed by other algorithms, you can load the data as shown in the figure above. Basically, the templated is shown below:</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1"></a>Rscript SCTK_runQC.R \</span>
<span id="cb15-2"><a href="#cb15-2"></a><span class="op">-</span>b <span class="op">/</span>base<span class="op">/</span>path \</span>
<span id="cb15-3"><a href="#cb15-3"></a><span class="op">-</span>P Preprocessing_Algorithm \</span>
<span id="cb15-4"><a href="#cb15-4"></a><span class="op">-</span>s SampleName \</span>
<span id="cb15-5"><a href="#cb15-5"></a><span class="op">-</span>o Output_Directory \</span>
<span id="cb15-6"><a href="#cb15-6"></a><span class="op">-</span>S <span class="ot">TRUE</span> \</span>
<span id="cb15-7"><a href="#cb15-7"></a><span class="op">-</span>F R,Python,FlatFile,HTAN</span></code></pre></div>
</div>
</div>
<div id="gene-sets" class="section level2">
<h2 class="hasAnchor">
<a href="#gene-sets" class="anchor"></a>Gene sets</h2>
<p>Quantifying the level of gene sets can be useful quality control. For example, the percentage of counts from mitochondrial genes can be an indicator or cell stress or death.</p>
<p>Users can pass a <a href="http://software.broadinstitute.org/cancer/software/gsea/wiki/index.php/Data_formats">GMT</a> file (with -g/–gmt argument) to the pipeline with one row for each gene set. The first column should be the name of the gene set (e.g. mito). The second column for each gene set in the GMT file (i.e. the description) should contain the location of where to look for the matching IDs in the data. If set to ‘rownames’, then the gene set IDs will be matched with the row IDs of the data matrix. If a character string or an integer index is supplied, then gene set IDs will be matched to IDs the that column of feature table. Gene sets with mitochondrial genes can be found <a href="https://github.com/compbiomed/singleCellTK/tree/master/exec">here</a>.</p>
</div>
</div>
<div id="outputs" class="section level1">
<h1 class="hasAnchor">
<a href="#outputs" class="anchor"></a>Outputs</h1>
<p>The output directory is created under the path specified by -o/–directory argument. Each sample is stored in the subdirectory (named by -s/–sample argument) within this output direcotry. Within each sample directory, the each output format will be separated into subdirectories. The output file hierarchy is shown below:</p>
<pre><code>(root; output directory)
├── level3Meta.csv
├── level4Meta.csv
├── sample1_dropletQC.html
├── sample1_cellQC.html
└── sample1 
    ├── R
    |   ├── sample1_Droplets.rds
    |   └── sample1_Cells.rds
    ├── Python
    |   ├── Droplets
    |   |   └── sample1.h5ad
    |   └── Cells
    |       └── sample1.h5ad
    ├── FlatFile
    |   ├── Droplets
    |   |   ├── assays
    |   |   |   └── sample1_counts.mtx.gz
    |   |   ├── metadata
    |   |   |   └── sample1_metadata.rds
    |   |   ├── sample1_colData.txt.gz
    |   |   └── sample1_rowData.txt.gz 
    |   └── Cells
    |       ├── assays
    |       |   └── sample1_counts.mtx.gz
    |       ├── metadata
    |       |   └── sample1_metadata.rds
    |       ├── reducedDims
    |       |   ├──sample1_decontX_UMAP.txt.gz
    |       |   ├──sample1_scrublet_TSNE.txt.gz
    |       |   └──sample1_scrublet_UMAP.txt.gz
    |       ├── sample1_colData.txt.gz
    |       └── sample1_rowData.txt.gz 
    └── sample1_QCparameters.yaml</code></pre>
</div>
<div id="documentation-of-tools-that-are-currently-available-within-the-pipeline" class="section level1">
<h1 class="hasAnchor">
<a href="#documentation-of-tools-that-are-currently-available-within-the-pipeline" class="anchor"></a>Documentation of tools that are currently available within the pipeline:</h1>
<div id="empty-droplet-detection" class="section level4">
<h4 class="hasAnchor">
<a href="#empty-droplet-detection" class="anchor"></a>Empty droplet detection:</h4>
<ul>
<li>
<a href="https://rdrr.io/github/MarioniLab/DropletUtils/man/emptyDrops.html">emptyDrops</a> from the package <a href="https://bioconductor.org/packages/release/bioc/html/DropletUtils.html">DropletUtils</a>
</li>
<li>
<a href="https://rdrr.io/github/MarioniLab/DropletUtils/man/barcodeRanks.html">barcodeRanks</a> from the package <a href="https://bioconductor.org/packages/release/bioc/html/DropletUtils.html">DropletUtils</a>
</li>
</ul>
</div>
<div id="doublet-detection" class="section level4">
<h4 class="hasAnchor">
<a href="#doublet-detection" class="anchor"></a>Doublet Detection</h4>
<ul>
<li>
<a href="https://rdrr.io/github/MarioniLab/scran/man/doubletCells.html">doubletCells</a> from the package <a href="http://bioconductor.org/packages/release/bioc/html/scran.html">scran</a>
</li>
<li>
<a href="https://rdrr.io/bioc/scds/man/cxds.html">cxds</a>, <a href="https://rdrr.io/bioc/scds/man/bcds.html">bcds</a>, and <a href="https://rdrr.io/bioc/scds/man/cxds_bcds_hybrid.html">cxds_bcds_hybrid</a> from the package <a href="http://bioconductor.org/packages/release/bioc/html/scds.html">scds</a>
</li>
<li>
<a href="https://rdrr.io/github/chris-mcginnis-ucsf/DoubletFinder/man/doubletFinder.html">doubletFinder</a> from the package <a href="https://github.com/chris-mcginnis-ucsf/DoubletFinder">DoubletFinder</a>
</li>
<li>
<a href="https://bioconda.github.io/recipes/scrublet/README.html">Scrublet</a> from the package <a href="https://github.com/allonkleinlab/scrublet">scrublet</a>
</li>
</ul>
</div>
<div id="ambient-rna-detection" class="section level4">
<h4 class="hasAnchor">
<a href="#ambient-rna-detection" class="anchor"></a>Ambient RNA detection</h4>
<ul>
<li>
<a href="https://rdrr.io/bioc/celda/man/decontX.html">decontX</a> from the package <a href="https://bioconductor.org/packages/release/bioc/html/celda.html">celda</a>
</li>
</ul>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by David Jenkins, Vidya Akavoor, Salam Alabdullatif, Shruthi Bandyadka, Emma Briars, Xinyun Cao, Sebastian Carrasco Pro, Tyler Faits, Rui Hong, Mohammed Muzamil Khan, Yusuke Koga, Anastasia Leshchyk, Irzam Sarfraz, Yichen Wang, Zhe Wang, W. Evan Johnson, Joshua David Campbell.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
