<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Quality Control for droplet-level data • singleCellTK</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/yeti/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../../bootstrap-toc.css">
<script src="../../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../../pkgdown.css" rel="stylesheet">
<script src="../../pkgdown.js"></script><link href="../../extra.css" rel="stylesheet">
<script src="../../extra.js"></script><meta property="og:title" content="Quality Control for droplet-level data">
<meta property="og:description" content="singleCellTK">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../../index.html">singleCellTK</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">2.3.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="https://www.camplab.net/sctk">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../../articles/articles/installation.html">Installation</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Documentation
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Data</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../../articles/articles/import_data.html">Import Data</a>
        </li>
        <li>
          <a href="../../articles/articles/cmd_import_scRNAseq_data_as_SCE.html">Import Data for Commandline QC</a>
        </li>
        <li>
          <a href="../../articles/articles/import_genesets.html">Import Gene Sets</a>
        </li>
        <li>
          <a href="../../articles/articles/import_annotation.html">Annotation</a>
        </li>
        <li>
          <a href="../../articles/articles/export_data.html">Export Data</a>
        </li>
        <li>
          <a href="../../articles/articles/delete_data.html">Delete Data</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">QC &amp; Filtering</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../../articles/articles/cnsl_dropletqc.html">Droplet QC (R console)</a>
        </li>
        <li>
          <a href="../../articles/articles/cnsl_cellqc.html">Cell QC (R console)</a>
        </li>
        <li>
          <a href="../../articles/articles/ui_qc.html">Cell QC (Shiny UI)</a>
        </li>
        <li>
          <a href="../../articles/articles/filtering.html">Filtering</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Normalization &amp; Batch Correction</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../../articles/articles/cnsl_normalization.html">Normalization</a>
        </li>
        <li>
          <a href="../../articles/articles/batch_correction.html">Batch Correction</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Feature Selection &amp; Dimensionality Reduction</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../../articles/articles/cnsl_feature_selection.html">Feature Selection</a>
        </li>
        <li>
          <a href="../../articles/articles/cnsl_dimensionality_reduction.html">Dimensionality Reduction</a>
        </li>
        <li>
          <a href="../../articles/articles/cnsl_2d_embedding.html">2D Embedding</a>
        </li>
      </ul>
</li>
    <li>
      <a href="../../articles/articles/clustering.html">Clustering</a>
    </li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Differential Expression &amp; Cell Type Labeling</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../../articles/articles/differential_expression.html">Differential Expression</a>
        </li>
        <li>
          <a href="../../articles/articles/find_marker.html">Find Marker</a>
        </li>
        <li>
          <a href="../../articles/articles/cell_type_labeling.html">Cell Type Labeling</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Pathway &amp; Enrichment Analysis</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../../articles/articles/cnsl_enrichR.html">EnrichR</a>
        </li>
        <li>
          <a href="../../articles/articles/pathwayAnalysis.html">Pathway Analysis</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Curated Workflows</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../../articles/articles/cnsl_seurat_curated_workflow.html">Seurat</a>
        </li>
        <li>
          <a href="../../articles/articles/celda_curated_workflow.html">Celda</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Visualization</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../../articles/articles/visualization.html">General Visualization</a>
        </li>
        <li>
          <a href="../../articles/articles/heatmap.html">Generic Heatmap</a>
        </li>
      </ul>
</li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../../articles/articles/console_analysis_tutorial.html">Analysis of PBMC3K (R Console)</a>
    </li>
  </ul>
</li>
<li>
  <a href="../../articles/articles/cmd_qc.html">SCTK-QC</a>
</li>
<li>
  <a href="../../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/compbiomed/singleCellTK">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="https://twitter.com/camplab1">
    <span class="fab fa-twitter fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="cnsl_cellqc_files/header-attrs-2.9/header-attrs.js"></script><script src="cnsl_cellqc_files/kePrint-0.0.1/kePrint.js"></script><link href="cnsl_cellqc_files/lightable-0.0.1/lightable.css" rel="stylesheet">
<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Quality Control for droplet-level data</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/compbiomed/singleCellTK/blob/master/vignettes/articles/cnsl_cellqc.Rmd"><code>vignettes/articles/cnsl_cellqc.Rmd</code></a></small>
      <div class="hidden name"><code>cnsl_cellqc.Rmd</code></div>

    </div>

    
    
<style>
body {
text-align: justify}
</style>
<div id="introduction" class="section level2">
<h2 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h2>
<p>Performing comprehensive quality control (QC) is necessary to remove poor quality cells for downstream analysis of single-cell RNA sequencing (scRNA-seq) data. Therefore, assessment of the data is required, for which various QC algorithms have been developed. In singleCellTK (SCTK), we have written convenience functions for several of these tools. In this guide, we will demonstrate how to use these functions to perform quality control on cell data. (For definition of cell data, please refer <a href="cmd_qc.html#introduction-1">this documentation</a>.)</p>
<p>The package can be loaded using the <code>library</code> command.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://www.camplab.net/sctk/">singleCellTK</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org">dplyr</a></span><span class="op">)</span></code></pre></div>
<p><br></p>
</div>
<div id="running-qc-on-cell-filtered-data" class="section level2">
<h2 class="hasAnchor">
<a href="#running-qc-on-cell-filtered-data" class="anchor"></a>Running QC on cell-filtered data</h2>
<div id="load-pbmc-data-from-10x" class="section level3">
<h3 class="hasAnchor">
<a href="#load-pbmc-data-from-10x" class="anchor"></a>Load PBMC data from 10X</h3>
<p>We will use a filtered form of the PBMC 3K and 6K dataset from the package <a href="http://bioconductor.org/packages/release/data/experiment/html/TENxPBMCData.html">TENxPBMCData</a>, which is available from the <code><a href="../../reference/importExampleData.html">importExampleData()</a></code> function. We will combine these datasets together into a single <a href="https://rdrr.io/bioc/SingleCellExperiment/man/SingleCellExperiment.html">SingleCellExperiment</a> object.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pbmc3k</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/importExampleData.html">importExampleData</a></span><span class="op">(</span>dataset <span class="op">=</span> <span class="st">"pbmc3k"</span><span class="op">)</span>
<span class="va">pbmc6k</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/importExampleData.html">importExampleData</a></span><span class="op">(</span>dataset <span class="op">=</span> <span class="st">"pbmc6k"</span><span class="op">)</span>

<span class="va">pbmc.combined</span> <span class="op">&lt;-</span> <span class="fu">BiocGenerics</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/cbind.html">cbind</a></span><span class="op">(</span><span class="va">pbmc3k</span>, <span class="va">pbmc6k</span><span class="op">)</span>
<span class="va">sample.vector</span> <span class="op">=</span> <span class="fu">colData</span><span class="op">(</span><span class="va">pbmc.combined</span><span class="op">)</span><span class="op">$</span><span class="va">sample</span></code></pre></div>
<p>SCTK also supports the importing of single-cell data from the following platforms: <a href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/what-is-cell-ranger">10X CellRanger</a>, <a href="https://github.com/alexdobin/STAR/blob/master/docs/STARsolo.md">STARSolo</a>, <a href="https://www.kallistobus.tools/">BUSTools</a>, <a href="https://github.com/ambrosejcarr/seqc">SEQC</a>, <a href="https://github.com/hms-dbmi/dropEst">DropEST</a>, <a href="https://salmon.readthedocs.io/en/latest/alevin.html">Alevin</a>, as well as dataset already stored in <a href="https://rdrr.io/bioc/SingleCellExperiment/man/SingleCellExperiment.html">SingleCellExperiment</a> object and <a href="https://github.com/theislab/anndata">AnnData</a> object. To load your own input data, please refer <code>Function Reference for pre-processing tools</code> under <code>Console Analysis</code> section in <a href="import_data.html">Import data into SCTK</a> for detailed instruction.</p>
</div>
<div id="run-2d-embedding" class="section level3">
<h3 class="hasAnchor">
<a href="#run-2d-embedding" class="anchor"></a>Run 2D embedding</h3>
<p>SCTK utilizes 2D embedding techniques such as TSNE and UMAP for visualizing single-cell data. Users can modify the dimensions by adjusting the parameters within the function. The <code>logNorm</code> parameter should be set to <code>TRUE</code> for normalization prior to generating the 2D embedding.</p>
<p>The <code>sample</code> parameter may be specified if multiple samples exist in the <code>SingleCellExperiment</code> object. Here, we will use the sample vector stored in the <code>colData</code> slot of the <code>SingleCellExperiment</code> object.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># UMAP:</span>
<span class="va">pbmc.combined</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/getUMAP.html">getUMAP</a></span><span class="op">(</span>inSCE <span class="op">=</span> <span class="va">pbmc.combined</span>, useAssay <span class="op">=</span> <span class="st">"counts"</span>, logNorm <span class="op">=</span> <span class="cn">TRUE</span>, sample <span class="op">=</span> <span class="va">sample.vector</span><span class="op">)</span>

<span class="co"># TSNE (not run):</span>
<span class="co"># pbmc &lt;- getTSNE(inSCE=pbmc, useAssay="counts", logNorm = TRUE, sample = colData(pbmc)$sample)</span></code></pre></div>
</div>
<div id="run-cellqc" class="section level3">
<h3 class="hasAnchor">
<a href="#run-cellqc" class="anchor"></a>Run CellQC</h3>
<p>All of the droplet-based QC algorithms are able to be run under the wrapper function <code><a href="../../reference/runCellQC.html">runCellQC()</a></code>. By default all possible QC algorithms will be run.</p>
<p>Users may set a <code>sample</code> parameter if need to compare between multiple samples. Here, we will use the sample vector stored in the <code>SingleCellExperiment</code> object.</p>
<p>If users wishes, a list of gene sets can be applied to the function to determine the expression of a set of specific genes. A gene list imported into the <code>SingleCellExperiment</code> object using <code>importGeneSets*</code> functions can be set as <code>collectionName</code>. Additionally, a pre-made list of genes can be used to determine the level of gene expression per cell. A list containing gene identifiers may be set as <code>geneSetList</code>, or the user may instead use the <code>geneSetCollection</code> parameter to supply a <code>GeneSetCollection</code> object from the <a href="https://bioconductor.org/packages/release/bioc/html/GSEABase.html">GSEABase</a> package. Please also refer to <a href="import_genesets.html"><em>Import Genesets</em> documentation</a>.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pbmc.combined</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/importGeneSetsFromGMT.html">importGeneSetsFromGMT</a></span><span class="op">(</span>inSCE <span class="op">=</span> <span class="va">pbmc.combined</span>, collectionName <span class="op">=</span> <span class="st">"mito"</span>, file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata/mito_subset.gmt"</span>, package <span class="op">=</span> <span class="st">"singleCellTK"</span><span class="op">)</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">12345</span><span class="op">)</span>
<span class="va">pbmc.combined</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/runCellQC.html">runCellQC</a></span><span class="op">(</span><span class="va">pbmc.combined</span>, algorithms <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"QCMetrics"</span>, <span class="st">"scrublet"</span>, <span class="st">"scDblFinder"</span>, <span class="st">"cxds"</span>, <span class="st">"bcds"</span>, <span class="st">"cxds_bcds_hybrid"</span>, <span class="st">"doubletFinder"</span>, <span class="st">"decontX"</span><span class="op">)</span>, sample <span class="op">=</span> <span class="va">sample.vector</span>, collectionName <span class="op">=</span> <span class="st">"mito"</span><span class="op">)</span></code></pre></div>
<pre><code>## [20:17:50] WARNING: amalgamation/../src/learner.cc:1095: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'binary:logistic' was changed from 'error' to 'logloss'. Explicitly set eval_metric if you'd like to restore the old behavior.
## [20:18:20] WARNING: amalgamation/../src/learner.cc:1095: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'binary:logistic' was changed from 'error' to 'logloss'. Explicitly set eval_metric if you'd like to restore the old behavior.
## [20:18:36] WARNING: amalgamation/../src/learner.cc:1095: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'binary:logistic' was changed from 'error' to 'logloss'. Explicitly set eval_metric if you'd like to restore the old behavior.
## [20:19:04] WARNING: amalgamation/../src/learner.cc:1095: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'binary:logistic' was changed from 'error' to 'logloss'. Explicitly set eval_metric if you'd like to restore the old behavior.</code></pre>
<p><br> If users choose to only run a specific set of algorithms, they can specify which to run with the <code>algorithms</code> parameter. By default, the <code><a href="../../reference/runCellQC.html">runCellQC()</a></code> will run <code>"QCMetrics"</code>, <code>"scDblFinder"</code>, <code>"cxds"</code>, <code>"bcds"</code>, <code>"cxds_bcds_hybrid"</code> and <code>"decontX"</code> algorithms by default. Besides, <code>"scrublet"</code> and <code>"doubletFinder"</code> are supported if our users want to run them.</p>
<p>After running QC functions with SCTK, the output will be stored in the <code>colData</code> slot of the <code>SingleCellExperiment</code> object.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="fu">colData</span><span class="op">(</span><span class="va">pbmc.combined</span><span class="op">)</span>, <span class="fl">5</span><span class="op">)</span></code></pre></div>
<div style="border: 1px solid #ddd; padding: 5px; overflow-x: scroll; width:80%; ">
<table class="table table">
<thead><tr>
<th style="text-align:left;">
</th>
<th style="text-align:left;">
Sample
</th>
<th style="text-align:left;">
Barcode
</th>
<th style="text-align:left;">
Sequence
</th>
<th style="text-align:right;">
Library
</th>
<th style="text-align:left;">
Cell_ranger_version
</th>
<th style="text-align:left;">
Tissue_status
</th>
<th style="text-align:left;">
Barcode_type
</th>
<th style="text-align:left;">
Chemistry
</th>
<th style="text-align:left;">
Sequence_platform
</th>
<th style="text-align:left;">
Individual
</th>
<th style="text-align:left;">
Date_published
</th>
<th style="text-align:left;">
sample
</th>
<th style="text-align:right;">
sum
</th>
<th style="text-align:right;">
detected
</th>
<th style="text-align:right;">
percent.top_50
</th>
<th style="text-align:right;">
percent.top_100
</th>
<th style="text-align:right;">
percent.top_200
</th>
<th style="text-align:right;">
percent.top_500
</th>
<th style="text-align:right;">
subsets_mito_sum
</th>
<th style="text-align:right;">
subsets_mito_detected
</th>
<th style="text-align:right;">
subsets_mito_percent
</th>
<th style="text-align:right;">
total
</th>
<th style="text-align:right;">
scrublet_score
</th>
<th style="text-align:left;">
scrublet_call
</th>
<th style="text-align:left;">
scDblFinder_sample
</th>
<th style="text-align:left;">
scDblFinder_cluster
</th>
<th style="text-align:left;">
scDblFinder_class
</th>
<th style="text-align:right;">
scDblFinder_doublet_score
</th>
<th style="text-align:right;">
scDblFinder_weighted
</th>
<th style="text-align:right;">
scDblFinder_difficulty
</th>
<th style="text-align:right;">
scDblFinder_cxds_score
</th>
<th style="text-align:left;">
scDblFinder_mostLikelyOrigin
</th>
<th style="text-align:left;">
scDblFinder_originAmbiguous
</th>
<th style="text-align:right;">
doubletFinder_doublet_score_resolution_1.5
</th>
<th style="text-align:left;">
doubletFinder_doublet_label_resolution_1.5
</th>
<th style="text-align:right;">
scds_cxds_score
</th>
<th style="text-align:left;">
scds_cxds_call
</th>
<th style="text-align:right;">
scds_bcds_score
</th>
<th style="text-align:left;">
scds_bcds_call
</th>
<th style="text-align:right;">
scds_hybrid_score
</th>
<th style="text-align:left;">
scds_hybrid_call
</th>
<th style="text-align:right;">
decontX_contamination
</th>
<th style="text-align:left;">
decontX_clusters
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;">
pbmc3k_AAACATACAACCAC-1
</td>
<td style="text-align:left;">
pbmc3k
</td>
<td style="text-align:left;">
AAACATACAACCAC-1
</td>
<td style="text-align:left;">
AAACATACAACCAC
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:left;">
v1.1.0
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:left;">
GemCode
</td>
<td style="text-align:left;">
Chromium_v1
</td>
<td style="text-align:left;">
NextSeq500
</td>
<td style="text-align:left;">
HealthyDonor2
</td>
<td style="text-align:left;">
2016-05-26
</td>
<td style="text-align:left;">
pbmc3k
</td>
<td style="text-align:right;">
2421
</td>
<td style="text-align:right;">
781
</td>
<td style="text-align:right;">
47.74886
</td>
<td style="text-align:right;">
63.27964
</td>
<td style="text-align:right;">
74.96902
</td>
<td style="text-align:right;">
88.39323
</td>
<td style="text-align:right;">
73
</td>
<td style="text-align:right;">
10
</td>
<td style="text-align:right;">
3.015283
</td>
<td style="text-align:right;">
2421
</td>
<td style="text-align:right;">
0.1707317
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
pbmc3k
</td>
<td style="text-align:left;">
1
</td>
<td style="text-align:left;">
singlet
</td>
<td style="text-align:right;">
0.4827743
</td>
<td style="text-align:right;">
0.8104423
</td>
<td style="text-align:right;">
0.0510226
</td>
<td style="text-align:right;">
0.0303611
</td>
<td style="text-align:left;">
1+2
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:right;">
0.0138889
</td>
<td style="text-align:left;">
Singlet
</td>
<td style="text-align:right;">
22794.14
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:right;">
0.0149839
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:right;">
0.2109824
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:right;">
0.0330631
</td>
<td style="text-align:left;">
pbmc3k-1
</td>
</tr>
<tr>
<td style="text-align:left;">
pbmc3k_AAACATTGAGCTAC-1
</td>
<td style="text-align:left;">
pbmc3k
</td>
<td style="text-align:left;">
AAACATTGAGCTAC-1
</td>
<td style="text-align:left;">
AAACATTGAGCTAC
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:left;">
v1.1.0
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:left;">
GemCode
</td>
<td style="text-align:left;">
Chromium_v1
</td>
<td style="text-align:left;">
NextSeq500
</td>
<td style="text-align:left;">
HealthyDonor2
</td>
<td style="text-align:left;">
2016-05-26
</td>
<td style="text-align:left;">
pbmc3k
</td>
<td style="text-align:right;">
4903
</td>
<td style="text-align:right;">
1352
</td>
<td style="text-align:right;">
45.50275
</td>
<td style="text-align:right;">
61.02386
</td>
<td style="text-align:right;">
71.81318
</td>
<td style="text-align:right;">
82.62288
</td>
<td style="text-align:right;">
186
</td>
<td style="text-align:right;">
10
</td>
<td style="text-align:right;">
3.793596
</td>
<td style="text-align:right;">
4903
</td>
<td style="text-align:right;">
0.1034483
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
pbmc3k
</td>
<td style="text-align:left;">
4
</td>
<td style="text-align:left;">
singlet
</td>
<td style="text-align:right;">
0.4283143
</td>
<td style="text-align:right;">
0.5054636
</td>
<td style="text-align:right;">
0.0114767
</td>
<td style="text-align:right;">
0.1392021
</td>
<td style="text-align:left;">
1+4
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:right;">
0.1527778
</td>
<td style="text-align:left;">
Singlet
</td>
<td style="text-align:right;">
35689.05
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:right;">
0.9710556
</td>
<td style="text-align:left;">
TRUE
</td>
<td style="text-align:right;">
0.8471188
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:right;">
0.1377744
</td>
<td style="text-align:left;">
pbmc3k-2
</td>
</tr>
</tbody>
</table>
</div>
<p><br></p>
<details><summary><b>A summary of all outputs</b></summary><table class="table">
<colgroup>
<col width="29%">
<col width="45%">
<col width="13%">
<col width="10%">
</colgroup>
<thead><tr class="header">
<th>QC output</th>
<th align="left">Description</th>
<th align="left">Methods</th>
<th align="left">Package/Tool</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><code>sum</code></td>
<td align="left">Total counts</td>
<td align="left"><code><a href="../../reference/runPerCellQC.html">runPerCellQC()</a></code></td>
<td align="left"><a href="https://bioconductor.org/packages/release/bioc/html/scater.html">scater</a></td>
</tr>
<tr class="even">
<td><code>detected</code></td>
<td align="left">Total features</td>
<td align="left"><code><a href="../../reference/runPerCellQC.html">runPerCellQC()</a></code></td>
<td align="left"><a href="https://bioconductor.org/packages/release/bioc/html/scater.html">scater</a></td>
</tr>
<tr class="odd">
<td><code>percent_top</code></td>
<td align="left">% Expression coming from top features</td>
<td align="left"><code><a href="../../reference/runPerCellQC.html">runPerCellQC()</a></code></td>
<td align="left"><a href="https://bioconductor.org/packages/release/bioc/html/scater.html">scater</a></td>
</tr>
<tr class="even">
<td><code>subsets_*</code></td>
<td align="left">sum, detected, percent_top calculated on specified gene list</td>
<td align="left"><code><a href="../../reference/runPerCellQC.html">runPerCellQC()</a></code></td>
<td align="left"><a href="https://bioconductor.org/packages/release/bioc/html/scater.html">scater</a></td>
</tr>
<tr class="odd">
<td><code>scrublet_score</code></td>
<td align="left">Doublet score</td>
<td align="left"><code><a href="../../reference/runScrublet.html">runScrublet()</a></code></td>
<td align="left"><a href="https://github.com/swolock/scrublet">scrublet</a></td>
</tr>
<tr class="even">
<td><code>scrublet_call</code></td>
<td align="left">Doublet classification based on threshold</td>
<td align="left"><code><a href="../../reference/runScrublet.html">runScrublet()</a></code></td>
<td align="left"><a href="https://github.com/swolock/scrublet">scrublet</a></td>
</tr>
<tr class="odd">
<td><code>scDblFinder_doublet_score</code></td>
<td align="left">Doublet score</td>
<td align="left"><code><a href="../../reference/runScDblFinder.html">runScDblFinder()</a></code></td>
<td align="left"><a href="https://bioconductor.org/packages/release/bioc/html/scDblFinder.html">scDblFinder</a></td>
</tr>
<tr class="even">
<td><code>doubletFinder_doublet_score</code></td>
<td align="left">Doublet score</td>
<td align="left"><code><a href="../../reference/runDoubletFinder.html">runDoubletFinder()</a></code></td>
<td align="left"><a href="https://github.com/chris-mcginnis-ucsf/DoubletFinder">DoubletFinder</a></td>
</tr>
<tr class="odd">
<td><code>doubletFinder_doublet_label_resolution</code></td>
<td align="left">Doublet classification based on threshold</td>
<td align="left"><code><a href="../../reference/runDoubletFinder.html">runDoubletFinder()</a></code></td>
<td align="left"><a href="https://github.com/chris-mcginnis-ucsf/DoubletFinder">DoubletFinder</a></td>
</tr>
<tr class="even">
<td><code>scds_cxds_score</code></td>
<td align="left">Doublet score</td>
<td align="left"><code><a href="../../reference/runCxds.html">runCxds()</a></code></td>
<td align="left"><a href="https://www.bioconductor.org/packages/release/bioc/html/scds.html">SCDS</a></td>
</tr>
<tr class="odd">
<td><code>scds_cxds_call</code></td>
<td align="left">Doublet classification based on threshold</td>
<td align="left"><code><a href="../../reference/runCxds.html">runCxds()</a></code></td>
<td align="left"><a href="https://www.bioconductor.org/packages/release/bioc/html/scds.html">SCDS</a></td>
</tr>
<tr class="even">
<td><code>scds_bcds_score</code></td>
<td align="left">Doublet score</td>
<td align="left"><code><a href="../../reference/runBcds.html">runBcds()</a></code></td>
<td align="left"><a href="https://www.bioconductor.org/packages/release/bioc/html/scds.html">SCDS</a></td>
</tr>
<tr class="odd">
<td><code>scds_bcds_call</code></td>
<td align="left">Doublet classification based on threshold</td>
<td align="left"><code><a href="../../reference/runBcds.html">runBcds()</a></code></td>
<td align="left"><a href="https://www.bioconductor.org/packages/release/bioc/html/scds.html">SCDS</a></td>
</tr>
<tr class="even">
<td><code>scds_hybrid_score</code></td>
<td align="left">Doublet score</td>
<td align="left"><code><a href="../../reference/runCxdsBcdsHybrid.html">runCxdsBcdsHybrid()</a></code></td>
<td align="left"><a href="https://www.bioconductor.org/packages/release/bioc/html/scds.html">SCDS</a></td>
</tr>
<tr class="odd">
<td><code>scds_hybrid_call</code></td>
<td align="left">Doublet classification based on threshold</td>
<td align="left"><code><a href="../../reference/runCxdsBcdsHybrid.html">runCxdsBcdsHybrid()</a></code></td>
<td align="left"><a href="https://www.bioconductor.org/packages/release/bioc/html/scds.html">SCDS</a></td>
</tr>
<tr class="even">
<td><code>decontX_contamination</code></td>
<td align="left">Ambient RNA contamination</td>
<td align="left"><code><a href="../../reference/runDecontX.html">runDecontX()</a></code></td>
<td align="left"><a href="https://www.bioconductor.org/packages/release/bioc/html/celda.html">celda</a></td>
</tr>
<tr class="odd">
<td><code>decontX_clusters</code></td>
<td align="left">Clusters determined in dataset based on underlying algorithm</td>
<td align="left"><code><a href="../../reference/runDecontX.html">runDecontX()</a></code></td>
<td align="left"><a href="https://www.bioconductor.org/packages/release/bioc/html/celda.html">celda</a></td>
</tr>
</tbody>
</table></details><p><br></p>
<p>The names of the 2D embedding and dimension reduction matrices are stored in the <code>reducedDims</code> slot of the <code>SingleCellExperiment</code> object.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">reducedDims</span><span class="op">(</span><span class="va">pbmc.combined</span><span class="op">)</span></code></pre></div>
<pre><code>## List of length 6
## names(6): UMAP scrublet_TSNE ... decontX_pbmc3k_UMAP decontX_pbmc6k_UMAP</code></pre>
<div id="generating-a-summary-statistic-table" class="section level4">
<h4 class="hasAnchor">
<a href="#generating-a-summary-statistic-table" class="anchor"></a>Generating a summary statistic table</h4>
<p>The function <code><a href="../../reference/sampleSummaryStats.html">sampleSummaryStats()</a></code> may be used to generate a table containing the mean and median of the data per sample.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../../reference/sampleSummaryStats.html">sampleSummaryStats</a></span><span class="op">(</span><span class="va">pbmc.combined</span>, sample <span class="op">=</span> <span class="va">sample.vector</span><span class="op">)</span></code></pre></div>
<pre><code>##                           pbmc3k  pbmc6k   Total
## Number of Cells          2700.00 5419.00 8119.00
## Mean counts              2366.90 2027.60 2140.50
## Median counts            2197.00 1873.00 1988.00
## Mean features detected    846.99  748.06  780.96
## Median features detected  817.00  716.00  750.00</code></pre>
<p>If users choose to generate a table for all QC metrics generated through <code><a href="../../reference/runCellQC.html">runCellQC()</a></code>, they may set the <code>simple</code> parameter to <code>FALSE</code>.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../../reference/sampleSummaryStats.html">sampleSummaryStats</a></span><span class="op">(</span><span class="va">pbmc.combined</span>, sample <span class="op">=</span> <span class="va">sample.vector</span>, simple <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></code></pre></div>
<pre><code>##                                                           pbmc3k    pbmc6k
## Number of Cells                                        2700.0000 5419.0000
## Mean counts                                            2366.9000 2027.6000
## Median counts                                          2197.0000 1873.0000
## Mean features detected                                  846.9900  748.0600
## Median features detected                                817.0000  716.0000
## Scrublet - Number of doublets                            41.0000  132.0000
## Scrublet - Percentage of doublets                         1.5200    2.4400
## scDblFinder - Number of doublets                         83.0000  273.0000
## scDblFinder - Percentage of doublets                      3.0700    5.0400
## DoubletFinder - Number of doublets, Resolution 1.5      202.0000  406.0000
## DoubletFinder - Percentage of doublets, Resolution 1.5    7.4800    7.4900
## CXDS - Number of doublets                               132.0000  294.0000
## CXDS - Percentage of doublets                             4.8900    5.4300
## BCDS - Number of doublets                               151.0000  257.0000
## BCDS - Percentage of doublets                             5.5900    4.7400
## SCDS Hybrid - Number of doublets                        173.0000  301.0000
## SCDS Hybrid - Percentage of doublets                      6.4100    5.5500
## DecontX - Mean contamination                              0.0808    0.0609
## DecontX - Median contamination                            0.0550    0.0365
##                                                            Total
## Number of Cells                                        8119.0000
## Mean counts                                            2140.5000
## Median counts                                          1988.0000
## Mean features detected                                  780.9600
## Median features detected                                750.0000
## Scrublet - Number of doublets                           173.0000
## Scrublet - Percentage of doublets                         2.1300
## scDblFinder - Number of doublets                        356.0000
## scDblFinder - Percentage of doublets                      4.3800
## DoubletFinder - Number of doublets, Resolution 1.5      608.0000
## DoubletFinder - Percentage of doublets, Resolution 1.5    7.4900
## CXDS - Number of doublets                               426.0000
## CXDS - Percentage of doublets                             5.2500
## BCDS - Number of doublets                               408.0000
## BCDS - Percentage of doublets                             5.0300
## SCDS Hybrid - Number of doublets                        474.0000
## SCDS Hybrid - Percentage of doublets                      5.8400
## DecontX - Mean contamination                              0.0675
## DecontX - Median contamination                            0.0417</code></pre>
</div>
</div>
<div id="running-individual-qc-methods" class="section level3">
<h3 class="hasAnchor">
<a href="#running-individual-qc-methods" class="anchor"></a>Running individual QC methods</h3>
<p>Instead of running all quality control methods on the dataset at once, users may elect to execute QC methods individually. The parameters as well as the outputs to individual QC functions are described in detail as follows:</p>
<details><summary><b>General QC metrics</b></summary><details><summary><b>runPerCellQC</b></summary><p>SingleCellTK utilizes the <a href="https://bioconductor.org/packages/release/bioc/html/scater.html">scater</a> package to compute cell-level QC metrics. The wrapper function <code><a href="../../reference/runPerCellQC.html">runPerCellQC()</a></code> can be used to separately compute general QC metrics on its own.</p>
<ul>
<li>
<code>inSCE</code> parameter is the input SingleCellExperiment object.</li>
<li>
<code>useAssay</code> is the assay object that in the SingleCellExperiment object the user wishes to use.</li>
</ul>
<p>A list of gene sets can be applied to the function to determine the expression of a set of specific genes, as mentioned before. Please also refer to <a href="import_genesets.html"><em>Import Genesets</em> documentation</a>.</p>
<p>The QC outputs are <code>sum</code>, <code>detected</code>, and <code>percent_top_X</code>, stored as variables in <code>colData</code>.</p>
<ul>
<li>
<code>sum</code> contains the total number of counts for each cell.</li>
<li>
<code>detected</code> contains the total number of features for each cell.</li>
<li>
<code>percent_top_X</code> contains the percentage of the total counts that is made up by the expression of the top X genes for each cell.</li>
<li>The <code>subsets_</code> columns contain information for the specific gene list that was used. For instance, if a gene list containing mitochondrial genes named <code>"mito"</code> was used, <code>subsets_mito_sum</code> would contain the total number of mitochondrial counts for each cell.</li>
</ul>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pbmc.combined</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/runPerCellQC.html">runPerCellQC</a></span><span class="op">(</span>
  inSCE <span class="op">=</span> <span class="va">pbmc.combined</span>,
  useAssay <span class="op">=</span> <span class="st">"counts"</span>,
  collectionName <span class="op">=</span> <span class="st">"mito"</span><span class="op">)</span></code></pre></div>
  </details></details><details><summary><b>Doublet Detection</b></summary><p>Doublets hinder cell-type identification by appearing as a distinct transcriptomic state, and need to be removed for downstream analysis. SCTK contains various doublet detection tools that the user may choose from.</p>
  <details><summary><b>runScrublet</b></summary><p><a href="https://github.com/swolock/scrublet">Scrublet</a> aims to detect doublets by creating simulated doublets from combining transcriptomic profiles of existing cells in the dataset. The wrapper function <code><a href="../../reference/runScrublet.html">runScrublet()</a></code> can be used to separately run the Scrublet algorithm on its own.</p>
<ul>
<li>
<code>sample</code> indicates what sample each cell originates from. It can be set to <code>NULL</code> if all cells in the dataset came from the same sample.</li>
</ul>
<p>Scrublet also has a large set of parameters that the user can adjust, please see the function reference for detail, by clicking on the function name.</p>
<p>The Scrublet outputs include the following <code>colData</code> variables:</p>
<ul>
<li>
<code>scrublet_score</code>, which is a numeric variable of the likelihood that a cell is a doublet</li>
<li>
<code>scrublet_label</code>, which is the assignment of whether the cell is a doublet.</li>
</ul>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pbmc.combined</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/runScrublet.html">runScrublet</a></span><span class="op">(</span>
  inSCE <span class="op">=</span> <span class="va">pbmc.combined</span>,
  sample <span class="op">=</span> <span class="fu">colData</span><span class="op">(</span><span class="va">pbmc.combined</span><span class="op">)</span><span class="op">$</span><span class="va">sample</span>, 
  useAssay <span class="op">=</span> <span class="st">"counts"</span>
<span class="op">)</span></code></pre></div>
</details><details><summary><b>runScDblFinder</b></summary><p><a href="https://rdrr.io/bioc/scDblFinder/">ScDblFinder</a> is a doublet detection algorithm. ScDblFinder aims to detect doublets by creating a simulated doublet from existing cells and projecting it to the same PCA space as the cells. The wrapper function <code><a href="../../reference/runScDblFinder.html">runScDblFinder()</a></code> can be used to separately run the ScDblFinder algorithm on its own.</p>
<ul>
<li>
<code>nNeighbors</code> is the number of nearest neighbor used to calculate the density for doublet detection.</li>
<li>
<code>simDoublets</code> is used to determine the number of simulated doublets used for doublet detection.</li>
</ul>
<p>The output of ScDblFinder is a <code>scDblFinder_doublet_score</code>, which will be stored as a <code>colData</code> variable. The doublet score of a droplet will be higher if the it is deemed likely to be a doublet.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pbmc.combined</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/runScDblFinder.html">runScDblFinder</a></span><span class="op">(</span>inSCE <span class="op">=</span> <span class="va">pbmc.combined</span>, sample <span class="op">=</span> <span class="fu">colData</span><span class="op">(</span><span class="va">pbmc.combined</span><span class="op">)</span><span class="op">$</span><span class="va">sample</span>, useAssay <span class="op">=</span> <span class="st">"counts"</span><span class="op">)</span></code></pre></div>
</details><details><summary><b>runDoubletFinder</b></summary><p><a href="https://github.com/chris-mcginnis-ucsf/DoubletFinder">DoubletFinder</a> is a doublet detection algorithm which depends on the single cell analysis package <a href="https://satijalab.org/seurat/">Seurat</a>. The wrapper function <code><a href="../../reference/runDoubletFinder.html">runDoubletFinder()</a></code> can be used to separately run the DoubletFinder algorithm on its own.</p>
<ul>
<li>
<code>seuratRes</code> - <code><a href="../../reference/runDoubletFinder.html">runDoubletFinder()</a></code> relies on a parameter (in Seurat) called “resolution” to determine cells that may be doublets. Users will be able to manipulate the resolution parameter through <code>seuratRes</code>. If multiple numeric vectors are stored in <code>seuratRes</code>, there will be multiple label/scores.</li>
<li>
<code>seuratNfeatures</code> determines the number of features that is used in the <a href="https://satijalab.org/seurat/reference/findvariablefeatures"><code>FindVariableFeatures</code></a> function in Seurat.</li>
<li>
<code>seuratPcs</code> determines the number of dimensions used in the <a href="https://satijalab.org/seurat/reference/findneighbors"><code>FindNeighbors</code></a> function in Seurat.</li>
<li>
<code>formationRate</code> is the estimated doublet detection rate in the dataset. It aims to detect doublets by creating simulated doublets from combining transcriptomic profiles of existing cells in the dataset.</li>
</ul>
<p>The DoubletFinder outputs include the following <code>colData</code> variable:</p>
<ul>
<li>
<code>doubletFinder_doublet_score</code>, which is a numeric variable of the likelihood that a cell is a doublet</li>
<li>
<code>doubletFinder_doublet_label</code>, which is the assignment of whether the cell is a doublet.</li>
</ul>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pbmc.combined</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/runDoubletFinder.html">runDoubletFinder</a></span><span class="op">(</span>
  inSCE <span class="op">=</span> <span class="va">pbmc.combined</span>, useAssay <span class="op">=</span> <span class="st">"counts"</span>,
  sample <span class="op">=</span> <span class="fu">colData</span><span class="op">(</span><span class="va">pbmc.combined</span><span class="op">)</span><span class="op">$</span><span class="va">sample</span>, 
  seuratRes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1.0</span><span class="op">)</span>, seuratPcs <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">15</span>,
  seuratNfeatures <span class="op">=</span> <span class="fl">2000</span>,
  formationRate <span class="op">=</span> <span class="fl">0.075</span>, seed <span class="op">=</span> <span class="fl">12345</span>
<span class="op">)</span></code></pre></div>
</details><details><summary><b>runCXDS</b></summary><p>CXDS, or co-expression based doublet scoring, is an algorithm in the <a href="https://www.bioconductor.org/packages/release/bioc/html/scds.html">SCDS</a> package which employs a binomial model for the co-expression of pairs of genes to determine doublets. The wrapper function <code><a href="../../reference/runCxds.html">runCxds()</a></code> can be used to separately run the CXDS algorithm on its own.</p>
<ul>
<li>
<code>ntop</code> is the number of top variance genes to consider.</li>
<li>
<code>binThresh</code> is the minimum counts a gene needs to have to be included in the analysis.</li>
<li>
<code>verb</code> determines whether progress messages will be displayed or not.</li>
<li>
<code>retRes</code> will determine whether the gene pair results should be returned or not.</li>
<li>
<code>estNdbl</code> is the user estimated number of doublets.</li>
</ul>
<p>The output of <code><a href="../../reference/runCxds.html">runCxds()</a></code> is the doublet score, <code>scds_cxds_score</code>, which will be stored as a <code>colData</code> variable.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pbmc.combined</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/runCxds.html">runCxds</a></span><span class="op">(</span>
  inSCE <span class="op">=</span> <span class="va">pbmc.combined</span>, sample <span class="op">=</span> <span class="fu">colData</span><span class="op">(</span><span class="va">pbmc.combined</span><span class="op">)</span><span class="op">$</span><span class="va">sample</span>, 
  ntop <span class="op">=</span> <span class="fl">500</span>, binThresh <span class="op">=</span> <span class="fl">0</span>,
  verb <span class="op">=</span> <span class="cn">FALSE</span>, retRes <span class="op">=</span> <span class="cn">FALSE</span>, estNdbl <span class="op">=</span> <span class="cn">FALSE</span>
<span class="op">)</span></code></pre></div>
</details><details><summary><b>runBCDS</b></summary><p>BCDS, or binary classification based doublet scoring, is an algorithm in the <a href="https://www.bioconductor.org/packages/release/bioc/html/scds.html">SCDS</a> package which uses a binary classification approach to determine doublets. The wrapper function <code><a href="../../reference/runBcds.html">runBcds()</a></code> can be used to separately run the BCDS algorithm on its own.</p>
<ul>
<li>
<code>ntop</code> is the number of top variance genes to consider.</li>
<li>
<code>srat</code> is the ratio between original number of cells and simulated doublets.</li>
<li>
<code>nmax</code> is the maximum number of cycles that the algorithm should run through. If set to <code>"tune"</code>, this will be automatic.</li>
<li>
<code>varImp</code> determines if the variable importance should be returned or not.</li>
</ul>
<p>The output of <code><a href="../../reference/runBcds.html">runBcds()</a></code> is <code>scds_bcds_score</code>, which is the likelihood that a cell is a doublet and will be stored as a <code>colData</code> variable.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pbmc.combined</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/runBcds.html">runBcds</a></span><span class="op">(</span>
  inSCE <span class="op">=</span> <span class="va">pbmc.combined</span>, seed <span class="op">=</span> <span class="fl">12345</span>, sample <span class="op">=</span> <span class="fu">colData</span><span class="op">(</span><span class="va">pbmc.combined</span><span class="op">)</span><span class="op">$</span><span class="va">sample</span>, 
  ntop <span class="op">=</span> <span class="fl">500</span>, srat <span class="op">=</span> <span class="fl">1</span>, nmax <span class="op">=</span> <span class="st">"tune"</span>, varImp <span class="op">=</span> <span class="cn">FALSE</span>
<span class="op">)</span></code></pre></div>
</details><details><summary><b>runCxdsBcdsHybrid</b></summary><p>The CXDS-BCDS hybrid algorithm, uses both CXDS and BCDS algorithms from the <a href="https://www.bioconductor.org/packages/release/bioc/html/scds.html">SCDS</a> package. The wrapper function <code><a href="../../reference/runCxdsBcdsHybrid.html">runCxdsBcdsHybrid()</a></code> can be used to separately run the CXDS-BCDS hybrid algorithm on its own.</p>
<p>All parameters from the <code><a href="../../reference/runCxds.html">runCxds()</a></code> and <code><a href="../../reference/runBcds.html">runBcds()</a></code> functions may be applied to this function in the <code>cxdsArgs</code> and <code>bcdsArgs</code> parameters, respectively.</p>
<p>The output of <code><a href="../../reference/runCxdsBcdsHybrid.html">runCxdsBcdsHybrid()</a></code> is the doublet score, <code>scds_hybrid_score</code>, which will be stored as a <code>colData</code> variable.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pbmc.combined</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/runCxdsBcdsHybrid.html">runCxdsBcdsHybrid</a></span><span class="op">(</span>
  inSCE <span class="op">=</span> <span class="va">pbmc.combined</span>, sample <span class="op">=</span> <span class="fu">colData</span><span class="op">(</span><span class="va">pbmc.combined</span><span class="op">)</span><span class="op">$</span><span class="va">sample</span>, 
  seed <span class="op">=</span> <span class="fl">12345</span>, nTop <span class="op">=</span> <span class="fl">500</span>
<span class="op">)</span></code></pre></div>
  </details></details><details><summary><b>Ambient RNA Detection</b></summary><details><summary><b>runDecontX</b></summary><p>In droplet-based single cell technologies, ambient RNA that may have been released from apoptotic or damaged cells may get incorporated into another droplet, and can lead to contamination. <a href="https://rdrr.io/bioc/celda/man/decontX.html">decontX</a>, available from <a href="https://bioconductor.org/packages/release/bioc/html/celda.html">celda</a>, is a Bayesian method for the identification of the contamination level at a cellular level. The wrapper function <code><a href="../../reference/runDecontX.html">runDecontX()</a></code> can be used to separately run the DecontX algorithm on its own.</p>
<p>The outputs of <code><a href="../../reference/runDecontX.html">runDecontX()</a></code> are <code>decontX_contamination</code> and <code>decontX_clusters</code>.</p>
<ul>
<li>
<code>decontX_contamination</code> is a numeric vector which characterizes the level of contamination in each cell.</li>
<li>Clustering is performed as part of the <code><a href="../../reference/runDecontX.html">runDecontX()</a></code> algorithm. <code>decontX_clusters</code> is the resulting cluster assignment, which can also be labeled on the plot. For performing fine-tuned clustering in SCTK, please refer to <a href="clustering.html"><em>Clustering</em> documentation</a>
</li>
</ul>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>pbmc.combined <span class="ot">&lt;-</span> <span class="fu">runDecontX</span>(</span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">inSCE =</span> pbmc.combined, <span class="at">useAssay =</span> <span class="st">"counts"</span></span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">sample =</span> <span class="fu">colData</span>(pbmc.combined)<span class="sc">$</span>sample</span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div>
  </details></details>
</div>
<div id="plotting-qc-metrics" class="section level3">
<h3 class="hasAnchor">
<a href="#plotting-qc-metrics" class="anchor"></a>Plotting QC metrics</h3>
<p>Upon running <code><a href="../../reference/runCellQC.html">runCellQC()</a></code> or any individual QC methods, the QC outputs will need to be plotted. For each QC method, SCTK provides a specialized plotting function.</p>
<details><summary><b>General QC metrics</b></summary><details><summary><b>runPerCellQC</b></summary><p>The wrapper function <code><a href="../../reference/plotRunPerCellQCResults.html">plotRunPerCellQCResults()</a></code> can be used to plot the general QC outputs.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">runpercellqc.results</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/plotRunPerCellQCResults.html">plotRunPerCellQCResults</a></span><span class="op">(</span>inSCE <span class="op">=</span> <span class="va">pbmc.combined</span>, sample <span class="op">=</span> <span class="va">sample.vector</span>, combinePlot <span class="op">=</span> <span class="st">"all"</span>, axisSize <span class="op">=</span> <span class="fl">8</span>, axisLabelSize <span class="op">=</span> <span class="fl">9</span>, titleSize <span class="op">=</span> <span class="fl">20</span>, labelSamples<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">runpercellqc.results</span></code></pre></div>
<p><img src="cnsl_cellqc_files/figure-html/unnamed-chunk-3-1.png" width="960"></p>
  </details></details><details><summary><b>Doublet Detection</b></summary><details><summary><b>Scrublet</b></summary><p>The wrapper function <code><a href="../../reference/plotScrubletResults.html">plotScrubletResults()</a></code> can be used to plot the results from the Scrublet algorithm. Here, we will use the UMAP coordinates generated from <code><a href="../../reference/getUMAP.html">getUMAP()</a></code> in previous sections.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">reducedDims</span><span class="op">(</span><span class="va">pbmc.combined</span><span class="op">)</span></code></pre></div>
<pre><code>## List of length 6
## names(6): UMAP scrublet_TSNE ... decontX_pbmc3k_UMAP decontX_pbmc6k_UMAP</code></pre>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">scrublet.results</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/plotScrubletResults.html">plotScrubletResults</a></span><span class="op">(</span>
  inSCE <span class="op">=</span> <span class="va">pbmc.combined</span>,
  reducedDimName <span class="op">=</span> <span class="st">"UMAP"</span>,
  sample <span class="op">=</span> <span class="fu">colData</span><span class="op">(</span><span class="va">pbmc.combined</span><span class="op">)</span><span class="op">$</span><span class="va">sample</span>, 
  combinePlot <span class="op">=</span> <span class="st">"all"</span>,
  titleSize <span class="op">=</span> <span class="fl">10</span>,
  axisLabelSize <span class="op">=</span> <span class="fl">8</span>,
  axisSize <span class="op">=</span> <span class="fl">10</span>,
  legendSize <span class="op">=</span> <span class="fl">10</span>,
  legendTitleSize <span class="op">=</span> <span class="fl">10</span>
<span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">scrublet.results</span> </code></pre></div>
<p><img src="cnsl_cellqc_files/figure-html/plot_umap_scrublet-1.png" width="960"></p>
</details><details><summary><b>ScDblFinder</b></summary><p>The wrapper function <code><a href="../../reference/plotScDblFinderResults.html">plotScDblFinderResults()</a></code> can be used to plot the QC outputs from the ScDblFinder algorithm.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">scDblFinder.results</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/plotScDblFinderResults.html">plotScDblFinderResults</a></span><span class="op">(</span>
  inSCE <span class="op">=</span> <span class="va">pbmc.combined</span>, sample <span class="op">=</span> <span class="fu">colData</span><span class="op">(</span><span class="va">pbmc.combined</span><span class="op">)</span><span class="op">$</span><span class="va">sample</span>, 
  reducedDimName <span class="op">=</span> <span class="st">"UMAP"</span>, combinePlot <span class="op">=</span> <span class="st">"all"</span>,
  titleSize <span class="op">=</span> <span class="fl">13</span>,
  axisLabelSize <span class="op">=</span> <span class="fl">13</span>,
  axisSize <span class="op">=</span> <span class="fl">13</span>,
  legendSize <span class="op">=</span> <span class="fl">13</span>,
  legendTitleSize <span class="op">=</span> <span class="fl">13</span>
<span class="op">)</span></code></pre></div>
<p><br></p>
</details><details><summary><b>DoubletFinder</b></summary><p>The wrapper function <code><a href="../../reference/plotDoubletFinderResults.html">plotDoubletFinderResults()</a></code> can be used to plot the QC outputs from the DoubletFinder algorithm.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">doubletFinderResults</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/plotDoubletFinderResults.html">plotDoubletFinderResults</a></span><span class="op">(</span>
  inSCE <span class="op">=</span> <span class="va">pbmc.combined</span>,
  sample <span class="op">=</span> <span class="fu">colData</span><span class="op">(</span><span class="va">pbmc.combined</span><span class="op">)</span><span class="op">$</span><span class="va">sample</span>, 
  reducedDimName <span class="op">=</span> <span class="st">"UMAP"</span>,
  combinePlot <span class="op">=</span> <span class="st">"all"</span>,
  titleSize <span class="op">=</span> <span class="fl">13</span>,
  axisLabelSize <span class="op">=</span> <span class="fl">13</span>,
  axisSize <span class="op">=</span> <span class="fl">13</span>,
  legendSize <span class="op">=</span> <span class="fl">13</span>,
  legendTitleSize <span class="op">=</span> <span class="fl">13</span>
<span class="op">)</span></code></pre></div>
<p><br></p>
</details><details><summary><b>SCDS, CXDS</b></summary><p>The wrapper function <code><a href="../../reference/plotCxdsResults.html">plotCxdsResults()</a></code> can be used to plot the QC outputs from the CXDS algorithm.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">cxdsResults</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/plotCxdsResults.html">plotCxdsResults</a></span><span class="op">(</span>
  inSCE <span class="op">=</span> <span class="va">pbmc.combined</span>,
  sample <span class="op">=</span> <span class="fu">colData</span><span class="op">(</span><span class="va">pbmc.combined</span><span class="op">)</span><span class="op">$</span><span class="va">sample</span>, 
  reducedDimName <span class="op">=</span> <span class="st">"UMAP"</span>, combinePlot <span class="op">=</span> <span class="st">"all"</span>,
  titleSize <span class="op">=</span> <span class="fl">13</span>,
  axisLabelSize <span class="op">=</span> <span class="fl">13</span>,
  axisSize <span class="op">=</span> <span class="fl">13</span>,
  legendSize <span class="op">=</span> <span class="fl">13</span>,
  legendTitleSize <span class="op">=</span> <span class="fl">13</span>
<span class="op">)</span></code></pre></div>
<p><br></p>
</details><details><summary><b>SCDS, BCDS</b></summary><p>The wrapper function <code><a href="../../reference/plotBcdsResults.html">plotBcdsResults()</a></code> can be used to plot the QC outputs from the BCDS algorithm</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">bcdsResults</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/plotBcdsResults.html">plotBcdsResults</a></span><span class="op">(</span>
  inSCE <span class="op">=</span> <span class="va">pbmc.combined</span>,
  sample <span class="op">=</span> <span class="fu">colData</span><span class="op">(</span><span class="va">pbmc.combined</span><span class="op">)</span><span class="op">$</span><span class="va">sample</span>, 
  reducedDimName <span class="op">=</span> <span class="st">"UMAP"</span>, combinePlot <span class="op">=</span> <span class="st">"all"</span>,
  titleSize <span class="op">=</span> <span class="fl">13</span>,
  axisLabelSize <span class="op">=</span> <span class="fl">13</span>,
  axisSize <span class="op">=</span> <span class="fl">13</span>,
  legendSize <span class="op">=</span> <span class="fl">13</span>,
  legendTitleSize <span class="op">=</span> <span class="fl">13</span>
<span class="op">)</span></code></pre></div>
<p><br></p>
</details><details><summary><b>SCDS, CXDS-BCDS hybrid</b></summary><p>The wrapper function <code><a href="../../reference/plotScdsHybridResults.html">plotScdsHybridResults()</a></code> can be used to plot the QC outputs from the CXDS-BCDS hybrid algorithm.</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">bcdsCxdsHybridResults</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/plotScdsHybridResults.html">plotScdsHybridResults</a></span><span class="op">(</span>
  inSCE <span class="op">=</span> <span class="va">pbmc.combined</span>, sample <span class="op">=</span> <span class="fu">colData</span><span class="op">(</span><span class="va">pbmc.combined</span><span class="op">)</span><span class="op">$</span><span class="va">sample</span>, 
  reducedDimName <span class="op">=</span> <span class="st">"UMAP"</span>, combinePlot <span class="op">=</span> <span class="st">"all"</span>,
  titleSize <span class="op">=</span> <span class="fl">13</span>,
  axisLabelSize <span class="op">=</span> <span class="fl">13</span>,
  axisSize <span class="op">=</span> <span class="fl">13</span>,
  legendSize <span class="op">=</span> <span class="fl">13</span>,
  legendTitleSize <span class="op">=</span> <span class="fl">13</span>
<span class="op">)</span></code></pre></div>
<p><br></p>
</details></details><details><summary><b>Ambient RNA Detection</b></summary><details><summary><b>DecontX</b></summary><p>The wrapper function <code><a href="../../reference/plotDecontXResults.html">plotDecontXResults()</a></code> can be used to plot the QC outputs from the DecontX algorithm.</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">decontxResults</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/plotDecontXResults.html">plotDecontXResults</a></span><span class="op">(</span>
  inSCE <span class="op">=</span> <span class="va">pbmc.combined</span>, sample <span class="op">=</span> <span class="fu">colData</span><span class="op">(</span><span class="va">pbmc.combined</span><span class="op">)</span><span class="op">$</span><span class="va">sample</span>, 
  reducedDimName <span class="op">=</span> <span class="st">"UMAP"</span>, combinePlot <span class="op">=</span> <span class="st">"all"</span>,
  titleSize <span class="op">=</span> <span class="fl">8</span>,
  axisLabelSize <span class="op">=</span> <span class="fl">8</span>,
  axisSize <span class="op">=</span> <span class="fl">10</span>,
  legendSize <span class="op">=</span> <span class="fl">5</span>,
  legendTitleSize <span class="op">=</span> <span class="fl">7</span>,
  relWidths <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0.5</span>, <span class="fl">1</span>, <span class="fl">1</span><span class="op">)</span>,
  sampleRelWidths <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0.5</span>, <span class="fl">1</span>, <span class="fl">1</span><span class="op">)</span>,
  labelSamples <span class="op">=</span> <span class="cn">TRUE</span>,
  labelClusters <span class="op">=</span> <span class="cn">FALSE</span>
<span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">decontxResults</span></code></pre></div>
<p><img src="cnsl_cellqc_files/figure-html/plots_umap_decontx-1.png" width="864"></p>
  </details></details>
</div>
<div id="filtering-the-dataset" class="section level3">
<h3 class="hasAnchor">
<a href="#filtering-the-dataset" class="anchor"></a>Filtering the dataset</h3>
<p><code>SingleCellExperiment</code> objects can be subset by its <code>colData</code> using <code><a href="../../reference/subsetSCECols.html">subsetSCECols()</a></code>. The <code>colData</code> parameter takes in a character vector of expression(s) which will be used to identify a subset of cells using variables found in the <code>colData</code> of the <code>SingleCellExperiment</code> object. For example, if <code>x</code> is a numeric vector in <code>colData</code>, then setting <code>colData = "x &lt; 5"</code> will return a <code>SingleCellExperiment</code> object where all columns (cells) meet the condition that <code>x</code> is less than 5. The <code>index</code> parameter takes in a numeric vector of indices which should be kept, while <code>bool</code> takes in a logical vector of <code>TRUE</code> or <code>FALSE</code> which should be of the same length as the number of columns (cells) in the <code>SingleCellExperiment</code> object. Please refer to our <a href="filtering.html"><em>Filtering</em> documentation</a> for detail.</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">#Before filtering:</span>
<span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">pbmc.combined</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] 32738  8119</code></pre>
<p>Remove barcodes with high mitochondrial gene expression:</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pbmc.combined</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/subsetSCECols.html">subsetSCECols</a></span><span class="op">(</span><span class="va">pbmc.combined</span>, colData <span class="op">=</span> <span class="st">'subsets_mito_percent &lt; 20'</span><span class="op">)</span></code></pre></div>
<p>Remove detected doublets from Scrublet:</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pbmc.combined</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/subsetSCECols.html">subsetSCECols</a></span><span class="op">(</span><span class="va">pbmc.combined</span>, colData <span class="op">=</span> <span class="st">'scrublet_call == FALSE'</span><span class="op">)</span></code></pre></div>
<p>Remove cells with high levels of ambient RNA contamination:</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pbmc.combined</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/subsetSCECols.html">subsetSCECols</a></span><span class="op">(</span><span class="va">pbmc.combined</span>, colData <span class="op">=</span> <span class="st">'decontX_contamination &lt; 0.5'</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">#After filtering:</span>
<span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">pbmc.combined</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] 32738  7933</code></pre>
<p><br></p>
<blockquote>
<p>For performing QC on droplet-level raw count matrix with SCTK, please refer to our <a href="cnsl_dropletqc.html"><em>Droplet QC</em> documentation</a>.</p>
</blockquote>
 <details><summary><b>Session Information</b></summary><div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<pre><code>## R version 4.1.1 (2021-08-10)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Ubuntu 20.04.3 LTS
## 
## Matrix products: default
## BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.9.0
## LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.9.0
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8          LC_NUMERIC=C                 
##  [3] LC_TIME=en_US.UTF-8           LC_COLLATE=en_US.UTF-8       
##  [5] LC_MONETARY=en_US.UTF-8       LC_MESSAGES=en_US.UTF-8      
##  [7] LC_PAPER=en_US.UTF-8          LC_NAME=en_US.UTF-8          
##  [9] LC_ADDRESS=en_US.UTF-8        LC_TELEPHONE=en_US.UTF-8     
## [11] LC_MEASUREMENT=en_US.UTF-8    LC_IDENTIFICATION=en_US.UTF-8
## 
## attached base packages:
## [1] parallel  stats4    stats     graphics  grDevices utils     datasets 
## [8] methods   base     
## 
## other attached packages:
##  [1] TENxPBMCData_1.10.0         HDF5Array_1.20.0           
##  [3] rhdf5_2.36.0                dplyr_1.0.7                
##  [5] singleCellTK_2.3.2          DelayedArray_0.18.0        
##  [7] Matrix_1.3-4                SingleCellExperiment_1.14.1
##  [9] SummarizedExperiment_1.22.0 Biobase_2.52.0             
## [11] GenomicRanges_1.44.0        GenomeInfoDb_1.28.1        
## [13] IRanges_2.26.0              S4Vectors_0.30.0           
## [15] BiocGenerics_0.38.0         MatrixGenerics_1.4.0       
## [17] matrixStats_0.59.0         
## 
## loaded via a namespace (and not attached):
##   [1] MCMCprecision_0.4.0           rappdirs_0.3.3               
##   [3] scds_1.8.0                    scattermore_0.7              
##   [5] R.methodsS3_1.8.1             SeuratObject_4.0.2           
##   [7] ragg_1.1.3                    tidyr_1.1.3                  
##   [9] ggplot2_3.3.5                 bit64_4.0.5                  
##  [11] knitr_1.33                    irlba_2.3.3                  
##  [13] R.utils_2.10.1                data.table_1.14.0            
##  [15] rpart_4.1-15                  doParallel_1.0.16            
##  [17] KEGGREST_1.32.0               RCurl_1.98-1.3               
##  [19] generics_0.1.0                ScaledMatrix_1.0.0           
##  [21] cowplot_1.1.1                 RSQLite_2.2.7                
##  [23] combinat_0.0-8                RANN_2.6.1                   
##  [25] future_1.21.0                 bit_4.0.4                    
##  [27] webshot_0.5.2                 xml2_1.3.2                   
##  [29] spatstat.data_2.1-0           httpuv_1.6.1                 
##  [31] assertthat_0.2.1              viridis_0.6.1                
##  [33] xfun_0.24                     jquerylib_0.1.4              
##  [35] evaluate_0.14                 promises_1.2.0.1             
##  [37] fansi_0.5.0                   assertive.files_0.0-2        
##  [39] dbplyr_2.1.1                  igraph_1.2.6                 
##  [41] DBI_1.1.1                     htmlwidgets_1.5.3            
##  [43] spatstat.geom_2.2-2           purrr_0.3.4                  
##  [45] ellipsis_0.3.2                RSpectra_0.16-0              
##  [47] annotate_1.70.0               deldir_0.2-10                
##  [49] sparseMatrixStats_1.4.0       vctrs_0.3.8                  
##  [51] ROCR_1.0-11                   abind_1.4-5                  
##  [53] RcppEigen_0.3.3.9.1           cachem_1.0.5                 
##  [55] withr_2.4.2                   GSVAdata_1.28.0              
##  [57] sctransform_0.3.2             scran_1.20.1                 
##  [59] goftest_1.2-2                 svglite_2.0.0                
##  [61] cluster_2.1.2                 ExperimentHub_2.0.0          
##  [63] dotCall64_1.0-1               lazyeval_0.2.2               
##  [65] crayon_1.4.1                  labeling_0.4.2               
##  [67] edgeR_3.34.0                  pkgconfig_2.0.3              
##  [69] nlme_3.1-152                  vipor_0.4.5                  
##  [71] rlang_0.4.11                  globals_0.14.0               
##  [73] lifecycle_1.0.0               miniUI_0.1.1.1               
##  [75] filelock_1.0.2                dbscan_1.1-8                 
##  [77] BiocFileCache_2.0.0           enrichR_3.0                  
##  [79] rsvd_1.0.5                    AnnotationHub_3.0.1          
##  [81] rprojroot_2.0.2               polyclip_1.10-0              
##  [83] lmtest_0.9-38                 graph_1.70.0                 
##  [85] Rhdf5lib_1.14.2               zoo_1.8-9                    
##  [87] beeswarm_0.4.0                ggridges_0.5.3               
##  [89] rjson_0.2.20                  png_0.1-7                    
##  [91] viridisLite_0.4.0             bitops_1.0-7                 
##  [93] R.oo_1.24.0                   KernSmooth_2.23-20           
##  [95] spam_2.7-0                    rhdf5filters_1.4.0           
##  [97] pROC_1.17.0.1                 Biostrings_2.60.1            
##  [99] blob_1.2.1                    DelayedMatrixStats_1.14.0    
## [101] stringr_1.4.0                 parallelly_1.26.1            
## [103] gridGraphics_0.5-1            beachmat_2.8.0               
## [105] scales_1.1.1                  memoise_2.0.0                
## [107] GSEABase_1.54.0               magrittr_2.0.1               
## [109] plyr_1.8.6                    ica_1.0-2                    
## [111] zlibbioc_1.38.0               compiler_4.1.1               
## [113] kableExtra_1.3.4              dqrng_0.3.0                  
## [115] RColorBrewer_1.1-2            fitdistrplus_1.1-5           
## [117] XVector_0.32.0                listenv_0.8.0                
## [119] patchwork_1.1.1               pbapply_1.4-3                
## [121] MASS_7.3-54                   mgcv_1.8-36                  
## [123] tidyselect_1.1.1              stringi_1.7.2                
## [125] textshaping_0.3.5             highr_0.9                    
## [127] yaml_2.2.1                    assertive.numbers_0.0-2      
## [129] BiocSingular_1.8.1            locfit_1.5-9.4               
## [131] ggrepel_0.9.1                 grid_4.1.1                   
## [133] sass_0.4.0                    tools_4.1.1                  
## [135] future.apply_1.7.0            rstudioapi_0.13              
## [137] foreach_1.5.1                 bluster_1.2.1                
## [139] celda_1.8.1                   metapod_1.0.0                
## [141] gridExtra_2.3                 farver_2.1.0                 
## [143] assertive.types_0.0-3         Rtsne_0.15                   
## [145] DropletUtils_1.12.1           digest_0.6.27                
## [147] BiocManager_1.30.16           FNN_1.1.3                    
## [149] shiny_1.6.0                   Rcpp_1.0.7                   
## [151] scuttle_1.2.0                 BiocVersion_3.13.1           
## [153] later_1.2.0                   RcppAnnoy_0.0.18             
## [155] httr_1.4.2                    AnnotationDbi_1.54.1         
## [157] assertive.properties_0.0-4    colorspace_2.0-2             
## [159] rvest_1.0.0                   XML_3.99-0.6                 
## [161] fs_1.5.0                      tensor_1.5                   
## [163] reticulate_1.20               splines_4.1.1                
## [165] fields_12.5                   uwot_0.1.10                  
## [167] statmod_1.4.36                spatstat.utils_2.2-0         
## [169] pkgdown_1.6.1                 scater_1.20.1                
## [171] xgboost_1.4.1.1               plotly_4.9.4.1               
## [173] systemfonts_1.0.2             xtable_1.8-4                 
## [175] assertive.base_0.0-9          jsonlite_1.7.2               
## [177] R6_2.5.0                      pillar_1.6.1                 
## [179] htmltools_0.5.1.1             mime_0.11                    
## [181] glue_1.4.2                    fastmap_1.1.0                
## [183] BiocParallel_1.26.1           BiocNeighbors_1.10.0         
## [185] interactiveDisplayBase_1.30.0 codetools_0.2-18             
## [187] maps_3.3.0                    fishpond_1.8.0               
## [189] utf8_1.2.1                    lattice_0.20-44              
## [191] bslib_0.2.5.1                 spatstat.sparse_2.0-0        
## [193] tibble_3.1.2                  multipanelfigure_2.1.2       
## [195] curl_4.3.2                    ggbeeswarm_0.6.0             
## [197] leiden_0.3.8                  scDblFinder_1.6.0            
## [199] gtools_3.9.2                  magick_2.7.2                 
## [201] survival_3.2-13               limma_3.48.1                 
## [203] rmarkdown_2.9                 desc_1.3.0                   
## [205] munsell_0.5.0                 GenomeInfoDbData_1.2.6       
## [207] iterators_1.0.13              reshape2_1.4.4               
## [209] gtable_0.3.0                  spatstat.core_2.2-0          
## [211] Seurat_4.0.3</code></pre>
</details>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Yichen Wang, Irzam Sarfraz, Rui Hong, Yusuke Koga, Salam Alabdullatif, David Jenkins, Vidya Akavoor, Xinyun Cao, Shruthi Bandyadka, Anastasia Leshchyk, Tyler Faits, Mohammed Muzamil Khan, Zhe Wang, W. Evan Johnson, Joshua David Campbell.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
