<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Seurat Curated Workflow (Console) • singleCellTK</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/yeti/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../../bootstrap-toc.css">
<script src="../../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../../pkgdown.css" rel="stylesheet">
<script src="../../pkgdown.js"></script><meta property="og:title" content="Seurat Curated Workflow (Console)">
<meta property="og:description" content="singleCellTK">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../../index.html">singleCellTK</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">2.0.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../../articles/articles/introduction_to_SCTK.html">General Workflow</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Shell Command Line Analysis
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../../articles/articles/cmd_import_scRNAseq_data_as_SCE.html">Import data into SCTK</a>
    </li>
    <li>
      <a href="../../articles/articles/cmd_qc.html">Quality Control</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    R Console Analysis
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../../articles/articles/cnsl_import_scRNAseq_data_as_SCE.html">Import data into SCTK</a>
    </li>
    <li>
      <a href="../../articles/articles/cnsl_dropletqc.html">Quality Control of droplet data</a>
    </li>
    <li>
      <a href="../../articles/articles/cnsl_cellqc.html">Quality Control of cell data</a>
    </li>
    <li>
      <a href="../../articles/articles/cnsl_normalization.html">Normalization</a>
    </li>
    <li>
      <a href="../../articles/articles/cnsl_batch_correction.html">Batch Correction</a>
    </li>
    <li>
      <a href="../../articles/articles/cnsl_feature_selection.html">Feature Selection</a>
    </li>
    <li>
      <a href="../../articles/articles/cnsl_dimensionality_reduction.html">Dimensionality Reduction</a>
    </li>
    <li>
      <a href="../../articles/articles/cnsl_clustering.html">Clustering</a>
    </li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Differential Expression</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../../articles/articles/cnsl_differential_expression.html">Differential Expression</a>
        </li>
        <li>
          <a href="../../articles/articles/cnsl_find_marker.html">Find Marker via DE</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Enrichment Analysis</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../../articles/articles/cnsl_gsva.html">GSVA</a>
        </li>
        <li>
          <a href="../../articles/articles/cnsl_enrichR.html">EnrichR</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Visualization</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../../articles/articles/cnsl_visualization.html">General Visualization</a>
        </li>
        <li>
          <a href="../../articles/articles/cnsl_heatmap.html">Heatmap Visualization</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Curated Workflows</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../../articles/articles/cnsl_seurat_curated_workflow.html">Seurat</a>
        </li>
        <li>
          <a href="../../articles/articles/cnsl_celda_curated_workflow.html">Celda</a>
        </li>
      </ul>
</li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Shiny UI Analysis
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Data</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../../articles/articles/ui_import_data.html">Import Data</a>
        </li>
        <li>
          <a href="../../articles/articles/ui_import_genesets.html">Import Gene Sets</a>
        </li>
        <li>
          <a href="../../articles/articles/ui_import_annotation.html">Annotation</a>
        </li>
        <li>
          <a href="../../articles/articles/ui_export_data.html">Export Data</a>
        </li>
        <li>
          <a href="../../articles/articles/ui_delete_data.html">Delete Data</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">QC &amp; Filtering</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../../articles/articles/ui_qc.html">Quality Control</a>
        </li>
        <li>
          <a href="../../articles/articles/ui_filtering.html">Filtering</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Normalization &amp; Batch Correction</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../../articles/articles/ui_normalization.html">Normalization</a>
        </li>
        <li>
          <a href="../../articles/articles/ui_batch_correction.html">Batch Correction</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Feature Selection &amp; Dimensionality Reduction</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../../articles/articles/ui_feature_selection.html">Feature Selection</a>
        </li>
        <li>
          <a href="../../articles/articles/ui_dimensionality_reduction.html">Dimensionality Reduction</a>
        </li>
      </ul>
</li>
    <li>
      <a href="../../articles/articles/ui_clustering.html">Clustering</a>
    </li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Differential Expression &amp; Cell Type Labeling</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../../articles/articles/ui_differential_expression.html">Differential Expression</a>
        </li>
        <li>
          <a href="../../articles/articles/ui_find_marker.html">Find Marker</a>
        </li>
        <li>
          <a href="../../articles/articles/ui_cell_type_labeling.html">Cell Type Labeling</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Pathway &amp; Enrichment Analysis</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../../articles/articles/ui_gsva.html">GSVA</a>
        </li>
        <li>
          <a href="../../articles/articles/ui_enrichR.html">EnrichR</a>
        </li>
      </ul>
</li>
    <li>
      <a href="../../articles/articles/ui_sample_size_calc.html">Sample Size Calculator</a>
    </li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Curated Workflows</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../../articles/articles/ui_seurat_curated_workflow.html">Seurat</a>
        </li>
        <li>
          <a href="../../articles/articles/ui_celda_curated_workflow.html">Celda</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Visualization</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../../articles/articles/ui_cellViewer.html">cellViewer</a>
        </li>
        <li>
          <a href="../../articles/articles/ui_heatmap.html">Generic Heatmap</a>
        </li>
      </ul>
</li>
  </ul>
</li>
<li>
  <a href="../../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/compbiomed/singleCellTK">
    <span class="fas fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="cnsl_seurat_curated_workflow_files/header-attrs-2.7/header-attrs.js"></script><script src="cnsl_seurat_curated_workflow_files/htmlwidgets-1.5.3/htmlwidgets.js"></script><script src="cnsl_seurat_curated_workflow_files/plotly-binding-4.9.3/plotly.js"></script><script src="cnsl_seurat_curated_workflow_files/typedarray-0.1/typedarray.min.js"></script><script src="cnsl_seurat_curated_workflow_files/jquery-3.5.1/jquery.min.js"></script><link href="cnsl_seurat_curated_workflow_files/crosstalk-1.1.1/css/crosstalk.css" rel="stylesheet">
<script src="cnsl_seurat_curated_workflow_files/crosstalk-1.1.1/js/crosstalk.min.js"></script><link href="cnsl_seurat_curated_workflow_files/plotly-htmlwidgets-css-1.57.1/plotly-htmlwidgets.css" rel="stylesheet">
<script src="cnsl_seurat_curated_workflow_files/plotly-main-1.57.1/plotly-latest.min.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Seurat Curated Workflow (Console)</h1>
                        <h4 class="author">Irzam Sarfraz</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/compbiomed/singleCellTK/blob/master/vignettes/articles/cnsl_seurat_curated_workflow.Rmd"><code>vignettes/articles/cnsl_seurat_curated_workflow.Rmd</code></a></small>
      <div class="hidden name"><code>cnsl_seurat_curated_workflow.Rmd</code></div>

    </div>

    
    
<div id="introduction" class="section level2">
<h2 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h2>
<p><a href="https://satijalab.org/seurat/">Seurat</a> is an R package (<a href="https://www.nature.com/articles/nbt.4096">Butler et al., Nature Biotechnology 2018</a> &amp; <a href="https://www.cell.com/cell/fulltext/S0092-8674(19)30559-8">Stuart<em>, Butler</em>, et al., Cell 2019</a>) that offers various functions to perform analysis of scRNA-Seq data on the R console.</p>
<p>All methods provided by SCTK for Seurat workflow use a  object both as an input and output.</p>
<p>Using a sample dataset:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://compbiomed.github.io/sctk_docs/">singleCellTK</a></span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/importExampleData.html">importExampleData</a></span><span class="op">(</span><span class="st">'pbmc3k'</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sce</span></code></pre></div>
<pre><code>## class: SingleCellExperiment 
## dim: 32738 2700 
## metadata(0):
## assays(1): counts
## rownames(32738): MIR1302-10 FAM138A ... AC002321.2 AC002321.1
## rowData names(3): ENSEMBL_ID Symbol_TENx Symbol
## colnames(2700): pbmc3k_AAACATACAACCAC-1 pbmc3k_AAACATTGAGCTAC-1 ...
##   pbmc3k_TTTGCATGAGAGGC-1 pbmc3k_TTTGCATGCCTCAC-1
## colData names(12): Sample Barcode ... Date_published sample
## reducedDimNames(0):
## altExpNames(0):</code></pre>
</div>
<div id="normalize-data" class="section level2">
<h2 class="hasAnchor">
<a href="#normalize-data" class="anchor"></a>1. Normalize Data</h2>
<p>Once raw data is uploaded and stored in a SingleCellExperiment object, seuratNormalizeData function can be used to normalize the data. The method returns a SingleCellExperiment object with normalized data stored as a new assay in the input object.</p>
<p>Parameters to this function include useAssay (specify the assay that should be normalized), normAssayName (specify the new name of the normalized assay, defaults to “seuratNormData”), normalizationMethod (specify the normalization method to use, defaults to “LogNormalize”) and scaleFactor (defaults to 10000).</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/seuratNormalizeData.html">seuratNormalizeData</a></span><span class="op">(</span>inSCE <span class="op">=</span> <span class="va">sce</span>, useAssay <span class="op">=</span> <span class="st">"counts"</span>, normAssayName <span class="op">=</span> <span class="st">"seuratNormData"</span><span class="op">)</span></code></pre></div>
<pre><code>## Warning: replacing previous import 'spatstat.utils::RelevantNA' by
## 'spatstat.sparse::RelevantNA' when loading 'spatstat.core'</code></pre>
<pre><code>## Warning: replacing previous import 'spatstat.utils::grokIndexVector' by
## 'spatstat.sparse::grokIndexVector' when loading 'spatstat.core'</code></pre>
<pre><code>## Warning: replacing previous import 'spatstat.utils::fullIndexSequence' by
## 'spatstat.sparse::fullIndexSequence' when loading 'spatstat.core'</code></pre>
<pre><code>## Warning: replacing previous import 'spatstat.utils::positiveIndex' by
## 'spatstat.sparse::positiveIndex' when loading 'spatstat.core'</code></pre>
<pre><code>## Warning: replacing previous import 'spatstat.utils::replacementIndex' by
## 'spatstat.sparse::replacementIndex' when loading 'spatstat.core'</code></pre>
<pre><code>## Warning: replacing previous import 'spatstat.utils::RelevantEmpty' by
## 'spatstat.sparse::RelevantEmpty' when loading 'spatstat.core'</code></pre>
<pre><code>## Warning: replacing previous import 'spatstat.utils::isRelevantZero' by
## 'spatstat.sparse::isRelevantZero' when loading 'spatstat.core'</code></pre>
<pre><code>## Warning: replacing previous import 'spatstat.utils::logicalIndex' by
## 'spatstat.sparse::logicalIndex' when loading 'spatstat.core'</code></pre>
<pre><code>## Warning: replacing previous import 'spatstat.utils::RelevantZero' by
## 'spatstat.sparse::RelevantZero' when loading 'spatstat.core'</code></pre>
<pre><code>## Warning: replacing previous import 'spatstat.utils::strictIndexSequence' by
## 'spatstat.sparse::strictIndexSequence' when loading 'spatstat.core'</code></pre>
<pre><code>## Warning: Non-unique features (rownames) present in the input matrix, making
## unique</code></pre>
</div>
<div id="scale-data" class="section level2">
<h2 class="hasAnchor">
<a href="#scale-data" class="anchor"></a>2. Scale Data</h2>
<p>Normalized data can be scaled by using the seuratScaleData function that takes inputs a SingleCellExperiment object that has been normalized previously by the seuratNormalizeData function. Scaled assay is stored back in the input object.</p>
<p>Parameters include useAssay (specify the name of normalized assay), scaledAssayName (specify the new name for scaled assay, defaults to “seuratScaledData”), model (specify the method to use, defaults to “linear”), scale (specify if the data should be scaled, defaults to TRUE), center (specify if the data should be centered, defaults to TRUE) and scaleMax (specify the maximum clipping value, defaults to 10).</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/seuratScaleData.html">seuratScaleData</a></span><span class="op">(</span>inSCE <span class="op">=</span> <span class="va">sce</span>, useAssay <span class="op">=</span> <span class="st">"seuratNormData"</span>, scaledAssayName <span class="op">=</span> <span class="st">"seuratScaledData"</span><span class="op">)</span></code></pre></div>
<pre><code>## Warning: Non-unique features (rownames) present in the input matrix, making
## unique</code></pre>
<pre><code>## Centering and scaling data matrix</code></pre>
</div>
<div id="highly-variable-genes" class="section level2">
<h2 class="hasAnchor">
<a href="#highly-variable-genes" class="anchor"></a>3. Highly Variable Genes</h2>
<p>Highly variable genes can be identified by first using the seuratFindHVG function that computes that statistics against a selected HVG method in the rowData of input object. The genes can be identified by using the .seuratGetVariableFeatures function. The variable genes can be visualized using the seuratPlotHVG(vals$counts) method. Parameters for seuratFindHVG include useAssay (specify the name of the scaled assay, defaults to “seuratScaledData”) and hvgMethod (specify the method to use for variable genes computation, defaults to “vst”).</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/seuratFindHVG.html">seuratFindHVG</a></span><span class="op">(</span>inSCE <span class="op">=</span> <span class="va">sce</span>, useAssay <span class="op">=</span> <span class="st">"seuratScaledData"</span>, hvgMethod <span class="op">=</span> <span class="st">"vst"</span><span class="op">)</span></code></pre></div>
<pre><code>## Warning: Non-unique features (rownames) present in the input matrix, making
## unique</code></pre>
<pre><code>## Warning in eval(predvars, data, env): NaNs produced</code></pre>
<pre><code>## Warning in hvf.info$variance.expected[not.const] &lt;- 10^fit$fitted: number of
## items to replace is not a multiple of replacement length</code></pre>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="fu">singleCellTK</span><span class="fu">:::</span><span class="fu"><a href="../../reference/.seuratGetVariableFeatures.html">.seuratGetVariableFeatures</a></span><span class="op">(</span><span class="va">sce</span>, <span class="fl">100</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code>## Warning: Non-unique features (rownames) present in the input matrix, making
## unique</code></pre>
<pre><code>##   [1] "SREBF2"       "MIR4519"      "ANKS3"        "FAM122B"      "LSR"         
##   [6] "FCRL6"        "DNAJC17"      "HSDL2"        "HK3"          "SGSM3"       
##  [11] "INO80"        "VCPKMT"       "RPRD1A"       "TREM1"        "ING3"        
##  [16] "CSTA"         "IFRD2"        "CD83"         "KCMF1"        "ZCCHC6"      
##  [21] "LEMD2"        "RP11-244H3.4" "HAUS7"        "SLC25A1"      "KTI12"       
##  [26] "GPR137B"      "CBLL1"        "MDM1"         "SLC39A6"      "PLXNB2"      
##  [31] "LIPA"         "ANKRD36C"     "LONP2"        "TOMM40"       "SLC25A38"    
##  [36] "XIAP"         "RNF20"        "TFG"          "YIPF5"        "STYXL1"      
##  [41] "CLUAP1"       "PNPLA6"       "OSBPL9"       "TOB2"         "NRG1"        
##  [46] "PPAPDC1B"     "DNAJC9"       "CERK"         "PDLIM5"       "CXorf40A"    
##  [51] "NFRKB"        "CCDC65"       "XPNPEP3"      "PPIL3"        "MRPL17"      
##  [56] "CLASP2"       "UBXN7"        "RCN1"         "PAK1"         "XRCC1"       
##  [61] "OGFOD2"       "GMIP"         "IWS1"         "MAD2L1"       "STARD3NL"    
##  [66] "NUDT5"        "PPM1F"        "ACSS1"        "CD180"        "C14orf159"   
##  [71] "RASA2"        "FAM105A"      "ORAI2"        "ITSN2"        "LATS1"       
##  [76] "PTEN"         "PKN1"         "SCAF1"        "ZCCHC11"      "DNAJB11"     
##  [81] "WDR54"        "TMEM248"      "PPP1R35"      "NAPRT1"       "PRDX3"       
##  [86] "IL17RA"       "C21orf59"     "MRPS22"       "BCAP29"       "FAM160A2"    
##  [91] "NUP62"        "TMEM184B"     "RFWD2"        "GSK3B"        "ZNF354A"     
##  [96] "ITPK1"        "TM2D3"        "NFATC3"       "MAGED2"       "TRAF2"</code></pre>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../../reference/seuratPlotHVG.html">seuratPlotHVG</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></code></pre></div>
<pre><code>## Warning: Non-unique features (rownames) present in the input matrix, making
## unique</code></pre>
<pre><code>## Warning in self$trans$transform(x): NaNs produced</code></pre>
<pre><code>## Warning: Transformation introduced infinite values in continuous x-axis</code></pre>
<pre><code>## Warning: Removed 13847 rows containing missing values (geom_point).</code></pre>
<p><img src="cnsl_seurat_curated_workflow_files/figure-html/unnamed-chunk-6-1.png" width="700"></p>
</div>
<div id="dimensionality-reduction" class="section level2">
<h2 class="hasAnchor">
<a href="#dimensionality-reduction" class="anchor"></a>4. Dimensionality Reduction</h2>
<p>PCA or ICA can be computed using the seuratPCA and seuratICA functions respectively. Plots can be visualized using seuratReductionPlot, seuratElbowPlot, seuratComputeJackStraw, seuratJackStrawPlot, seuratComputeHeatmap.</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/seuratPCA.html">seuratPCA</a></span><span class="op">(</span>inSCE <span class="op">=</span> <span class="va">sce</span>, useAssay <span class="op">=</span> <span class="st">"seuratScaledData"</span>, reducedDimName <span class="op">=</span> <span class="st">"seuratPCA"</span>, nPCs <span class="op">=</span> <span class="fl">20</span><span class="op">)</span></code></pre></div>
<pre><code>## Warning: Non-unique features (rownames) present in the input matrix, making
## unique</code></pre>
<pre><code>## Warning in PrepDR(object = object, features = features, verbose = verbose): The
## following 4 features requested have not been scaled (running reduction without
## them): NDUFA7.1, PSMA2.1, CLN3.1, IER3IP1.1</code></pre>
<pre><code>## PC_ 1 
## Positive:  PTPRCAP, RPL10A, CD69, ACAP1, EVL, OCIAD2, LAT, TRAT1, AQP3, GIMAP5 
##     C12orf57, LEF1, MAL, SELL, SIT1, ARHGAP15, RPL22, CD8B, LBH, HSPA8 
##     MYC, LDLRAP1, CMPK1, PIM1, MZT2A, GIMAP7, TMEM123, SKAP1, STMN3, SVIP 
## Negative:  LST1, AIF1, LYZ, FCN1, S100A9, LGALS1, PSAP, S100A11, GRN, LILRA5 
##     CEBPB, RNF130, ANXA2, RAC1, FGR, IGSF6, HLA-DRA, CTSB, NUP214, FPR1 
##     MAFB, ALDH2, FGL2, RGS2, TGFBI, TALDO1, STX11, CSF3R, BLVRA, ATP6V0B 
## PC_ 2 
## Positive:  CST7, GNLY, GZMH, CCL4, KLRD1, CLIC3, ITGB2, ANXA1, RHOC, CD53 
##     HSPA8, GIMAP7, S1PR5, ID2, PLEKHF1, TPST2, LITAF, SH2D1B, XCL1, FCRL6 
##     ENO1, ZAP70, C1orf21, EVL, XBP1, CLIC1, CD63, PSMB9, KLRG1, GIMAP4 
## Negative:  HLA-DRA, LINC00926, VPREB3, S100A9, HLA-DQB1, HLA-DQA1, NCF1, HLA-DMA, CD79B, BANK1 
##     LYZ, PKIG, GNG7, S100A12, FCN1, MARCH1, PPAPDC1B, CD19, CD40, CSF3R 
##     MEF2C, SNX29P2, GAPT, CD22, NUP214, AIF1, ADAM28, IL8, FOSB, LST1 
## PC_ 3 
## Positive:  GIMAP7, ANXA1, GIMAP4, S100A9, S100A11, RGCC, GIMAP5, LEF1, TRAT1, AQP3 
##     MAL, LYZ, FCN1, S100A12, LAT, LGALS1, LDLRAP1, AIF1, CSF3R, IL8 
##     CD8B, C14orf64, RP11-664D1.1, NUP214, APBA2, NUCB2, BST1, PIM1, RGL4, CD44 
## Negative:  HLA-DQA1, CD79B, HLA-DQB1, LINC00926, HLA-DRA, VPREB3, BANK1, HLA-DMA, PKIG, CD40 
##     GNG7, CD19, PPAPDC1B, MEF2C, SNX2, MARCH1, CD22, ADAM28, CD180, SNX29P2 
##     STX7, LAT2, GAPT, PLAC8, TSPAN3, ORAI2, CD53, NCF1, ITM2C, PTPRCAP 
## PC_ 4 
## Positive:  RPL10A, RPL22, AQP3, C6orf48, MAL, LEF1, TMEM123, RGCC, TRAT1, CCDC109B 
##     CD44, HSPA8, ST13, MYC, CORO1B, RPL41, LDLRAP1, ANXA5, RP11-664D1.1, RCAN3 
##     TRADD, CRIP2, TSHZ2, ARHGAP15, MZT2A, CD69, TOMM20, CCNL1, EIF4B, ID3 
## Negative:  CST7, GNLY, CCL4, GZMH, KLRD1, CLIC3, S1PR5, C1orf21, XCL1, TPST2 
##     FCRL6, PLEKHF1, SH2D1B, CD63, ASCL2, KLRG1, NCR3, PLA2G16, SYNGR1, ID2 
##     ITGB2, ZBTB16, LPCAT1, YPEL1, AOAH, RGS3, CTSD, GNPTAB, FBXO6, CLIC1 
## PC_ 5 
## Positive:  CSF3R, ID1, NCF1, ALDH2, GRN, S100A12, S100A9, PLBD1, LYZ, NRG1 
##     BST1, NFKBIA, HMGB2, METTL9, CATSPER1, CSF2RA, IL8, NUP214, FCN1, CD33 
##     HLA-DMA, ITGAM, TMEM205, SYNGR1, BNIP3L, FPR1, TREM1, ARHGAP26, MEGF9, GM2A 
## Negative:  HES4, CKB, RHOC, CTSL, SIGLEC10, CXCL16, TCF7L2, UNC119, CUX1, SLC2A6 
##     CEBPB, ARRB1, LYST, DRAP1, SPN, NAAA, LILRA5, PHF19, ASAH1, CSTB 
##     SEC14L1, PCGF5, LST1, GPR137B, PLXNB2, PRADC1, PIK3AP1, RAB8A, CD300E, PTP4A2</code></pre>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../../reference/seuratReductionPlot.html">seuratReductionPlot</a></span><span class="op">(</span>inSCE <span class="op">=</span> <span class="va">sce</span>, useReduction <span class="op">=</span> <span class="st">"pca"</span><span class="op">)</span></code></pre></div>
<pre><code>## Warning: Non-unique features (rownames) present in the input matrix, making
## unique</code></pre>
<p><img src="cnsl_seurat_curated_workflow_files/figure-html/unnamed-chunk-7-1.png" width="700"></p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../../reference/seuratElbowPlot.html">seuratElbowPlot</a></span><span class="op">(</span>inSCE <span class="op">=</span> <span class="va">sce</span><span class="op">)</span></code></pre></div>
<pre><code>## Warning: Non-unique features (rownames) present in the input matrix, making
## unique</code></pre>
<div id="htmlwidget-344f8d8bc8cfa1ac530c" style="width:700px;height:432.632880098888px;" class="plotly html-widget"></div>
<script type="application/json" data-for="htmlwidget-344f8d8bc8cfa1ac530c">{"x":{"data":[{"x":[1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20],"y":[5.76021670451582,4.06389618799781,3.19290003430594,2.95648273106818,2.27491048170988,1.95821981187464,1.9084479843415,1.87443181515443,1.8492825938378,1.84543226650672,1.84004207660626,1.83535040452516,1.83053137282114,1.82329173764463,1.82138359536113,1.81641004565679,1.81164493612067,1.80695033853469,1.80279606763886,1.80127285848148],"text":["Dimension: 1 <br />Standard Deviation: 5.8 <br />Is Significant? ","Dimension: 2 <br />Standard Deviation: 4.1 <br />Is Significant? ","Dimension: 3 <br />Standard Deviation: 3.2 <br />Is Significant? ","Dimension: 4 <br />Standard Deviation: 3 <br />Is Significant? ","Dimension: 5 <br />Standard Deviation: 2.3 <br />Is Significant? ","Dimension: 6 <br />Standard Deviation: 2 <br />Is Significant? ","Dimension: 7 <br />Standard Deviation: 1.9 <br />Is Significant? ","Dimension: 8 <br />Standard Deviation: 1.9 <br />Is Significant? ","Dimension: 9 <br />Standard Deviation: 1.8 <br />Is Significant? ","Dimension: 10 <br />Standard Deviation: 1.8 <br />Is Significant? ","Dimension: 11 <br />Standard Deviation: 1.8 <br />Is Significant? ","Dimension: 12 <br />Standard Deviation: 1.8 <br />Is Significant? ","Dimension: 13 <br />Standard Deviation: 1.8 <br />Is Significant? ","Dimension: 14 <br />Standard Deviation: 1.8 <br />Is Significant? ","Dimension: 15 <br />Standard Deviation: 1.8 <br />Is Significant? ","Dimension: 16 <br />Standard Deviation: 1.8 <br />Is Significant? ","Dimension: 17 <br />Standard Deviation: 1.8 <br />Is Significant? ","Dimension: 18 <br />Standard Deviation: 1.8 <br />Is Significant? ","Dimension: 19 <br />Standard Deviation: 1.8 <br />Is Significant? ","Dimension: 20 <br />Standard Deviation: 1.8 <br />Is Significant? "],"type":"scatter","mode":"markers","marker":{"autocolorscale":false,"color":"rgba(0,0,0,1)","opacity":1,"size":5.66929133858268,"symbol":"circle","line":{"width":1.88976377952756,"color":"rgba(0,0,0,1)"}},"hoveron":"points","showlegend":false,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null}],"layout":{"margin":{"t":31.049537260322,"r":9.29846409298464,"b":54.2359216521858,"l":40.5147364051474},"font":{"color":"rgba(0,0,0,1)","family":"","size":18.5969281859693},"xaxis":{"domain":[0,1],"automargin":true,"type":"linear","autorange":false,"range":[0.0499999999999999,20.95],"tickmode":"array","ticktext":["5","10","15","20"],"tickvals":[5,10,15,20],"categoryorder":"array","categoryarray":["5","10","15","20"],"nticks":null,"ticks":"outside","tickcolor":"rgba(0,0,0,1)","ticklen":4.64923204649232,"tickwidth":0.66417600664176,"showticklabels":true,"tickfont":{"color":"rgba(0,0,0,1)","family":"","size":15.9402241594022},"tickangle":-0,"showline":true,"linecolor":"rgba(0,0,0,1)","linewidth":0.66417600664176,"showgrid":false,"gridcolor":null,"gridwidth":0,"zeroline":false,"anchor":"y","title":{"text":"PC","font":{"color":"rgba(0,0,0,1)","family":"","size":18.5969281859693}},"hoverformat":".2f"},"yaxis":{"domain":[0,1],"automargin":true,"type":"linear","autorange":false,"range":[1.60332566617976,5.95816389681754],"tickmode":"array","ticktext":["2","3","4","5"],"tickvals":[2,3,4,5],"categoryorder":"array","categoryarray":["2","3","4","5"],"nticks":null,"ticks":"outside","tickcolor":"rgba(0,0,0,1)","ticklen":4.64923204649232,"tickwidth":0.66417600664176,"showticklabels":true,"tickfont":{"color":"rgba(0,0,0,1)","family":"","size":15.9402241594022},"tickangle":-0,"showline":true,"linecolor":"rgba(0,0,0,1)","linewidth":0.66417600664176,"showgrid":false,"gridcolor":null,"gridwidth":0,"zeroline":false,"anchor":"x","title":{"text":"Standard Deviation","font":{"color":"rgba(0,0,0,1)","family":"","size":18.5969281859693}},"hoverformat":".2f"},"shapes":[{"type":"rect","fillcolor":null,"line":{"color":null,"width":0,"linetype":[]},"yref":"paper","xref":"paper","x0":0,"x1":1,"y0":0,"y1":1}],"showlegend":false,"legend":{"bgcolor":null,"bordercolor":null,"borderwidth":0,"font":{"color":"rgba(0,0,0,1)","family":"","size":15.9402241594022}},"hovermode":"closest","barmode":"relative"},"config":{"doubleClick":"reset","showSendToCloud":false},"source":"A","attrs":{"3fdc4b8211d4":{"x":{},"y":{},"type":"scatter"}},"cur_data":"3fdc4b8211d4","visdat":{"3fdc4b8211d4":["function (y) ","x"]},"highlight":{"on":"plotly_click","persistent":false,"dynamic":false,"selectize":false,"opacityDim":0.2,"selected":{"opacity":1},"debounce":0},"shinyEvents":["plotly_hover","plotly_click","plotly_selected","plotly_relayout","plotly_brushed","plotly_brushing","plotly_clickannotation","plotly_doubleclick","plotly_deselect","plotly_afterplot","plotly_sunburstclick"],"base_url":"https://plot.ly"},"evals":[],"jsHooks":[]}</script><div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/seuratComputeJackStraw.html">seuratComputeJackStraw</a></span><span class="op">(</span>inSCE <span class="op">=</span> <span class="va">sce</span>, useAssay <span class="op">=</span> <span class="st">"seuratScaledData"</span>, dims <span class="op">=</span> <span class="fl">20</span><span class="op">)</span></code></pre></div>
<pre><code>## Warning: Non-unique features (rownames) present in the input matrix, making
## unique</code></pre>
<pre><code>## Warning: Invalid name supplied, making object name syntactically valid. New
## object name is Seurat..JackStraw.RNA.pca; see ?make.names for more details on
## syntax validity</code></pre>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../../reference/seuratJackStrawPlot.html">seuratJackStrawPlot</a></span><span class="op">(</span>inSCE <span class="op">=</span> <span class="va">sce</span>, dims <span class="op">=</span> <span class="fl">20</span><span class="op">)</span></code></pre></div>
<pre><code>## Warning: Non-unique features (rownames) present in the input matrix, making
## unique</code></pre>
<pre><code>## Warning: Removed 33322 rows containing missing values (geom_point).</code></pre>
<p><img src="cnsl_seurat_curated_workflow_files/figure-html/unnamed-chunk-7-3.png" width="700"></p>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../../reference/seuratComputeHeatmap.html">seuratComputeHeatmap</a></span><span class="op">(</span>inSCE <span class="op">=</span> <span class="va">sce</span>, useAssay <span class="op">=</span> <span class="st">"seuratScaledData"</span>, useReduction <span class="op">=</span> <span class="st">"pca"</span>, nfeatures <span class="op">=</span> <span class="fl">20</span>, dims <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></code></pre></div>
<pre><code>## Warning: Non-unique features (rownames) present in the input matrix, making
## unique</code></pre>
<p><img src="cnsl_seurat_curated_workflow_files/figure-html/unnamed-chunk-7-4.png" width="700"></p>
</div>
<div id="tsneumap" class="section level2">
<h2 class="hasAnchor">
<a href="#tsneumap" class="anchor"></a>5. tSNE/UMAP</h2>
<p>seuratRunTSNE/seuratRunUMAP can be used to compute tSNE/UMAP statistics and store into the input object. Parameters to both functions include inSCE(input SCE object), useReduction (specify the reduction to use i.e. pca or ica), reducedDimName (name of this new reduction) and dims (number of dims to use). seuratReductionPlot can be used to visualize the results.</p>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/seuratRunTSNE.html">seuratRunTSNE</a></span><span class="op">(</span>inSCE <span class="op">=</span> <span class="va">sce</span>, useReduction <span class="op">=</span> <span class="st">"pca"</span>, reducedDimName <span class="op">=</span> <span class="st">"seuratTSNE"</span>, dims <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></code></pre></div>
<pre><code>## Warning: Non-unique features (rownames) present in the input matrix, making
## unique</code></pre>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../../reference/seuratReductionPlot.html">seuratReductionPlot</a></span><span class="op">(</span><span class="va">sce</span>, <span class="st">"tsne"</span><span class="op">)</span></code></pre></div>
<pre><code>## Warning: Non-unique features (rownames) present in the input matrix, making
## unique</code></pre>
<p><img src="cnsl_seurat_curated_workflow_files/figure-html/unnamed-chunk-8-1.png" width="700"></p>
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/seuratRunUMAP.html">seuratRunUMAP</a></span><span class="op">(</span>inSCE <span class="op">=</span> <span class="va">sce</span>, useReduction <span class="op">=</span> <span class="st">"pca"</span>, reducedDimName <span class="op">=</span> <span class="st">"seuratUMAP"</span>, dims <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></code></pre></div>
<pre><code>## Warning: Non-unique features (rownames) present in the input matrix, making
## unique</code></pre>
<pre><code>## Warning: The default method for RunUMAP has changed from calling Python UMAP via reticulate to the R-native UWOT using the cosine metric
## To use Python UMAP via reticulate, set umap.method to 'umap-learn' and metric to 'correlation'
## This message will be shown once per session</code></pre>
<div class="sourceCode" id="cb55"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../../reference/seuratReductionPlot.html">seuratReductionPlot</a></span><span class="op">(</span><span class="va">sce</span>, <span class="st">"umap"</span><span class="op">)</span></code></pre></div>
<pre><code>## Warning: Non-unique features (rownames) present in the input matrix, making
## unique</code></pre>
<p><img src="cnsl_seurat_curated_workflow_files/figure-html/unnamed-chunk-10-1.png" width="700"></p>
</div>
<div id="clustering" class="section level2">
<h2 class="hasAnchor">
<a href="#clustering" class="anchor"></a>6. Clustering</h2>
<p>seuratFindClusters function can be used to compute the clusters, which can later be plotted through the seuratReductionPlot method with cluster labels. The parameters to the function include inSCE (input SCE object), useAssay (name of the scaled assay), useReduction (specify which reduction to use i.e. pca or ica), dims (number of dims to use) and the algorithm (either louvain, multilevel or SLM).</p>
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/seuratFindClusters.html">seuratFindClusters</a></span><span class="op">(</span>inSCE <span class="op">=</span> <span class="va">sce</span>, useAssay <span class="op">=</span> <span class="st">"seuratScaledData"</span>, useReduction <span class="op">=</span> <span class="st">"pca"</span>, dims <span class="op">=</span> <span class="fl">10</span>, algorithm <span class="op">=</span> <span class="st">"louvain"</span><span class="op">)</span></code></pre></div>
<pre><code>## Warning: Non-unique features (rownames) present in the input matrix, making
## unique</code></pre>
<pre><code>## Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck
## 
## Number of nodes: 2700
## Number of edges: 98075
## 
## Running Louvain algorithm...
## Maximum modularity in 10 random starts: 0.7929
## Number of communities: 9
## Elapsed time: 0 seconds</code></pre>
<p>seuratReductionPlot can then be used to plot all reductions previously computed with cluster labels:</p>
<div class="sourceCode" id="cb60"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../../reference/seuratReductionPlot.html">seuratReductionPlot</a></span><span class="op">(</span><span class="va">sce</span>, <span class="st">"pca"</span><span class="op">)</span></code></pre></div>
<pre><code>## Warning: Non-unique features (rownames) present in the input matrix, making
## unique</code></pre>
<p><img src="cnsl_seurat_curated_workflow_files/figure-html/unnamed-chunk-12-1.png" width="700"></p>
<div class="sourceCode" id="cb62"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../../reference/seuratReductionPlot.html">seuratReductionPlot</a></span><span class="op">(</span><span class="va">sce</span>, <span class="st">"tsne"</span><span class="op">)</span></code></pre></div>
<pre><code>## Warning: Non-unique features (rownames) present in the input matrix, making
## unique</code></pre>
<p><img src="cnsl_seurat_curated_workflow_files/figure-html/unnamed-chunk-12-2.png" width="700"></p>
<div class="sourceCode" id="cb64"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../../reference/seuratReductionPlot.html">seuratReductionPlot</a></span><span class="op">(</span><span class="va">sce</span>, <span class="st">"umap"</span><span class="op">)</span></code></pre></div>
<pre><code>## Warning: Non-unique features (rownames) present in the input matrix, making
## unique</code></pre>
<p><img src="cnsl_seurat_curated_workflow_files/figure-html/unnamed-chunk-12-3.png" width="700"></p>
</div>
<div id="find-markers" class="section level2">
<h2 class="hasAnchor">
<a href="#find-markers" class="anchor"></a>7. Find Markers</h2>
<p>Marker genes can be identified using the seuratFindMarers function. This function can either use one specified column from colData of the input object as a group variable if all groups from that variable are to be used ( parameter) or users can manually specify the cells included in one group vs cells included in the second group ( and  parameter).</p>
<div class="sourceCode" id="cb66"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/seuratFindMarkers.html">seuratFindMarkers</a></span><span class="op">(</span>inSCE <span class="op">=</span> <span class="va">sce</span>, allGroup <span class="op">=</span> <span class="st">"Seurat_louvain_Resolution0.8"</span><span class="op">)</span></code></pre></div>
<pre><code>## Warning: Non-unique features (rownames) present in the input matrix, making
## unique</code></pre>
<div class="sourceCode" id="cb68"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="fu">metadata</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">[[</span><span class="st">"seuratMarkers"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code>##   gene.id         p_val avg_log2FC pct.1 pct.2     p_val_adj cluster
## 1    IL32 7.519719e-155 -35.841637 0.940 0.433 2.461806e-150       1
## 2     LTB 1.838828e-128  16.025684 0.963 0.619 6.019954e-124       1
## 3    CD3D 6.447790e-113 -17.042316 0.901 0.401 2.110878e-108       1
## 4    IL7R 1.823098e-112  -4.101946 0.743 0.295 5.968458e-108       1
## 5    LDHB  2.572198e-96  10.334754 0.930 0.589  8.420861e-92       1
## 6    CD3E  1.041419e-85  14.705010 0.803 0.380  3.409397e-81       1</code></pre>
<p>The marker genes identified can be visualized through one of the available plots from ridge plot, violin plot, feature plot, dot plot and heatmap plot. All marker genes visualizations can be plotted through the wrapper function , which must be supplied the SCE object (markers previously computed), name of the scaled assay, type of the plot (available options are , , ,  and ), features that should be plotted (character vector) and the grouping variable that is available in the colData slot of the input object. An additional parameter  decides in how many columns should the visualizations be plotted.</p>
<div class="sourceCode" id="cb70"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../../reference/seuratGenePlot.html">seuratGenePlot</a></span><span class="op">(</span>
        inSCE <span class="op">=</span> <span class="va">sce</span>,
        scaledAssayName <span class="op">=</span> <span class="st">"seuratScaledData"</span>,
        plotType <span class="op">=</span> <span class="st">"ridge"</span>,
        features <span class="op">=</span> <span class="fu">metadata</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">[[</span><span class="st">"seuratMarkers"</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">gene.id</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span>,
        groupVariable <span class="op">=</span> <span class="st">"Seurat_louvain_Resolution0.8"</span>,
        ncol <span class="op">=</span> <span class="fl">2</span>
      <span class="op">)</span></code></pre></div>
<pre><code>## Warning: Non-unique features (rownames) present in the input matrix, making
## unique</code></pre>
<pre><code>## Picking joint bandwidth of 0.374</code></pre>
<pre><code>## Picking joint bandwidth of 0.637</code></pre>
<pre><code>## Picking joint bandwidth of 0.282</code></pre>
<pre><code>## Picking joint bandwidth of 0.191</code></pre>
<p><img src="cnsl_seurat_curated_workflow_files/figure-html/unnamed-chunk-15-1.png" width="700"></p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by David Jenkins, Vidya Akavoor, Salam Alabdullatif, Shruthi Bandyadka, Emma Briars, Xinyun Cao, Sebastian Carrasco Pro, Tyler Faits, Rui Hong, Mohammed Muzamil Khan, Yusuke Koga, Anastasia Leshchyk, Irzam Sarfraz, Yichen Wang, Zhe Wang, W. Evan Johnson, Joshua David Campbell.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
