<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Quality Control in R console • singleCellTK</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/yeti/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../../bootstrap-toc.css">
<script src="../../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../../pkgdown.css" rel="stylesheet">
<script src="../../pkgdown.js"></script><link href="../../extra.css" rel="stylesheet">
<script src="../../extra.js"></script><meta property="og:title" content="Quality Control in R console">
<meta property="og:description" content="singleCellTK">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../../index.html">singleCellTK</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">2.2.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../../articles/articles/introduction_to_SCTK.html">General Workflow</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Shell Command Line Analysis
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../../articles/articles/cmd_import_scRNAseq_data_as_SCE.html">Import data into SCTK</a>
    </li>
    <li>
      <a href="../../articles/articles/cmd_qc.html">Quality Control</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    R Console Analysis
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../../articles/articles/cnsl_import_scRNAseq_data_as_SCE.html">Import data into SCTK</a>
    </li>
    <li>
      <a href="../../articles/articles/cnsl_dropletqc.html">Quality Control of droplet data</a>
    </li>
    <li>
      <a href="../../articles/articles/cnsl_cellqc.html">Quality Control of cell data</a>
    </li>
    <li>
      <a href="../../articles/articles/cnsl_normalization.html">Normalization</a>
    </li>
    <li>
      <a href="../../articles/articles/cnsl_batch_correction.html">Batch Correction</a>
    </li>
    <li>
      <a href="../../articles/articles/cnsl_feature_selection.html">Feature Selection</a>
    </li>
    <li>
      <a href="../../articles/articles/cnsl_dimensionality_reduction.html">Dimensionality Reduction</a>
    </li>
    <li>
      <a href="../../articles/articles/cnsl_clustering.html">Clustering</a>
    </li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Differential Expression</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../../articles/articles/cnsl_differential_expression.html">Differential Expression</a>
        </li>
        <li>
          <a href="../../articles/articles/cnsl_find_marker.html">Find Marker via DE</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Enrichment Analysis</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../../articles/articles/cnsl_gsva.html">GSVA</a>
        </li>
        <li>
          <a href="../../articles/articles/cnsl_enrichR.html">EnrichR</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Visualization</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../../articles/articles/cnsl_visualization.html">General Visualization</a>
        </li>
        <li>
          <a href="../../articles/articles/cnsl_heatmap.html">Heatmap Visualization</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Curated Workflows</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../../articles/articles/cnsl_seurat_curated_workflow.html">Seurat</a>
        </li>
        <li>
          <a href="../../articles/articles/cnsl_celda_curated_workflow.html">Celda</a>
        </li>
      </ul>
</li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Shiny UI Analysis
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Data</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../../articles/articles/ui_import_data.html">Import Data</a>
        </li>
        <li>
          <a href="../../articles/articles/ui_import_genesets.html">Import Gene Sets</a>
        </li>
        <li>
          <a href="../../articles/articles/ui_import_annotation.html">Annotation</a>
        </li>
        <li>
          <a href="../../articles/articles/ui_export_data.html">Export Data</a>
        </li>
        <li>
          <a href="../../articles/articles/ui_delete_data.html">Delete Data</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">QC &amp; Filtering</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../../articles/articles/ui_qc.html">Quality Control</a>
        </li>
        <li>
          <a href="../../articles/articles/ui_filtering.html">Filtering</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Normalization &amp; Batch Correction</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../../articles/articles/ui_normalization.html">Normalization</a>
        </li>
        <li>
          <a href="../../articles/articles/ui_batch_correction.html">Batch Correction</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Feature Selection &amp; Dimensionality Reduction</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../../articles/articles/ui_feature_selection.html">Feature Selection</a>
        </li>
        <li>
          <a href="../../articles/articles/ui_dimensionality_reduction.html">Dimensionality Reduction</a>
        </li>
      </ul>
</li>
    <li>
      <a href="../../articles/articles/ui_clustering.html">Clustering</a>
    </li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Differential Expression &amp; Cell Type Labeling</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../../articles/articles/ui_differential_expression.html">Differential Expression</a>
        </li>
        <li>
          <a href="../../articles/articles/ui_find_marker.html">Find Marker</a>
        </li>
        <li>
          <a href="../../articles/articles/ui_cell_type_labeling.html">Cell Type Labeling</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Pathway &amp; Enrichment Analysis</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../../articles/articles/ui_gsva.html">GSVA</a>
        </li>
        <li>
          <a href="../../articles/articles/ui_enrichR.html">EnrichR</a>
        </li>
      </ul>
</li>
    <li>
      <a href="../../articles/articles/ui_sample_size_calc.html">Sample Size Calculator</a>
    </li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Curated Workflows</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../../articles/articles/ui_seurat_curated_workflow.html">Seurat</a>
        </li>
        <li>
          <a href="../../articles/articles/ui_celda_curated_workflow.html">Celda</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Visualization</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../../articles/articles/ui_cellViewer.html">cellViewer</a>
        </li>
        <li>
          <a href="../../articles/articles/ui_heatmap.html">Generic Heatmap</a>
        </li>
      </ul>
</li>
  </ul>
</li>
<li>
  <a href="../../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/compbiomed/singleCellTK">
    <span class="fa fa-github"></span>
     
  </a>
</li>
<li>
  <a href="https://twitter.com/camplab1">
    <span class="fa fa-twitter"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><link href="cnsl_cellqc_files/anchor-sections-1.0/anchor-sections.css" rel="stylesheet">
<script src="cnsl_cellqc_files/anchor-sections-1.0/anchor-sections.js"></script><script src="cnsl_cellqc_files/kePrint-0.0.1/kePrint.js"></script><link href="cnsl_cellqc_files/lightable-0.0.1/lightable.css" rel="stylesheet">
<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Quality Control in R console</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/compbiomed/singleCellTK/blob/master/vignettes/articles/cnsl_cellqc.Rmd"><code>vignettes/articles/cnsl_cellqc.Rmd</code></a></small>
      <div class="hidden name"><code>cnsl_cellqc.Rmd</code></div>

    </div>

    
    
<style>
body {
text-align: justify}
</style>
<div id="introduction" class="section level2">
<h2 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h2>
<p>Performing comprehensive quality control (QC) is necessary to remove poor quality cells for downstream analysis of single-cell RNA sequencing (scRNA-seq) data. Within droplet-based scRNA-seq data, droplets containing cells must be differentiated from empty droplets. Therefore, assessment of the data is required, for which various QC algorithms have been developed. In singleCellTK, we have written convenience functions for several of these tools. In this guide, we will demonstrate how to use these functions to perform quality control on unfiltered, droplet-level data.</p>
<p>The package can be loaded using the <code>library</code> command.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://compbiomed.github.io/sctk_docs/">singleCellTK</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org">dplyr</a></span><span class="op">)</span></code></pre></div>
</div>
<div id="running-quality-control-of-cell-filtered-single-cell-data-with-singlecelltk" class="section level2">
<h2 class="hasAnchor">
<a href="#running-quality-control-of-cell-filtered-single-cell-data-with-singlecelltk" class="anchor"></a>Running quality control of cell-filtered single cell data with singleCellTK</h2>
<div id="load-pbmc-data-from-10x" class="section level3">
<h3 class="hasAnchor">
<a href="#load-pbmc-data-from-10x" class="anchor"></a>Load PBMC data from 10X</h3>
<p>We will use a filtered form of the PBMC 3K and 6K dataset from the package <a href="http://bioconductor.org/packages/release/data/experiment/html/TENxPBMCData.html">TENxPBMCData</a>, which is available from the <code>importExampleData</code> function. We will combine these datasets together into a single SingleCellExperiment object.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pbmc3k</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/importExampleData.html">importExampleData</a></span><span class="op">(</span>dataset <span class="op">=</span> <span class="st">"pbmc3k"</span><span class="op">)</span>
<span class="va">pbmc6k</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/importExampleData.html">importExampleData</a></span><span class="op">(</span>dataset <span class="op">=</span> <span class="st">"pbmc6k"</span><span class="op">)</span>

<span class="va">pbmc.combined</span> <span class="op">&lt;-</span> <span class="fu">BiocGenerics</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/cbind.html">cbind</a></span><span class="op">(</span><span class="va">pbmc3k</span>, <span class="va">pbmc6k</span><span class="op">)</span>
<span class="va">sample.vector</span> <span class="op">=</span> <span class="fu">colData</span><span class="op">(</span><span class="va">pbmc.combined</span><span class="op">)</span><span class="op">$</span><span class="va">sample</span></code></pre></div>
<p>singleCellTK also supports the importing of single-cell data from the following platforms: <a href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/what-is-cell-ranger">10X CellRanger</a>, <a href="https://github.com/alexdobin/STAR/blob/master/docs/STARsolo.md">STARSolo</a>, <a href="https://www.kallistobus.tools/">BUSTools</a>, <a href="https://github.com/ambrosejcarr/seqc">SEQC</a>, <a href="https://github.com/hms-dbmi/dropEst">DropEST</a>, and <a href="https://github.com/theislab/anndata">AnnData</a>. To load your own input data, please refer <a href="./import_scRNAseq_data_as_SCE_Console.html">Import data into SCTK</a> for deailed instruction.</p>
</div>
<div id="run-dimensionality-reduction" class="section level3">
<h3 class="hasAnchor">
<a href="#run-dimensionality-reduction" class="anchor"></a>Run Dimensionality Reduction</h3>
<p>SCTK utilizes dimensionality reduction techniques such as TSNE and UMAP for visualizing single-cell data. The user can modify the dimensions by adjusting the parameters within the function. The <code>logNorm</code> parameter should be set to TRUE for normalization prior to running dimensionality reduction.</p>
<p>The <code>sample</code> parameter may be specified if multiple samples exist in the SingleCellExperiment object. Here, we will use the sample vector stored in the colData of the SingleCellExperiment object.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># UMAP:</span>
<span class="va">pbmc.combined</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/getUMAP.html">getUMAP</a></span><span class="op">(</span>inSCE <span class="op">=</span> <span class="va">pbmc.combined</span>, useAssay <span class="op">=</span> <span class="st">"counts"</span>, logNorm <span class="op">=</span> <span class="cn">TRUE</span>, sample <span class="op">=</span> <span class="va">sample.vector</span><span class="op">)</span>

<span class="co"># TSNE (not run):</span>
<span class="co"># pbmc &lt;- getTSNE(inSCE=pbmc, useAssay="counts", logNorm = TRUE, sample = colData(pbmc)$sample)</span></code></pre></div>
</div>
<div id="runcellqc" class="section level3">
<h3 class="hasAnchor">
<a href="#runcellqc" class="anchor"></a>runCellQC</h3>
<p>All of the droplet-based QC algorithms are able to be run under the wrapper function <code>runCellQC</code>. By default all possible QC algorithms will be run.</p>
<p>Users may set a <code>sample</code> parameter if you would like to compare between multiple samples. Here, we will use the sample vector stored in the SingleCellExperiment object.</p>
<p>If the user wishes, a list of gene sets can be applied to the function to determine the expression of a set of specific genes. A gene list imported into the SingleCellExperiment object using <code>importGeneSets</code> functions can be set as <code>collectionName</code>. Additionally, a pre-made list of genes can be used to determine the level of gene expression per cell. A list containing gene symbols may be set as <code>geneSetList</code>, or the user may instead use the <code>geneSetCollection</code> parameter to supply a <code>GeneSetCollection</code> object from the <a href="https://bioconductor.org/packages/release/bioc/html/GSEABase.html">GSEABase</a> package.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pbmc.combined</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/importGeneSetsFromGMT.html">importGeneSetsFromGMT</a></span><span class="op">(</span>inSCE <span class="op">=</span> <span class="va">pbmc.combined</span>, collectionName <span class="op">=</span> <span class="st">"mito"</span>, file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata/mito_subset.gmt"</span>, package <span class="op">=</span> <span class="st">"singleCellTK"</span><span class="op">)</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">12345</span><span class="op">)</span>
<span class="va">pbmc.combined</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/runCellQC.html">runCellQC</a></span><span class="op">(</span><span class="va">pbmc.combined</span>, sample <span class="op">=</span> <span class="va">sample.vector</span>, collectionName <span class="op">=</span> <span class="st">"mito"</span><span class="op">)</span></code></pre></div>
<p><br> If users choose to only run a specific set of algorithms, they can specify which to run with the <code>algorithms</code> parameter.</p>
<p>When quality control functions are run in singleCellTK, the output of the function is stored in the <code>colData</code> slot of the <code>SingleCellExperiment</code> object.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="fu">colData</span><span class="op">(</span><span class="va">pbmc.combined</span><span class="op">)</span>, <span class="fl">5</span><span class="op">)</span></code></pre></div>
<div style="border: 1px solid #ddd; padding: 5px; overflow-x: scroll; width:80%; ">
<table class="table table">
<thead><tr>
<th style="text-align:left;">
</th>
<th style="text-align:left;">
Sample
</th>
<th style="text-align:left;">
Barcode
</th>
<th style="text-align:left;">
Sequence
</th>
<th style="text-align:right;">
Library
</th>
<th style="text-align:left;">
Cell_ranger_version
</th>
<th style="text-align:left;">
Tissue_status
</th>
<th style="text-align:left;">
Barcode_type
</th>
<th style="text-align:left;">
Chemistry
</th>
<th style="text-align:left;">
Sequence_platform
</th>
<th style="text-align:left;">
Individual
</th>
<th style="text-align:left;">
Date_published
</th>
<th style="text-align:left;">
sample
</th>
<th style="text-align:right;">
sum
</th>
<th style="text-align:right;">
detected
</th>
<th style="text-align:right;">
percent.top_50
</th>
<th style="text-align:right;">
percent.top_100
</th>
<th style="text-align:right;">
percent.top_200
</th>
<th style="text-align:right;">
percent.top_500
</th>
<th style="text-align:right;">
subsets_mito_sum
</th>
<th style="text-align:right;">
subsets_mito_detected
</th>
<th style="text-align:right;">
subsets_mito_percent
</th>
<th style="text-align:right;">
total
</th>
<th style="text-align:right;">
scrublet_score
</th>
<th style="text-align:left;">
scrublet_call
</th>
<th style="text-align:left;">
scDblFinder_sample
</th>
<th style="text-align:right;">
scDblFinder_cluster
</th>
<th style="text-align:left;">
scDblFinder_class
</th>
<th style="text-align:right;">
scDblFinder_doublet_score
</th>
<th style="text-align:right;">
scDblFinder_ratio
</th>
<th style="text-align:right;">
scDblFinder_weighted
</th>
<th style="text-align:left;">
scDblFinder_nearestClass
</th>
<th style="text-align:right;">
scDblFinder_difficulty
</th>
<th style="text-align:right;">
scDblFinder_cxds_score
</th>
<th style="text-align:left;">
scDblFinder_mostLikelyOrigin
</th>
<th style="text-align:left;">
scDblFinder_originAmbiguous
</th>
<th style="text-align:right;">
doubletFinder_doublet_score_resolution_1.5
</th>
<th style="text-align:left;">
doubletFinder_doublet_label_resolution_1.5
</th>
<th style="text-align:right;">
scds_cxds_score
</th>
<th style="text-align:left;">
scds_cxds_call
</th>
<th style="text-align:right;">
scds_bcds_score
</th>
<th style="text-align:left;">
scds_bcds_call
</th>
<th style="text-align:right;">
scds_hybrid_score
</th>
<th style="text-align:left;">
scds_hybrid_call
</th>
<th style="text-align:right;">
decontX_contamination
</th>
<th style="text-align:left;">
decontX_clusters
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;">
pbmc3k_AAACATACAACCAC-1
</td>
<td style="text-align:left;">
pbmc3k
</td>
<td style="text-align:left;">
AAACATACAACCAC-1
</td>
<td style="text-align:left;">
AAACATACAACCAC
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:left;">
v1.1.0
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:left;">
GemCode
</td>
<td style="text-align:left;">
Chromium_v1
</td>
<td style="text-align:left;">
NextSeq500
</td>
<td style="text-align:left;">
HealthyDonor2
</td>
<td style="text-align:left;">
2016-05-26
</td>
<td style="text-align:left;">
pbmc3k
</td>
<td style="text-align:right;">
2421
</td>
<td style="text-align:right;">
781
</td>
<td style="text-align:right;">
47.74886
</td>
<td style="text-align:right;">
63.27964
</td>
<td style="text-align:right;">
74.96902
</td>
<td style="text-align:right;">
88.39323
</td>
<td style="text-align:right;">
73
</td>
<td style="text-align:right;">
10
</td>
<td style="text-align:right;">
3.015283
</td>
<td style="text-align:right;">
2421
</td>
<td style="text-align:right;">
0.2394366
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
pbmc3k
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:left;">
singlet
</td>
<td style="text-align:right;">
0.5120444
</td>
<td style="text-align:right;">
0.86
</td>
<td style="text-align:right;">
0.8746153
</td>
<td style="text-align:left;">
artificialDoublet
</td>
<td style="text-align:right;">
0.1069034
</td>
<td style="text-align:right;">
0.0248409
</td>
<td style="text-align:left;">
1+2
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:right;">
0.0138889
</td>
<td style="text-align:left;">
Singlet
</td>
<td style="text-align:right;">
22794.14
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:right;">
0.0023788
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:right;">
0.208113
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:right;">
0.0331599
</td>
<td style="text-align:left;">
pbmc3k-1
</td>
</tr>
<tr>
<td style="text-align:left;">
pbmc3k_AAACATTGAGCTAC-1
</td>
<td style="text-align:left;">
pbmc3k
</td>
<td style="text-align:left;">
AAACATTGAGCTAC-1
</td>
<td style="text-align:left;">
AAACATTGAGCTAC
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:left;">
v1.1.0
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:left;">
GemCode
</td>
<td style="text-align:left;">
Chromium_v1
</td>
<td style="text-align:left;">
NextSeq500
</td>
<td style="text-align:left;">
HealthyDonor2
</td>
<td style="text-align:left;">
2016-05-26
</td>
<td style="text-align:left;">
pbmc3k
</td>
<td style="text-align:right;">
4903
</td>
<td style="text-align:right;">
1352
</td>
<td style="text-align:right;">
45.50275
</td>
<td style="text-align:right;">
61.02386
</td>
<td style="text-align:right;">
71.81318
</td>
<td style="text-align:right;">
82.62288
</td>
<td style="text-align:right;">
186
</td>
<td style="text-align:right;">
10
</td>
<td style="text-align:right;">
3.793596
</td>
<td style="text-align:right;">
4903
</td>
<td style="text-align:right;">
0.0935252
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
pbmc3k
</td>
<td style="text-align:right;">
4
</td>
<td style="text-align:left;">
singlet
</td>
<td style="text-align:right;">
0.0793516
</td>
<td style="text-align:right;">
0.52
</td>
<td style="text-align:right;">
0.5519875
</td>
<td style="text-align:left;">
artificialDoublet
</td>
<td style="text-align:right;">
0.0338731
</td>
<td style="text-align:right;">
0.1137545
</td>
<td style="text-align:left;">
1+4
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:right;">
0.1527778
</td>
<td style="text-align:left;">
Singlet
</td>
<td style="text-align:right;">
35689.05
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:right;">
0.8127224
</td>
<td style="text-align:left;">
TRUE
</td>
<td style="text-align:right;">
1.110518
</td>
<td style="text-align:left;">
TRUE
</td>
<td style="text-align:right;">
0.1390662
</td>
<td style="text-align:left;">
pbmc3k-2
</td>
</tr>
</tbody>
</table>
</div>
<p><br></p>
<p>A summary of all outputs is shown below:</p>
<table class="table">
<colgroup>
<col width="29%">
<col width="45%">
<col width="13%">
<col width="10%">
</colgroup>
<thead><tr class="header">
<th>QC output</th>
<th align="left">Description</th>
<th align="left">Methods</th>
<th align="left">Package/Tool</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>sum</td>
<td align="left">Total counts</td>
<td align="left">runPerCellQC</td>
<td align="left">scater</td>
</tr>
<tr class="even">
<td>detected</td>
<td align="left">Total features</td>
<td align="left">runPerCellQC</td>
<td align="left">scater</td>
</tr>
<tr class="odd">
<td>percent_top</td>
<td align="left">% Expression coming from top features</td>
<td align="left">runPerCellQC</td>
<td align="left">scater</td>
</tr>
<tr class="even">
<td>subsets_</td>
<td align="left">sum, detected, percent_top calculated on specified gene list</td>
<td align="left">runPerCellQC</td>
<td align="left">scater</td>
</tr>
<tr class="odd">
<td>scrublet_score</td>
<td align="left">Doublet score</td>
<td align="left">runScrublet</td>
<td align="left">scrublet</td>
</tr>
<tr class="even">
<td>scrublet_call</td>
<td align="left">Doublet classification based on threshold</td>
<td align="left">runScrublet</td>
<td align="left">scrublet</td>
</tr>
<tr class="odd">
<td>scDblFinder_doublet_score</td>
<td align="left">Doublet score</td>
<td align="left">runScDblFinder</td>
<td align="left">scran</td>
</tr>
<tr class="even">
<td>doubletFinder_doublet_score</td>
<td align="left">Doublet score</td>
<td align="left">runDoubletFinder</td>
<td align="left">DoubletFinder</td>
</tr>
<tr class="odd">
<td>doubletFinder_doublet_label_resolution</td>
<td align="left">Doublet classification based on threshold</td>
<td align="left">runDoubletFinder</td>
<td align="left">DoubletFinder</td>
</tr>
<tr class="even">
<td>scds_cxds_score</td>
<td align="left">Doublet score</td>
<td align="left">runCxds</td>
<td align="left">SCDS</td>
</tr>
<tr class="odd">
<td>scds_cxds_call</td>
<td align="left">Doublet classification based on threshold</td>
<td align="left">runCxds</td>
<td align="left">SCDS</td>
</tr>
<tr class="even">
<td>scds_bcds_score</td>
<td align="left">Doublet score</td>
<td align="left">runBcds</td>
<td align="left">SCDS</td>
</tr>
<tr class="odd">
<td>scds_bcds_call</td>
<td align="left">Doublet classification based on threshold</td>
<td align="left">runBcds</td>
<td align="left">SCDS</td>
</tr>
<tr class="even">
<td>scds_hybrid_score</td>
<td align="left">Doublet score</td>
<td align="left">runCxdsBcdsHybrid</td>
<td align="left">SCDS</td>
</tr>
<tr class="odd">
<td>scds_hybrid_call</td>
<td align="left">Doublet classification based on threshold</td>
<td align="left">runCxdsBcdsHybrid</td>
<td align="left">SCDS</td>
</tr>
<tr class="even">
<td>decontX_contamination</td>
<td align="left">Ambient RNA contamination</td>
<td align="left">runDecontX</td>
<td align="left">celda</td>
</tr>
<tr class="odd">
<td>decontX_clusters</td>
<td align="left">Clusters determined in dataset based on underlying algorithm</td>
<td align="left">runDecontX</td>
<td align="left">celda</td>
</tr>
</tbody>
</table>
<p>The names of the reducedDims of the <code>SingleCellExperiment</code> object are stored in the <code>reducedDims</code> slot.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">reducedDims</span><span class="op">(</span><span class="va">pbmc.combined</span><span class="op">)</span></code></pre></div>
<pre><code>## List of length 6
## names(6): UMAP scrublet_TSNE ... decontX_pbmc3k_UMAP decontX_pbmc6k_UMAP</code></pre>
<div id="generating-a-summary-statistic-table" class="section level4">
<h4 class="hasAnchor">
<a href="#generating-a-summary-statistic-table" class="anchor"></a>Generating a summary statistic table</h4>
<p>The function <code>sampleSummaryStats</code> may be used to generate a table containing the mean and median of the data per sample.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../../reference/sampleSummaryStats.html">sampleSummaryStats</a></span><span class="op">(</span><span class="va">pbmc.combined</span>, sample <span class="op">=</span> <span class="va">sample.vector</span><span class="op">)</span></code></pre></div>
<pre><code>##                           pbmc3k  pbmc6k   Total
## Number of Cells          2700.00 5419.00 8119.00
## Mean counts              2366.90 2027.60 2140.50
## Median counts            2197.00 1873.00 1988.00
## Mean features detected    846.99  748.06  780.96
## Median features detected  817.00  716.00  750.00</code></pre>
<p>If users choose to generate a table for all QC metrics generated through <code>runCellQC</code>, they may set the <code>simple</code> parameter to <code>FALSE</code>.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../../reference/sampleSummaryStats.html">sampleSummaryStats</a></span><span class="op">(</span><span class="va">pbmc.combined</span>, sample <span class="op">=</span> <span class="va">sample.vector</span>, simple <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></code></pre></div>
<pre><code>##                                                           pbmc3k    pbmc6k
## Number of Cells                                        2700.0000 5419.0000
## Mean counts                                            2366.9000 2027.6000
## Median counts                                          2197.0000 1873.0000
## Mean features detected                                  846.9900  748.0600
## Median features detected                                817.0000  716.0000
## Scrublet - Number of doublets                            44.0000  132.0000
## Scrublet - Percentage of doublets                         1.6300    2.4400
## scDblFinder - Number of doublets                         55.0000  255.0000
## scDblFinder - Percentage of doublets                      2.0400    4.7100
## DoubletFinder - Number of doublets, Resolution 1.5      202.0000  406.0000
## DoubletFinder - Percentage of doublets, Resolution 1.5    7.4800    7.4900
## CXDS - Number of doublets                               132.0000  294.0000
## CXDS - Percentage of doublets                             4.8900    5.4300
## BCDS - Number of doublets                               160.0000  278.0000
## BCDS - Percentage of doublets                             5.9300    5.1300
## SCDS Hybrid - Number of doublets                        180.0000  309.0000
## SCDS Hybrid - Percentage of doublets                      6.6700    5.7000
## DecontX - Mean contamination                              0.0809    0.0605
## DecontX - Median contamination                            0.0552    0.0365
##                                                            Total
## Number of Cells                                        8119.0000
## Mean counts                                            2140.5000
## Median counts                                          1988.0000
## Mean features detected                                  780.9600
## Median features detected                                750.0000
## Scrublet - Number of doublets                           176.0000
## Scrublet - Percentage of doublets                         2.1700
## scDblFinder - Number of doublets                        310.0000
## scDblFinder - Percentage of doublets                      3.8200
## DoubletFinder - Number of doublets, Resolution 1.5      608.0000
## DoubletFinder - Percentage of doublets, Resolution 1.5    7.4900
## CXDS - Number of doublets                               426.0000
## CXDS - Percentage of doublets                             5.2500
## BCDS - Number of doublets                               438.0000
## BCDS - Percentage of doublets                             5.3900
## SCDS Hybrid - Number of doublets                        489.0000
## SCDS Hybrid - Percentage of doublets                      6.0200
## DecontX - Mean contamination                              0.0673
## DecontX - Median contamination                            0.0417</code></pre>
<p><br></p>
</div>
</div>
<div id="plotting-qc-metrics" class="section level3">
<h3 class="hasAnchor">
<a href="#plotting-qc-metrics" class="anchor"></a>Plotting QC metrics</h3>
<p>Upon running runCellQC, the QC outputs will need to be plotted. For each QC method, singleCellTK contains specialized plotting functions that may be used for the generation of QC plots.</p>
<div id="general-qc-metrics" class="section level4">
<h4 class="hasAnchor">
<a href="#general-qc-metrics" class="anchor"></a>General QC metrics</h4>
<div id="runpercellqc" class="section level5">
<h5 class="hasAnchor">
<a href="#runpercellqc" class="anchor"></a><strong>runPerCellQC</strong>
</h5>
<p>SingleCellTK utilizes the <a href="https://bioconductor.org/packages/release/bioc/html/scater.html">scater</a> package to compute cell-level QC metrics.</p>
<p>The wrapper function plotRunPerCellQCResults can be used to plot the general QC outputs.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">runpercellqc.results</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/plotRunPerCellQCResults.html">plotRunPerCellQCResults</a></span><span class="op">(</span>inSCE <span class="op">=</span> <span class="va">pbmc.combined</span>, sample <span class="op">=</span> <span class="va">sample.vector</span>, combinePlot <span class="op">=</span> <span class="st">"all"</span>, axisSize <span class="op">=</span> <span class="fl">8</span>, axisLabelSize <span class="op">=</span> <span class="fl">9</span>, titleSize <span class="op">=</span> <span class="fl">20</span>, labelSamples<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">runpercellqc.results</span></code></pre></div>
<p><img src="cnsl_cellqc_files/figure-html/unnamed-chunk-2-1.png" width="960"></p>
</div>
</div>
<div id="doublet-detection" class="section level4">
<h4 class="hasAnchor">
<a href="#doublet-detection" class="anchor"></a>Doublet detection</h4>
<p>Doublets hinder cell-type identification by appearing as a distinct transcriptomic state, and need to be removed for downstream analysis. singleCellTK contains various doublet detection tools that the user may choose from.</p>
<div id="scrublet" class="section level5">
<h5 class="hasAnchor">
<a href="#scrublet" class="anchor"></a><strong>Scrublet</strong>
</h5>
<p><a href="https://github.com/swolock/scrublet/blob/master/examples/scrublet_basics.ipynb">Scrublet</a> aims to detect doublets by creating simulated doublets from combining transcriptomic profiles of existing cells in the dataset.</p>
<p>The wrapper function <code>plotScrubletResults</code> can be used to plot the results from the Scrublet algorithm.</p>
<p>The names of the reducedDims of the <code>SingleCellExperiment</code> object are stored in the <code>reducedDims</code> slot. Here, we will use the UMAP coordinates generated from <code>getUMAP</code> above.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">reducedDims</span><span class="op">(</span><span class="va">pbmc.combined</span><span class="op">)</span></code></pre></div>
<pre><code>## List of length 6
## names(6): UMAP scrublet_TSNE ... decontX_pbmc3k_UMAP decontX_pbmc6k_UMAP</code></pre>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">scrublet.results</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/plotScrubletResults.html">plotScrubletResults</a></span><span class="op">(</span>
  inSCE <span class="op">=</span> <span class="va">pbmc.combined</span>,
  reducedDimName <span class="op">=</span> <span class="st">"UMAP"</span>,
  sample <span class="op">=</span> <span class="fu">colData</span><span class="op">(</span><span class="va">pbmc.combined</span><span class="op">)</span><span class="op">$</span><span class="va">sample</span>, 
  combinePlot <span class="op">=</span> <span class="st">"all"</span>,
  titleSize <span class="op">=</span> <span class="fl">10</span>,
  axisLabelSize <span class="op">=</span> <span class="fl">8</span>,
  axisSize <span class="op">=</span> <span class="fl">10</span>,
  legendSize <span class="op">=</span> <span class="fl">10</span>,
  legendTitleSize <span class="op">=</span> <span class="fl">10</span>
<span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">scrublet.results</span> </code></pre></div>
<p><img src="cnsl_cellqc_files/figure-html/plot_umap_scrublet-1.png" width="960"></p>
</div>
<div id="scdblfinder" class="section level5">
<h5 class="hasAnchor">
<a href="#scdblfinder" class="anchor"></a><strong>ScDblFinder</strong>
</h5>
<p><a href="https://rdrr.io/bioc/scDblFinder/">ScDblFinder</a> is a doublet detection algorithm. ScDblFinder aims to detect doublets by creating a simulated doublet from existing cells and projecting it to the same PCA space as the cells.</p>
<p>The wrapper function plotScDblFinderResults can be used to plot the QC outputs from the ScDblFinder algorithm.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">scDblFinder.results</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/plotScDblFinderResults.html">plotScDblFinderResults</a></span><span class="op">(</span>
  inSCE <span class="op">=</span> <span class="va">pbmc.combined</span>, sample <span class="op">=</span> <span class="fu">colData</span><span class="op">(</span><span class="va">pbmc.combined</span><span class="op">)</span><span class="op">$</span><span class="va">sample</span>, 
  reducedDimName <span class="op">=</span> <span class="st">"UMAP"</span>, combinePlot <span class="op">=</span> <span class="st">"all"</span>,
  titleSize <span class="op">=</span> <span class="fl">13</span>,
  axisLabelSize <span class="op">=</span> <span class="fl">13</span>,
  axisSize <span class="op">=</span> <span class="fl">13</span>,
  legendSize <span class="op">=</span> <span class="fl">13</span>,
  legendTitleSize <span class="op">=</span> <span class="fl">13</span>
<span class="op">)</span></code></pre></div>
<p><br></p>
</div>
<div id="doubletfinder" class="section level5">
<h5 class="hasAnchor">
<a href="#doubletfinder" class="anchor"></a><strong>DoubletFinder</strong>
</h5>
<p><a href="https://github.com/chris-mcginnis-ucsf/DoubletFinder">DoubletFinder</a> is a doublet detection algorithm which depends on the single cell analysis package Seurat.</p>
<p>The wrapper function plotDoubletFinderResults can be used to plot the QC outputs from the DoubletFinder algorithm.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">doubletFinderResults</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/plotDoubletFinderResults.html">plotDoubletFinderResults</a></span><span class="op">(</span>
  inSCE <span class="op">=</span> <span class="va">pbmc.combined</span>,
  sample <span class="op">=</span> <span class="fu">colData</span><span class="op">(</span><span class="va">pbmc.combined</span><span class="op">)</span><span class="op">$</span><span class="va">sample</span>, 
  reducedDimName <span class="op">=</span> <span class="st">"UMAP"</span>,
  combinePlot <span class="op">=</span> <span class="st">"all"</span>,
  titleSize <span class="op">=</span> <span class="fl">13</span>,
  axisLabelSize <span class="op">=</span> <span class="fl">13</span>,
  axisSize <span class="op">=</span> <span class="fl">13</span>,
  legendSize <span class="op">=</span> <span class="fl">13</span>,
  legendTitleSize <span class="op">=</span> <span class="fl">13</span>
<span class="op">)</span></code></pre></div>
<p><br></p>
</div>
<div id="scds-cxds" class="section level5">
<h5 class="hasAnchor">
<a href="#scds-cxds" class="anchor"></a><strong>SCDS, CXDS</strong>
</h5>
<p>CXDS, or co-expression based doublet scoring, is an algorithm in the <a href="https://www.bioconductor.org/packages/release/bioc/html/scds.html">SCDS</a> package which employs a binomial model for the co-expression of pairs of genes to determine doublets.</p>
<p>The wrapper function <code>plotCxdsResults</code> can be used to plot the QC outputs from the CXDS algorithm.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">cxdsResults</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/plotCxdsResults.html">plotCxdsResults</a></span><span class="op">(</span>
  inSCE <span class="op">=</span> <span class="va">pbmc.combined</span>,
  sample <span class="op">=</span> <span class="fu">colData</span><span class="op">(</span><span class="va">pbmc.combined</span><span class="op">)</span><span class="op">$</span><span class="va">sample</span>, 
  reducedDimName <span class="op">=</span> <span class="st">"UMAP"</span>, combinePlot <span class="op">=</span> <span class="st">"all"</span>,
  titleSize <span class="op">=</span> <span class="fl">13</span>,
  axisLabelSize <span class="op">=</span> <span class="fl">13</span>,
  axisSize <span class="op">=</span> <span class="fl">13</span>,
  legendSize <span class="op">=</span> <span class="fl">13</span>,
  legendTitleSize <span class="op">=</span> <span class="fl">13</span>
<span class="op">)</span></code></pre></div>
<p><br></p>
</div>
<div id="scds-bcds" class="section level5">
<h5 class="hasAnchor">
<a href="#scds-bcds" class="anchor"></a><strong>SCDS, BCDS</strong>
</h5>
<p>BCDS, or binary classification based doublet scoring, is an algorithm in the <a href="https://www.bioconductor.org/packages/release/bioc/html/scds.html">SCDS</a> package which uses a binary classification approach to determine doublets.</p>
<p>The wrapper function plotBCDSResults can be used to plot the QC outputs from the BCDS algorithm</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">bcdsResults</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/plotBcdsResults.html">plotBcdsResults</a></span><span class="op">(</span>
  inSCE <span class="op">=</span> <span class="va">pbmc.combined</span>,
  sample <span class="op">=</span> <span class="fu">colData</span><span class="op">(</span><span class="va">pbmc.combined</span><span class="op">)</span><span class="op">$</span><span class="va">sample</span>, 
  reducedDimName <span class="op">=</span> <span class="st">"UMAP"</span>, combinePlot <span class="op">=</span> <span class="st">"all"</span>,
  titleSize <span class="op">=</span> <span class="fl">13</span>,
  axisLabelSize <span class="op">=</span> <span class="fl">13</span>,
  axisSize <span class="op">=</span> <span class="fl">13</span>,
  legendSize <span class="op">=</span> <span class="fl">13</span>,
  legendTitleSize <span class="op">=</span> <span class="fl">13</span>
<span class="op">)</span></code></pre></div>
<p><br></p>
</div>
<div id="scds-cxds-bcds-hybrid" class="section level5">
<h5 class="hasAnchor">
<a href="#scds-cxds-bcds-hybrid" class="anchor"></a><strong>SCDS, CXDS-BCDS hybrid</strong>
</h5>
<p>The CXDS-BCDS hybrid algorithm, uses both CXDS and BCDS algorithms from the <a href="https://www.bioconductor.org/packages/release/bioc/html/scds.html">SCDS</a> package.</p>
<p>The wrapper function plotScdsHybridResults can be used to plot the QC outputs from the CXDS-BCDS hybrid algorithm.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">bcdsCxdsHybridResults</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/plotScdsHybridResults.html">plotScdsHybridResults</a></span><span class="op">(</span>
  inSCE <span class="op">=</span> <span class="va">pbmc.combined</span>, sample <span class="op">=</span> <span class="fu">colData</span><span class="op">(</span><span class="va">pbmc.combined</span><span class="op">)</span><span class="op">$</span><span class="va">sample</span>, 
  reducedDimName <span class="op">=</span> <span class="st">"UMAP"</span>, combinePlot <span class="op">=</span> <span class="st">"all"</span>,
  titleSize <span class="op">=</span> <span class="fl">13</span>,
  axisLabelSize <span class="op">=</span> <span class="fl">13</span>,
  axisSize <span class="op">=</span> <span class="fl">13</span>,
  legendSize <span class="op">=</span> <span class="fl">13</span>,
  legendTitleSize <span class="op">=</span> <span class="fl">13</span>
<span class="op">)</span></code></pre></div>
<p><br></p>
</div>
</div>
<div id="ambient-rna-detection" class="section level4">
<h4 class="hasAnchor">
<a href="#ambient-rna-detection" class="anchor"></a>Ambient RNA detection</h4>
<div id="decontx" class="section level5">
<h5 class="hasAnchor">
<a href="#decontx" class="anchor"></a><strong>DecontX</strong>
</h5>
<p>In droplet-based single cell technologies, ambient RNA that may have been released from apoptotic or damaged cells may get incorporated into another droplet, and can lead to contamination. <a href="https://rdrr.io/bioc/celda/man/decontX.html">decontX</a>, available from the <a href="https://bioconductor.org/packages/release/bioc/html/celda.html">celda</a>, is a Bayesian method for the identification of the contamination level at a cellular level.</p>
<p>The wrapper function plotDecontXResults can be used to plot the QC outputs from the DecontX algorithm.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">decontxResults</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/plotDecontXResults.html">plotDecontXResults</a></span><span class="op">(</span>
  inSCE <span class="op">=</span> <span class="va">pbmc.combined</span>, sample <span class="op">=</span> <span class="fu">colData</span><span class="op">(</span><span class="va">pbmc.combined</span><span class="op">)</span><span class="op">$</span><span class="va">sample</span>, 
  reducedDimName <span class="op">=</span> <span class="st">"UMAP"</span>, combinePlot <span class="op">=</span> <span class="st">"all"</span>,
  titleSize <span class="op">=</span> <span class="fl">8</span>,
  axisLabelSize <span class="op">=</span> <span class="fl">8</span>,
  axisSize <span class="op">=</span> <span class="fl">10</span>,
  legendSize <span class="op">=</span> <span class="fl">5</span>,
  legendTitleSize <span class="op">=</span> <span class="fl">7</span>,
  relWidths <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0.5</span>, <span class="fl">1</span>, <span class="fl">1</span><span class="op">)</span>,
  sampleRelWidths <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0.5</span>, <span class="fl">1</span>, <span class="fl">1</span><span class="op">)</span>,
  labelSamples <span class="op">=</span> <span class="cn">TRUE</span>,
  labelClusters <span class="op">=</span> <span class="cn">FALSE</span>
<span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">decontxResults</span></code></pre></div>
<p><img src="cnsl_cellqc_files/figure-html/plots_umap_decontx-1.png" width="864"></p>
</div>
</div>
</div>
<div id="filtering-the-dataset" class="section level3">
<h3 class="hasAnchor">
<a href="#filtering-the-dataset" class="anchor"></a>Filtering the dataset</h3>
<p>SingleCellExperiment objects can be subset by its colData using subsetSCECols. The colData parameter takes in an expression in character vector form which will be used to identify a subset of columns using variables found in the colData of the SingleCellExperiment object. For example, if x is a numeric vector in colData, then “x &lt; 5” will return all columns with x less than 5. The index parameter takes in a vector of indices which should be kept, while bool takes in a vector of TRUE or FALSE which should be the same length as the number of columns in the SingleCellExperiment object.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">#Before filtering:</span>
<span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">pbmc.combined</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] 32738  8119</code></pre>
<p>Remove barcodes with high mitochondrial gene expression:</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pbmc.combined</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/subsetSCECols.html">subsetSCECols</a></span><span class="op">(</span><span class="va">pbmc.combined</span>, colData <span class="op">=</span> <span class="st">'subsets_mito_percent &lt; 20'</span><span class="op">)</span></code></pre></div>
<p>Remove detected doublets from Scrublet:</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pbmc.combined</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/subsetSCECols.html">subsetSCECols</a></span><span class="op">(</span><span class="va">pbmc.combined</span>, colData <span class="op">=</span> <span class="st">'scrublet_call == FALSE'</span><span class="op">)</span></code></pre></div>
<p>Remove cells with high levels of ambient RNA contamination:</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pbmc.combined</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/subsetSCECols.html">subsetSCECols</a></span><span class="op">(</span><span class="va">pbmc.combined</span>, colData <span class="op">=</span> <span class="st">'decontX_contamination &lt; 0.5'</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">#After filtering:</span>
<span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">pbmc.combined</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] 32738  7932</code></pre>
</div>
<div id="running-individual-qc-methods" class="section level3">
<h3 class="hasAnchor">
<a href="#running-individual-qc-methods" class="anchor"></a>Running individual QC methods</h3>
<p>Instead of running all quality control methods on the dataset at once, users may elect to execute QC methods individually. The parameters as well as the outputs to individual QC functions are described in detail as follows:</p>
<div id="runpercellqc-1" class="section level4">
<h4 class="hasAnchor">
<a href="#runpercellqc-1" class="anchor"></a>runPerCellQC</h4>
<p>The wrapper function <code>runPerCellQC</code> can be used to separately compute QC metrics on its own.</p>
<p>In this function, the <code>inSCE</code> parameter is the input SingleCellExperiment object, while the <code>useAssay</code> parameter is the assay object that in the SingleCellExperiment object the user wishes to use.</p>
<p>If the user wishes, a list of gene sets can be applied to the function to determine the expression of a set of specific genes. A gene list imported into the SingleCellExperiment object using <code>importGeneSets</code> functions can be set as <code>collectionName</code>. Additionally, a pre-made list of genes can be used to determine the level of gene expression per cell. A list containing gene symbols may be set as <code>geneSetList</code>, or the user may instead use the <code>geneSetCollection</code> parameter to supply a <code>GeneSetCollection</code> object from the <a href="https://bioconductor.org/packages/release/bioc/html/GSEABase.html">GSEABase</a> package.</p>
<p>The QC outputs are <code>sum</code>, <code>detected</code>, and <code>percent_top_X</code>.</p>
<p><code>sum</code> contains the total number of counts for each cell.</p>
<p><code>detected</code> contains the total number of features for each cell.</p>
<p><code>percent_top_X</code> contains the percentage of the total counts that is made up by the expression of the top X genes for each cell.</p>
<p>The <code>subsets_</code> columns contain information for the specific gene list that was used. For instance, if a gene list containing mitochondrial genes named <code>mito</code> was used, <code>subsets_mito_sum</code> would contains the total number of mitochondrial counts for each cell.</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pbmc.combined</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/runPerCellQC.html">runPerCellQC</a></span><span class="op">(</span>
  inSCE <span class="op">=</span> <span class="va">pbmc.combined</span>,
  useAssay <span class="op">=</span> <span class="st">"counts"</span>,
  collectionName <span class="op">=</span> <span class="st">"mito"</span><span class="op">)</span></code></pre></div>
</div>
<div id="runscrublet" class="section level4">
<h4 class="hasAnchor">
<a href="#runscrublet" class="anchor"></a>runScrublet</h4>
<p>The wrapper function <code>runScrublet</code> can be used to separately run the Scrublet algorithm on its own.</p>
<p>The <code>sample</code> parameter indicates what sample each cell originated from. It can be set to <code>NULL</code> if all cells in the dataset came from the same sample.</p>
<p>Scrublet also has a large set of parameters that the user can adjust; please refer to the Scrublet website for more details.</p>
<p>The Scrublet outputs are <code>scrublet_score</code>, which is a numeric variable of the likelihood that a cell is a doublet, and the <code>scrublet_label</code>, which is the assignment of whether the cell is a doublet.</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pbmc.combined</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/runScrublet.html">runScrublet</a></span><span class="op">(</span>
  inSCE <span class="op">=</span> <span class="va">pbmc.combined</span>,
  sample <span class="op">=</span> <span class="fu">colData</span><span class="op">(</span><span class="va">pbmc.combined</span><span class="op">)</span><span class="op">$</span><span class="va">sample</span>, 
  useAssay <span class="op">=</span> <span class="st">"counts"</span>
<span class="op">)</span></code></pre></div>
</div>
<div id="runscdblfinder" class="section level4">
<h4 class="hasAnchor">
<a href="#runscdblfinder" class="anchor"></a>runScDblFinder</h4>
<p>The wrapper function <code>runBarcodeRankDrops</code> can be used to separately run the ScDblFinder algorithm on its own. The <code>nNeighbors</code> parameter is the number of nearest neighbor used to calculate the density for doublet detection. <code>simDoublets</code> is used to determine the number of simulated doublets used for doublet detection.</p>
<p>The output of ScDblFinder is a <code>scDblFinder_doublet_score</code>. The doublet score of a droplet will be higher if the it is deemed likely to be a doublet.</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pbmc.combined</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/runScDblFinder.html">runScDblFinder</a></span><span class="op">(</span>inSCE <span class="op">=</span> <span class="va">pbmc.combined</span>, sample <span class="op">=</span> <span class="fu">colData</span><span class="op">(</span><span class="va">pbmc.combined</span><span class="op">)</span><span class="op">$</span><span class="va">sample</span>, useAssay <span class="op">=</span> <span class="st">"counts"</span><span class="op">)</span></code></pre></div>
</div>
<div id="rundoubletfinder" class="section level4">
<h4 class="hasAnchor">
<a href="#rundoubletfinder" class="anchor"></a>runDoubletFinder</h4>
<p>The wrapper function <code>runDoubletFinder</code> can be used to separately run the DoubletFinder algorithm on its own. <code>runDoubletFinder</code> relies on a parameter (in Seurat) called resolution to determine cells that may be doublets. Users will be able to manipulate the resolution parameter through <code>seuratRes</code>. If multiple numeric vectors are stored in <code>seuratRes</code>, there will be multiple label/scores. The <code>seuratNfeatures</code> parameter determines the number of features that is used in the <code>FindVariableFeatures</code> function in Seurat. <code>seuratPcs</code> parameter determines the number of dimensions used in the <code>FindNeighbors</code> function in Seurat. The <code>formationRate</code> parameter is the estimated doublet detection rate in the dataset. aims to detect doublets by creating simulated doublets from combining transcriptomic profiles of existing cells in the dataset.</p>
<p>The DoubletFinder outputs are <code>doubletFinder_doublet_score</code>, which is a numeric variable of the likelihood that a cell is a doublet, and the <code>doubletFinder_doublet_label</code>, which is the assignment of whether the cell is a doublet.</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pbmc.combined</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/runDoubletFinder.html">runDoubletFinder</a></span><span class="op">(</span>
  inSCE <span class="op">=</span> <span class="va">pbmc.combined</span>, useAssay <span class="op">=</span> <span class="st">"counts"</span>,
  sample <span class="op">=</span> <span class="fu">colData</span><span class="op">(</span><span class="va">pbmc.combined</span><span class="op">)</span><span class="op">$</span><span class="va">sample</span>, 
  seuratRes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1.0</span><span class="op">)</span>, seuratPcs <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">15</span>,
  seuratNfeatures <span class="op">=</span> <span class="fl">2000</span>,
  formationRate <span class="op">=</span> <span class="fl">0.075</span>, seed <span class="op">=</span> <span class="fl">12345</span>
<span class="op">)</span></code></pre></div>
</div>
<div id="runcxds" class="section level4">
<h4 class="hasAnchor">
<a href="#runcxds" class="anchor"></a>runCXDS</h4>
<p>The wrapper function <code>runCxds</code> can be used to separately run the CXDS algorithm on its own. In <code>runCxds</code>, the <code>ntop</code> parameter is the number of top variance genes to consider. The <code>binThresh</code> parameter is the minimum counts a gene needs to have to be included in the analysis. <code>verb</code> determines whether progress messages will be displayed or not. <code>retRes</code> will determine whether the gene pair results should be returned or not. The user may set the estimated number of doublets with estNdbl.</p>
<p>The output of <code>runCxds</code> is the doublet score, <code>scds_cxds_score</code>.</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pbmc.combined</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/runCxds.html">runCxds</a></span><span class="op">(</span>
  inSCE <span class="op">=</span> <span class="va">pbmc.combined</span>, sample <span class="op">=</span> <span class="fu">colData</span><span class="op">(</span><span class="va">pbmc.combined</span><span class="op">)</span><span class="op">$</span><span class="va">sample</span>, 
  ntop <span class="op">=</span> <span class="fl">500</span>, binThresh <span class="op">=</span> <span class="fl">0</span>,
  verb <span class="op">=</span> <span class="cn">FALSE</span>, retRes <span class="op">=</span> <span class="cn">FALSE</span>, estNdbl <span class="op">=</span> <span class="cn">FALSE</span>
<span class="op">)</span></code></pre></div>
</div>
<div id="runbcds" class="section level4">
<h4 class="hasAnchor">
<a href="#runbcds" class="anchor"></a>runBCDS</h4>
<p>The wrapper function <code>runBcds</code> can be used to separately run the BCDS algorithm on its own. In <code>runBcds</code>, the <code>ntop</code> parameter is the number of top variance genes to consider. The <code>srat</code> parameter is the ratio between original number of cells and simulated doublets. The <code>nmax</code> parameter is the maximum number of cycles that the algorithm should run through. If set to <code>tune</code>, this will be automatic. The <code>varImp</code> parameter determines if the variable importance should be returned or not.</p>
<p>The output of <code>runBcds</code> is scds_bcds_score, which is the likelihood that a cell is a doublet.</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pbmc.combined</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/runBcds.html">runBcds</a></span><span class="op">(</span>
  inSCE <span class="op">=</span> <span class="va">pbmc.combined</span>, seed <span class="op">=</span> <span class="fl">12345</span>, sample <span class="op">=</span> <span class="fu">colData</span><span class="op">(</span><span class="va">pbmc.combined</span><span class="op">)</span><span class="op">$</span><span class="va">sample</span>, 
  ntop <span class="op">=</span> <span class="fl">500</span>, srat <span class="op">=</span> <span class="fl">1</span>, nmax <span class="op">=</span> <span class="st">"tune"</span>, varImp <span class="op">=</span> <span class="cn">FALSE</span>
<span class="op">)</span></code></pre></div>
</div>
<div id="runcxdsbcdshybrid" class="section level4">
<h4 class="hasAnchor">
<a href="#runcxdsbcdshybrid" class="anchor"></a>runCxdsBcdsHybrid</h4>
<p>The wrapper function <code>runCxdsBcdsHybrid</code> can be used to separately run the CXDS-BCDS hybrid algorithm on its own.</p>
<p>All parameters from the <code>runBCDS</code> and <code>runBCDS</code> functions may be applied to this function in the cxdsArgs and bcdsArgs parameters, respectively. The output of <code>runCxdsBcdsHybrid</code> is the doublet score, <code>scds_hybrid_score</code>.</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pbmc.combined</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/runCxdsBcdsHybrid.html">runCxdsBcdsHybrid</a></span><span class="op">(</span>
  inSCE <span class="op">=</span> <span class="va">pbmc.combined</span>, sample <span class="op">=</span> <span class="fu">colData</span><span class="op">(</span><span class="va">pbmc.combined</span><span class="op">)</span><span class="op">$</span><span class="va">sample</span>, 
  seed <span class="op">=</span> <span class="fl">12345</span>, nTop <span class="op">=</span> <span class="fl">500</span>
<span class="op">)</span></code></pre></div>
</div>
<div id="rundecontx" class="section level4">
<h4 class="hasAnchor">
<a href="#rundecontx" class="anchor"></a>runDecontX</h4>
<p>The outputs of <code>runDecontX</code> are <code>decontX_contamination</code> and <code>decontX_clusters</code>.</p>
<p>The wrapper function <code>runDecontX</code> can be used to separately run the DecontX algorithm on its own. <code>decontX_contamination</code> is a numeric vector which characterizes the level of contamination in each cell. Clustering is performed as part of the <code>runDecontX</code> algorithm. <code>decontX_clusters</code> is the resulting cluster assignment, which can also be labeled on the plot.</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pbmc.combined</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/runDecontX.html">runDecontX</a></span><span class="op">(</span><span class="va">pbmc.combined</span>, sample <span class="op">=</span> <span class="fu">colData</span><span class="op">(</span><span class="va">pbmc.combined</span><span class="op">)</span><span class="op">$</span><span class="va">sample</span>, useAssay <span class="op">=</span> <span class="st">"counts"</span><span class="op">)</span></code></pre></div>
</div>
</div>
</div>
<div id="session-information" class="section level1">
<h1 class="hasAnchor">
<a href="#session-information" class="anchor"></a>Session Information</h1>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<pre><code>## R version 4.0.2 (2020-06-22)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: CentOS Linux 7 (Core)
## 
## Matrix products: default
## BLAS:   /share/pkg.7/r/4.0.2/install/lib/libopenblasp-r0.3.7.so
## LAPACK: /share/pkg.7/r/4.0.2/install/lib/liblapack.so.3.9.0
## 
## locale:
## [1] C
## 
## attached base packages:
## [1] parallel  stats4    stats     graphics  grDevices utils     datasets 
## [8] methods   base     
## 
## other attached packages:
##  [1] TENxPBMCData_1.8.0          HDF5Array_1.18.0           
##  [3] rhdf5_2.34.0                dplyr_1.0.2                
##  [5] singleCellTK_2.2.0          DelayedArray_0.16.0        
##  [7] Matrix_1.2-18               SingleCellExperiment_1.12.0
##  [9] SummarizedExperiment_1.20.0 Biobase_2.50.0             
## [11] GenomicRanges_1.42.0        GenomeInfoDb_1.26.0        
## [13] IRanges_2.24.0              S4Vectors_0.28.0           
## [15] BiocGenerics_0.36.0         MatrixGenerics_1.2.0       
## [17] matrixStats_0.57.0         
## 
## loaded via a namespace (and not attached):
##   [1] rappdirs_0.3.1                MCMCprecision_0.4.0          
##   [3] scds_1.6.0                    scattermore_0.7              
##   [5] R.methodsS3_1.8.1             SeuratObject_4.0.0           
##   [7] ragg_0.4.0                    tidyr_1.1.2                  
##   [9] ggplot2_3.3.2                 bit64_4.0.5                  
##  [11] knitr_1.30                    irlba_2.3.3                  
##  [13] R.utils_2.10.1                data.table_1.13.2            
##  [15] rpart_4.1-15                  RCurl_1.98-1.2               
##  [17] doParallel_1.0.16             generics_0.1.0               
##  [19] cowplot_1.1.0                 RSQLite_2.2.1                
##  [21] RANN_2.6.1                    combinat_0.0-8               
##  [23] future_1.20.1                 bit_4.0.4                    
##  [25] webshot_0.5.2                 xml2_1.3.2                   
##  [27] spatstat.data_2.0-0           httpuv_1.5.4                 
##  [29] assertthat_0.2.1              viridis_0.5.1                
##  [31] xfun_0.19                     evaluate_0.14                
##  [33] promises_1.1.1                assertive.files_0.0-2        
##  [35] dbplyr_2.0.0                  igraph_1.2.6                 
##  [37] DBI_1.1.0                     htmlwidgets_1.5.2            
##  [39] spatstat.geom_1.65-5          purrr_0.3.4                  
##  [41] ellipsis_0.3.1                RSpectra_0.16-0              
##  [43] annotate_1.68.0               deldir_0.2-3                 
##  [45] sparseMatrixStats_1.2.0       vctrs_0.3.5                  
##  [47] ROCR_1.0-11                   abind_1.4-5                  
##  [49] RcppEigen_0.3.3.7.0           withr_2.3.0                  
##  [51] GSVAdata_1.26.0               sctransform_0.3.2            
##  [53] scran_1.18.1                  goftest_1.2-2                
##  [55] cluster_2.1.0                 ExperimentHub_1.16.0         
##  [57] dotCall64_1.0-0               lazyeval_0.2.2               
##  [59] crayon_1.3.4                  labeling_0.4.2               
##  [61] edgeR_3.32.0                  pkgconfig_2.0.3              
##  [63] nlme_3.1-148                  vipor_0.4.5                  
##  [65] rlang_0.4.8                   globals_0.13.1               
##  [67] lifecycle_0.2.0               miniUI_0.1.1.1               
##  [69] dbscan_1.1-5                  BiocFileCache_1.14.0         
##  [71] enrichR_2.1                   rsvd_1.0.3                   
##  [73] AnnotationHub_2.22.0          rprojroot_2.0.2              
##  [75] polyclip_1.10-0               lmtest_0.9-38                
##  [77] graph_1.68.0                  Rhdf5lib_1.12.0              
##  [79] zoo_1.8-8                     beeswarm_0.2.3               
##  [81] ggridges_0.5.2                png_0.1-7                    
##  [83] viridisLite_0.3.0             rjson_0.2.20                 
##  [85] bitops_1.0-6                  R.oo_1.24.0                  
##  [87] KernSmooth_2.23-16            spam_2.5-1                   
##  [89] rhdf5filters_1.2.0            pROC_1.16.2                  
##  [91] blob_1.2.1                    DelayedMatrixStats_1.12.0    
##  [93] stringr_1.4.0                 parallelly_1.21.0            
##  [95] gridGraphics_0.5-0            beachmat_2.6.4               
##  [97] scales_1.1.1                  memoise_1.1.0                
##  [99] GSEABase_1.52.0               magrittr_2.0.1               
## [101] plyr_1.8.6                    ica_1.0-2                    
## [103] zlibbioc_1.36.0               compiler_4.0.2               
## [105] kableExtra_1.3.1              dqrng_0.2.1                  
## [107] RColorBrewer_1.1-2            fitdistrplus_1.1-1           
## [109] XVector_0.30.0                listenv_0.8.0                
## [111] patchwork_1.1.0               pbapply_1.4-3                
## [113] MASS_7.3-51.6                 mgcv_1.8-31                  
## [115] tidyselect_1.1.0              stringi_1.5.3                
## [117] textshaping_0.2.1             highr_0.8                    
## [119] yaml_2.2.1                    assertive.numbers_0.0-2      
## [121] BiocSingular_1.6.0            locfit_1.5-9.4               
## [123] ggrepel_0.8.2                 grid_4.0.2                   
## [125] tools_4.0.2                   future.apply_1.6.0           
## [127] rstudioapi_0.13               bluster_1.0.0                
## [129] foreach_1.5.1                 celda_1.7.7                  
## [131] gridExtra_2.3                 farver_2.0.3                 
## [133] assertive.types_0.0-3         Rtsne_0.15                   
## [135] DropletUtils_1.10.0           digest_0.6.27                
## [137] BiocManager_1.30.10           FNN_1.1.3                    
## [139] shiny_1.5.0                   Rcpp_1.0.6                   
## [141] scuttle_1.0.4                 BiocVersion_3.12.0           
## [143] later_1.2.0                   RcppAnnoy_0.0.18             
## [145] httr_1.4.2                    AnnotationDbi_1.52.0         
## [147] assertive.properties_0.0-4    colorspace_2.0-0             
## [149] rvest_0.3.6                   XML_3.99-0.5                 
## [151] fs_1.5.0                      tensor_1.5                   
## [153] reticulate_1.18               splines_4.0.2                
## [155] fields_11.6                   uwot_0.1.9                   
## [157] statmod_1.4.35                spatstat.utils_2.0-0         
## [159] pkgdown_1.6.1                 scater_1.18.3                
## [161] xgboost_1.2.0.1               plotly_4.9.2.1               
## [163] systemfonts_0.3.2             xtable_1.8-4                 
## [165] assertive.base_0.0-7          jsonlite_1.7.1               
## [167] R6_2.5.0                      pillar_1.4.6                 
## [169] htmltools_0.5.1.1             mime_0.9                     
## [171] glue_1.4.2                    fastmap_1.0.1                
## [173] BiocParallel_1.24.1           BiocNeighbors_1.8.1          
## [175] interactiveDisplayBase_1.28.0 codetools_0.2-16             
## [177] maps_3.3.0                    fishpond_1.6.0               
## [179] lattice_0.20-41               spatstat.sparse_1.2-1        
## [181] tibble_3.0.4                  multipanelfigure_2.1.2       
## [183] curl_4.3                      ggbeeswarm_0.6.0             
## [185] leiden_0.3.5                  scDblFinder_1.4.0            
## [187] gtools_3.8.2                  magick_2.5.0                 
## [189] survival_3.1-12               limma_3.46.0                 
## [191] rmarkdown_2.5                 desc_1.2.0                   
## [193] munsell_0.5.0                 GenomeInfoDbData_1.2.4       
## [195] iterators_1.0.13              reshape2_1.4.4               
## [197] gtable_0.3.0                  spatstat.core_1.65-5         
## [199] Seurat_4.0.1</code></pre>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by David Jenkins, Vidya Akavoor, Salam Alabdullatif, Shruthi Bandyadka, Emma Briars, Xinyun Cao, Sebastian Carrasco Pro, Tyler Faits, Rui Hong, Mohammed Muzamil Khan, Yusuke Koga, Anastasia Leshchyk, Irzam Sarfraz, Yichen Wang, Zhe Wang, W. Evan Johnson, Joshua David Campbell.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
