<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Quality Control in R console • singleCellTK</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/yeti/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../../bootstrap-toc.css">
<script src="../../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../../pkgdown.css" rel="stylesheet">
<script src="../../pkgdown.js"></script><link href="../../extra.css" rel="stylesheet">
<script src="../../extra.js"></script><meta property="og:title" content="Quality Control in R console">
<meta property="og:description" content="singleCellTK">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../../index.html">singleCellTK</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">2.2.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../../articles/articles/installation.html">Installation</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Documentation
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Data</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../../articles/articles/import_data.html">Import Data</a>
        </li>
        <li>
          <a href="../../articles/articles/cmd_import_scRNAseq_data_as_SCE.html">Import Data in Commandline</a>
        </li>
        <li>
          <a href="../../articles/articles/import_genesets.html">Import Gene Sets</a>
        </li>
        <li>
          <a href="../../articles/articles/import_annotation.html">Annotation</a>
        </li>
        <li>
          <a href="../../articles/articles/export_data.html">Export Data</a>
        </li>
        <li>
          <a href="../../articles/articles/delete_data.html">Delete Data</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">QC &amp; Filtering</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../../articles/articles/cmd_qc.html">QC with commandline</a>
        </li>
        <li>
          <a href="../../articles/articles/cnsl_cellqc.html">Cell QC in R console</a>
        </li>
        <li>
          <a href="../../articles/articles/cnsl_dropletqc.html">Droplet QC in R console</a>
        </li>
        <li>
          <a href="../../articles/articles/ui_qc.html">QC in the UI</a>
        </li>
        <li>
          <a href="../../articles/articles/filtering.html">Filtering</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Normalization &amp; Batch Correction</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../../articles/articles/cnsl_normalization.html">Normalization</a>
        </li>
        <li>
          <a href="../../articles/articles/batch_correction.html">Batch Correction</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Feature Selection &amp; Dimensionality Reduction</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../../articles/articles/cnsl_feature_selection.html">Feature Selection</a>
        </li>
        <li>
          <a href="../../articles/articles/cnsl_dimensionality_reduction.html">Dimensionality Reduction</a>
        </li>
        <li>
          <a href="../../articles/articles/cnsl_2d_embedding.html">2D Embedding</a>
        </li>
      </ul>
</li>
    <li>
      <a href="../../articles/articles/clustering.html">Clustering</a>
    </li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Differential Expression &amp; Cell Type Labeling</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../../articles/articles/differential_expression.html">Differential Expression</a>
        </li>
        <li>
          <a href="../../articles/articles/find_marker.html">Find Marker</a>
        </li>
        <li>
          <a href="../../articles/articles/cell_type_labeling.html">Cell Type Labeling</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Pathway &amp; Enrichment Analysis</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../../articles/articles/cnsl_enrichR.html">EnrichR</a>
        </li>
        <li>
          <a href="../../articles/articles/pathwayAnalysis.html">Pathway Analysis</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Curated Workflows</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../../articles/articles/cnsl_seurat_curated_workflow.html">Seurat</a>
        </li>
        <li>
          <a href="../../articles/articles/celda_curated_workflow.html">Celda</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Visualization</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../../articles/articles/visualization.html">General Visualization</a>
        </li>
        <li>
          <a href="../../articles/articles/heatmap.html">Generic Heatmap</a>
        </li>
      </ul>
</li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Tutorial
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../../articles/articles/console_analysis_tutorial.html">Console Analysis Tutorial</a>
    </li>
  </ul>
</li>
<li>
  <a href="../../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/compbiomed/singleCellTK">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="https://twitter.com/camplab1">
    <span class="fab fa-twitter fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="cnsl_dropletqc_files/header-attrs-2.7/header-attrs.js"></script><script src="cnsl_dropletqc_files/kePrint-0.0.1/kePrint.js"></script><link href="cnsl_dropletqc_files/lightable-0.0.1/lightable.css" rel="stylesheet">
<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Quality Control in R console</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/compbiomed/singleCellTK/blob/master/vignettes/articles/cnsl_dropletqc.Rmd"><code>vignettes/articles/cnsl_dropletqc.Rmd</code></a></small>
      <div class="hidden name"><code>cnsl_dropletqc.Rmd</code></div>

    </div>

    
    
<style>
body {
text-align: justify}
</style>
<div id="introduction" class="section level2">
<h2 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h2>
<p>Performing comprehensive quality control (QC) is necessary to remove poor quality cells for downstream analysis of single-cell RNA sequencing (scRNA-seq) data. Within droplet-based scRNA-seq data, droplets containing cells must be differentiated from empty droplets. Therefore, assessment of the data is required, for which various QC algorithms have been developed. In singleCellTK, we have written convenience functions for several of these tools. In this guide, we will demonstrate how to use these functions to perform quality control on unfiltered, droplet-level data.</p>
<p>The package can be loaded using the <code>library</code> command.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://compbiomed.github.io/sctk_docs/">singleCellTK</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org">dplyr</a></span><span class="op">)</span></code></pre></div>
</div>
<div id="running-quality-control-of-droplet-raw-count-data-with-singlecelltk" class="section level2">
<h2 class="hasAnchor">
<a href="#running-quality-control-of-droplet-raw-count-data-with-singlecelltk" class="anchor"></a>Running quality control of droplet raw count data with singleCellTK</h2>
<div id="load-pbmc4k-data-from-10x" class="section level3">
<h3 class="hasAnchor">
<a href="#load-pbmc4k-data-from-10x" class="anchor"></a>Load PBMC4k data from 10X</h3>
<p>singleCellTK takes in a <code>SingleCellExperiment</code> object from the <a href="https://bioconductor.org/packages/release/bioc/html/SingleCellExperiment.html">SingleCellExperiment</a> package. We will utilize the 10X PBMC 4K dataset as an example. For the quality control of droplet-based counts data, we will install the dataset from the 10X Genomics website using the <code>BiocFileCache</code>(<a href="https://www.bioconductor.org/packages/release/bioc/html/BiocFileCache.html" class="uri">https://www.bioconductor.org/packages/release/bioc/html/BiocFileCache.html</a>) package.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Install BiocFileCache if is it not already</span>
<span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocFileCache"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span>
  <span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span>
    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span>
  <span class="op">}</span>
  <span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span><span class="op">(</span><span class="st">"BiocFileCache"</span><span class="op">)</span>
<span class="op">}</span>

<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="st">"BiocFileCache"</span><span class="op">)</span>
<span class="va">bfc</span> <span class="op">&lt;-</span> <span class="fu">BiocFileCache</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocFileCache/man/BiocFileCache-class.html">BiocFileCache</a></span><span class="op">(</span><span class="st">"raw_data"</span>, ask <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>
<span class="va">raw.path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocFileCache/man/BiocFileCache-class.html">bfcrpath</a></span><span class="op">(</span><span class="va">bfc</span>, <span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span><span class="op">(</span>
  <span class="st">"http://cf.10xgenomics.com/samples"</span>,
  <span class="st">"cell-exp/2.1.0/pbmc4k/pbmc4k_raw_gene_bc_matrices.tar.gz"</span>
<span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/untar.html">untar</a></span><span class="op">(</span><span class="va">raw.path</span>, exdir <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html">tempdir</a></span><span class="op">(</span><span class="op">)</span>, <span class="st">"pbmc4k"</span><span class="op">)</span><span class="op">)</span>

<span class="va">fname</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html">tempdir</a></span><span class="op">(</span><span class="op">)</span>, <span class="st">"pbmc4k/raw_gene_bc_matrices/GRCh38"</span><span class="op">)</span>
<span class="va">pbmc4k.droplet</span> <span class="op">&lt;-</span> <span class="fu">DropletUtils</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/DropletUtils/man/read10xCounts.html">read10xCounts</a></span><span class="op">(</span><span class="va">fname</span>, col.names <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>

<span class="co">### change the sample column</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="fu">colData</span><span class="op">(</span><span class="va">pbmc4k.droplet</span><span class="op">)</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"sample"</span>, <span class="st">"Barcode"</span><span class="op">)</span>
<span class="fu">colData</span><span class="op">(</span><span class="va">pbmc4k.droplet</span><span class="op">)</span><span class="op">$</span><span class="va">sample</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="st">"pbmc4k"</span>, <span class="fu"><a href="https://rdrr.io/r/base/nrow.html">ncol</a></span><span class="op">(</span><span class="fu">colData</span><span class="op">(</span><span class="va">pbmc4k.droplet</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>singleCellTK also supports the importing of single-cell data from the following platforms: <a href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/what-is-cell-ranger">10X CellRanger</a>, <a href="https://github.com/alexdobin/STAR/blob/master/docs/STARsolo.md">STARSolo</a>, <a href="https://www.kallistobus.tools/">BUSTools</a>, <a href="https://github.com/ambrosejcarr/seqc">SEQC</a>, <a href="https://github.com/hms-dbmi/dropEst">DropEST</a>, and <a href="https://github.com/theislab/anndata">AnnData</a>. To load your own input data, please refer <a href="import_data.html">Import data into SCTK</a> for deailed instruction.</p>
</div>
<div id="rundropletqc" class="section level3">
<h3 class="hasAnchor">
<a href="#rundropletqc" class="anchor"></a>runDropletQC</h3>
<p>All droplet-based QC functions are able to be run under the wrapper function runDropletQC. By default all possible QC algorithms will be run.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pbmc4k.droplet</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/runDropletQC.html">runDropletQC</a></span><span class="op">(</span><span class="va">pbmc4k.droplet</span><span class="op">)</span></code></pre></div>
<pre><code>## Sun Sep 26 13:41:24 2021 ... Running 'perCellQCMetrics'</code></pre>
<pre><code>## Sun Sep 26 13:41:24 2021 ... Running 'emptyDrops'</code></pre>
<pre><code>## Sun Sep 26 13:41:55 2021 ... Running 'barcodeRanks'</code></pre>
<p>If users choose to only run a specific set of algorithms, they can specify which to run with the <code>algorithms</code> parameter.</p>
<p>When quality control functions are run in singleCellTK, the output of the function is stored in the colData slot of the SingleCellExperiment object.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="fu">colData</span><span class="op">(</span><span class="va">pbmc4k.droplet</span><span class="op">)</span>, <span class="fl">5</span><span class="op">)</span></code></pre></div>
<div style="border: 1px solid #ddd; padding: 5px; overflow-x: scroll; width:80%; ">
<table class="table table">
<thead><tr>
<th style="text-align:left;">
</th>
<th style="text-align:left;">
sample
</th>
<th style="text-align:left;">
Barcode
</th>
<th style="text-align:right;">
sum
</th>
<th style="text-align:right;">
detected
</th>
<th style="text-align:right;">
percent.top_50
</th>
<th style="text-align:right;">
percent.top_100
</th>
<th style="text-align:right;">
percent.top_200
</th>
<th style="text-align:right;">
percent.top_500
</th>
<th style="text-align:right;">
total
</th>
<th style="text-align:right;">
dropletUtils_emptyDrops_total
</th>
<th style="text-align:right;">
dropletUtils_emptyDrops_logprob
</th>
<th style="text-align:right;">
dropletUtils_emptyDrops_pvalue
</th>
<th style="text-align:left;">
dropletUtils_emptyDrops_limited
</th>
<th style="text-align:right;">
dropletUtils_emptyDrops_fdr
</th>
<th style="text-align:right;">
dropletUtils_BarcodeRank_Knee
</th>
<th style="text-align:right;">
dropletUtils_BarcodeRank_Inflection
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;">
CTCTACGCATGAGCGA-1
</td>
<td style="text-align:left;">
pbmc4k
</td>
<td style="text-align:left;">
CTCTACGCATGAGCGA-1
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
NaN
</td>
<td style="text-align:right;">
NaN
</td>
<td style="text-align:right;">
NaN
</td>
<td style="text-align:right;">
NaN
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
NA
</td>
<td style="text-align:right;">
NA
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:right;">
NA
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
CCGTGGATCCCTCTTT-1
</td>
<td style="text-align:left;">
pbmc4k
</td>
<td style="text-align:left;">
CCGTGGATCCCTCTTT-1
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
100.00000
</td>
<td style="text-align:right;">
100
</td>
<td style="text-align:right;">
100
</td>
<td style="text-align:right;">
100
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
NA
</td>
<td style="text-align:right;">
NA
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:right;">
NA
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
TGGTTAGAGCGATTCT-1
</td>
<td style="text-align:left;">
pbmc4k
</td>
<td style="text-align:left;">
TGGTTAGAGCGATTCT-1
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
NaN
</td>
<td style="text-align:right;">
NaN
</td>
<td style="text-align:right;">
NaN
</td>
<td style="text-align:right;">
NaN
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
NA
</td>
<td style="text-align:right;">
NA
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:right;">
NA
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
AGAATAGCACCCATGG-1
</td>
<td style="text-align:left;">
pbmc4k
</td>
<td style="text-align:left;">
AGAATAGCACCCATGG-1
</td>
<td style="text-align:right;">
88
</td>
<td style="text-align:right;">
79
</td>
<td style="text-align:right;">
67.04545
</td>
<td style="text-align:right;">
100
</td>
<td style="text-align:right;">
100
</td>
<td style="text-align:right;">
100
</td>
<td style="text-align:right;">
88
</td>
<td style="text-align:right;">
88
</td>
<td style="text-align:right;">
NA
</td>
<td style="text-align:right;">
NA
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:right;">
NA
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
GACTGCGCATGTCCTC-1
</td>
<td style="text-align:left;">
pbmc4k
</td>
<td style="text-align:left;">
GACTGCGCATGTCCTC-1
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
NaN
</td>
<td style="text-align:right;">
NaN
</td>
<td style="text-align:right;">
NaN
</td>
<td style="text-align:right;">
NaN
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
NA
</td>
<td style="text-align:right;">
NA
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:right;">
NA
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
</tr>
</tbody>
</table>
</div>
<p><br></p>
<p>A summary of all outputs is shown below:</p>
<table class="table">
<colgroup>
<col width="27%">
<col width="46%">
<col width="14%">
<col width="10%">
</colgroup>
<thead><tr class="header">
<th align="left">QC output</th>
<th align="left">Description</th>
<th align="left">Methods</th>
<th align="left">Package/Tool</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">dropletUtils_emptyDrops_total</td>
<td align="left">Total counts</td>
<td align="left">runEmptyDrops</td>
<td align="left">DropletUtils</td>
</tr>
<tr class="even">
<td align="left">dropletUtils_emptyDrops_logprob</td>
<td align="left">The log-probability of droplet being empty</td>
<td align="left">runEmptyDrops</td>
<td align="left">DropletUtils</td>
</tr>
<tr class="odd">
<td align="left">dropletUtils_emptyDrops_pvalue</td>
<td align="left">Monte Carlo p-value of droplet being empty</td>
<td align="left">runEmptyDrops</td>
<td align="left">DropletUtils</td>
</tr>
<tr class="even">
<td align="left">dropletUtils_emptyDrops_limited</td>
<td align="left">Whether a lower p-value could be obtained by increasing niters</td>
<td align="left">runEmptyDrops</td>
<td align="left">DropletUtils</td>
</tr>
<tr class="odd">
<td align="left">dropletUtils_emptyDrops_fdr</td>
<td align="left">p-value of droplet being empty, corrected for false detection rate</td>
<td align="left">runEmptyDrops</td>
<td align="left">DropletUtils</td>
</tr>
<tr class="even">
<td align="left">dropletUtils_BarcodeRank_Knee</td>
<td align="left">Whether total UMI count value is higher than knee point</td>
<td align="left">runBarcodeRankDrops</td>
<td align="left">DropletUtils</td>
</tr>
<tr class="odd">
<td align="left">dropletUtils_BarcodeRank_Inflection</td>
<td align="left">Whether total UMI count value is higher than inflection point</td>
<td align="left">runBarcodeRankDrops</td>
<td align="left">DropletUtils</td>
</tr>
</tbody>
</table>
</div>
<div id="plotting-qc-metrics" class="section level3">
<h3 class="hasAnchor">
<a href="#plotting-qc-metrics" class="anchor"></a>Plotting QC metrics</h3>
<p>Upon running runDropletQC, the QC outputs will need to be plotted. For each QC method, singleCellTK contains specialized plotting functions that may be used for the generation of QC plots.</p>
<div id="emptydrops" class="section level4">
<h4 class="hasAnchor">
<a href="#emptydrops" class="anchor"></a>EmptyDrops</h4>
<p>It is crucial to distinguish the data occurring from real cells and empty droplets containing ambient RNA. SCTK employs the <a href="https://rdrr.io/github/MarioniLab/DropletUtils/man/emptyDrops.html">EmptyDrops</a> algorithm from the <a href="https://bioconductor.org/packages/release/bioc/html/DropletUtils.html">DropletUtils</a> package to test for empty droplets.</p>
<p>The wrapper function <code>plotEmptyDropsResults</code> can be used to plot the results from the EmptyDrops algorithm. This will visualize the empty droplets, by plotting the total UMI counts against the log probability for each barcode.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">emptyDropsResults</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/plotEmptyDropsResults.html">plotEmptyDropsResults</a></span><span class="op">(</span>
  inSCE <span class="op">=</span> <span class="va">pbmc4k.droplet</span>,
  axisLabelSize <span class="op">=</span> <span class="fl">20</span>,
  sample <span class="op">=</span> <span class="cn">NULL</span>,
  fdrCutoff <span class="op">=</span> <span class="fl">0.01</span>,
  dotSize <span class="op">=</span> <span class="fl">0.5</span>,
  defaultTheme <span class="op">=</span> <span class="cn">TRUE</span>
<span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">emptyDropsResults</span><span class="op">$</span><span class="va">scatterEmptyDrops</span></code></pre></div>
<p><img src="cnsl_dropletqc_files/figure-html/scatter_emptydrops-1.png" width="768" style="display: block; margin: auto;"><br> Data points are colored by FDR values, where we see a small portion of the dataset contains barcodes that do not meet the threshold.</p>
</div>
<div id="barcoderanks" class="section level4">
<h4 class="hasAnchor">
<a href="#barcoderanks" class="anchor"></a>BarcodeRanks</h4>
<p><a href="https://rdrr.io/bioc/DropletUtils/man/barcodeRanks.html">BarcodeRanks</a> from the <a href="https://bioconductor.org/packages/release/bioc/html/DropletUtils.html">DropletUtils</a> package computes barcode rank statistics and identifies the knee and inflection points on the total count curve. The knee and inflection points on the curve represent the difference between empty droplets and cell-containing droplets with much more RNA. The lower parameter is again the lower bound of the total UMI count, in which all barcodes below the lower bound are assumed to be empty droplets.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../../reference/plotBarcodeRankScatter.html">plotBarcodeRankScatter</a></span><span class="op">(</span>
  inSCE <span class="op">=</span> <span class="va">pbmc4k.droplet</span>,
  title <span class="op">=</span> <span class="st">"BarcodeRanks Rank Plot"</span>,
  legendSize <span class="op">=</span> <span class="fl">14</span>
<span class="op">)</span></code></pre></div>
<p><img src="cnsl_dropletqc_files/figure-html/run_kneeplot_barcoderank-1.png" width="768" style="display: block; margin: auto;"></p>
<p>The total UMI count of each barcode is plotted against its rank, where we see a steep dropoff of UMI counts around the inflection point, where we see a separation between cell containing and empty droplets.</p>
</div>
</div>
<div id="filtering-the-dataset" class="section level3">
<h3 class="hasAnchor">
<a href="#filtering-the-dataset" class="anchor"></a>Filtering the dataset</h3>
<p>SingleCellExperiment objects can be subset by its colData using <code>subsetSCECols</code>. The <code>colData</code> parameter takes in an expression in character vector form which will be used to identify a subset of columns using variables found in the colData of the SingleCellExperiment object. For example, if x is a numeric vector in colData, then “x &lt; 5” will return all columns with x less than 5. The <code>index</code> parameter takes in a vector of indices which should be kept, while <code>bool</code> takes in a vector of TRUE or FALSE which should be the same length as the number of columns in the SingleCellExperiment object.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">#Before filtering:</span>
<span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">pbmc4k.droplet</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] 33694 60000</code></pre>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pbmc4k.droplet</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/subsetSCECols.html">subsetSCECols</a></span><span class="op">(</span><span class="va">pbmc4k.droplet</span>, colData <span class="op">=</span> <span class="st">'dropletUtils_BarcodeRank_Inflection == 1'</span><span class="op">)</span>
<span class="va">pbmc4k.droplet</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/subsetSCECols.html">subsetSCECols</a></span><span class="op">(</span><span class="va">pbmc4k.droplet</span>, colData <span class="op">=</span> <span class="st">'!is.na(pbmc4k.droplet$dropletUtils_emptyDrops_fdr)'</span><span class="op">)</span>
<span class="va">pbmc4k.droplet</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/subsetSCECols.html">subsetSCECols</a></span><span class="op">(</span><span class="va">pbmc4k.droplet</span>, colData <span class="op">=</span> <span class="st">'pbmc4k.droplet$dropletUtils_emptyDrops_fdr &lt; 0.01'</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">#After filtering:</span>
<span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">pbmc4k.droplet</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] 33694   334</code></pre>
<p>We can compare the average total UMI counts per cell before and after cell filtration:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/plotSCEViolinColData.html">plotSCEViolinColData</a></span><span class="op">(</span><span class="va">pbmc4k.droplet.prefilt</span>, coldata <span class="op">=</span> <span class="st">"sum"</span>, summary <span class="op">=</span> <span class="st">"mean"</span>, title <span class="op">=</span> <span class="st">"Pre-filter"</span>, ylab <span class="op">=</span> <span class="st">"Total counts"</span><span class="op">)</span>
<span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/plotSCEViolinColData.html">plotSCEViolinColData</a></span><span class="op">(</span><span class="va">pbmc4k.droplet</span>, coldata <span class="op">=</span> <span class="st">"sum"</span>, summary <span class="op">=</span> <span class="st">"mean"</span>, title <span class="op">=</span> <span class="st">"Post-filter"</span>, ylab <span class="op">=</span> <span class="st">"Total counts"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="fu">cowplot</span><span class="fu">::</span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html">plot_grid</a></span><span class="op">(</span><span class="va">p1</span>, <span class="va">p2</span>, ncol <span class="op">=</span> <span class="fl">2</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="cnsl_dropletqc_files/figure-html/Count%20Violin%20after%20filtering-1.png" width="576" style="display: block; margin: auto;"></p>
<p>For additional cell-filtered quality control with singleCellTK, please refer to “Running quality control of cell-filtered single cell data with singleCellTK”.</p>
</div>
<div id="running-individual-qc-methods" class="section level3">
<h3 class="hasAnchor">
<a href="#running-individual-qc-methods" class="anchor"></a>Running individual QC methods</h3>
<p>Instead of running all quality control methods on the dataset at once, users may elect to execute QC methods individually. The parameters as well as the outputs to individual QC functions are described in detail as follows:</p>
<div id="runemptydrops" class="section level4">
<h4 class="hasAnchor">
<a href="#runemptydrops" class="anchor"></a>runEmptyDrops</h4>
<p>The wrapper function <code>runEmptyDrops</code> can be used to separately run the EmptyDrops algorithm on its own.</p>
<p>In runEmptyDrops, the <code>lower</code> parameter is the lower bound of the total UMI count, in which all barcodes below the lower bound are assumed to be empty droplets. The <code>niters</code> parameter is the number of iterations the function will run for the calculation. <code>testAmbient</code> indicates whether results should be returned for barcodes that have a total UMI count below what is specified in <code>lower</code>.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pbmc4k.droplet</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/runEmptyDrops.html">runEmptyDrops</a></span><span class="op">(</span>
  inSCE <span class="op">=</span> <span class="va">pbmc4k.droplet</span>,
  useAssay <span class="op">=</span> <span class="st">"counts"</span>,
  lower <span class="op">=</span> <span class="fl">100</span>,
  niters <span class="op">=</span> <span class="fl">10000</span>
<span class="op">)</span></code></pre></div>
</div>
<div id="runbarcoderankdrops" class="section level4">
<h4 class="hasAnchor">
<a href="#runbarcoderankdrops" class="anchor"></a>runBarcodeRankDrops</h4>
<p>The wrapper function <code>runBarcodeRankDrops</code> can be used to separately run the BarcodeRanks algorithm on its own.</p>
<p>The <code>lower</code> parameter is again the lower bound of the total UMI count, in which all barcodes below the lower bound are assumed to be empty droplets.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pbmc4k.droplet</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/runBarcodeRankDrops.html">runBarcodeRankDrops</a></span><span class="op">(</span>
  inSCE <span class="op">=</span> <span class="va">pbmc4k.droplet</span>,
  useAssay <span class="op">=</span> <span class="st">"counts"</span>,
  fitBounds <span class="op">=</span> <span class="cn">NULL</span>, df <span class="op">=</span> <span class="fl">20</span>
<span class="op">)</span></code></pre></div>
<p><br><br></p>
</div>
</div>
</div>
<div id="session-information" class="section level1">
<h1 class="hasAnchor">
<a href="#session-information" class="anchor"></a>Session Information</h1>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<pre><code>## R version 4.0.4 (2021-02-15)
## Platform: x86_64-w64-mingw32/x64 (64-bit)
## Running under: Windows 10 x64 (build 19042)
## 
## Matrix products: default
## 
## locale:
## [1] LC_COLLATE=Chinese (Simplified)_China.936 
## [2] LC_CTYPE=Chinese (Simplified)_China.936   
## [3] LC_MONETARY=Chinese (Simplified)_China.936
## [4] LC_NUMERIC=C                              
## [5] LC_TIME=Chinese (Simplified)_China.936    
## 
## attached base packages:
## [1] parallel  stats4    stats     graphics  grDevices utils     datasets 
## [8] methods   base     
## 
## other attached packages:
##  [1] BiocFileCache_1.14.0        dbplyr_2.1.0               
##  [3] dplyr_1.0.5                 singleCellTK_2.2.1         
##  [5] DelayedArray_0.16.2         Matrix_1.3-3               
##  [7] SingleCellExperiment_1.12.0 SummarizedExperiment_1.20.0
##  [9] Biobase_2.50.0              GenomicRanges_1.42.0       
## [11] GenomeInfoDb_1.26.2         IRanges_2.24.1             
## [13] S4Vectors_0.28.1            BiocGenerics_0.36.0        
## [15] MatrixGenerics_1.2.1        matrixStats_0.58.0         
## 
## loaded via a namespace (and not attached):
##   [1] ggbeeswarm_0.6.0          colorspace_2.0-0         
##   [3] ellipsis_0.3.1            rprojroot_2.0.2          
##   [5] GSVAdata_1.26.0           scuttle_1.0.4            
##   [7] XVector_0.30.0            BiocNeighbors_1.8.2      
##   [9] fs_1.5.0                  rstudioapi_0.13          
##  [11] farver_2.1.0              bit64_4.0.5              
##  [13] fansi_0.4.2               xml2_1.3.2               
##  [15] R.methodsS3_1.8.1         sparseMatrixStats_1.2.1  
##  [17] cachem_1.0.4              knitr_1.31               
##  [19] scater_1.18.6             jsonlite_1.7.2           
##  [21] R.oo_1.24.0               HDF5Array_1.18.1         
##  [23] compiler_4.0.4            httr_1.4.2               
##  [25] dqrng_0.2.1               assertthat_0.2.1         
##  [27] fastmap_1.1.0             limma_3.46.0             
##  [29] BiocSingular_1.6.0        htmltools_0.5.1.1        
##  [31] tools_4.0.4               rsvd_1.0.3               
##  [33] gtable_0.3.0              glue_1.4.2               
##  [35] GenomeInfoDbData_1.2.4    rappdirs_0.3.3           
##  [37] Rcpp_1.0.6                jquerylib_0.1.3          
##  [39] pkgdown_1.6.1             vctrs_0.3.6              
##  [41] rhdf5filters_1.2.0        svglite_2.0.0            
##  [43] DelayedMatrixStats_1.12.3 xfun_0.21                
##  [45] stringr_1.4.0             beachmat_2.6.4           
##  [47] rvest_0.3.6               lifecycle_1.0.0          
##  [49] irlba_2.3.3               gtools_3.8.2             
##  [51] edgeR_3.32.1              zlibbioc_1.36.0          
##  [53] scales_1.1.1              ragg_1.1.1               
##  [55] rhdf5_2.34.0              yaml_2.2.1               
##  [57] curl_4.3                  memoise_2.0.0            
##  [59] reticulate_1.18           gridExtra_2.3            
##  [61] ggplot2_3.3.3             sass_0.3.1               
##  [63] stringi_1.5.3             RSQLite_2.2.3            
##  [65] highr_0.8                 desc_1.3.0               
##  [67] BiocParallel_1.24.1       rlang_0.4.10             
##  [69] pkgconfig_2.0.3           systemfonts_1.0.1        
##  [71] bitops_1.0-6              evaluate_0.14            
##  [73] lattice_0.20-41           purrr_0.3.4              
##  [75] Rhdf5lib_1.12.1           labeling_0.4.2           
##  [77] cowplot_1.1.1             bit_4.0.4                
##  [79] tidyselect_1.1.0          magrittr_2.0.1           
##  [81] R6_2.5.0                  generics_0.1.0           
##  [83] fishpond_1.6.0            DBI_1.1.1                
##  [85] pillar_1.5.1              withr_2.4.1              
##  [87] RCurl_1.98-1.2            tibble_3.1.0             
##  [89] crayon_1.4.1              DropletUtils_1.10.3      
##  [91] utf8_1.1.4                rmarkdown_2.7            
##  [93] viridis_0.5.1             locfit_1.5-9.4           
##  [95] grid_4.0.4                blob_1.2.1               
##  [97] webshot_0.5.2             digest_0.6.27            
##  [99] R.utils_2.10.1            textshaping_0.3.1        
## [101] munsell_0.5.0             beeswarm_0.3.1           
## [103] viridisLite_0.3.0         kableExtra_1.3.4         
## [105] vipor_0.4.5               bslib_0.2.4</code></pre>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by David Jenkins, Vidya Akavoor, Salam Alabdullatif, Shruthi Bandyadka, Emma Briars, Xinyun Cao, Sebastian Carrasco Pro, Tyler Faits, Rui Hong, Mohammed Muzamil Khan, Yusuke Koga, Anastasia Leshchyk, Irzam Sarfraz, Yichen Wang, Zhe Wang, W. Evan Johnson, Joshua David Campbell.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
