<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Differential Expression • singleCellTK</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/yeti/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../../bootstrap-toc.css">
<script src="../../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../../pkgdown.css" rel="stylesheet">
<script src="../../pkgdown.js"></script><link href="../../extra.css" rel="stylesheet">
<script src="../../extra.js"></script><meta property="og:title" content="Differential Expression">
<meta property="og:description" content="singleCellTK">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../../index.html">singleCellTK</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">2.2.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../../articles/articles/introduction_to_SCTK.html">General Workflow</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Shell Command Line Analysis
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../../articles/articles/cmd_import_scRNAseq_data_as_SCE.html">Import data into SCTK</a>
    </li>
    <li>
      <a href="../../articles/articles/cmd_qc.html">Quality Control</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    R Console Analysis
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../../articles/articles/cnsl_import_scRNAseq_data_as_SCE.html">Import data into SCTK</a>
    </li>
    <li>
      <a href="../../articles/articles/cnsl_dropletqc.html">Quality Control of droplet data</a>
    </li>
    <li>
      <a href="../../articles/articles/cnsl_cellqc.html">Quality Control of cell data</a>
    </li>
    <li>
      <a href="../../articles/articles/cnsl_normalization.html">Normalization</a>
    </li>
    <li>
      <a href="../../articles/articles/cnsl_batch_correction.html">Batch Correction</a>
    </li>
    <li>
      <a href="../../articles/articles/cnsl_feature_selection.html">Feature Selection</a>
    </li>
    <li>
      <a href="../../articles/articles/cnsl_dimensionality_reduction.html">Dimensionality Reduction</a>
    </li>
    <li>
      <a href="../../articles/articles/cnsl_clustering.html">Clustering</a>
    </li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Differential Expression</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../../articles/articles/cnsl_differential_expression.html">Differential Expression</a>
        </li>
        <li>
          <a href="../../articles/articles/cnsl_find_marker.html">Find Marker via DE</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Enrichment Analysis</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../../articles/articles/cnsl_gsva.html">GSVA</a>
        </li>
        <li>
          <a href="../../articles/articles/cnsl_enrichR.html">EnrichR</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Visualization</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../../articles/articles/cnsl_visualization.html">General Visualization</a>
        </li>
        <li>
          <a href="../../articles/articles/cnsl_heatmap.html">Heatmap Visualization</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Curated Workflows</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../../articles/articles/cnsl_seurat_curated_workflow.html">Seurat</a>
        </li>
        <li>
          <a href="../../articles/articles/cnsl_celda_curated_workflow.html">Celda</a>
        </li>
      </ul>
</li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Shiny UI Analysis
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Data</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../../articles/articles/ui_import_data.html">Import Data</a>
        </li>
        <li>
          <a href="../../articles/articles/ui_import_genesets.html">Import Gene Sets</a>
        </li>
        <li>
          <a href="../../articles/articles/ui_import_annotation.html">Annotation</a>
        </li>
        <li>
          <a href="../../articles/articles/ui_export_data.html">Export Data</a>
        </li>
        <li>
          <a href="../../articles/articles/ui_delete_data.html">Delete Data</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">QC &amp; Filtering</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../../articles/articles/ui_qc.html">Quality Control</a>
        </li>
        <li>
          <a href="../../articles/articles/ui_filtering.html">Filtering</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Normalization &amp; Batch Correction</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../../articles/articles/ui_normalization.html">Normalization</a>
        </li>
        <li>
          <a href="../../articles/articles/ui_batch_correction.html">Batch Correction</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Feature Selection &amp; Dimensionality Reduction</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../../articles/articles/ui_feature_selection.html">Feature Selection</a>
        </li>
        <li>
          <a href="../../articles/articles/ui_dimensionality_reduction.html">Dimensionality Reduction</a>
        </li>
      </ul>
</li>
    <li>
      <a href="../../articles/articles/ui_clustering.html">Clustering</a>
    </li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Differential Expression &amp; Cell Type Labeling</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../../articles/articles/ui_differential_expression.html">Differential Expression</a>
        </li>
        <li>
          <a href="../../articles/articles/ui_find_marker.html">Find Marker</a>
        </li>
        <li>
          <a href="../../articles/articles/ui_cell_type_labeling.html">Cell Type Labeling</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Pathway &amp; Enrichment Analysis</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../../articles/articles/ui_gsva.html">GSVA</a>
        </li>
        <li>
          <a href="../../articles/articles/ui_enrichR.html">EnrichR</a>
        </li>
      </ul>
</li>
    <li>
      <a href="../../articles/articles/ui_sample_size_calc.html">Sample Size Calculator</a>
    </li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Curated Workflows</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../../articles/articles/ui_seurat_curated_workflow.html">Seurat</a>
        </li>
        <li>
          <a href="../../articles/articles/ui_celda_curated_workflow.html">Celda</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Visualization</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../../articles/articles/ui_cellViewer.html">cellViewer</a>
        </li>
        <li>
          <a href="../../articles/articles/ui_heatmap.html">Generic Heatmap</a>
        </li>
      </ul>
</li>
  </ul>
</li>
<li>
  <a href="../../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/compbiomed/singleCellTK">
    <span class="fas fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="differential_expression_files/header-attrs-2.7/header-attrs.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Differential Expression</h1>
                        <h4 class="author">Yichen Wang</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/compbiomed/singleCellTK/blob/master/vignettes/articles/differential_expression.Rmd"><code>vignettes/articles/differential_expression.Rmd</code></a></small>
      <div class="hidden name"><code>differential_expression.Rmd</code></div>

    </div>

    
    

<meta name="viewport" content="width=device-width, initial-scale=1">
<div id="introduction" class="section level2">
<h2 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h2>
<p>Single Cell Toolkit (SCTK) performs differential expression (DE) analysis in a group-VS-group way. The groups to be compared can be set by selecting the individual cells or by using existing categorical cell annotation. SCTK supports five methods, including wilcoxon test <span class="citation">[1]</span>, MAST <span class="citation">[2]</span>, Limma <span class="citation">[3]</span>, DESeq2 <span class="citation">[4]</span>, and ANOVA. Users can perform multiple sets of DE analysis with unique analysis name set each time.</p>
</div>
<div id="workflow-guide" class="section level2">
<h2 class="hasAnchor">
<a href="#workflow-guide" class="anchor"></a>Workflow Guide</h2>
<div class="tab">
  <button class="tablinks" onclick="openTab(event, 'interactive')" id="ia-button">Interactive Analysis</button>
  <button class="tablinks" onclick="openTab(event, 'console')" id="console-button">Console Analysis</button>
</div>

<div id="interactive" class="tabcontent">
<p><strong>Entry of the Panel</strong></p>
<p>From anywhere of the UI, the panel for DE can be accessed from the top navigation panel at the circled tab shown below.</p>
<p><img src="ui_screenshots/differential_expression/de_ui_entry.png" alt="Entry"><br></p>
<p>The UI is constructed with mainly four parts: <strong>1.</strong> Assay input; <strong>2.</strong> Condition selection; <strong>3.</strong> Parameter settings; <strong>4.</strong> Result and other visualization.</p>
<p><strong>Assay Input</strong></p>
<p>A SingleCellExperiment (SCE) object, which stores all the expression, reduction, and metadata, is active behind the interface. The assay needed here is a matrix that contains the expression information for all cells and all features (genes), and is saved in the <code>assay</code> slot of the SCE object. The selection for this should be made at the top of the panel, as shown in the screenshot below.</p>
<p><img src="ui_screenshots/differential_expression/de_ui_assay.png" style="border: 1px solid rgba(0,0,0,0.2)"></p>
<p><strong>Condition Selection</strong></p>
<p>Usually in terms of a DE analysis, we have two necessary conditions that have to be defined. One is the condition of interests and the other one is the condition of control. The condition of interests will then be compared to the condition of control via computational methods and statistical result will then be returned. The selection for a condition here means deciding which cells should be grouped together as one condition.</p>
<p>SCTK provides reasonably flexible approaches to setting the conditions:</p>
<details><summary><b>Categorical Selection</b></summary><p>The first approach is designed for the fastest use, where conditions of interests and of control can be simply defined by categorical variables of the same class in the cell annotation (i.e. within one single column in <code>colData</code> of the background SCE object).</p>
<p><img src="ui_screenshots/differential_expression/de_ui_condMeth1.png" style="border: 1px solid rgba(0,0,0,0.2)"></p>
<p>First, users need to choose one single option from the selection list <strong>“Choose Annotation Class”</strong>. Then, in the UI, there will be two columns for each condition. For each of them, users can make selections for which categories should belong to the condition, at the selection input <strong>“Select Condition(s)”</strong>. One or more selections are acceptable. When selections are made for one condition while no selection for the other, all the categories but those already used for the former will go to the latter. The text boxes, <strong>“Cells selected”</strong>, are only for showing users what are selected. While the text span below the boxes summaries the total number of cells selected.<br><strong>TODO check if text box actually is editable</strong></p>
<p>In the example in the figure shown above, we intend to define the conditions based on the clustering result generated by Seurat’s louvain clustering, with resolution 0.8, <code>"Seurat_louvain_Resolution0.8"</code>. Then we define that the cells assigned to “cluster 2” are of our interests, while the cells assigned to both “cluster 4” and “cluster 5” should be a control against “cluster 2.” (Same as the example in the <a href="differential_expression.html#example">R console tutorial for DE</a>)</p>
</details><details><summary><b>Data Table Firtering Selection</b></summary><p>The second approach allows users to make the most use of all the cell annotations available. Meanwhile, single selection on each cell is also enabled to ensure the highest flexibility, though can be relatively ineffective.</p>
<p><img src="ui_screenshots/differential_expression/de_ui_condMeth2.png" style="border: 1px solid rgba(0,0,0,0.2)"></p>
<p>In this approach, we utilize a data table, where filters can be applied, for the flexible definition on each condition. By default, all classes of annotations are displayed. In the selection input <strong>“Columns to display”</strong>, users can select one or more classes annotations, which should be displayed and used for filtering. The blank box under each column title is where the filters are applied. For categorical columns, one or more selections on available variables can be made; for numeric columns (with continuous values), value range can be set. All filters applied are not technically making any selection on cells, but adjusting which cells to display in the table. This can be inferred from the first text span summary below the table area. After users applied all necessary filters, click on <strong>“Add all filtered”</strong> button below the table area to technically make the selection. Meanwhile, Each row in the table is clickable to make a single selection/cancellation on cells. The advantage of the approach is that conditions defined by necessarily multiple classes of annotations can be fulfilled.</p>
<p>Similarly as the first approach, in the figure, we are making an equivalent condition definition.</p>
</details><details><summary><b>Select by Pasting a list</b></summary><p>The final way allows users without any useful annotation but still know which cells are interesting to perform the analysis. It is just by easily pasting a list of cell identifiers into the text box for each condition. Note that the cell identifiers used must be the default cell IDs in the background SCE object (i.e. <code>colnames</code>). The input text should be formatted by one ID per line, splitted with no symbol. A summary text span below each text box will be dynamically updated after pasting, for the number of valid IDs found in the input.</p>
<p><img src="ui_screenshots/differential_expression/de_ui_condMeth3.png" style="border: 1px solid rgba(0,0,0,0.2)"></p>
</details><br><p><strong>NOTE</strong> that the “Name of Condition” fields are always required. Though default text is placed, it is <strong>highly recommanded</strong> that users put easily understandable namings there, in order to avoid the confusion when multiple sets of analysis are performed, and to keep the automatic legends on DE specific plottings clean as well.</p>
<details><summary><b>Other Parameters</b></summary><p><img src="ui_screenshots/differential_expression/de_ui_param.png" style="border: 1px solid rgba(0,0,0,0.2)"></p>
<ul>
<li>
<strong>Method</strong>. At the selection input <strong>“Choose analysis method”</strong>, where “MAST,” “DESeq2,” “Limma” and “ANOVA” are available. Note that these method requires different types of input assay, so please pay attention to the red notification.</li>
<li>
<strong>FDR cutoff</strong>. At the numeric input <strong>“Output FDR less than”</strong>. As the label indicates, the cutoff set here will rule out DE genes with FDR (false discovery rate) value more than the cutoff from the result.</li>
<li>
<strong>Log2FC cutoff</strong>. At the numeric input <strong>“Output Log2FC Absolute value greater than”</strong>. As the label indicates, the cutoff set here will rule out DE genes with absolute value of Log2FC (logged fold change) smaller than the cutoff from the result.</li>
<li>
<strong>Whether only up-regulation</strong>. At the checkbox input <strong>“Only up-regulated genes”</strong>. By checking this option, only DE genes with positive log2FC value will remain in the result. And in the DE analysis within SCTK, “up-regulation” will always mean a positive log2FC tested from the condition of interests against the condition of control.</li>
<li>
<strong>Covariate</strong>. At the selection input <strong>“Select Covariates”</strong>. A covariate is another set of categorization on the cells involved in the analysis, used for modeling. In the UI, we only allow setting the covariates by using the cell annotations saved in background (i.e. <code>colData</code> of the SCE object). Multiple selections are acceptable.</li>
<li>
<strong>Name of an analysis</strong>. At the text input <strong>“Name of Differential Expression Analysis:”</strong>. This is a required field for all DE analysis. Similarly to the naming of each condition, when performing batches of analysis, users will need unique identifiers to recognize each set of result, and to avoid confusion in downstream analysis.</li>
</ul></details><br><p><strong>Result Visualization</strong></p>
<p><strong>Gene Table</strong></p>
<p><img src="ui_screenshots/differential_expression/de_ui_res1.png" style="border: 1px solid rgba(0,0,0,0.2)"></p>
<p>In this tab, there will be a table with all the genes that are thought to be differentially expressed in the condition of interest against the condition of control, with high significance and passed all filter parameters. The gene names (default identifier in the background object, not necessarily a gene symbol), p-values, log2FC values, and FDR values will be present. The table will be saved in background by the name of the analysis, entered before running. Users can also download the table in comma-splitted value (CSV) format, by clicking <strong>“Download Result Table”</strong> button.</p>
<p><strong>Violin Plot</strong></p>
<p><img src="ui_screenshots/differential_expression/de_ui_res2.png" style="border: 1px solid rgba(0,0,0,0.2)"></p>
<p>In the violin plot tab, the UI will present a standard violin plot showing how expression differs in the top N selected genes between cells in both conditions.</p>
<p><strong>Linear Modeling Plot</strong></p>
<p><img src="ui_screenshots/differential_expression/de_ui_res3.png" style="border: 1px solid rgba(0,0,0,0.2)"></p>
<p>In the linear modeling plot tab, an analysis of covariance (ANCOVA) is done, showing how expression differs in the top N selected genes between cells in both conditions.</p>
<p><strong>Heatmap</strong></p>
<p><img src="ui_screenshots/differential_expression/de_ui_res4.png" style="border: 1px solid rgba(0,0,0,0.2)"></p>
<p>This tab enables a limited heatmap visualization for the DE analysis selected. Here SCTK will automatically group the cells to the two conditions where they are assigned to, and group the genes by the log2FC values, which indicate whether a gene is up-regulated or down-regulated.</p>
<details><summary><b>DE Heatmap Parameters</b></summary><ul>
<li>Further filtering by the statistics can be applied through numeric inputs <strong>“Absolute log2FC value greater than”</strong> and <strong>“FDR value less than”</strong>.</li>
<li>Additional labeling can be added from background by multi-selection inputs <strong>“Additional cell annotation”</strong> and <strong>“Additional feature annotation”</strong>.</li>
<li>The heatmap is by default splitted by the grouping “condition” and “regulation,” mentioned above. And this can only be achieved by selecting these two option in the multi-selection inputs <strong>“Split columns by”</strong> and <strong>“Split rows by”</strong>, respectively. Technically, these options can be canceled while other options can also be added, as long as users wish.</li>
</ul></details><br><p>Furthermore, if more sophisticated settings on the DE heatmap are needed, users can move to the <a href="ui_heatmap.html">generic heatmap viewer</a> and make use of the “import from analysis” functionality.</p>
</div>

<div id="console" class="tabcontent">
<p>The differential expression analysis can be easily performed on any preprocessed SCE dataset. Here we introduce the workflow of using the generic wrapper <code><a href="../../reference/runDEAnalysis.html">runDEAnalysis()</a></code>.</p>
<p><strong>Basic Parameters</strong></p>
<p>The most basic parameters include:</p>
<ul>
<li>The input SCE object - <code>inSCE</code>.</li>
<li>The method listed in the table above to use - <code>method</code>.</li>
<li>The assay to perform the analysis on - <code>useAssay</code>
</li>
</ul>
<p><strong>Other required parameter</strong></p>
<p><strong>1. Conditions</strong></p>
<p>The R script functions we have allow users to perform differential expression analysis with relatively flexible condition setting. In terms of the condition of interests and the condition for control, comparison groups can be set by giving either one or more categories under a column in <code>colData</code>, or a prepared index vector as long as the indices are able to subset the input SCE object.</p>
<p>Additionally, only specifying the condition of interests is allowed and the control will then be set to all the other cells, which turns to a biomarker finding analysis.</p>
<p>Since the conditions can be set by using indices or annotations, we have two groups of parameters listed below. Note that only one way of setting can be used at one time.</p>
<ul>
<li>Use <code>index1</code> and <code>index2</code> for index style setting</li>
<li>Use <code>class</code> to specify the annotation vector by either directly giving a vector of proper length or giving a column name of <code>colData</code>. Then use <code>classGroup1</code>, <code>classGroup2</code> to specify which categories in <code>class</code> are of interests.</li>
</ul>
<p><strong>2. Namings</strong></p>
<p>Something that might be special in our workflow is that, users will be required to specify name strings for:</p>
<ul>
<li>The group of interests - <code>groupName1</code>.</li>
<li>The group of control - <code>groupName2</code>.</li>
<li>The analysis run itself - <code>analysisName</code>.</li>
</ul>
<p>The reason is we assume that users are likely to perform multiple batches of analysis for a single dataset (e.g. group1 Vs. group2 and then group1 Vs. group3), and we hope to have everything stored in one SCE without leaving users any confusion when they look back on it. Also, the namings are also used by DE analysis related plotting functions, so that legends can be well annotated.</p>
<details><summary><b>Example</b></summary><p>To demonstrate a simple and clear example, here we use the “PBMC-3k” dataset from “10X” which can be easily imported with SCTK functions.</p>
<details><summary><b>Preprocessing</b></summary><p>The preprocessing only includes necessary steps to get cluster labels (i.e. QC and filtering are excluded), via using SCTK’s Seurat Curated Workflow.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://compbiomed.github.io/sctk_docs/">singleCellTK</a></span><span class="op">)</span>
<span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/importExampleData.html">importExampleData</a></span><span class="op">(</span><span class="st">"pbmc3k"</span><span class="op">)</span>
<span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/scaterlogNormCounts.html">scaterlogNormCounts</a></span><span class="op">(</span><span class="va">sce</span>, <span class="st">"logcounts"</span><span class="op">)</span>
<span class="co"># Go through the Seurat curated workflow to get basic clusters</span>
<span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/seuratNormalizeData.html">seuratNormalizeData</a></span><span class="op">(</span>inSCE <span class="op">=</span> <span class="va">sce</span>, useAssay <span class="op">=</span> <span class="st">"counts"</span><span class="op">)</span>
<span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/seuratFindHVG.html">seuratFindHVG</a></span><span class="op">(</span>inSCE <span class="op">=</span> <span class="va">sce</span>, useAssay <span class="op">=</span> <span class="st">"seuratNormData"</span><span class="op">)</span>
<span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/seuratScaleData.html">seuratScaleData</a></span><span class="op">(</span>inSCE <span class="op">=</span> <span class="va">sce</span>, useAssay <span class="op">=</span> <span class="st">"seuratNormData"</span><span class="op">)</span>
<span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/seuratPCA.html">seuratPCA</a></span><span class="op">(</span>inSCE <span class="op">=</span> <span class="va">sce</span>, useAssay <span class="op">=</span> <span class="st">"seuratScaledData"</span><span class="op">)</span>
<span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/seuratFindClusters.html">seuratFindClusters</a></span><span class="op">(</span>inSCE <span class="op">=</span> <span class="va">sce</span>, useAssay <span class="op">=</span> <span class="st">"seuratScaledData"</span><span class="op">)</span>
<span class="co"># Optional visualization</span>
<span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/seuratRunUMAP.html">seuratRunUMAP</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span>
<span class="fu"><a href="../../reference/plotSCEDimReduceColData.html">plotSCEDimReduceColData</a></span><span class="op">(</span>inSCE <span class="op">=</span> <span class="va">sce</span>, 
                        colorBy <span class="op">=</span> <span class="st">"Seurat_louvain_Resolution0.8"</span>, 
                        conditionClass <span class="op">=</span> <span class="st">"factor"</span>, 
                        reducedDimName <span class="op">=</span> <span class="st">"seuratUMAP"</span><span class="op">)</span></code></pre></div>
</details><br><div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Condition chosen basing on the UMAP</span>
<span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/runDEAnalysis.html">runDEAnalysis</a></span><span class="op">(</span>inSCE <span class="op">=</span> <span class="va">sce</span>, 
                        method <span class="op">=</span> <span class="st">"MAST"</span>, 
                        useAssay <span class="op">=</span> <span class="st">"logcounts"</span>,
                        class <span class="op">=</span> <span class="st">"Seurat_louvain_Resolution0.8"</span>, 
                        classGroup1 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">2</span><span class="op">)</span>, classGroup2 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">4</span>, <span class="fl">5</span><span class="op">)</span>,
                        groupName1 <span class="op">=</span> <span class="st">"c2"</span>, groupName2 <span class="op">=</span> <span class="st">"c4_c5"</span>, 
                        analysisName <span class="op">=</span> <span class="st">"c2_VS_c4_c5"</span><span class="op">)</span></code></pre></div>
</details><br><p><strong>Results</strong></p>
<p>The results are saved in the <code>metadata</code> slot of the returned SCE object</p>
<details><summary><b>Result list structure</b></summary><pre><code>matadata(sce)
          |- $info1
          |- $info2
          |- ...
          |- $diffExp
                  |-$AnalysisName1
                  |-$AnalysisName2
                  |-...
                  |-$c2_VS_c3_c8
                             |-$useAssay    = "logcounts"
                             |-$groupNames  = c("c2", "c3_c8")
                             |-$select
                             |      |-$ix1  = c(FALSE, TRUE,  FALSE, ...)
                             |      |-$ix2  = c(FALSE, FALSE, FALSE, ...)
                             |
                             |        (Two logical vectors both having `ncol(inSCE)` values,
                             |         specifies which cells are selected for "c2" or "c3_c8")
                             |
                             |-$annotation  = "Seurat_louvain_Resolution0.8"
                             |-$result      = (the `data.frame` of top DEG table, shown below)
                             |-$method      = "MAST"</code></pre>
</details><br><p>To fetch the result as a table of the Top differential expressed genes and the statistics:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">DEG</span> <span class="op">&lt;-</span> <span class="fu">metadata</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">$</span><span class="va">diffExp</span><span class="op">$</span><span class="va">c2_VS_c4_c5</span><span class="op">$</span><span class="va">result</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">DEG</span><span class="op">)</span></code></pre></div>
<p>To visualize the result in a heatmap:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../../reference/plotDEGHeatmap.html">plotDEGHeatmap</a></span><span class="op">(</span><span class="va">sce</span>, useResult <span class="op">=</span> <span class="st">"c2_VS_c4_c5"</span>, log2fcThreshold <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></code></pre></div>
<p>For other approaches of visualizing top differential expressed genes, refer to <code><a href="../../reference/plotDEGRegression.html">plotDEGRegression()</a></code> and <code><a href="../../reference/plotDEGViolin.html">plotDEGViolin()</a></code>, which are borrowed from <a href="https://www.bioconductor.org/packages/release/bioc/vignettes/MAST/inst/doc/MAITAnalysis.html#401_visualization_of_50_most_differentially_expressed_genes">MAST’s tutorial</a>.</p>
</div>
<script>
document.getElementById("ia-button").click();
</script>
</div>
<div id="references" class="section level2 unnumbered">
<h2 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h2>
<div id="refs" class="references csl-bib-body">
<div id="ref-Aaron2016" class="csl-entry">
<div class="csl-left-margin">[1] </div>
<div class="csl-right-inline">J. C. M. Aaron T. L. Lun Davis J. McCarthy, <span>“<span class="nocase">A step-by-step workflow for low-level analysis of single-cell RNA-seq data with Bioconductor</span>,”</span> <em>F1000Research</em>, 2016, doi: <a href="https://doi.org/10.12688/f1000research.9501.2">10.12688/f1000research.9501.2</a>.</div>
</div>
<div id="ref-Finak2015" class="csl-entry">
<div class="csl-left-margin">[2] </div>
<div class="csl-right-inline">M. Y. Greg Finak Andrew McDavid, <span>“<span class="nocase">MAST: a flexible statistical framework for assessing transcriptional changes and characterizing heterogeneity in single-cell RNA sequencing data</span>,”</span> <em>Genome Biology</em>, vol. 16, no. 278, Dec. 2015, doi: <a href="https://doi.org/10.1186/s13059-015-0844-5">10.1186/s13059-015-0844-5</a>.</div>
</div>
<div id="ref-Ritchie2015" class="csl-entry">
<div class="csl-left-margin">[3] </div>
<div class="csl-right-inline">D. W. Matthew E. Ritchie Belinda Phipson, <span>“<span class="nocase">limma powers differential expression analyses for RNA-sequencing and microarray studies</span>,”</span> <em>Nucleic Acids Research</em>, vol. 43, p. e47, Apr. 2015, doi: <a href="https://doi.org/10.1093/nar/gkv007">10.1093/nar/gkv007</a>.</div>
</div>
<div id="ref-Love2014" class="csl-entry">
<div class="csl-left-margin">[4] </div>
<div class="csl-right-inline">S. A. Michael I Love Wolfgang Huber, <span>“<span class="nocase">Moderated estimation of fold change and dispersion for RNA-seq data with DESeq2</span>,”</span> <em>Genome Biology</em>, vol. 15, no. 550, Dec. 2014, doi: <a href="https://doi.org/10.1186/s13059-014-0550-8">10.1186/s13059-014-0550-8</a>.</div>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by David Jenkins, Vidya Akavoor, Salam Alabdullatif, Shruthi Bandyadka, Emma Briars, Xinyun Cao, Sebastian Carrasco Pro, Tyler Faits, Rui Hong, Mohammed Muzamil Khan, Yusuke Koga, Anastasia Leshchyk, Irzam Sarfraz, Yichen Wang, Zhe Wang, W. Evan Johnson, Joshua David Campbell.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
