<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Installation • singleCellTK</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/yeti/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../../bootstrap-toc.css">
<script src="../../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../../pkgdown.css" rel="stylesheet">
<script src="../../pkgdown.js"></script><link href="../../extra.css" rel="stylesheet">
<script src="../../extra.js"></script><meta property="og:title" content="Installation">
<meta property="og:description" content="singleCellTK">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../../index.html">singleCellTK</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">2.2.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../../articles/articles/installation.html">Installation</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Documentation
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Data</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../../articles/articles/import_data.html">Import Data</a>
        </li>
        <li>
          <a href="../../articles/articles/cmd_import_scRNAseq_data_as_SCE.html">Import Data in Commandline</a>
        </li>
        <li>
          <a href="../../articles/articles/import_genesets.html">Import Gene Sets</a>
        </li>
        <li>
          <a href="../../articles/articles/import_annotation.html">Annotation</a>
        </li>
        <li>
          <a href="../../articles/articles/export_data.html">Export Data</a>
        </li>
        <li>
          <a href="../../articles/articles/delete_data.html">Delete Data</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">QC &amp; Filtering</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../../articles/articles/cmd_qc.html">QC with commandline</a>
        </li>
        <li>
          <a href="../../articles/articles/cnsl_cellqc.html">Cell QC in R console</a>
        </li>
        <li>
          <a href="../../articles/articles/cnsl_dropletqc.html">Droplet QC in R console</a>
        </li>
        <li>
          <a href="../../articles/articles/ui_qc.html">QC in the UI</a>
        </li>
        <li>
          <a href="../../articles/articles/filtering.html">Filtering</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Normalization &amp; Batch Correction</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../../articles/articles/cnsl_normalization.html">Normalization</a>
        </li>
        <li>
          <a href="../../articles/articles/batch_correction.html">Batch Correction</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Feature Selection &amp; Dimensionality Reduction</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../../articles/articles/cnsl_feature_selection.html">Feature Selection</a>
        </li>
        <li>
          <a href="../../articles/articles/cnsl_dimensionality_reduction.html">Dimensionality Reduction</a>
        </li>
        <li>
          <a href="../../articles/articles/cnsl_2d_embedding.html">2D Embedding</a>
        </li>
      </ul>
</li>
    <li>
      <a href="../../articles/articles/clustering.html">Clustering</a>
    </li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Differential Expression &amp; Cell Type Labeling</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../../articles/articles/differential_expression.html">Differential Expression</a>
        </li>
        <li>
          <a href="../../articles/articles/find_marker.html">Find Marker</a>
        </li>
        <li>
          <a href="../../articles/articles/cell_type_labeling.html">Cell Type Labeling</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Pathway &amp; Enrichment Analysis</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../../articles/articles/cnsl_enrichR.html">EnrichR</a>
        </li>
        <li>
          <a href="../../articles/articles/pathwayAnalysis.html">Pathway Analysis</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Curated Workflows</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../../articles/articles/cnsl_seurat_curated_workflow.html">Seurat</a>
        </li>
        <li>
          <a href="../../articles/articles/celda_curated_workflow.html">Celda</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Visualization</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../../articles/articles/visualization.html">General Visualization</a>
        </li>
        <li>
          <a href="../../articles/articles/heatmap.html">Generic Heatmap</a>
        </li>
      </ul>
</li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Tutorial
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../../articles/articles/console_analysis_tutorial.html">Console Analysis Tutorial</a>
    </li>
  </ul>
</li>
<li>
  <a href="../../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/compbiomed/singleCellTK">
    <span class="fas fa-github"></span>
     
  </a>
</li>
<li>
  <a href="https://twitter.com/camplab1">
    <span class="fas fa-twitter"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="installation_files/header-attrs-2.7/header-attrs.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Installation</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/compbiomed/singleCellTK/blob/master/vignettes/articles/installation.Rmd"><code>vignettes/articles/installation.Rmd</code></a></small>
      <div class="hidden name"><code>installation.Rmd</code></div>

    </div>

    
    
<div id="quick-overview" class="section level2">
<h2 class="hasAnchor">
<a href="#quick-overview" class="anchor"></a>Quick Overview</h2>
<ol style="list-style-type: decimal">
<li>
<strong>R (&gt;=4.0)</strong>.</li>
</ol>
<p>SingleCellTK is an R based package, allowing users to run analysis in system command-line, R console, and R Shiny user interface. To get your machine fully equipped with SCTK, you must have <strong>R (&gt;=4.0)</strong> pre-installed. <a href="#install-r-4-0-">Install R</a></p>
<ol start="2" style="list-style-type: decimal">
<li><strong>singleCellTK</strong></li>
</ol>
<p>With R successfully installed, users can now start to install SingleCellTK to your local. For quick start, BioConductor users can directly requisite our released version with the following command:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span>

<span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span><span class="op">(</span><span class="st">"singleCellTK"</span><span class="op">)</span></code></pre></div>
<p><a href="#install-sctk">See detail</a>.</p>
<ol start="3" style="list-style-type: decimal">
<li><strong>Python environment</strong></li>
</ol>
<p>Additionally, SCTK wraps a number of useful toolkits, where a part of them are Python based. To enable this part of functionality, users need to get Python (3.x) and a set of Python packages built. The fast <code>pip</code> command to install all necessary dependencies is here:</p>
<pre class="shell"><code>pip install scipy numpy astroid six scrublet scanpy bbknn scanorama</code></pre>
<p><a href="#install-python-dependencies">Detail for Python Environment establishing</a></p>
<p>For users who would like to perform console interactive analysis, we strongly suggest that you have <a href="https://www.rstudio.com/products/rstudio/download/#download">RStudio</a> installed.</p>
</div>
<div id="detail" class="section level2">
<h2 class="hasAnchor">
<a href="#detail" class="anchor"></a>Detail</h2>
<div id="install-r-4-0" class="section level3">
<h3 class="hasAnchor">
<a href="#install-r-4-0" class="anchor"></a>Install R (&gt;=4.0)</h3>
<p>Users can go to <a href="https://www.r-project.org/"><strong>R Official Site</strong></a> to choose the preferred installer.</p>
<p>In the site, users will need to first <strong>specify a CRAN mirror</strong>. CRAN is a network of servers around the world that store approved R dependencies. Specifying a mirror that is the nearest to you helps minimize network load.</p>
<p><img src="ui_screenshots/installation/RChooseMirror.png" alt="chooseMirror"><br></p>
<p>Linux, Windows and MacOS users will be able to find the corresponding installer at the top of the page after choosing a mirror.</p>
<p><img src="ui_screenshots/installation/RSystem.png" alt="chooseMirror"><br></p>
</div>
<div id="install-sctk" class="section level3">
<h3 class="hasAnchor">
<a href="#install-sctk" class="anchor"></a>Install SCTK</h3>
<div id="latest-release-bioconductor" class="section level4">
<h4 class="hasAnchor">
<a href="#latest-release-bioconductor" class="anchor"></a>Latest Release (BioConductor)</h4>
<p>Our latest stable release is available on BioConductor. BioConductor is a distribution platform for biology-related packages. The page for SCTK can be found <a href="https://bioconductor.org/packages/release/bioc/html/singleCellTK.html">here</a></p>
<p>Run the following command in <strong>R console</strong> to install SCTK from BioConductor:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span>

<span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span><span class="op">(</span><span class="st">"singleCellTK"</span><span class="op">)</span></code></pre></div>
</div>
<div id="developing-version-github" class="section level4">
<h4 class="hasAnchor">
<a href="#developing-version-github" class="anchor"></a>Developing version (GitHub)</h4>
<p>We occasionally push new developmental updates of our source code on GitHub.</p>
<p>The simplest way to install the developing version is as followed:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html">requireNamespace</a></span><span class="op">(</span><span class="st">"devtools"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span>
    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"devtools"</span><span class="op">)</span>
<span class="op">}</span>
<span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://devtools.r-lib.org//reference/remote-reexports.html">install_github</a></span><span class="op">(</span><span class="st">"compbiomed/singleCellTK@devel"</span><span class="op">)</span></code></pre></div>
<p>Meanwhile, the source code can be downloaded <a href="https://github.com/compbiomed/singleCellTK/releases">Here</a></p>
<p><img src="ui_screenshots/installation/sourceCodeAssets.png" alt="chooseMirror"><br></p>
<p>To install from source code, you need to first <strong>download the tarball file (ending with <code>.tar.gz</code>)</strong>, which is a compressed file for the whole directory. Then <strong>in R console</strong>, run the following command:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># REPLACE {PATH} with the path you place the downloaded tarball file</span>
<span class="co"># REPLACE {VERSION} with the version number in the filename</span>
<span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"{PATH}/singleCellTK_{VERSION}.tar.gz"</span>, repos <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span></code></pre></div>
</div>
</div>
<div id="install-python-dependencies" class="section level3">
<h3 class="hasAnchor">
<a href="#install-python-dependencies" class="anchor"></a>Install Python Dependencies</h3>
<p>SCTK can be successfully loaded without a Python installation. While there are some parts of our analysis that support multiple methods, some options among them would require a live Python environment. For example, the quality control and batch correction sections.</p>
<div id="non-python-users" class="section level4">
<h4 class="hasAnchor">
<a href="#non-python-users" class="anchor"></a>Non-Python users</h4>
<p>If you have not had Python installed before, and do not have a plan to use Python. We have prepared a simple command to minimize the workload for you to clear the barrier. <strong>After installing SCTK</strong> in your R console:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://compbiomed.github.io/sctk_docs/">singleCellTK</a></span><span class="op">)</span>
<span class="fu"><a href="../../reference/sctkPythonInstallConda.html">sctkPythonInstallConda</a></span><span class="op">(</span><span class="op">)</span>
<span class="co">### Alternatively</span>
<span class="co"># sctkPythonInstallVirtualEnv()</span></code></pre></div>
</div>
<div id="work-with-python" class="section level4">
<h4 class="hasAnchor">
<a href="#work-with-python" class="anchor"></a>Work with Python</h4>
<p>If you would like to start to work with Python or are already an experienced Python programmer, here we skip the approach to the Python installation. If you have a standalone Python installation, it is not necessary to run the command above because it is likely to create a redundant Miniconda copy in your machine. For the list of dependencies that are required, please refer to the <code>pip</code> command at the top of this page, and you can also replace <code>pip</code> with <code>conda</code>.</p>
<p>If you would like to work in a Python virtual environment (“venv”) or a Conda environment, and have already created one, please run the environment selecting function in R console before starting to analyze your data. See the next paragraph.</p>
</div>
<div id="for-all-users" class="section level4">
<h4 class="hasAnchor">
<a href="#for-all-users" class="anchor"></a>For all users</h4>
<p>If you ran <code><a href="../../reference/sctkPythonInstallConda.html">sctkPythonInstallConda()</a></code> (or <code><a href="../../reference/sctkPythonInstallVirtualEnv.html">sctkPythonInstallVirtualEnv()</a></code>) directly, or have created your own SCTK specifc environment (or venv), it is important that you <strong>activate the environment before starting</strong> to perform any analyze. Otherwise, it will be hard for R to locate the dependencies you have installed.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">### For Python venv user</span>
<span class="fu"><a href="../../reference/selectSCTKVirtualEnvironment.html">selectSCTKVirtualEnvironment</a></span><span class="op">(</span><span class="st">"{NAME of your VENV}"</span><span class="op">)</span>
<span class="co">### For conda environment user</span>
<span class="fu"><a href="../../reference/selectSCTKConda.html">selectSCTKConda</a></span><span class="op">(</span><span class="st">"{NAME of your VENV}"</span><span class="op">)</span></code></pre></div>
</div>
</div>
</div>
<div id="troubleshooting-installation" class="section level2">
<h2 class="hasAnchor">
<a href="#troubleshooting-installation" class="anchor"></a>Troubleshooting Installation</h2>
<div id="bioconductor-versioning" class="section level3">
<h3 class="hasAnchor">
<a href="#bioconductor-versioning" class="anchor"></a>BioConductor Versioning</h3>
<p>For the majority of users, the commands above will install the latest version of the singleCellTK without any errors.</p>
<p>Rarely, you may encounter an error due to previously installed versions of some packages that are required for the singleCellTK. If you encounter an error during installation, use the commands below to check the version of Bioconductor that is installed:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span>
<span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/version.html">version</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<p>If the version number is not 3.6 or higher, you must upgrade Bioconductor to install the toolkit:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<p>After you install Bioconductor 3.6 or higher, you should be able to install the toolkit. If you still encounter an error, ensure your Bioconductor packages are up to date by running the following command.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/valid.html">valid</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<p>If the command above does not return <code>TRUE</code>, run the following command to update your R packages:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<p>Then, try to install the toolkit again.</p>
</div>
<div id="python-and-reticulate" class="section level3">
<h3 class="hasAnchor">
<a href="#python-and-reticulate" class="anchor"></a>Python and Reticulate</h3>
<p>SCTK adopts <a href="https://rstudio.github.io/reticulate/">reticulate</a> to connect R with Python. If the specific functions in SCTK, which requires Python dependencies, do not properly start to run after you have installed the dependencies or restart the R session, please try to specify the correct Python executable/binary file in a <strong>new R session</strong>. (here “new” means no Python or reticulate related operations have been performed)</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/rstudio/reticulate">reticulate</a></span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/reticulate/man/use_python.html">use_python</a></span><span class="op">(</span><span class="st">"{PATH to the correct Python exexutable/binary file}"</span><span class="op">)</span></code></pre></div>
<p>You can always use <code><a href="https://rdrr.io/pkg/reticulate/man/py_config.html">reticulate::py_config()</a></code> to check what specific Python version you are using and what other versions are available. <code><a href="https://rdrr.io/pkg/reticulate/man/py_config.html">py_config()</a></code> configures the R-to-Python connection at once, so you will have to restart the R session if you need to change the configuration with <code><a href="https://rdrr.io/pkg/reticulate/man/use_python.html">use_python()</a></code> after you have the check done.</p>
<p>If you still encounter an error, please <a href="https://github.com/compbiomed/singleCellTK/issues">raise an Issue</a> at our GitHub repository and we’d be happy to help.</p>
</div>
</div>
<div id="develop-singlecelltk" class="section level2">
<h2 class="hasAnchor">
<a href="#develop-singlecelltk" class="anchor"></a>Develop singleCellTK</h2>
<p>To contribute to singleCellTK, follow these steps:</p>
<p><strong>Note:</strong> Development of the singleCellTK is done using the latest version of R.</p>
<ol style="list-style-type: decimal">
<li>Fork the repo using the “Fork” button at the top-right corner of our <a href="https://github.com/compbiomed/singleCellTK">GitHub repository page</a>.</li>
<li>Download a local copy of <strong>your forked</strong> repository</li>
</ol>
<pre class="shell"><code>git clone https://github.com/{username}/singleCellTK.git</code></pre>
<ol start="3" style="list-style-type: decimal">
<li>Open Rstudio</li>
<li>Go to “File” -&gt; “New Project” -&gt; “Existing Directory” and select your git repository directory</li>
</ol>
<p>You can then make your changes and test your code using the Rstudio build tools. There is a lot of information about building packages available here: <a href="http://r-pkgs.had.co.nz/">http://r-pkgs.had.co.nz/</a>.</p>
<p>Information about building shiny packages is available here: <a href="http://shiny.rstudio.com/tutorial/">http://shiny.rstudio.com/tutorial/</a>.</p>
<p>When you are ready to upload your changes, commit them locally, push them to your forked repo, and make a pull request to the compbiomed repository.</p>
<p>Report bugs and request features on our <a href="https://github.com/compbiomed/singleCellTK/issues">GitHub issue tracker</a>.</p>
<p>Join us on <a href="https://compbiomed.slack.com/">slack</a>!</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by David Jenkins, Vidya Akavoor, Salam Alabdullatif, Shruthi Bandyadka, Emma Briars, Xinyun Cao, Sebastian Carrasco Pro, Tyler Faits, Rui Hong, Mohammed Muzamil Khan, Yusuke Koga, Anastasia Leshchyk, Irzam Sarfraz, Yichen Wang, Zhe Wang, W. Evan Johnson, Joshua David Campbell.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
