<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Comprehensive and Interactive Analysis of Single Cell RNA-Seq Data • singleCellTK</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/yeti/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><link href="extra.css" rel="stylesheet">
<script src="extra.js"></script><meta property="og:title" content="Comprehensive and Interactive Analysis of Single Cell RNA-Seq Data">
<meta property="og:description" content="The Single Cell Toolkit (SCTK) in the singleCellTK package provides an interface to popular tools for importing, quality control, analysis, and visualization of single cell RNA-seq data. SCTK allows users to seamlessly integrate tools from various packages at different stages of the analysis workflow. A general “a la carte” workflow gives users the ability access to multiple methods for data importing, calculation of general QC metrics, doublet detection, ambient RNA estimation and removal, filtering, normalization, batch correction or integration, dimensionality reduction, 2-D embedding, clustering, marker detection, differential expression, cell type labeling, pathway analysis, and data exporting. Curated workflows can be used to run Seurat and Celda. Streamlined quality control can be performed on the command line using the SCTK-QC pipeline. Users can analyze their data using commands in the R console or by using an interactive Shiny Graphical User Interface (GUI). Specific analyses or entire workflows can be summarized and shared with comprehensive HTML reports generated by Rmarkdown. Additional documentation and vignettes can be found at camplab.net/sctk.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">singleCellTK</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">2.4.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="https://www.camplab.net/sctk">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="articles/articles/installation.html">Installation</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Documentation
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Data</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="articles/articles/import_data.html">Import Data</a>
        </li>
        <li>
          <a href="articles/articles/cmd_import_scRNAseq_data_as_SCE.html">Import Data for Commandline QC</a>
        </li>
        <li>
          <a href="articles/articles/import_genesets.html">Import Gene Sets</a>
        </li>
        <li>
          <a href="articles/articles/import_annotation.html">Annotation</a>
        </li>
        <li>
          <a href="articles/articles/export_data.html">Export Data</a>
        </li>
        <li>
          <a href="articles/articles/delete_data.html">Delete Data</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">QC &amp; Filtering</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="articles/articles/cnsl_dropletqc.html">Droplet QC (R console)</a>
        </li>
        <li>
          <a href="articles/articles/cnsl_cellqc.html">Cell QC (R console)</a>
        </li>
        <li>
          <a href="articles/articles/ui_qc.html">Cell QC (Shiny UI)</a>
        </li>
        <li>
          <a href="articles/articles/filtering.html">Filtering</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Normalization &amp; Batch Correction</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="articles/articles/cnsl_normalization.html">Normalization</a>
        </li>
        <li>
          <a href="articles/articles/batch_correction.html">Batch Correction</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Feature Selection &amp; Dimensionality Reduction</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="articles/articles/cnsl_feature_selection.html">Feature Selection</a>
        </li>
        <li>
          <a href="articles/articles/cnsl_dimensionality_reduction.html">Dimensionality Reduction</a>
        </li>
        <li>
          <a href="articles/articles/cnsl_2d_embedding.html">2D Embedding</a>
        </li>
      </ul>
</li>
    <li>
      <a href="articles/articles/clustering.html">Clustering</a>
    </li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Differential Expression &amp; Cell Type Labeling</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="articles/articles/differential_expression.html">Differential Expression</a>
        </li>
        <li>
          <a href="articles/articles/find_marker.html">Find Marker</a>
        </li>
        <li>
          <a href="articles/articles/cell_type_labeling.html">Cell Type Labeling</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Pathway &amp; Enrichment Analysis</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="articles/articles/cnsl_enrichR.html">EnrichR</a>
        </li>
        <li>
          <a href="articles/articles/pathwayAnalysis.html">Pathway Analysis</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Curated Workflows</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="articles/articles/cnsl_seurat_curated_workflow.html">Seurat</a>
        </li>
        <li>
          <a href="articles/articles/celda_curated_workflow.html">Celda</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Visualization</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="articles/articles/visualization.html">General Visualization</a>
        </li>
        <li>
          <a href="articles/articles/heatmap.html">Generic Heatmap</a>
        </li>
      </ul>
</li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/articles/console_analysis_tutorial.html">Analysis of PBMC3K (R Console)</a>
    </li>
  </ul>
</li>
<li>
  <a href="articles/articles/cmd_qc.html">SCTK-QC</a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/compbiomed/singleCellTK">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="https://twitter.com/camplab1">
    <span class="fab fa-twitter fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
<div id="single-cell-tk" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#single-cell-tk" class="anchor"></a>Single Cell TK</h1></div>
<p><!-- badges: start --> <a href="https://github.com/compbiomed/singleCellTK/actions"><img src="https://github.com/compbiomed/singleCellTK/workflows/R-CMD-check/badge.svg" alt="R-CMD-check"></a> <a href="https://codecov.io/gh/compbiomed/singleCellTK"><img src="https://codecov.io/gh/compbiomed/singleCellTK/branch/devel/graph/badge.svg" alt="codecov"></a> <!-- badges: end --></p>
<p>The Single Cell ToolKit (SCTK) is an analysis platform that provides an <strong>R interface to several popular single-cell RNA-sequencing (scRNAseq) data preprocessing, quality control, analysis, and visualization tools</strong>. SCTK imports raw or filtered counts from various scRNAseq preprocessing tools such as 10x CellRanger, BUStools, Optimus, STARSolo, and more. By integrating several publicly available tools written in R or Python, SCTK can be used to perform extensive quality control including doublet detection, ambient RNA removal, and batch effect correction. SCTK integrates analysis workflows from popular tools such as Seurat and Bioconductor/OSCA into a single unified framework. Results from various workflows can be summarized and easily shared using comprehensive HTML reports. Lastly, data can be exported to Seurat or AnnData object to allow for seamless integration with other downstream analysis workflows. More information about the toolkit can be found at the toolkit <a href="https://camplab.net/sctk/">homepage</a>.</p>
<div id="installation" class="section level2">
<h2 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h2>
<p>Detailed instruction on how to install SCTK and additional dependencies are available at our homepage: <a href="https://camplab.net/sctk/" class="uri">https://camplab.net/sctk/</a></p>
</div>
<div id="features" class="section level2">
<h2 class="hasAnchor">
<a href="#features" class="anchor"></a>Features</h2>
<p>SCTK offers multiple ways to analyze your scRNAseq data both through the R console, commandline (QC) and graphical user interface (GUI) with the ability to use a large number of algorithms from both R &amp; Python integrated within the toolkit.</p>
<div id="console-analysis" class="section level4">
<h4 class="hasAnchor">
<a href="#console-analysis" class="anchor"></a>Console Analysis</h4>
<p>Traditional analysis of scRNAseq data can be performed in the R console using wrapper functions for a multitude of tools and algorithms.</p>
</div>
<div id="interactive-analysis" class="section level4">
<h4 class="hasAnchor">
<a href="#interactive-analysis" class="anchor"></a>Interactive Analysis</h4>
<p>The Shiny APP allows users without programming experience to easily analyze their scRNAseq data with a GUI.</p>
</div>
<div id="reports" class="section level4">
<h4 class="hasAnchor">
<a href="#reports" class="anchor"></a>Reports</h4>
<p>Comprehensive HTML reports developed with RMarkdown allows users to document, explore, and share their analyses.</p>
</div>
<div id="interoperability" class="section level4">
<h4 class="hasAnchor">
<a href="#interoperability" class="anchor"></a>Interoperability</h4>
<p>Tools from both R and Python can be seamlessly integrated within the same analysis workflow.</p>
</div>
</div>
<div id="report-issues" class="section level2">
<h2 class="hasAnchor">
<a href="#report-issues" class="anchor"></a>Report Issues</h2>
<p>If you face any difficulty in installing or have identified a bug in the toolkit, please feel free to open up an <a href="https://github.com/compbiomed/singleCellTK/issues">Issue</a> on GitHub. Questions about how to best analyze your scRNA-seq data can be asked in the <a href="https://github.com/compbiomed/singleCellTK/discussions">Discussions</a> page on GitHub.</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="links">
<h2>Links</h2>
<ul class="list-unstyled">
<li>Download from BIOC at <br><a href="https://www.bioconductor.org/packages/singleCellTK">https://​www.bioconductor.org/​packages/​singleCellTK</a>
</li>
<li>Browse source code at <br><a href="https://github.com/compbiomed/singleCellTK/">https://​github.com/​compbiomed/​singleCellTK/​</a>
</li>
<li>Report a bug at <br><a href="https://github.com/compbiomed/singleCellTK/issues">https://​github.com/​compbiomed/​singleCellTK/​issues</a>
</li>
</ul>
</div>
<div class="license">
<h2>License</h2>
<ul class="list-unstyled">
<li>
<a href="https://opensource.org/licenses/mit-license.php">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a>
</li>
</ul>
</div>
<div class="developers">
<h2>Developers</h2>
<ul class="list-unstyled">
<li>Yichen Wang <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0003-4347-5199" target="orcid.widget" aria-label="ORCID"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Irzam Sarfraz <br><small class="roles"> Author </small>  </li>
<li>Rui Hong <br><small class="roles"> Author </small>  </li>
<li>Yusuke Koga <br><small class="roles"> Author </small>  </li>
<li>Salam Alabdullatif <br><small class="roles"> Author </small>  </li>
<li>David Jenkins <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0002-7451-4288" target="orcid.widget" aria-label="ORCID"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Vidya Akavoor <br><small class="roles"> Author </small>  </li>
<li>Xinyun Cao <br><small class="roles"> Author </small>  </li>
<li>Shruthi Bandyadka <br><small class="roles"> Author </small>  </li>
<li>Anastasia Leshchyk <br><small class="roles"> Author </small>  </li>
<li>Tyler Faits <br><small class="roles"> Author </small>  </li>
<li>Mohammed Muzamil Khan <br><small class="roles"> Author </small>  </li>
<li>Zhe Wang <br><small class="roles"> Author </small>  </li>
<li>W. Evan Johnson <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0002-6247-6595" target="orcid.widget" aria-label="ORCID"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Joshua David Campbell <br><small class="roles"> Author </small>  </li>
</ul>
</div>

  </div>
</div>


      <footer><div class="copyright">
  <p>Developed by Yichen Wang, Irzam Sarfraz, Rui Hong, Yusuke Koga, Salam Alabdullatif, David Jenkins, Vidya Akavoor, Xinyun Cao, Shruthi Bandyadka, Anastasia Leshchyk, Tyler Faits, Mohammed Muzamil Khan, Zhe Wang, W. Evan Johnson, Joshua David Campbell.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
