<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Generation of comprehensive quality control metrics with SCTK • singleCellTK</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Generation of comprehensive quality control metrics with SCTK">
<meta property="og:description" content="singleCellTK">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">singleCellTK</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.9.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/introduction_to_SCTK.html">General Workflow</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Shell Command Line Analysis
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/import_scRNAseq_data_as_SCE_Commandline.html">Import data into SCTK</a>
    </li>
    <li>
      <a href="../articles/run_qcCommandline.Rmd">Quality Control</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    R Console Analysis
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/import_scRNAseq_data_as_SCE.html">Import data into SCTK</a>
    </li>
    <li>
      <a href="../articles/run_qcConsole.html">Quality Control</a>
    </li>
    <li>
      <a href="../articles/normalization.html">Normalization</a>
    </li>
    <li>
      <a href="../articles/batch_correction.html">Batch Correction</a>
    </li>
    <li>
      <a href="../articles/variability.html">Variability</a>
    </li>
    <li>
      <a href="../articles/feature_selection.html">Feature Selection</a>
    </li>
    <li>
      <a href="../articles/differential_expression.html">Differential Expression</a>
    </li>
    <li>
      <a href="../articles/find_marker.html">Find Marker via DE</a>
    </li>
    <li>
      <a href="../articles/visualize_results.html">Visualize Results</a>
    </li>
    <li>
      <a href="../articles/heatmap.html">Heatmap Visualization</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Shiny UI Analysis
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/import_scRNAseq_data_as_SCE.html">I am Place Holder 1</a>
    </li>
    <li>
      <a href="../articles/import_scRNAseq_data_as_SCE.html">I am Place Holder 2</a>
    </li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/compbiomed/singleCellTK">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Generation of comprehensive quality control metrics with SCTK</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/compbiomed/singleCellTK/blob/master/vignettes/run_qcCommandline.Rmd"><code>vignettes/run_qcCommandline.Rmd</code></a></small>
      <div class="hidden name"><code>run_qcCommandline.Rmd</code></div>

    </div>

    
    
<p>This pipeline will import data from single-cell preprocessing algorithms (e.g. CellRanger), generate various quality control metrics (e.g. doublet scores), and output results in standard data containers (e.g. SingleCellExperiment). Both the original droplet matrix and the filtered cell matrix will be processed. This pipeline is focused on single cell data generated from microfluidic devices (e.g. 10X).</p>
<div id="specifications" class="section level2">
<h2 class="hasAnchor">
<a href="#specifications" class="anchor"></a>Specifications</h2>
<ul>
<li>The pipeline is currently written in the R language. Users will need to install R version 3.6.2 (or higher) in order to run all of the required packages.</li>
<li>For importing files from the HCA Optimus pipeline, the “scipy” module needs to be installed in the default version of Python on the system.</li>
<li>The pipeline depends on some Python package for Python &gt; 3.0.0. User will need to install Python 3.6.3 (or higher) in order to install the proper version of Python packages.</li>
</ul>
</div>
<div id="installation" class="section level2">
<h2 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h2>
<div id="docker-users" class="section level3">
<h3 class="hasAnchor">
<a href="#docker-users" class="anchor"></a>Docker users</h3>
<p>If you are using docker image to run the pipeline, please skip this section and refer sections Running the pipeline and Docker and Singularity Images. Docker image has all dependencies installed properly so that you can use it directly on any machine.</p>
</div>
<div id="install-python-packages-and-dependencies" class="section level3">
<h3 class="hasAnchor">
<a href="#install-python-packages-and-dependencies" class="anchor"></a>Install Python packages and dependencies</h3>
<p>This pipeline depends on Python package described below. If these packages is not installed or not at the proper version, please install them by running the following code:</p>
<div class="sourceCode"><html><body><pre class="r">pip3 install --user h5py==2.9.0
pip3 install --user Scrublet
pip3 install --user virtualenv
pip3 install --user scanpy</pre></body></html></div>
</div>
<div id="install-r-packages-and-dependencies" class="section level3">
<h3 class="hasAnchor">
<a href="#install-r-packages-and-dependencies" class="anchor"></a>Install R packages and dependencies</h3>
<p>The script will automatically try to install the “singleCellTK” package from Bioconductor if not available. However, currently this code is only located on a development branch which needs to be installed from GitHub:</p>
<div class="sourceCode"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">devtools</span>)
<span class="fu"><a href="https://rdrr.io/pkg/devtools/man/remote-reexports.html">install_github</a></span>(<span class="st">"compbiomed/singleCellTK@devel"</span>)</pre></body></html></div>
</div>
</div>
<div id="data-formats" class="section level2">
<h2 class="hasAnchor">
<a href="#data-formats" class="anchor"></a>Data formats</h2>
<p>For each tool, the pipeline expect the data in a certain format within a specific directory structure. For example, data generated with CellRanger V3 is expected to be under a sample folder (specified with the “-s” flag) and a base folder (specified with the “-b” flag). Some tools prepend the sample name onto the output files instead of making a separate subdirectory (e.g. BUStools and SEQC). The combination of –base_path (“-b”) and –sample (“-s”) should specify the location of the data files: For a comprehensive description of expected input directory structure of import functions, refer documentation <a href="./import_scRNAseq_data_as_SCE.html">here</a></p>
</div>
<div id="arguments" class="section level2">
<h2 class="hasAnchor">
<a href="#arguments" class="anchor"></a>Arguments</h2>
<div id="required-arguments" class="section level3">
<h3 class="hasAnchor">
<a href="#required-arguments" class="anchor"></a>Required arguments</h3>
<p>The required arguments are as follows:</p>
<pre><code>-b, --basePath (required). Base path for the output from the preprocessing algorithm

-P, --preproc (required). Algorithm used for preprocessing. One of One of 'CellRangerV2', 'CellRangerV3', 'BUStools', 'STARSolo', 'SEQC', 'Optimus', 'DropEst', 'SceRDS', 'CountMatrix' and 'AnnData'.

-s, --sample (required). Name of the sample. This will be prepended to the cell barcodes.

-o, --directory (required). Output directory. A new subdirectory will be created with the name "sample". R, Python, and FlatFile directories will be created under the "sample" directory containing the data containers with QC metrics. Default ".". More information about output directory structure is explained in "Understanding outputs" section below.

-F, --outputFormat (required). The output format of this QC pipeline. Currently, it supports R (RDS), FlatFile, Python (AnnData) and HTAN (manifest files that meets HTAN requirement).

-S, --splitSample (required). Save a SingleCellExperiment object for each sample. Default is TRUE. If FALSE, the data of all samples will be combined into one SingleCellExperiment object and this object will be outputed.</code></pre>
</div>
<div id="optional-arguments" class="section level3">
<h3 class="hasAnchor">
<a href="#optional-arguments" class="anchor"></a>Optional arguments</h3>
<p>The optional arguments are as follows. Their usage depend on type of data and user-defined behaviour.</p>
<pre><code>-g, --gmt (optional). GMT file containing gene sets for quality control.

-t, --delim (optional, required when -g is specified). Delimiter used in GMT file. Default "\t".

-G, --genome (optional). The name of genome reference. This is only required for CellRangerV2 data.

-y, --yaml (optional). YAML file used to specify parameters of QC functions called by singleCellTK QC pipeline. Please check "Specify parameters for QC algorithms" section for details.

-c, --cellData (optional). The full path of the RDS file or Matrix file of the cell matrix. This would be use only when --preproc is SceRDS or CountMatrix.

-r, --rawData (optional). The full path of the RDS file or Matrix file of the droplet matrix. This would be provided only when --preproc is SceRDS or CountMatrix.

-C, --cellPath (optional). The directory contains matrix.mtx.gz, features.tsv.gz and barcodes.tsv.gz files originally generated by 10x CellrangerV2 or CellrangerV3 (files in the filtered_feature_bc_matrix directory). This argument only works when --preproc is CellRangerV2 or CellRangerV3. Default is NULL. If 'base_path' is NULL, 'cellPath' or 'rawPath' should be specified.

-R, --rawPath (optional). The directory contains matrix.mtx.gz, features.tsv.gz and barcodes.tsv.gz files originally generated by 10x CellrangerV2 or CellrangerV3 (files in the raw_feature_bc_matrix directory). This argument only works when --preproc is CellRangerV2 or CellRangerV3. Default is NULL. If 'base_path' is NULL, 'cellPath' or 'rawPath' should be specified.

-d, --dataType. Type of data as input. Default is Both, which means taking both droplet and cell matrix as input. If set as 'Droplet', it will only processes droplet data. If set as 'Cell', it will only processes cell data.

-D, --detectCells. Detect cells from droplet matrix. Default is FALSE. This argument is only eavluated when -d is 'Droplet'. If set as TRUE, cells will be detected and cell matrixed will be subset from the droplet matrix. Also, quality control will be performed on the detected cell matrix.

-m, --cellDetectMethod. Methods to detect cells from droplet matrix. Default is 'EmptyDrops'. This argument is only eavluated when -D is 'TRUE'. Other options could be 'Knee' or 'Inflection'. More information is provided in the section below.

-n, --numCores. Number of cores used to run the pipeline. By default is 1. Parallel computing is enabled if -n is greater than 1.

-T, --parallelType. Type of parallel computing used for parallel computing. Parallel computing used in this pipeline depends on "BiocParallel" package. Default is 'MulticoreParam'. It can be 'MulticoreParam' or 'SnowParam'. This argument will be evaluated only when numCores &gt; 1.</code></pre>
</div>
</div>
<div id="running-the-pipeline" class="section level2">
<h2 class="hasAnchor">
<a href="#running-the-pipeline" class="anchor"></a>Running the pipeline</h2>
<p>To run the pipeline script, users will need to download the ‘SCTK_runQC.R’ <a href="https://github.com/compbiomed/singleCellTK/blob/devel/exec/SCTK_runQC.R">here</a>.</p>
<div id="methods-to-load-cellrangerv2-or-cellrangerv3-data-set" class="section level3">
<h3 class="hasAnchor">
<a href="#methods-to-load-cellrangerv2-or-cellrangerv3-data-set" class="anchor"></a>Methods to load CellrangerV2 or CellrangerV3 data set</h3>
<p>This pipeline enables different ways to import CellrangerV2/CellrangerV3 data for flexibility.</p>
<ol style="list-style-type: decimal">
<li>If the cellranger data set is saved in the default <a href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/output/overview">cellranger output directory</a>, you can load the data by running following code:</li>
</ol>
<p>For CellRangerV3:</p>
<div class="sourceCode"><html><body><pre class="r">Rscript SCTK_runQC.R \
-b /base/path \
-P CellRangerV3 \
-s SampleName \
-o Output_Directory \
-S TRUE \
-F R,Python,FlatFile,HTAN</pre></body></html></div>
<p>For CellRangerV2, the reference used by cellranger needs to be specified by -G/–genome:</p>
<div class="sourceCode"><html><body><pre class="r">Rscript SCTK_runQC.R \
-b /base/path \
-P CellRangerV2 \
-s SampleName \
-o Output_Directory \
-S TRUE \
-G hg19 \
-F R,Python,FlatFile,HTAN</pre></body></html></div>
<ol start="2" style="list-style-type: decimal">
<li>If the cellranger count output have been moved out of the default <a href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/output/overview">cellranger output directory</a>, you can specified input using -R and -C:</li>
</ol>
<div class="sourceCode"><html><body><pre class="r">Rscript SCTK_runQC.R \
-P CellRangerV2 \
-C /path/to/cell/matrix \
-R /path/to/droplet/matrix \
-s SampleName \
-o Output_Directory \
-S TRUE \
-F R,Python,FlatFile,HTAN</pre></body></html></div>
<p>In this case, you must skip -b arguments and you can also skip -G argument for CellRangerV2 data.</p>
</div>
<div id="methods-to-import-data-set-stored-in-rds-file-or-txt-file-matrix" class="section level3">
<h3 class="hasAnchor">
<a href="#methods-to-import-data-set-stored-in-rds-file-or-txt-file-matrix" class="anchor"></a>Methods to import data set stored in RDS file or txt file (matrix)</h3>
<ol style="list-style-type: decimal">
<li>If your data in stored as a SingleCellExperiment in RDS file, singleCellTK also supports that type of input. To run quality control with RDS file as input, run the following code:</li>
</ol>
<div class="sourceCode"><html><body><pre class="r">Rscript SCTK_runQC.R \
-P SceRDS \
-s Samplename \
-o Output_Directory \
-S TRUE \
-F R,Python,FlatFile,HTAN \
-r /path/to/rds/file/droplet.RDS \
-c /path/to/rds/file/cell.RDS</pre></body></html></div>
<ol start="2" style="list-style-type: decimal">
<li>If your input is stored in txt file as a matrix, which has barcodes as colnames and genes as rownames, run the following code to start the quality control pipeline:</li>
</ol>
<div class="sourceCode"><html><body><pre class="r">Rscript SCTK_runQC.R \
-P CountMatrix \
-s Samplename \
-o Output_Directory \
-S TRUE \
-r /path/to/matrix/file/droplet.txt \
-c path/to/matrix/file/cell.txt</pre></body></html></div>
</div>
<div id="methods-to-run-pipeline-on-data-set-generated-by-other-algorithms" class="section level3">
<h3 class="hasAnchor">
<a href="#methods-to-run-pipeline-on-data-set-generated-by-other-algorithms" class="anchor"></a>Methods to run pipeline on data set generated by other algorithms</h3>
<p>If your data is preprocessed by other algorithms, you can load the data as shown in the figure above. Basically, the templated is shown below:</p>
<div class="sourceCode"><html><body><pre class="r">Rscript SCTK_runQC.R \
-b /base/path \
-P Preprocessing_Algorithm \
-s SampleName \
-o Output_Directory \
-S TRUE \
-F R,Python,FlatFile,HTAN</pre></body></html></div>
</div>
<div id="methods-to-run-pipeline-on-only-droplet-count-or-cell-count-data" class="section level3">
<h3 class="hasAnchor">
<a href="#methods-to-run-pipeline-on-only-droplet-count-or-cell-count-data" class="anchor"></a>Methods to run pipeline on only droplet count or cell count data</h3>
<p>User can choose to run QC pipeline on only one of the droplet count or cell count matrix, instead of running on both. In this case, the pipeline will only take single input and perform quality control on it. One of the example is shown below:</p>
<div class="sourceCode"><html><body><pre class="r">Rscript SCTK_runQC.R \
-b /base/path \
-P Preprocessing_Algorithm \
-s SampleName \
-o Output_Directory \
-S TRUE \
-F R,Python,FlatFile,HTAN \
-d Droplet \
-D TRUE \
-m EmptyDrops</pre></body></html></div>
<p>-d argument is used to specify which count matrix to used in the QC pipeline. Default is “Both”, which means take both droplet and cell count data as input.</p>
<p>If -d argument is set as “Droplet”, the QC pipeline will only take droplet count matrix as input and perform quality control. You can choose whether to detect cells from the droplet matrix by setting -D as TRUE. If yes, cell count matrix will be detected and the pipeline will also perform quality control on this matrix and output the result. You could further define the method used to detect cells from droplet matrix by setting -m argument. -m could be one of “EmptyDrops”, “Knee” or “Inflection”. “EmptyDrops” will keep cells that pass the “runEmptyDrops” function test. “Knee” and “Inflection” will keep cells that pass the knee or inflection point returned from “runBarcodeRankDrops” function.</p>
<p>If -d argument is set as “Cell”, the QC pipeline will only take cell count matrix as input and perform quality control. A figure showing the analysis steps and outputs of different inputs is shown below:</p>
<center>
<img src="qc_singleInput.png" alt="QC single input"> 
</center>
</div>
</div>
<div id="specify-parameters-for-qc-algorithms" class="section level2">
<h2 class="hasAnchor">
<a href="#specify-parameters-for-qc-algorithms" class="anchor"></a>Specify parameters for QC algorithms</h2>
<p>User can specify parameters for QC algorithms in this pipeline with a yaml file (supplied with -y/–yamlFile argument). The current supported QC algorithms including doublet dection (bcds, cxds, cxds_bcds_hybrid, doubletFinder, doubletCells and scrublet), decontamination (decontX), emptyDrop detection (emptyDrops) and barcodeRankDrops (barcodeRanks). A summary of each function is shown below:</p>
<center>
<img src="qc_yamlParameters.png" alt="QC yaml parameters"> 
</center>
<p>An example of QC parameters yaml file is shown below:</p>
<pre><code>---
Params:   ### should not be omitted
  bcds:
    ntop: 600

  cxds:
    ntop: 600

  cxds_bcds_hybrid:
    nTop: 600

  decontX:
    maxIter: 600

  emptyDrops:
    lower: 50
    niters: 5000
    testAmbient: True

  barcodeRanks:
    lower: 50</code></pre>
<p>The format of yaml file can be found <a href="https://docs.ansible.com/ansible/latest/reference_appendices/YAMLSyntax.html">here</a>. The parameters should be consistent with the parameters of each QC function in singleCellTK package. Parameters that are not defined in this yaml file will use the default value. Please refer <a href="../reference/index.html">reference</a> for detailed information about arguments in each QC function.</p>
</div>
<div id="analyzing-genes-sets" class="section level2">
<h2 class="hasAnchor">
<a href="#analyzing-genes-sets" class="anchor"></a>Analyzing genes sets</h2>
<p>Quantifying the level of gene sets can be useful quality control. For example, the percentage of counts from mitochondrial genes can be an indicator or cell stress or death.</p>
<p>Users can pass a <a href="http://software.broadinstitute.org/cancer/software/gsea/wiki/index.php/Data_formats">GMT</a> file (with -g/–gmt argument) to the pipeline with one row for each gene set. The first column should be the name of the gene set (e.g. mito). The second column for each gene set in the GMT file (i.e. the description) should contain the location of where to look for the matching IDs in the data. If set to ‘rownames’, then the gene set IDs will be matched with the row IDs of the data matrix. If a character string or an integer index is supplied, then gene set IDs will be matched to IDs the that column of feature table.</p>
</div>
<div id="parallel-computing" class="section level2">
<h2 class="hasAnchor">
<a href="#parallel-computing" class="anchor"></a>Parallel computing</h2>
<p>SCTK QC pipeline enables parallel computing to speed up the analysis. Parallel computing is enabled by setting -n/–numCores &gt; 1. The -n/–numCores is used to set the number of cores used for the pipeline.</p>
<p>The backend of parallel computing is supported by “BiocParallel” package. Therefore, users can select different types of parallel evaluation by setting -T/–parallelType argument. Default is “MulticoreParam”. Currently, “MulticoreParam” and “SnowParam” is supported for -T argument. However, “MulticoreParam” is not supported by Windows system. Winows user can choose “SnowParam” as the backend of parallel computing.</p>
</div>
<div id="understanding-outputs" class="section level2">
<h2 class="hasAnchor">
<a href="#understanding-outputs" class="anchor"></a>Understanding outputs</h2>
<p>The output directory is created under the path specified by -o/–directory argument. Each sample is stoed in the subdirectory (named by -s/–sample argument) within this output direcotry. Within each sample directory, the each output format will be separated into subdirectories. The output file hierarchy is shown below:</p>
<pre><code>(root; output directory)
├── level3Meta.csv
├── level4Meta.csv
└── sample1 
    ├── R
    |   ├── sample1_Droplets.rds
    |   └── sample1_Cells.rds
    ├── Python
    |   ├── Droplets
    |   |   └── sample1.h5ad
    |   └── Cells
    |       └── sample1.h5ad
    ├── FlatFile
    |   ├── Droplets
    |   |   ├── assays
    |   |   |   └── sample1_counts.mtx.gz
    |   |   ├── metadata
    |   |   |   └── sample1_metadata.rds
    |   |   ├── sample1_colData.txt.gz
    |   |   └── sample1_rowData.txt.gz 
    |   └── Cells
    |       ├── assays
    |       |   └── sample1_counts.mtx.gz
    |       ├── metadata
    |       |   └── sample1_metadata.rds
    |       ├── reducedDims
    |       |   ├──sample1_decontX_UMAP.txt.gz
    |       |   ├──sample1_scrublet_TSNE.txt.gz
    |       |   └──sample1_scrublet_UMAP.txt.gz
    |       ├── sample1_colData.txt.gz
    |       └── sample1_rowData.txt.gz 
    └── sample1_QCparameters.yaml</code></pre>
</div>
<div id="docker-and-singularity-images" class="section level2">
<h2 class="hasAnchor">
<a href="#docker-and-singularity-images" class="anchor"></a>Docker and Singularity Images</h2>
<p>singleCellTK is available to use with both Docker and Singularity. This container be used as a portal to the singleCellTK command line interface. The work was done in R. The singleCellTK docker image is available from <a href="https://hub.docker.com/r/campbio/sctk_qc">Docker Hub</a>.</p>
<div id="docker" class="section level3">
<h3 class="hasAnchor">
<a href="#docker" class="anchor"></a>Docker</h3>
<p>If you have not used docker before, you can follow the instruction to install and set up docker in <a href="https://docs.docker.com/docker-for-windows/">Windows</a>, <a href="https://docs.docker.com/docker-for-mac/">Mac</a> or <a href="https://runnable.com/docker/install-docker-on-linux">Linux</a>.</p>
<p>The Docker image can be obtained by running:</p>
<pre><code>docker pull campbio/sctk_qc</code></pre>
<p>Docker image will not access files in your host file system by default. To get access to the files on your machine, you can properly set up a mount volume. Noted that the transcriptome data and GMT file needed to be accessible to the container via mounted volume. In the below example, mount volumn is enabled for accessing input and output directory using argument -v. The transcriptome and GMT files inside <em>/path/to/data</em> is now available in <em>/SCTK_docker</em> folder inside the docker. To learn more about mounted volumes, please check out <a href="https://docs.docker.com/storage/volumes/">this post</a>.</p>
<p>The usage of each argument is the same as running command line analysis. Here is an example code to perform quanlity control of CellRangerV3 data singleCellTK docker:</p>
<pre><code>docker run --rm -v /path/to/data:/SCTK_docker \
-it campbio/sctk_qc:1.7.5 \
-b /SCTK_docker/cellranger \
-P CellRangerV3 \
-s pbmc_100x100 \
-o /SCTK_docker/result/tenx_v3_pbmc \
-g /SCTK_docker/mitochondrial_human_symbol.gmt \
-S TRUE \
-F R,Python,Flatfile,HTAN</code></pre>
</div>
<div id="singularity" class="section level3">
<h3 class="hasAnchor">
<a href="#singularity" class="anchor"></a>Singularity</h3>
<p>The Singulatiry image can easily be built using Docker Hub as a source:</p>
<pre><code>singularity pull docker://campbio/sctk_qc:1.7.5</code></pre>
<p>The usage of singleCellTK Singularity image is very similar to that of Docker. In Singularity 3.0+, the mount volume is <a href="https://singularity.lbl.gov/docs-mount">automatically overlaid</a>. However, you can use argument –bind/-B to specify your own mount volume. The example is shown as below:</p>
<pre><code>singularity run sctk_qc_1.7.5.sif \
-b ./cellranger \
-P CellRangerV3 \
-s pbmc_100x100 \
-o ./result/tenx_v3_pbmc \
-g ./mitochondrial_human_symbol.gmt</code></pre>
<p>The code above assumed that the dataset is in your current directory, which is automatically mounted by Singularity.</p>
<p>If you run Singularity image on BU SCC，it’s recommended to re-set the home directory to mount. Singularity will mount ‘~’ by default on BU SCC, which contains your personal R library folder. If we don’t re-set the home to mount, singularity will try to load R libraries which are not compiled inside the singularity image and cause some conflicts. You can point to some “sanitized home”, which is different from ‘~’ on SCC, using argument <a href="https://singularity.lbl.gov/faq#solution-1-specify-the-home-to-mount">-H/–home</a>.</p>
<p>Also, you might want to specify cpu architecture when run the Singularity image on BU SCC using ‘#$ -l cpu_arch=broadwell|haswell|skylake|cascadelake’ command. Because the python packages are compiled by SIMD instructions that are only compatible on these cpu architectures. One of the example is shown below:</p>
<pre><code>#!/bin/bash
#$ -cwd
#$ -j y
#$ -P camplab
#$ -pe omp 16
#$ -l cpu_arch=broadwell|haswell|skylake|cascadelake

singularity run --home=/PathToSanitizedHome \
--bind /PathToData:/data sctk_qc_1.7.5.sif \
-P CellRangerV3 \
-s gencodev34_pbmc_1k_v3 \
-b /data/gencodev34_pbmc_1k_v3
-o /data/result/gencodev34_pbmc_1k_v3 \
-S TRUE \
-F R,Python,FlatFile,HTAN \
-n 15 \
-T MulticoreParam</code></pre>
</div>
</div>
<div id="documentation-of-tools-that-are-currently-available-within-the-pipeline" class="section level2">
<h2 class="hasAnchor">
<a href="#documentation-of-tools-that-are-currently-available-within-the-pipeline" class="anchor"></a>Documentation of tools that are currently available within the pipeline:</h2>
<div id="empty-droplet-detection" class="section level4">
<h4 class="hasAnchor">
<a href="#empty-droplet-detection" class="anchor"></a>Empty droplet detection:</h4>
<ul>
<li>
<a href="https://rdrr.io/github/MarioniLab/DropletUtils/man/emptyDrops.html">emptyDrops</a> from the package <a href="https://bioconductor.org/packages/release/bioc/html/DropletUtils.html">DropletUtils</a>
</li>
<li>
<a href="https://rdrr.io/github/MarioniLab/DropletUtils/man/barcodeRanks.html">barcodeRanks</a> from the package <a href="https://bioconductor.org/packages/release/bioc/html/DropletUtils.html">DropletUtils</a>
</li>
</ul>
</div>
<div id="doublet-detection" class="section level4">
<h4 class="hasAnchor">
<a href="#doublet-detection" class="anchor"></a>Doublet Detection</h4>
<ul>
<li>
<a href="https://rdrr.io/github/MarioniLab/scran/man/doubletCells.html">doubletCells</a> from the package <a href="http://bioconductor.org/packages/release/bioc/html/scran.html">scran</a>
</li>
<li>
<a href="https://rdrr.io/bioc/scds/man/cxds.html">cxds</a>, <a href="https://rdrr.io/bioc/scds/man/bcds.html">bcds</a>, and <a href="https://rdrr.io/bioc/scds/man/cxds_bcds_hybrid.html">cxds_bcds_hybrid</a> from the package <a href="http://bioconductor.org/packages/release/bioc/html/scds.html">scds</a>
</li>
<li>
<a href="https://rdrr.io/github/chris-mcginnis-ucsf/DoubletFinder/man/doubletFinder.html">doubletFinder</a> from the package <a href="https://github.com/chris-mcginnis-ucsf/DoubletFinder">DoubletFinder</a>
</li>
<li>
<a href="https://bioconda.github.io/recipes/scrublet/README.html">Scrublet</a> from the package <a href="https://github.com/allonkleinlab/scrublet">scrublet</a>
</li>
</ul>
</div>
<div id="ambient-rna-detection" class="section level4">
<h4 class="hasAnchor">
<a href="#ambient-rna-detection" class="anchor"></a>Ambient RNA detection</h4>
<ul>
<li>
<a href="https://rdrr.io/bioc/celda/man/decontX.html">decontX</a> from the package <a href="https://bioconductor.org/packages/release/bioc/html/celda.html">celda</a>
</li>
</ul>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Joshua D. Campbell, David Jenkins, Tyler Faits, Mohammed Muzamil Khan, Zhe Wang, Emma Briars, Sebastian Carrasco Pro, W. Evan Johnson, Yusuke Koga, Shruthi Bandyadka, Anastasia Leshchyk, Rui Hong, Yichen Wang.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
