<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Scanpy Curated Workflow • singleCellTK</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/yeti/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<script src="../extra.js"></script><meta property="og:title" content="Scanpy Curated Workflow">
<meta property="og:description" content="singleCellTK">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">singleCellTK</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.8.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="https://www.camplab.net/sctk" class="external-link">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/installation.html">Installation</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/01_import_and_qc_tutorial.html">Importing and QC - PBMC3K</a>
    </li>
    <li>
      <a href="../articles/02_a_la_carte_workflow.html">A La Carte Workflow - PBMC3K</a>
    </li>
    <li>
      <a href="../articles/seurat_curated_workflow.html">Seurat Curated Workflow - PBMC3K</a>
    </li>
    <li>
      <a href="../articles/celda_curated_workflow.html">Celda Curated Workflow - PBMC3K</a>
    </li>
    <li>
      <a href="../articles/scanpy_curated_workflow.html">Scanpy Curated Workflow - PBMC3K</a>
    </li>
  </ul>
</li>
<li>
  <a href="../articles/cmd_qc.html">SCTK-QC</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Documentation
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Data</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/import_data.html">Import Data</a>
        </li>
        <li>
          <a href="../articles/import_genesets.html">Import Gene Sets</a>
        </li>
        <li>
          <a href="../articles/import_annotation.html">Annotation</a>
        </li>
        <li>
          <a href="../articles/export_data.html">Export Data</a>
        </li>
        <li>
          <a href="../articles/delete_data.html">Delete Data</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">QC &amp; Filtering</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/cnsl_dropletqc.html">Droplet QC (R console)</a>
        </li>
        <li>
          <a href="../articles/cnsl_cellqc.html">Cell QC (R console)</a>
        </li>
        <li>
          <a href="../articles/ui_qc.html">Cell QC (Shiny UI)</a>
        </li>
        <li>
          <a href="../articles/filtering.html">Filtering</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Normalization &amp; Batch Correction</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/normalization.html">Normalization</a>
        </li>
        <li>
          <a href="../articles/batch_correction.html">Batch Correction</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Feature Selection &amp; Dimensionality Reduction</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/feature_selection.html">Feature Selection</a>
        </li>
        <li>
          <a href="../articles/dimensionality_reduction.html">Dimensionality Reduction</a>
        </li>
        <li>
          <a href="../articles/2d_embedding.html">2D Embedding</a>
        </li>
      </ul>
</li>
    <li>
      <a href="../articles/clustering.html">Clustering</a>
    </li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Differential Expression &amp; Cell Type Labeling</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/differential_expression.html">Differential Expression</a>
        </li>
        <li>
          <a href="../articles/find_marker.html">Find Marker</a>
        </li>
        <li>
          <a href="../articles/cell_type_labeling.html">Cell Type Labeling</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Pathway &amp; Enrichment Analysis</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/enrichR.html">EnrichR</a>
        </li>
        <li>
          <a href="../articles/pathwayAnalysis.html">Pathway Analysis</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Trajectory Analysis</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/trajectoryAnalysis.html">TSCAN</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Curated Workflows</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/seurat_curated_workflow.html">Seurat</a>
        </li>
        <li>
          <a href="../articles/celda_curated_workflow.html">Celda</a>
        </li>
      </ul>
</li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Visualization</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/visualization.html">General Visualization</a>
        </li>
        <li>
          <a href="../articles/heatmap.html">Generic Heatmap</a>
        </li>
      </ul>
</li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">NEWS</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/compbiomed/singleCellTK" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="https://twitter.com/camplab1" class="external-link">
    <span class="fab fa-twitter fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Scanpy Curated Workflow</h1>
                        <h4 data-toc-skip class="author">Nida Pervaiz,
Irzam Sarfraz</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/compbiomed/singleCellTK/blob/HEAD/vignettes/articles/scanpy_curated_worklfow.Rmd" class="external-link"><code>vignettes/articles/scanpy_curated_worklfow.Rmd</code></a></small>
      <div class="hidden name"><code>scanpy_curated_worklfow.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p><strong>singleCellTK</strong> integrates functions from <a href="https://scanpy.readthedocs.io" class="external-link"><strong>Scanpy</strong></a> Scanpy
is a scalable toolkit for analyzing single-cell gene expression data
built jointly with anndata. It includes preprocessing, visualization,
clustering, trajectory inference and differential expression testing.
The Python-based implementation efficiently deals with datasets of more
than one million cells. It is an easy to use streamlined workflow using
both the shiny user interface as well as the R console. The shiny
application contains a separate tab that lets the users run the steps of
the workflow in a sequential manner with ability to visualize through
interactive plots from within the application. On the R console, the
toolkit offers wrapper functions that use the <a href="https://bioconductor.org/packages/release/bioc/html/SingleCellExperiment.html" class="external-link">SingleCellExperiment</a>
<span class="citation">[1]</span> object as the input and the output.
All computations from the wrapper functions are stored within this
object for further manipulation.</p>
<p>To view detailed instructions on how to use the workflow, please
select ‘Interactive Analysis’ for using the workflow in shiny
application or ‘Console Analysis’ for using these methods on R console
from the tabs below:</p>

<meta name="viewport" content="width=device-width, initial-scale=1">
</div>
<div class="section level2">
<h2 id="workflow-guide">Workflow Guide<a class="anchor" aria-label="anchor" href="#workflow-guide"></a>
</h2>
<div class="tab">
  <button class="tablinks" onclick="openTab(event, 'interactive')" id="ia-button">Interactive Analysis</button>
  <button class="tablinks" onclick="openTab(event, 'console')" id="console-button">Console Analysis</button>
</div>

<div id="interactive" class="tabcontent">
<p>In this tutorial example, we illustrate all the steps of the curated
workflow and focus on the options available to manipulate and customize
the steps of the workflow as per user requirements. To initiate the
<code>Scanpy</code> workflow, click on the ‘Curated Workflows’ from the
top menu and select <code>Scanpy</code>:</p>
<p>NOTE: This tutorial assumes that the data has already been uploaded
via the upload tab of the toolkit and filtered before using the
workflow.</p>
<p><strong>1. Normalize Data</strong> <br> Assuming that the data has
been uploaded via the Upload tab of the toolkit, the first step for the
analysis of the data is the Normalization of data. For this purpose, any
assay available in the uploaded data can be used.</p>
<ol style="list-style-type: decimal">
<li>Open up the ‘Normalize’ tab by clicking on it.</li>
<li>Select the <code>assay</code> to normalize from the dropdown
menu.</li>
<li>Specific parameters</li>
<li>Press the ‘Normalize’ button to start the normalization
process.</li>
</ol>
<p><img src="ui_screenshots/scanpy/scanpy_1_normalize_annotated.png"><br></p>
<p><strong>2. Highly Variable Genes</strong> <br> Identification of the
highly variable genes is core to the <code>Scanpy</code> workflow and
these highly variable genes are used throughout the remaining workflow.
<code>Scanpy</code> provides three methods for variable genes
identification i.e. <code>seurat</code>, <code>cell_ranger</code> and
<code>seurat_v3</code>.</p>
<ol style="list-style-type: decimal">
<li>Open up the ‘Highly Variable Genes’ tab.</li>
<li>Select method for computation of highly variable genes from
<code>seurat</code>, <code>cell_ranger</code> and
<code>seurat_v3</code>.</li>
<li>Input the number of genes that should be identified as variable.
Default is <code>2000</code>.</li>
<li>Specify parameters.</li>
<li>Press ‘Find HVG’ button to compute the variable genes.</li>
<li>Graph that plots each gene and its relationship based upon the
selected model in (2), and identifies if a gene is highly variable or
not.</li>
</ol>
<p><img src="ui_screenshots/scanpy/scanpy_3_hvg_annotated.png"><br></p>
<p><strong>3. Dimensionality Reduction</strong> <br><code>Scanpy</code>
workflow offers <code>PCA</code> for dimensionality reduction and the
components from the method can be used in the downstream analysis.
Moreover, several plots are available for the user to inspect the output
of the dimensionality reduction such as the standard ‘PCA Plot’, ‘PCA
Gene Ranking’ and ‘PCA Variance’.</p>
<ol style="list-style-type: decimal">
<li>Open up the ‘Dimensionality Reduction’ tab.</li>
<li>Select method for computation of PCA.</li>
<li>Input the number of components to compute.</li>
<li>Press the ‘Run PCA’ button to start processing.</li>
<li>Select the number of computed components that should be used in the
downstream analysis e.g. in ‘tSNE/UMAP’ computation or with
‘Clustering’.</li>
<li>The plot area from where all computed plots can be viewed by the
user.</li>
</ol>
<p><img src="ui_screenshots/scanpy/scanpy_dimensionalityReduction_annotated.png"><br></p>
<p><strong>4. tSNE/UMAP</strong> <br> ‘tSNE’ and ‘UMAP’ can be computed
and plotted once components are available from ‘Dimensionality
Reduction’ tab.</p>
<ol style="list-style-type: decimal">
<li>Open up the ‘tSNE/UMAP’ tab.</li>
<li>Select ‘tSNE’ or ‘UMAP’ sub-tab.</li>
<li>Select a reduction method. Only methods that are computed previously
in the ‘Dimensionality Reduction’ tab are available.</li>
<li>Set parameters.</li>
<li>Press ‘Run tSNE’ or ‘Run UMAP’ button to start processing.</li>
<li>‘tSNE’ or ‘UMAP’ plot depending upon the selected computation.</li>
</ol>
<p><img src="ui_screenshots/scanpy/scanpy_tsneUmap.png"><br></p>
<p><strong>5. Clustering</strong> <br> Cluster labels can be generated
for all cells/samples using one of the computed reduction method. Plots
are automatically re-computed with cluster labels. The available methods
for clustering as provided by <code>Scanpy</code> include
<code>louvain</code> and <code>leiden</code>.</p>
<ol style="list-style-type: decimal">
<li>Open up the ‘Clustering’ tab.</li>
<li>Select number of components to use from PCA previously
computed.</li>
<li>Select clustering method from <code>louvain</code> and
<code>leiden</code>.</li>
<li>Set parameters.</li>
<li>Press ‘Find Clusters’ button to start processing.</li>
<li>Re-computed plots with cluster labels. Only those plots are
available that have previously been computed.</li>
</ol>
<p><img src="ui_screenshots/scanpy/scanpy_6_clustering.png"><br></p>
<p><strong>6. Find Markers</strong> <br> ‘Find Markers’ tab can be used
to identify and visualize the marker genes using on of the provided
visualization methods. The tab offers identification of markers between
two selected phenotype groups or between all groups and can be decided
at the time of the computation. Furthermore, markers that are conserved
between two phenotype groups can also be identified. Visualizations such
as Dot Plot, Violin Plot, Matrix Plot and Heatmap Plot can be used to
visualize the individual marker genes.</p>
<ol style="list-style-type: decimal">
<li>Select “Find Markers” tab.</li>
<li>Select if you want to identify marker genes against all groups in a
biological variable or between two pre-defined groups.</li>
<li>Set parameters.</li>
<li>Press “Find Markers” button to run marker identification.</li>
<li>Filters can be applied on the table.</li>
</ol>
<p><img src="ui_screenshots/scanpy/scanpy_findmarkers_1.png"><br></p>
<ol start="6" style="list-style-type: decimal">
<li>Heatmap plot can be visualized for all genes populated in the table
against all biological groups in the selected phenotype variable.</li>
</ol>
<p><img src="ui_screenshots/scanpy/scanpy_findmarkers_2.png"><br></p>
<p><strong>7. Downstream Analysis</strong> <br> Once all steps of the
Scanpy workflow are completed, users can further analyze the data by
directly going to the various downstream analysis options (Differential
Expression, Marker Selection &amp; Pathway Analysis) from within the
Scanpy workflow.</p>
</div>
<div id="console" class="tabcontent">
<p>All methods provided by SCTK for Scanpy workflow use a
<code>SingleCellExperiment</code> object both as an input and
output.</p>
<p>Using a sample dataset:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://www.camplab.net/sctk/" class="external-link">singleCellTK</a></span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/importExampleData.html">importExampleData</a></span><span class="op">(</span><span class="st">'pbmc3k'</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## class: SingleCellExperiment </span></span>
<span><span class="co">## dim: 32738 2700 </span></span>
<span><span class="co">## metadata(0):</span></span>
<span><span class="co">## assays(1): counts</span></span>
<span><span class="co">## rownames(32738): MIR1302-10 FAM138A ... AC002321.2 AC002321.1</span></span>
<span><span class="co">## rowData names(3): ENSEMBL_ID Symbol_TENx Symbol</span></span>
<span><span class="co">## colnames(2700): pbmc3k_AAACATACAACCAC-1 pbmc3k_AAACATTGAGCTAC-1 ...</span></span>
<span><span class="co">##   pbmc3k_TTTGCATGAGAGGC-1 pbmc3k_TTTGCATGCCTCAC-1</span></span>
<span><span class="co">## colData names(12): Sample Barcode ... Date_published sample</span></span>
<span><span class="co">## reducedDimNames(0):</span></span>
<span><span class="co">## mainExpName: NULL</span></span>
<span><span class="co">## altExpNames(0):</span></span></code></pre>
<p><strong>1. Normalize Data</strong> <br> Once raw data is uploaded and
stored in a <code>SingleCellExperiment</code> object,
<code><a href="../reference/runScanpyNormalizeData.html">runScanpyNormalizeData()</a></code> function can be used to normalize
the data. The method returns a <code>SingleCellExperiment</code> object
with normalized data stored as a new assay in the input object.</p>
<p>Parameters to this function include <code>useAssay</code> (specify
the assay that should be normalized), <code>normAssayName</code>
(specify the new name of the normalized assay, defaults to
<code>"scanpyNormData"</code>), <code>targetSum</code> (defaults to
<code>1</code>) and <code>maxFraction</code> (defaults to
<code>0.05</code>).</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runScanpyNormalizeData.html">runScanpyNormalizeData</a></span><span class="op">(</span>inSCE <span class="op">=</span> <span class="va">sce</span>, useAssay <span class="op">=</span> <span class="st">"counts"</span>, normAssayName <span class="op">=</span> <span class="st">"scanpyNormData"</span><span class="op">)</span></span></code></pre></div>
<p><strong>2. Scale Data</strong> <br> Normalized data can be scaled by
using the <code><a href="../reference/runScanpyScaleData.html">runScanpyScaleData()</a></code> function that takes input a
<code>SingleCellExperiment</code> object that has been normalized
previously by the <code><a href="../reference/runScanpyNormalizeData.html">runScanpyNormalizeData()</a></code> function. Scaled
assay is stored back in the input object.</p>
<p>Parameters include <code>useAssay</code> (specify the name of
normalized assay) and <code>scaledAssayName</code> (specify the new name
for scaled assay, defaults to <code>"scanpyScaledData"</code>).</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runScanpyScaleData.html">runScanpyScaleData</a></span><span class="op">(</span>inSCE <span class="op">=</span> <span class="va">sce</span>, useAssay <span class="op">=</span> <span class="st">"scanpyNormData"</span>, scaledAssayName <span class="op">=</span> <span class="st">"scanpyScaledData"</span><span class="op">)</span></span></code></pre></div>
<p><strong>3. Highly Variable Genes</strong> <br> Highly variable genes
can be identified by first using the <code><a href="../reference/runScanpyFindHVG.html">runScanpyFindHVG()</a></code>
function that computes that statistics against a selected HVG method in
the rowData of input object. The variable genes can be visualized using
the <code><a href="../reference/plotScanpyHVG.html">plotScanpyHVG()</a></code> method. <br></p>
<p>Parameters for <code><a href="../reference/runScanpyFindHVG.html">runScanpyFindHVG()</a></code> include
<code>useAssay</code> (specify the name of the log normalized assay,
defaults to <code>"scanpyNormData"</code>) and <code>method()</code>
(specify the method to use for variable genes computation, defaults to
<code>"seurat"</code>). The <code>getTopHVG</code> function can be used
to select the top most variable genes and store it back in the SCE as an
altExp</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runScanpyFindHVG.html">runScanpyFindHVG</a></span><span class="op">(</span>inSCE <span class="op">=</span> <span class="va">sce</span>, useAssay <span class="op">=</span> <span class="st">"scanpyNormData"</span>, method <span class="op">=</span> <span class="st">"seurat"</span>, minMean <span class="op">=</span> <span class="fl">0.0125</span>, maxMean <span class="op">=</span> <span class="fl">3</span>, maxDisp <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="../reference/getTopHVG.html">getTopHVG</a></span><span class="op">(</span><span class="va">sce</span>, method <span class="op">=</span> <span class="st">"seurat"</span>, hvgNumber <span class="op">=</span> <span class="fl">10</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] "PPBP"         "FCRL2"        "CEP128"       "DOK3"         "NFE2L2"      </span></span>
<span><span class="co">##  [6] "YPEL2"        "ARVCF"        "UBE2D4"       "FAM210B"      "CTB-113I20.2"</span></span></code></pre>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotScanpyHVG.html">plotScanpyHVG</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span></code></pre></div>
<p><strong>4. Dimensionality Reduction</strong> <br> PCA can be computed
using the <code><a href="../reference/runScanpyPCA.html">runScanpyPCA()</a></code> function. Plots can be visualized
using <code><a href="../reference/plotScanpyPCA.html">plotScanpyPCA()</a></code>,
<code><a href="../reference/plotScanpyPCAGeneRanking.html">plotScanpyPCAGeneRanking()</a></code> and
<code><a href="../reference/plotScanpyPCAVariance.html">plotScanpyPCAVariance()</a></code>.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runScanpyPCA.html">runScanpyPCA</a></span><span class="op">(</span>inSCE <span class="op">=</span> <span class="va">sce</span>, useAssay <span class="op">=</span> <span class="st">"scanpyScaledData"</span>, reducedDimName <span class="op">=</span> <span class="st">"scanpyPCA"</span>, nPCs <span class="op">=</span> <span class="fl">20</span>, method <span class="op">=</span> <span class="st">"auto"</span>, use_highly_variable <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/plotScanpyPCA.html">plotScanpyPCA</a></span><span class="op">(</span>inSCE <span class="op">=</span> <span class="va">sce</span>, reducedDimName <span class="op">=</span> <span class="st">"scanpyPCA"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/plotScanpyPCAGeneRanking.html">plotScanpyPCAGeneRanking</a></span><span class="op">(</span>inSCE <span class="op">=</span> <span class="va">sce</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/plotScanpyPCAVariance.html">plotScanpyPCAVariance</a></span><span class="op">(</span>inSCE <span class="op">=</span> <span class="va">sce</span>, nPCs <span class="op">=</span> <span class="fl">20</span><span class="op">)</span></span></code></pre></div>
<p><strong>5. tSNE/UMAP</strong> <br><code><a href="../reference/runScanpyTSNE.html">runScanpyTSNE()</a></code> and
<code><a href="../reference/runScanpyUMAP.html">runScanpyUMAP()</a></code> can be used to compute tSNE/UMAP statistics
and store into the input object. Parameters to both functions include
<code>inSCE</code> (input SCE object), <code>useReducedDim</code>
(specify the reduction to use i.e. <code>"pca"</code>),
<code>reducedDimName</code> (name of this new reduction) and
<code>dims</code> (number of dims to use).
<code><a href="../reference/plotScanpyEmbedding.html">plotScanpyEmbedding()</a></code> can be used to visualize the
results.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runScanpyTSNE.html">runScanpyTSNE</a></span><span class="op">(</span>inSCE <span class="op">=</span> <span class="va">sce</span>, useReducedDim <span class="op">=</span> <span class="st">"scanpyPCA"</span>, reducedDimName <span class="op">=</span> <span class="st">"scanpyTSNE"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runScanpyUMAP.html">runScanpyUMAP</a></span><span class="op">(</span>inSCE <span class="op">=</span> <span class="va">sce</span>, useReducedDim <span class="op">=</span> <span class="st">"scanpyPCA"</span>, reducedDimName <span class="op">=</span> <span class="st">"scanpyUMAP"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotScanpyEmbedding.html">plotScanpyEmbedding</a></span><span class="op">(</span><span class="va">sce</span>, reducedDimName <span class="op">=</span> <span class="st">"scanpyTSNE"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/plotScanpyEmbedding.html">plotScanpyEmbedding</a></span><span class="op">(</span><span class="va">sce</span>, reducedDimName <span class="op">=</span> <span class="st">"scanpyUMAP"</span><span class="op">)</span></span></code></pre></div>
<p><strong>6. Clustering</strong> <br><code><a href="../reference/runScanpyFindClusters.html">runScanpyFindClusters()</a></code> function can be used to compute the
clusters, which can later be plotted through the
<code><a href="../reference/plotScanpyEmbedding.html">plotScanpyEmbedding()</a></code> method with cluster labels. The
parameters to the function include <code>inSCE</code> (input SCE
object), <code>useAssay</code> (name of the scaled assay),
<code>useReducedDim</code> (specify which reduction to use
i.e. <code>"scanpyPCA"</code>), <code>dims</code> (number of dims to
use) and the method (either <code>"louvain"</code> or
<code>"leiden"</code>).</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runScanpyFindClusters.html">runScanpyFindClusters</a></span><span class="op">(</span>inSCE <span class="op">=</span> <span class="va">sce</span>, useAssay <span class="op">=</span> <span class="st">"scanpyScaledData"</span>, useReducedDim <span class="op">=</span> <span class="st">"scanpyPCA"</span>, method <span class="op">=</span> <span class="st">"leiden"</span>, resolution <span class="op">=</span> <span class="fl">0.8</span>, nNeighbors <span class="op">=</span> <span class="fl">15</span>, dims <span class="op">=</span> <span class="fl">10</span>, cor_method <span class="op">=</span> <span class="st">"pearson"</span><span class="op">)</span></span></code></pre></div>
<p><code><a href="../reference/plotScanpyEmbedding.html">plotScanpyEmbedding()</a></code> can then be used to plot all
reductions previously computed with cluster labels:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotScanpyEmbedding.html">plotScanpyEmbedding</a></span><span class="op">(</span><span class="va">sce</span>, reducedDimName <span class="op">=</span> <span class="st">"scanpyPCA"</span>, color <span class="op">=</span> <span class="st">'Scanpy_leiden_0.8'</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/plotScanpyEmbedding.html">plotScanpyEmbedding</a></span><span class="op">(</span><span class="va">sce</span>, reducedDimName <span class="op">=</span> <span class="st">"scanpyTSNE"</span>, color <span class="op">=</span> <span class="st">'Scanpy_leiden_0.8'</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/plotScanpyEmbedding.html">plotScanpyEmbedding</a></span><span class="op">(</span><span class="va">sce</span>, reducedDimName <span class="op">=</span> <span class="st">"scanpyUMAP"</span>, color <span class="op">=</span> <span class="st">'Scanpy_leiden_0.8'</span><span class="op">)</span></span></code></pre></div>
<p><strong>7. Find Markers</strong> <br> Marker genes can be identified
using the <code><a href="../reference/runScanpyFindMarkers.html">runScanpyFindMarkers()</a></code> function. This function
can use one specified column from <code>colData</code> of the input
object from which the user can select the groups versus which marker
genes need to be obtained. User can either compare all group of cells
versus the rest or can can manually specify the cells included in one
group vs cells included in the second group (<code>group1</code> and
<code>group2</code> parameter).</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runScanpyFindMarkers.html">runScanpyFindMarkers</a></span><span class="op">(</span>inSCE <span class="op">=</span> <span class="va">sce</span>, colDataName <span class="op">=</span> <span class="st">"Scanpy_leiden_0.8"</span>, nGenes <span class="op">=</span> <span class="fl">10</span>, test <span class="op">=</span> <span class="st">"t-test"</span>, corr_method <span class="op">=</span> <span class="st">"benjamini-hochberg"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu">metadata</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">[[</span><span class="st">"scanpyMarkersTable"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##      Gene findMarker_cluster   Log2_FC        Pvalue   zscore</span></span>
<span><span class="co">## 01  RPS27                  0 0.7101832 1.169946e-125 28.92589</span></span>
<span><span class="co">## 02  RPS12                  0 0.7652162 1.675553e-100 25.67202</span></span>
<span><span class="co">## 03  RPS25                  0 0.7343966  7.038085e-91 23.83379</span></span>
<span><span class="co">## 04  RPL32                  0 0.6419957  3.152299e-86 23.30648</span></span>
<span><span class="co">## 05   RPS6                  0 0.6920167  1.457690e-86 23.08136</span></span>
<span><span class="co">## 06 EEF1A1                  0 0.5713664  6.820436e-87 23.05413</span></span></code></pre>
<p>The marker genes identified can be visualized through one of the
available plots from <code>matrix plot</code>, <code>violin plot</code>,
<code>dot plot</code> and <code>heatmap plot</code>.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotScanpyMarkerGenes.html">plotScanpyMarkerGenes</a></span><span class="op">(</span><span class="va">sce</span>, groups <span class="op">=</span> <span class="st">'0'</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/plotScanpyMarkerGenesViolin.html">plotScanpyMarkerGenesViolin</a></span><span class="op">(</span><span class="va">sce</span>, groups <span class="op">=</span> <span class="st">'0'</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/plotScanpyMarkerGenesHeatmap.html">plotScanpyMarkerGenesHeatmap</a></span><span class="op">(</span><span class="va">sce</span>, groupBy <span class="op">=</span> <span class="st">'Scanpy_leiden_0.8'</span>, nGenes <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span>
<span><span class="co">#plotScanpyMarkerGenesDotPlot(sce, groupBy = 'Scanpy_leiden_0.8', nGenes = 10)</span></span>
<span><span class="co">#plotScanpyMarkerGenesMatrixPlot(sce, groupBy = 'Scanpy_leiden_0.8', nGenes = 10)</span></span></code></pre></div>
      </div>
<script>
document.getElementById("ia-button").click();
</script>
</div>
<div class="section level2 unnumbered">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body">
<div id="ref-Amezquita2020" class="csl-entry">
<div class="csl-left-margin">[1] </div>
<div class="csl-right-inline">R.
A. Amezquita <em>et al.</em>, <span>“<span class="nocase">Orchestrating
single-cell analysis with Bioconductor</span>,”</span> <em>Nature
Methods</em>, vol. 17, no. 2, pp. 137–145, Feb. 2020, doi: <a href="https://doi.org/10.1038/s41592-019-0654-x" class="external-link">10.1038/s41592-019-0654-x</a>.</div>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Yichen Wang, Irzam Sarfraz, Rui Hong, Yusuke Koga, Salam Alabdullatif, Nida Pervaiz, David Jenkins, Vidya Akavoor, Xinyun Cao, Shruthi Bandyadka, Anastasia Leshchyk, Tyler Faits, Mohammed Muzamil Khan, Zhe Wang, W. Evan Johnson, Joshua David Campbell.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.5.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
