<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Import Datasets - UI • singleCellTK</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/yeti/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Import Datasets - UI">
<meta property="og:description" content="singleCellTK">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">singleCellTK</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">2.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/introduction_to_SCTK.html">General Workflow</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Shell Command Line Analysis
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/import_scRNAseq_data_as_SCE_Commandline.html">Import data into SCTK</a>
    </li>
    <li>
      <a href="../articles/run_qcCommandline.html">Quality Control</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    R Console Analysis
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/import_scRNAseq_data_as_SCE_Console.html">Import data into SCTK</a>
    </li>
    <li>
      <a href="../articles/run_qcConsole.html">Quality Control</a>
    </li>
    <li>
      <a href="../articles/normalization.html">Normalization</a>
    </li>
    <li>
      <a href="../articles/batch_correction.html">Batch Correction</a>
    </li>
    <li>
      <a href="../articles/variability.html">Variability</a>
    </li>
    <li>
      <a href="../articles/feature_selection.html">Feature Selection</a>
    </li>
    <li>
      <a href="../articles/clustering.html">Clustering</a>
    </li>
    <li>
      <a href="../articles/differential_expression.html">Differential Expression</a>
    </li>
    <li>
      <a href="../articles/find_marker.html">Find Marker via DE</a>
    </li>
    <li>
      <a href="../articles/visualize_results.html">Visualize Results</a>
    </li>
    <li>
      <a href="../articles/heatmap.html">Heatmap Visualization</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Shiny UI Analysis
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/ui_import_datasets.html">Import Datasets</a>
    </li>
    <li>
      <a href="../articles/ui_batch_correction.html">Batch Correction</a>
    </li>
    <li>
      <a href="../articles/ui_clustering.html">Clustering</a>
    </li>
    <li>
      <a href="../articles/ui_differential_expression.html">Differential Expression</a>
    </li>
    <li>
      <a href="../articles/ui_find_marker.html">Find Marker</a>
    </li>
    <li>
      <a href="../articles/ui_heatmap.html">Generic Heatmap</a>
    </li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/compbiomed/singleCellTK">
    <span class="fas fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="ui_import_datasets_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Import Datasets - UI</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/compbiomed/singleCellTK/blob/master/vignettes/ui_import_datasets.Rmd"><code>vignettes/ui_import_datasets.Rmd</code></a></small>
      <div class="hidden name"><code>ui_import_datasets.Rmd</code></div>

    </div>

    
    
<div id="introduction" class="section level2">
<h2 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h2>
<p>This page is the landing page of the Single Cell ToolKit (SCTK) application. Here SCTK allows users to import multiple datasets at once, and the datasets can be in various forms, including the output from preprocessing tools (listed later), the text files of count matrices and annotations, the RDS file saving a SingleCellExperiment object or the example dataset ready online.<br>
The UI contains mainly three sections for importing and another section for data summary. Through the first and the second sections, users can choose the type of a single data source and specify the location of the data from this single source, respectively. A table in the third section summarizes files and samples that are specified from all sources. Finally, all data can be imported by clicking on the import button right below this table.</p>
<p><img src="import_ui_workflow.png" alt="workflow"><br></p>
</div>
<div id="importing-single-cell-datasets" class="section level2">
<h2 class="hasAnchor">
<a href="#importing-single-cell-datasets" class="anchor"></a>Importing Single-Cell Datasets</h2>
<p><img src="import_ui_source.png" style="border: 1px solid rgba(0,0,0,0.2)"></p>
<p>For a single source of data, users need to choose one option from the radio buttons listed in the screenshot. The next step will depend on the option chosen.</p>
<div id="import-from-a-preprocessing-tool" class="section level3">
<h3 class="hasAnchor">
<a href="#import-from-a-preprocessing-tool" class="anchor"></a>Import from a preprocessing tool</h3>
<p><img src="import_ui_source1.png" alt="source1"><br></p>
<p>SCTK supports importing data from the following pre-processing tools in our UI workflow:<br>
- <a href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/what-is-cell-ranger">CellRanger</a> (versions 2 and 3)<br>
- <a href="https://github.com/alexdobin/STAR/blob/master/docs/STARsolo.md">STARSolo</a><br>
- <a href="https://github.com/BUStools/bustools">BUStools</a><br>
- <a href="https://github.com/ambrosejcarr/seqc">Seqc</a><br>
- <a href="https://data.humancellatlas.org/pipelines/optimus-workflow">Optimus</a></p>
<p>Choose the specific preprocessing tool that matches your dataset, and then press “<strong>Add a Sample</strong>”. A corresponding dialog modal will pop out with description that instructs you to select the directory that stores your data.</p>
<div id="cell-ranger" class="section level4">
<h4 class="hasAnchor">
<a href="#cell-ranger" class="anchor"></a>Cell Ranger</h4>
<p>The import procedure for Cell Ranger versions 2 and 3 are the same. To import a Cell Ranger sample, the user must select how they would like to import the sample: select a base directory that holds multiple sample directories, select a single sample directory, or select a directory holding the data files for a sample. Depending on this selection, the user must then provide the apprpriate directory paths and sample names and then click “OK.”</p>
<p><img src="import_ui_cellRanger.png" alt="cell ranger"><br></p>
</div>
<div id="starsolo" class="section level4">
<h4 class="hasAnchor">
<a href="#starsolo" class="anchor"></a>STARsolo</h4>
<p>To import a STARsolo sample, the user must select the parent directory of the <strong><code>"Gene/"</code></strong> directory for the sample they want to import, name the sample, and click “OK.”</p>
<p><img src="import_ui_STARsolo.png" alt="STARsolo"><br></p>
</div>
<div id="bustools" class="section level4">
<h4 class="hasAnchor">
<a href="#bustools" class="anchor"></a>BUStools</h4>
<p>To import a BUStools sample, the user must point to the <strong><code>"genecount/"</code></strong> directory for the sample they want to import, name the sample, and click “OK.”</p>
<p><img src="import_ui_BUStools.png" alt="BUStools"><br></p>
</div>
<div id="seqc" class="section level4">
<h4 class="hasAnchor">
<a href="#seqc" class="anchor"></a>SEQC</h4>
<p>To import a SEQC sample, the user must point to the directory that contains the <strong>barcodes, gene, and matrix files</strong> for the sample they want to import and name the sample. This tool additionally needs the user to specify a <strong>“Sample ID”</strong> which is what the name of the sample is as it appears in the names of its data files (e.g. for a file called <code>"pbmc_1k_sparse_counts_barcodes.csv"</code> the sample ID is <code>"pbmc_1k"</code>). Then the user must click “OK.”</p>
<p><img src="import_ui_SEQC.png" alt="SEQC"><br></p>
</div>
<div id="optimus" class="section level4">
<h4 class="hasAnchor">
<a href="#optimus" class="anchor"></a>Optimus</h4>
<p>To import an Optimus sample, the user must select the directory that contains the following four directories for the sample they want to import: <strong><code>"call-MergeCountFiles"</code></strong>, <strong><code>"call-MergeCellMetrics"</code></strong>, <strong><code>"call-MergeGeneMetrics"</code></strong>, <strong><code>"call-RunEmptyDrops"</code></strong>. Then the user must prove a sample name and click “OK.”</p>
<p><img src="import_ui_Optimus.png" alt="Optimus"><br></p>
</div>
</div>
<div id="import-from-flat-files-csv-tsv" class="section level3">
<h3 class="hasAnchor">
<a href="#import-from-flat-files-csv-tsv" class="anchor"></a>Import from flat files (CSV, TSV)</h3>
<p>To import tab-separated data, the user must provide <strong>a counts file</strong> (.csv or .tsv) and specify what type of input assay it is. The user can optionally also provide <strong>cell annotation</strong> and <strong>feature</strong> files (also .csv or .tsv). The user must then click <strong>“Add to Sample List”</strong> to add it to the table of samples.<br>
After choosing a single file for uploading, user must wait until the progress bar runs to the right end and says “Upload complete”. Sometimes the progress bar runs to the end but it takes some time for the system to finalize reading a large file.</p>
<p><img src="import_ui_source2.png" alt="Flat files"><br></p>
</div>
<div id="upload-singlecellexperiment-or-seurat-object-stored-in-an-rds-file" class="section level3">
<h3 class="hasAnchor">
<a href="#upload-singlecellexperiment-or-seurat-object-stored-in-an-rds-file" class="anchor"></a>Upload <code>SingleCellExperiment</code> or <code>Seurat</code> object stored in an RDS file</h3>
<p>RDS is a file format that stores a single object from R environment. To import a single-cell dataset stored in a <strong><code>SingleCellExperiment</code></strong> (SCE) object from an RDS file, the user can simply select which <strong>RDS</strong> file to import and click <strong>“Add to Sample List”</strong>.<br>
After choosing a single file for uploading, user must wait until the progress bar runs to the right end and says “Upload complete”. Sometimes the progress bar runs to the end but it takes some time for the system to finalize reading a large file.</p>
<p><img src="import_ui_source3.png" style="border: 1px solid rgba(0,0,0,0.2)"></p>
</div>
<div id="import-example-datasets" class="section level3">
<h3 class="hasAnchor">
<a href="#import-example-datasets" class="anchor"></a>Import example datasets</h3>
<p>Users can also opt to import and work with published example datasets. To do so, the user must select which dataset they would like to use from the dropdown menu and click <strong>“Add to Sample List”</strong>. For more information about the chosen dataset, users can click on the hyperlink of the title and open the reference website in a browser.</p>
<p><img src="import_ui_source4.png" style="border: 1px solid rgba(0,0,0,0.2)"></p>
</div>
<div id="complete-upload" class="section level3">
<h3 class="hasAnchor">
<a href="#complete-upload" class="anchor"></a>Complete Upload</h3>
<p>After successfully choosing each source of dataset, it will be inserted to a new line of the table at the buttom of this UI page, in “<strong>3. Import</strong>” section. This is a wait-list for all the samples of interests. To finally import all datasets of interests, users must click on <strong>“Import”</strong> button.<br>
Before importing, users can click on the “<code>[X]</code>” button on the right to remove the selection of the corresponding single dataset, or click on <strong>“Clear Samples”</strong> to reset all selections. This only clears this wait-list for importing, but will not work on any dataset already imported.<br>
Note that it is normal that in the “Location” field, some paths might happen to point to a temporary cache.</p>
<p><img src="import_ui_sampleList.png" style="border: 1px solid rgba(0,0,0,0.2)"></p>
</div>
</div>
<div id="importing-gene-sets" class="section level2">
<h2 class="hasAnchor">
<a href="#importing-gene-sets" class="anchor"></a>Importing Gene Sets</h2>
<p><img src="import_ui_gs_entry.png" alt="gs entry"><br></p>
<p>Users can import gene sets to use in later steps of the toolkit’s pipeline. The UIs are implemented in the page that can be entered through the links, as shown in the screenshot above.<br>
There are three ways to import gene sets, and they are described below.</p>
<div id="import-gene-sets-from-a-gmt-file" class="section level3">
<h3 class="hasAnchor">
<a href="#import-gene-sets-from-a-gmt-file" class="anchor"></a>Import Gene Sets from a GMT File</h3>
<p>The first option is to upload the GMT file detailing a gene set. For this option, users must provide a GMT file through the file input <strong>“Choose GMT File”</strong>; a text collection name at <strong>“Collection Name”</strong>. If the collection name entered already exists in current session, the gene set from the GMT file will be added to that collection. Otherwise, a new collection with the given name will be created. Finally, users must click <strong>“Upload”</strong> to finish the procedure.</p>
<p><img src="import_ui_gs_gmt.png" style="border: 1px solid rgba(0,0,0,0.2)"></p>
</div>
<div id="select-from-a-database" class="section level3">
<h3 class="hasAnchor">
<a href="#select-from-a-database" class="anchor"></a>Select from a Database</h3>
<p>SCTK also has a database of gene sets that users can choose from. For this option, the user must simply select the gene sets they would like to upload and click <strong>“Upload”</strong>. A brief description of each gene set is provided in the checklist.</p>
<p><img src="import_ui_gs_db.png" style="border: 1px solid rgba(0,0,0,0.2)"></p>
</div>
<div id="paste-in-gene-sets" class="section level3">
<h3 class="hasAnchor">
<a href="#paste-in-gene-sets" class="anchor"></a>Paste in Gene Sets</h3>
<p>Users also have the option to paste their gene set into a text box. In order to do so, the user must paste their gene set into the provided text box and then provide a collection name to create a new collection. If there are already gene sets imported, the user can choose an existing collection to add their pasted gene set into. Then the user must click “Upload.”</p>
<p><img src="import_ui_gs_paste.png" style="border: 1px solid rgba(0,0,0,0.2)"></p>
</div>
<div id="map-gene-sets-to-specific-feature-ids" class="section level3">
<h3 class="hasAnchor">
<a href="#map-gene-sets-to-specific-feature-ids" class="anchor"></a>Map Gene Sets to Specific Feature IDs</h3>
<p>There is an additional selection input labeled with <strong>“Location within SCE object where the gene identifiers in should be mapped”</strong>, placed above the <strong>“Upload”</strong> button as already shown in screenshots above. This specifies the feature IDs stored in the imported single-cell dataset, to which the gene identifiers in the uploaded gene sets will be mapped. For example, if the uploaded gene sets are presented by gene symbols, users must make sure the selection here is pointing to gene symbols instead of ENSEMBL IDs. Most of the time, “rownames” is the default feature ID of the imported single-cell dataset.</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by David Jenkins, Vidya Akavoor, Salam Alabdullatif, Shruthi Bandyadka, Emma Briars, Xinyun Cao, Sebastian Carrasco Pro, Tyler Faits, Rui Hong, Mohammed Muzamil Khan, Yusuke Koga, Anastasia Leshchyk, Irzam Sarfraz, Yichen Wang, Zhe Wang, W. Evan Johnson, Joshua David Campbell.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
