<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Create a SingleCellExperiment object from files — importFromFiles • singleCellTK</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/yeti/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous" />


<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Create a SingleCellExperiment object from files — importFromFiles" />
<meta property="og:description" content="Creates a SingleCellExperiment object from a counts file in various formats.
and a file of annotation information, ." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">singleCellTK</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.7.6</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../articles/v01-Introduction_to_singleCellTK.html">Introduction to singleCellTK</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Interactive Shiny Analysis
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/v03-tab01_Upload.html">1: Upload</a>
    </li>
    <li>
      <a href="../articles/v04-tab02_Data-Summary-and-Filtering.html">2: Data Summary and Filtering</a>
    </li>
    <li>
      <a href="../articles/v05-tab03_Visualization-and-Clustering.html">3: Visualization and Clustering</a>
    </li>
    <li>
      <a href="../articles/v06-tab04_Batch-Correction.html">4: Batch Correction</a>
    </li>
    <li>
      <a href="../articles/v07-tab05_Differential-Expression.html">5: Differential Expression</a>
    </li>
    <li>
      <a href="../articles/v08-tab06_Pathway-Activity-Analysis.html">6: Pathway Activity Analysis</a>
    </li>
    <li>
      <a href="../articles/v09-tab07_Sample-Size.html">7: Sample Size</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Command Line Analysis
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/v02-Processing_and_Visualizing_Data_in_the_SingleCellTK.html">Processing and Visualizing Data</a>
    </li>
    <li>
      <a href="../articles/v10-Aligning_and_Quantifying_scRNA-Seq_Data.html">Aligning and Quantifying scRNA-Seq Data</a>
    </li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/compbiomed/singleCellTK">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Create a SingleCellExperiment object from files</h1>
    <small class="dont-index">Source: <a href='https://github.com/compbiomed/singleCellTK/blob/master/R/importFromFiles.R'><code>R/importFromFiles.R</code></a></small>
    <div class="hidden name"><code>importFromFiles.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Creates a SingleCellExperiment object from a counts file in various formats.
and a file of annotation information, .</p>
    </div>

    <pre class="usage"><span class='fu'>importFromFiles</span>(
  <span class='kw'>assayFile</span>,
  annotFile = <span class='kw'>NULL</span>,
  featureFile = <span class='kw'>NULL</span>,
  assayName = <span class='st'>"counts"</span>,
  inputDataFrames = <span class='fl'>FALSE</span>,
  class = <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='st'>"Matrix"</span>, <span class='st'>"matrix"</span>),
  delayedArray = <span class='fl'>FALSE</span>,
  annotFileHeader = <span class='fl'>FALSE</span>,
  annotFileRowName = <span class='fl'>1</span>,
  annotFileSep = <span class='st'>"\t"</span>,
  featureHeader = <span class='fl'>FALSE</span>,
  featureRowName = <span class='fl'>1</span>,
  featureSep = <span class='st'>"\t"</span>,
  gzipped = <span class='st'>"auto"</span>
)</pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>assayFile</th>
      <td><p>The path to a  file in .mtx, .txt, .csv, .tab, or .tsv format.</p></td>
    </tr>
    <tr>
      <th>annotFile</th>
      <td><p>The path to a text file that contains columns of annotation
information for each sample in the assayFile. This file should have the same
number of rows as there are columns in the assayFile. If multiple samples are
represented in these files, this should be denoted by a column called <code>'sample'</code>
within the <code>annotFile</code>.</p></td>
    </tr>
    <tr>
      <th>featureFile</th>
      <td><p>The path to a text file that contains columns of
annotation information for each gene in the count matrix. This file should
have the same genes in the same order as assayFile. This is optional.</p></td>
    </tr>
    <tr>
      <th>assayName</th>
      <td><p>The name of the assay that you are uploading. The default
is "counts".</p></td>
    </tr>
    <tr>
      <th>inputDataFrames</th>
      <td><p>If TRUE, assayFile and annotFile are read as data
frames instead of file paths. The default is FALSE.</p></td>
    </tr>
    <tr>
      <th>class</th>
      <td><p>Character. The class of the expression matrix stored in the SCE
object. Can be one of "Matrix" (as returned by
<a href='https://rdrr.io/pkg/Matrix/man/externalFormats.html'>readMM</a> function), or "matrix" (as returned by
<a href='https://rdrr.io/r/base/matrix.html'>matrix</a> function). Default "Matrix".</p></td>
    </tr>
    <tr>
      <th>delayedArray</th>
      <td><p>Boolean. Whether to read the expression matrix as
<a href='https://rdrr.io/pkg/DelayedArray/man/DelayedArray-class.html'>DelayedArray</a> object or not. Default <code>TRUE</code>.</p></td>
    </tr>
    <tr>
      <th>annotFileHeader</th>
      <td><p>Whether there's a header (colnames) in the cell annotation file. Default is FALSE</p></td>
    </tr>
    <tr>
      <th>annotFileRowName</th>
      <td><p>Which column is used as the rownames for the cell annotation file. Default is 1 (first column).</p></td>
    </tr>
    <tr>
      <th>annotFileSep</th>
      <td><p>Separater used for the cell annotation file. Default is "\t".</p></td>
    </tr>
    <tr>
      <th>featureHeader</th>
      <td><p>Whether there's a header (colnames) in the feature annotation file. Default is FALSE</p></td>
    </tr>
    <tr>
      <th>featureRowName</th>
      <td><p>Which column is used as the rownames for the feature annotation file. Default is 1 (first column).</p></td>
    </tr>
    <tr>
      <th>featureSep</th>
      <td><p>Separater used for the feature annotation file. Default is "\t".</p></td>
    </tr>
    <tr>
      <th>gzipped</th>
      <td><p>Whether the input file is gzipped. Default is "auto" and it will automatically detect whether the file is gzipped. Other options is TRUE or FALSE.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>a SingleCellExperiment object</p>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by David Jenkins, Tyler Faits, Mohammed Muzamil Khan, Zhe Wang, Emma Briars, Sebastian Carrasco Pro, W. Evan Johnson.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


