<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Function reference â€¢ singleCellTK</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/yeti/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous" />


<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Function reference" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-index">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">singleCellTK</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.7.6</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../articles/v01-Introduction_to_singleCellTK.html">Introduction to singleCellTK</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Interactive Shiny Analysis
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/v03-tab01_Upload.html">1: Upload</a>
    </li>
    <li>
      <a href="../articles/v04-tab02_Data-Summary-and-Filtering.html">2: Data Summary and Filtering</a>
    </li>
    <li>
      <a href="../articles/v05-tab03_Visualization-and-Clustering.html">3: Visualization and Clustering</a>
    </li>
    <li>
      <a href="../articles/v06-tab04_Batch-Correction.html">4: Batch Correction</a>
    </li>
    <li>
      <a href="../articles/v07-tab05_Differential-Expression.html">5: Differential Expression</a>
    </li>
    <li>
      <a href="../articles/v08-tab06_Pathway-Activity-Analysis.html">6: Pathway Activity Analysis</a>
    </li>
    <li>
      <a href="../articles/v09-tab07_Sample-Size.html">7: Sample Size</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Command Line Analysis
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/v02-Processing_and_Visualizing_Data_in_the_SingleCellTK.html">Processing and Visualizing Data</a>
    </li>
    <li>
      <a href="../articles/v10-Aligning_and_Quantifying_scRNA-Seq_Data.html">Aligning and Quantifying scRNA-Seq Data</a>
    </li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/compbiomed/singleCellTK">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index">

    <colgroup>
      
      <col class="alias" />
      <col class="title" />
    </colgroup>

    <tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-shiny-app" class="hasAnchor"><a href="#section-shiny-app" class="anchor"></a>Shiny App</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="singleCellTK.html">singleCellTK()</a></code> </p>
        </td>
        <td><p>Run the single cell analysis app</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-data-summary-and-filtering" class="hasAnchor"><a href="#section-data-summary-and-filtering" class="anchor"></a>Data Summary and Filtering</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="filterSCData.html">filterSCData()</a></code> </p>
        </td>
        <td><p>Filter Genes and Samples from a Single Cell Object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="convertGeneIDs.html">convertGeneIDs()</a></code> </p>
        </td>
        <td><p>Convert Gene IDs</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-dimensionality-reduction-and-clustering" class="hasAnchor"><a href="#section-dimensionality-reduction-and-clustering" class="anchor"></a>Dimensionality Reduction and Clustering</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="getPCA.html">getPCA()</a></code> </p>
        </td>
        <td><p>getPCA
A wrapper to <a href='https://rdrr.io/pkg/scater/man/runPCA.html'>runPCA</a> function to compute principal component analysis (PCA) from a given <code>SingleCellExperiment</code> object.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getTSNE.html">getTSNE()</a></code> </p>
        </td>
        <td><p>Run t-SNE dimensionality reduction method on the assay data.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getBiomarker.html">getBiomarker()</a></code> </p>
        </td>
        <td><p>Given a list of genes and a SingleCellExperiment object, return
the binary or continuous expression of the genes.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotBiomarker.html">plotBiomarker()</a></code> </p>
        </td>
        <td><p>Given a set of genes, return a ggplot of expression
values.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotPCA.html">plotPCA()</a></code> </p>
        </td>
        <td><p>Plot PCA run data from its components.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotTSNE.html">plotTSNE()</a></code> </p>
        </td>
        <td><p>Plot t-SNE plot on dimensionality reduction data run from t-SNE method.</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-batch-correction" class="hasAnchor"><a href="#section-batch-correction" class="anchor"></a>Batch Correction</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="plotBatchVariance.html">plotBatchVariance()</a></code> </p>
        </td>
        <td><p>Plot the percent of the variation that is explained by batch and condition
in the data</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotSCEBatchFeatureMean.html">plotSCEBatchFeatureMean()</a></code> </p>
        </td>
        <td><p>Plot mean feature value in each batch of a SingleCellExperiment object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="runComBat.html">runComBat()</a></code> </p>
        </td>
        <td><p>Apply ComBat batch effect correction method to SingleCellExperiment object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="runBBKNN.html">runBBKNN()</a></code> </p>
        </td>
        <td><p>Apply BBKNN batch effect correction method to SingleCellExperiment object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="runHarmony.html">runHarmony()</a></code> </p>
        </td>
        <td><p>Apply Harmony batch effect correction method to SingleCellExperiment object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="runLimmaBC.html">runLimmaBC()</a></code> </p>
        </td>
        <td><p>Apply Limma's batch effect correction method to SingleCellExperiment object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="runSeurat3Integration.html">runSeurat3Integration()</a></code> </p>
        </td>
        <td><p>Apply Integration batch effect correction method from Seurat v3 to
SingleCellExperiment object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="runFastMNN.html">runFastMNN()</a></code> </p>
        </td>
        <td><p>Apply a fast version of the mutual nearest neighbors (MNN) batch effect
correction method to SingleCellExperiment object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="runLIGER.html">runLIGER()</a></code> </p>
        </td>
        <td><p>Apply LIGER batch effect correction method to SingleCellExperiment object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="runMNNCorrect.html">runMNNCorrect()</a></code> </p>
        </td>
        <td><p>Apply the mutual nearest neighbors (MNN) batch effect correction method to
SingleCellExperiment object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="runSCANORAMA.html">runSCANORAMA()</a></code> </p>
        </td>
        <td><p>Apply the mutual nearest neighbors (MNN) batch effect correction method to
SingleCellExperiment object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="runSCMerge.html">runSCMerge()</a></code> </p>
        </td>
        <td><p>Apply scMerge batch effect correction method to SingleCellExperiment object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="runZINBWaVE.html">runZINBWaVE()</a></code> </p>
        </td>
        <td><p>Apply ZINBWaVE Batch effect correction method to SingleCellExperiment object</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-differential-expression" class="hasAnchor"><a href="#section-differential-expression" class="anchor"></a>Differential Expression</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="runDESeq2.html">runDESeq2()</a></code> </p>
        </td>
        <td><p>Perform differential expression analysis on SCE with DESeq2.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="runLimmaDE.html">runLimmaDE()</a></code> </p>
        </td>
        <td><p>Perform differential expression analysis on SCE with Limma.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="runANOVA.html">runANOVA()</a></code> </p>
        </td>
        <td><p>Perform differential expression analysis on SCE with ANOVA</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="runMAST.html">runMAST()</a></code> </p>
        </td>
        <td><p>Perform differential expression analysis on SCE with MAST</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotDEGViolin.html">plotDEGViolin()</a></code> </p>
        </td>
        <td><p>plot the violin plot to show visualize the expression distribution of DEGs
identified by differential expression analysis</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotDEGRegression.html">plotDEGRegression()</a></code> </p>
        </td>
        <td><p>plot the linear regression to show visualize the expression the of DEGs
identified by differential expression analysis</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotDEGHeatmap.html">plotDEGHeatmap()</a></code> </p>
        </td>
        <td><p>Heatmap visualization of DEG result</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="thresholdGenes.html">thresholdGenes()</a></code> </p>
        </td>
        <td><p>MAST Identify adaptive thresholds</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="findMarkerDiffExp.html">findMarkerDiffExp()</a></code> </p>
        </td>
        <td><p>Find the marker gene set for each cluster
With an input SingleCellExperiment object and specifying the clustering
labels, this function iteratively call the differential expression analysis
on each cluster against all the others.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotMarkerDiffExp.html">plotMarkerDiffExp()</a></code> </p>
        </td>
        <td><p>Plot a heatmap to visualize the result of <code>findMarkerDiffExp</code></p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-pathway-activity-analysis" class="hasAnchor"><a href="#section-pathway-activity-analysis" class="anchor"></a>Pathway Activity Analysis</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="gsvaSCE.html">gsvaSCE()</a></code> <code><a href="gsvaSCE.html">gsvaPlot()</a></code> </p>
        </td>
        <td><p>Run GSVA analysis on a SingleCellExperiment object</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-sample-size" class="hasAnchor"><a href="#section-sample-size" class="anchor"></a>Sample Size</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="calcEffectSizes.html">calcEffectSizes()</a></code> </p>
        </td>
        <td><p>Finds the effect sizes for all genes in the original dataset, regardless of
significance.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="generateSimulatedData.html">generateSimulatedData()</a></code> </p>
        </td>
        <td><p>Generates a single simulated dataset, bootstrapping from the input counts
matrix.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="iterateSimulations.html">iterateSimulations()</a></code> </p>
        </td>
        <td><p>Returns significance data from a snapshot.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="subDiffEx.html">subDiffEx()</a></code> <code><a href="subDiffEx.html">subDiffExttest()</a></code> <code><a href="subDiffEx.html">subDiffExANOVA()</a></code> </p>
        </td>
        <td><p>Passes the output of generateSimulatedData() to differential expression
tests, picking either t-tests or ANOVA for data with only two conditions or
multiple conditions, respectively.</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-data" class="hasAnchor"><a href="#section-data" class="anchor"></a>Data</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="mouseBrainSubsetSCE.html">mouseBrainSubsetSCE</a></code> </p>
        </td>
        <td><p>Example Single Cell RNA-Seq data in SingleCellExperiment Object, GSE60361
subset</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-sctkexperiment" class="hasAnchor"><a href="#section-sctkexperiment" class="anchor"></a>SCtkExperiment</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-alignment" class="hasAnchor"><a href="#section-alignment" class="anchor"></a>Alignment</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="alignSingleCellData.html">alignSingleCellData()</a></code> </p>
        </td>
        <td><p>Align Single Cell RNA-Seq Data and Create a SingleCellExperiment Object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="parseRsubreadLogs.html">parseRsubreadLogs()</a></code> </p>
        </td>
        <td><p>Parse Rsubread Logs for Mapping and Feature Count Statistics</p></td>
      </tr>
    </tbody>
    </table>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by David Jenkins, Tyler Faits, Mohammed Muzamil Khan, Zhe Wang, Emma Briars, Sebastian Carrasco Pro, W. Evan Johnson.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


